{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "03df5b93",
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a44bd48b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# pip install --upgrade google-api-python-client\n",
    "from google.auth.transport.requests import Request\n",
    "from google.oauth2.credentials import Credentials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0e8629c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sudo pip install google-auth-oauthlib\n",
    "from google_auth_oauthlib.flow import InstalledAppFlow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b6a8be40",
   "metadata": {},
   "outputs": [],
   "source": [
    "from linkml_runtime.utils.schemaview import SchemaView"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fcfa26b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from shutil import copyfile\n",
    "import os\n",
    "import os.path\n",
    "\n",
    "import re\n",
    "\n",
    "import json\n",
    "import yaml\n",
    "\n",
    "from linkml_runtime.linkml_model.meta import SlotDefinition, ClassDefinition\n",
    "\n",
    "from linkml.generators.yamlgen import YAMLGenerator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c97ec101",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b219e2ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# AttributeError: module 'pyparsing' has no attribute 'downcaseTokens'\n",
    "# ERROR: httplib2 0.20.1 has requirement pyparsing<3,>=2.4.2, but you'll have pyparsing 3.0.3 which is incompatible.\n",
    "from googleapiclient.discovery import build"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cd18d196",
   "metadata": {},
   "outputs": [],
   "source": [
    "blank_row = {\n",
    "    \"Ontology ID\": \"\",\n",
    "    \"parent class\": \"\",\n",
    "    \"label\": \"\",\n",
    "    \"datatype\": \"\",\n",
    "    \"source\": \"\",\n",
    "    \"data status\": \"\",\n",
    "    \"requirement\": \"\",\n",
    "    \"min value\": \"\",\n",
    "    \"max value\": \"\",\n",
    "    \"capitalize\": \"\",\n",
    "    \"pattern\": \"\",\n",
    "    \"description\": \"\",\n",
    "    \"guidance\": \"\",\n",
    "    \"examples\": \"\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dde1aadd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# required columns can be asserted without putting them in a section entitled \"required\" etc.\n",
    "#   ie it might be possible to use the section for something orthogonal\n",
    "required_sections = [\"sample identification\", \"required\", \"required where applicable\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "513fde5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dh_template_root = \"../../DataHarmonizer/template/\"\n",
    "dh_template_prefix = dh_template_root + \"IoT_\"\n",
    "dh_template_suffix = \"/data.tsv\"\n",
    "ref_temp_filename = \"reference_template.html\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "af4ee814",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for making a template based on MIxS slots\n",
    "#   from https://github.com/cmungall/mixs-source\n",
    "#   (GenomicsStandardsConsortium/mixs-source might be slightly out of sync)\n",
    "\n",
    "# make sure to regenrate this form tiemto time\n",
    "mixs_yaml_file = \"../../mixs-source/model/schema/mixs.yaml\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0e440a0",
   "metadata": {},
   "source": [
    "Both use same sheet: 1QDeeUcDqXes69Y2RjU2aWgOpCVWo5OVsBX9MKmMqi_o\n",
    "\n",
    "```\n",
    "                Mungall      GSC\n",
    "mixs6.tsv       345753674    750683809\n",
    "mixs6_core.tsv  567040283    178015749\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f174bc71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Montana's \"Index-of-Terms\" sheet is in XLSX format:\n",
    "#   https://docs.google.com/spreadsheets/d/1R5w5tcIGm9dDRhC4H6NmhcHV7J5zx7FQ/edit#gid=1133203354\n",
    "# I copied and converted to Google Sheet format in MAM@lbl.gov google drive\n",
    "#   https://docs.google.com/spreadsheets/d/1lj4OuEE4IYwy2v7RzcG79lHjNdFwmDETMDTDaRAWojY/edit#gid=1133203354\n",
    "\n",
    "# added mixs_6_slot_name column\n",
    "#   blank if no corresponding mixs_6_slot\n",
    "#   different from \"name\" column if \"name\" column can be repaired\n",
    "\n",
    "# there's also a \"Controlled Terms\" tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "44bc71be",
   "metadata": {},
   "outputs": [],
   "source": [
    "SCOPES = [\"https://www.googleapis.com/auth/spreadsheets.readonly\"]\n",
    "\n",
    "# The ID and range of a sample spreadsheet.\n",
    "IOT_SPREADSHEET_ID = \"1lj4OuEE4IYwy2v7RzcG79lHjNdFwmDETMDTDaRAWojY\"\n",
    "IOT_RANGE_NAME = \"Glossary of terms!A1:Z\"\n",
    "\n",
    "CV_RANGE_NAME = \"Controlled Terms!A1:Z\"\n",
    "\n",
    "# resuing google_api_credentials.json from https://github.com/cancerDHC/sheet2linkml\n",
    "client_secret_flie = \"../../sheet2linkml/google_api_credentials.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d08c5599",
   "metadata": {},
   "outputs": [],
   "source": [
    "creds = None\n",
    "# The file token.json stores the user's access and refresh tokens, and is\n",
    "# created automatically when the authorization flow completes for the first\n",
    "# time.\n",
    "if os.path.exists(\"token.json\"):\n",
    "    creds = Credentials.from_authorized_user_file(\"token.json\", SCOPES)\n",
    "# If there are no (valid) credentials available, let the user log in.\n",
    "if not creds or not creds.valid:\n",
    "    if creds and creds.expired and creds.refresh_token:\n",
    "        creds.refresh(Request())\n",
    "    else:\n",
    "        flow = InstalledAppFlow.from_client_secrets_file(client_secret_flie, SCOPES)\n",
    "        creds = flow.run_local_server(port=0)\n",
    "    # Save the credentials for the next run\n",
    "    with open(\"token.json\", \"w\") as token:\n",
    "        token.write(creds.to_json())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1a0d6631",
   "metadata": {},
   "outputs": [],
   "source": [
    "service = build(\"sheets\", \"v4\", credentials=creds)\n",
    "\n",
    "# Call the Sheets API\n",
    "sheet = service.spreadsheets()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e4d72c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = (\n",
    "    sheet.values().get(spreadsheetId=IOT_SPREADSHEET_ID, range=IOT_RANGE_NAME).execute()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fb3aa315",
   "metadata": {},
   "outputs": [],
   "source": [
    "iot_glossary = pd.DataFrame(result[\"values\"], columns=result[\"values\"][0]).drop(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "72e7ee83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column Header</th>\n",
       "      <th>name</th>\n",
       "      <th>mixs_6_slot_name</th>\n",
       "      <th>different?</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Guidance</th>\n",
       "      <th>Expected Value</th>\n",
       "      <th>syntax</th>\n",
       "      <th>Category</th>\n",
       "      <th>Associated Packages</th>\n",
       "      <th>Origin</th>\n",
       "      <th>Notes</th>\n",
       "      <th>GitHub Ticket</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Unique ID</td>\n",
       "      <td>unique_ID</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>A globally unique identifier assigned to the b...</td>\n",
       "      <td>Field REQUIRED for ALL sample submission. Opti...</td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>all</td>\n",
       "      <td></td>\n",
       "      <td>v6 has unique ID and sample name synonymous. B...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sample Name</td>\n",
       "      <td>samp_name</td>\n",
       "      <td>samp_name</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>A local identifier or name that for the materi...</td>\n",
       "      <td>ID that is present on the shipped sample. Huma...</td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>all</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>v6</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Analysis/Data Type</td>\n",
       "      <td>investigation_type</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td>This field is constrained to contain only a se...</td>\n",
       "      <td></td>\n",
       "      <td>drop down selection list</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>all</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>enumeration values are EMSL origin, MIxS has d...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sample Type</td>\n",
       "      <td>package</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td>MIxS Package</td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>all</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>This is assumed by the package, but I think sh...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Source Material ID</td>\n",
       "      <td>source_mat_ID</td>\n",
       "      <td>source_mat_id</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>A unique identifier assigned to a material sam...</td>\n",
       "      <td>A unique identifier assigned to an original ma...</td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>all</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>New v6 definition is better to indicate bio sa...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>wastewater type</td>\n",
       "      <td>wastewater_type</td>\n",
       "      <td>wastewater_type</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>the origin of wastewater such as human waste, ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td></td>\n",
       "      <td>water_extract_biosolid</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>water content</td>\n",
       "      <td>water_content</td>\n",
       "      <td>water_content</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>Water content measurement</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{float}</td>\n",
       "      <td></td>\n",
       "      <td>soil; sediment; water_extract_soil</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>water content method</td>\n",
       "      <td>water_content_meth</td>\n",
       "      <td>water_cont_soil_meth</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>Reference or method used in determining the wa...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{PMID}|{DOI}|{URL}</td>\n",
       "      <td></td>\n",
       "      <td>soil; sediment</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>water current</td>\n",
       "      <td>water_current</td>\n",
       "      <td>water_current</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>Measurement of magnitude and direction of flow...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{float} {unit}</td>\n",
       "      <td></td>\n",
       "      <td>pore_water; water</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>water temperature regimen</td>\n",
       "      <td>water_temp_regm</td>\n",
       "      <td>water_temp_regm</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>Information about treatment involving an expos...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>measurement value {value} {unit}; treatment in...</td>\n",
       "      <td></td>\n",
       "      <td>plant_associated</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>248 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Column Header                name      mixs_6_slot_name  \\\n",
       "1                    Unique ID           unique_ID                         \n",
       "2                  Sample Name           samp_name             samp_name   \n",
       "3           Analysis/Data Type  investigation_type                         \n",
       "4                  Sample Type             package                         \n",
       "5           Source Material ID       source_mat_ID         source_mat_id   \n",
       "..                         ...                 ...                   ...   \n",
       "244            wastewater type     wastewater_type       wastewater_type   \n",
       "245              water content       water_content         water_content   \n",
       "246       water content method  water_content_meth  water_cont_soil_meth   \n",
       "247              water current       water_current         water_current   \n",
       "248  water temperature regimen     water_temp_regm       water_temp_regm   \n",
       "\n",
       "    different?                                         Definition  \\\n",
       "1        FALSE  A globally unique identifier assigned to the b...   \n",
       "2         TRUE  A local identifier or name that for the materi...   \n",
       "3        FALSE                                                      \n",
       "4        FALSE                                                      \n",
       "5         TRUE  A unique identifier assigned to a material sam...   \n",
       "..         ...                                                ...   \n",
       "244       TRUE  the origin of wastewater such as human waste, ...   \n",
       "245       TRUE                          Water content measurement   \n",
       "246      FALSE  Reference or method used in determining the wa...   \n",
       "247       TRUE  Measurement of magnitude and direction of flow...   \n",
       "248       TRUE  Information about treatment involving an expos...   \n",
       "\n",
       "                                              Guidance Expected Value  \\\n",
       "1    Field REQUIRED for ALL sample submission. Opti...                  \n",
       "2    ID that is present on the shipped sample. Huma...                  \n",
       "3    This field is constrained to contain only a se...                  \n",
       "4                                         MIxS Package                  \n",
       "5    A unique identifier assigned to an original ma...                  \n",
       "..                                                 ...            ...   \n",
       "244                                                                     \n",
       "245                                                                     \n",
       "246                                                                     \n",
       "247                                                                     \n",
       "248                                                                     \n",
       "\n",
       "                                                syntax               Category  \\\n",
       "1                                               {text}  sample identification   \n",
       "2                                               {text}  sample identification   \n",
       "3                             drop down selection list  sample identification   \n",
       "4                                               {text}  sample identification   \n",
       "5                                               {text}  sample identification   \n",
       "..                                                 ...                    ...   \n",
       "244                                             {text}                          \n",
       "245                                            {float}                          \n",
       "246                                 {PMID}|{DOI}|{URL}                          \n",
       "247                                     {float} {unit}                          \n",
       "248  measurement value {value} {unit}; treatment in...                          \n",
       "\n",
       "                    Associated Packages Origin  \\\n",
       "1                                   all          \n",
       "2                                   all   MIxS   \n",
       "3                                   all   MIxS   \n",
       "4                                   all   MIxS   \n",
       "5                                   all   MIxS   \n",
       "..                                  ...    ...   \n",
       "244              water_extract_biosolid   MIxS   \n",
       "245  soil; sediment; water_extract_soil   MIxS   \n",
       "246                      soil; sediment   MIxS   \n",
       "247                   pore_water; water   MIxS   \n",
       "248                    plant_associated   MIxS   \n",
       "\n",
       "                                                 Notes GitHub Ticket  \n",
       "1    v6 has unique ID and sample name synonymous. B...          None  \n",
       "2                                                   v6          None  \n",
       "3    enumeration values are EMSL origin, MIxS has d...          None  \n",
       "4    This is assumed by the package, but I think sh...          None  \n",
       "5    New v6 definition is better to indicate bio sa...          None  \n",
       "..                                                 ...           ...  \n",
       "244                                               None          None  \n",
       "245                                               None          None  \n",
       "246                                               None          None  \n",
       "247                                               None          None  \n",
       "248                                               None          None  \n",
       "\n",
       "[248 rows x 13 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iot_glossary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fdd0cc10",
   "metadata": {},
   "outputs": [],
   "source": [
    "controlled_terms = (\n",
    "    sheet.values().get(spreadsheetId=IOT_SPREADSHEET_ID, range=CV_RANGE_NAME).execute()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "77017485",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctdf = pd.DataFrame(\n",
    "    controlled_terms[\"values\"], columns=controlled_terms[\"values\"][0]\n",
    ").drop(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5a167f63",
   "metadata": {},
   "outputs": [],
   "source": [
    "ct_dol = {k: [i for i in v if i] for (k, v) in ctdf.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ffe50e4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'investigation_type': ['genome',\n",
       "  'transcriptome',\n",
       "  'proteome',\n",
       "  'metabolome',\n",
       "  'lipidome',\n",
       "  'organic matter',\n",
       "  'imaging- light',\n",
       "  'imaging- electron',\n",
       "  'imaging- ion',\n",
       "  'chemical speciation/mapping',\n",
       "  'molecular structure'],\n",
       " 'growth_facil': ['field',\n",
       "  'field_incubation',\n",
       "  'growth_chamber',\n",
       "  'open_top_chamber',\n",
       "  'greenhouse',\n",
       "  'experimental_garden',\n",
       "  'lab_incubation',\n",
       "  'other'],\n",
       " 'store_cond': ['fresh', 'frozen', 'lyophilized', 'other'],\n",
       " 'samp_store_temp': ['4 degree Celsius',\n",
       "  '-20 degree Celsius',\n",
       "  '-80 degree Celsius',\n",
       "  'room temperature',\n",
       "  'other'],\n",
       " 'plant_struc': ['stem',\n",
       "  'leaf',\n",
       "  'root',\n",
       "  'fine_roots',\n",
       "  'whole_plant',\n",
       "  'rhizodeposits'],\n",
       " 'growth_medium': ['media_composition',\n",
       "  'broth',\n",
       "  'agar_plate',\n",
       "  '3D-structure',\n",
       "  'plate',\n",
       "  'bioreactor'],\n",
       " 'biol_stat': ['wild',\n",
       "  'natural',\n",
       "  'semi-natural',\n",
       "  'inbred line',\n",
       "  \"breeder's line\",\n",
       "  'hybrid',\n",
       "  'clonal selection',\n",
       "  'mutant'],\n",
       " 'biotic_relationship': ['free living',\n",
       "  'parasitism',\n",
       "  'commensalism',\n",
       "  'symbiotic',\n",
       "  'mutualism'],\n",
       " 'cur_land_use': ['Badlands',\n",
       "  'Cities',\n",
       "  'Conifers (e.g. pine,spruce,fir,cypress)',\n",
       "  'Crop Trees (nuts,fruit,christmas trees,nursery trees)',\n",
       "  'Farmstead',\n",
       "  'Gravel',\n",
       "  'Hardwoods (e.g. oak,hickory,elm,aspen)',\n",
       "  'Hayland',\n",
       "  'Horticultural Plants (e.g. tulips)',\n",
       "  'Industrial Areas',\n",
       "  'Intermixed Hardwood and Conifers',\n",
       "  'Marshlands (grass,sedges,rushes)',\n",
       "  'Meadows (grasses,alfalfa,fescue,bromegrass,timothy)',\n",
       "  'Mines/Quarries',\n",
       "  'Mudflats',\n",
       "  'Oil Waste Areas',\n",
       "  'Pastureland (grasslands used for livestock grazing)',\n",
       "  'Permanent Snow or Ice',\n",
       "  'Rainforest (evergreen forest receiving >406 cm annual rainfall)',\n",
       "  'Rangeland',\n",
       "  'Roads/Railroads',\n",
       "  'Rock',\n",
       "  'Row Crops',\n",
       "  'Saline Seeps',\n",
       "  'Salt Flats',\n",
       "  'Sand',\n",
       "  'Shrub Crops (blueberries,nursery ornamentals,filberts)',\n",
       "  'Shrub Land (e.g. mesquite,sage-brush,creosote bush,shrub oak,eucalyptus)',\n",
       "  'Small Grains',\n",
       "  'Successional Shrub Land (tree saplings,hazels,sumacs,chokecherry,shrub dogwoods,blackberries)',\n",
       "  'Swamp (permanent or semi-permanent water body dominated by woody plants)',\n",
       "  'Tropical (e.g. mangrove,palms)',\n",
       "  'Tundra (mosses,lichens)',\n",
       "  'Vegetable Crops',\n",
       "  'Vine Crops (grapes)'],\n",
       " 'drainage_class': ['Excessively Drained',\n",
       "  'Moderately Well',\n",
       "  'Poorly',\n",
       "  'Somewhat Poorly',\n",
       "  'Very Poorly',\n",
       "  'Well'],\n",
       " 'fao_class': ['Acrisols',\n",
       "  'Andosols',\n",
       "  'Arenosols',\n",
       "  'Cambisols',\n",
       "  'Chernozems',\n",
       "  'Ferralsols',\n",
       "  'Fluvisols',\n",
       "  'Gleysols',\n",
       "  'Greyzems',\n",
       "  'Gypsisols',\n",
       "  'Histosols',\n",
       "  'Kastanozems',\n",
       "  'Lithosols',\n",
       "  'Luvisols',\n",
       "  'Nitosols',\n",
       "  'Phaeozems',\n",
       "  'Planosols',\n",
       "  'Podzols',\n",
       "  'Podzoluvisols',\n",
       "  'Rankers',\n",
       "  'Regosols',\n",
       "  'Rendzinas',\n",
       "  'Solonchaks',\n",
       "  'Solonetz',\n",
       "  'Vertisols',\n",
       "  'Yermosols'],\n",
       " 'growth_habit': ['erect', 'semi-erect', 'spreading', 'prostrate'],\n",
       " 'horizon': ['A Horizon',\n",
       "  'B Horizon',\n",
       "  'C Horizon',\n",
       "  'E Horizon',\n",
       "  'O Horizon',\n",
       "  'Permafrost',\n",
       "  'R Layer'],\n",
       " 'oxy_stat_samp': ['aerobic',\n",
       "  'anaerobic',\n",
       "  'anoxic',\n",
       "  'facultative',\n",
       "  'microaerophilic',\n",
       "  'microanaerobe',\n",
       "  'obligate aerobe',\n",
       "  'obligate anaerobe'],\n",
       " 'plant_sex': ['Androdioecious',\n",
       "  'Androecious',\n",
       "  'Androgynomonoecious',\n",
       "  'Andromonoecious',\n",
       "  'Bisexual',\n",
       "  'Dichogamous',\n",
       "  'Diclinous',\n",
       "  'Dioecious',\n",
       "  'Gynodioecious',\n",
       "  'Gynoecious',\n",
       "  'Gynomonoecious',\n",
       "  'Gynoecious',\n",
       "  'Gynomonoecious',\n",
       "  'Hermaphroditic',\n",
       "  'Imperfect',\n",
       "  'Monoclinous',\n",
       "  'Monoecious',\n",
       "  'Perfect',\n",
       "  'Polygamodioecious',\n",
       "  'Polygamomonoecious',\n",
       "  'Polygamous',\n",
       "  'Protandrous',\n",
       "  'Protogynous',\n",
       "  'Subandroecious',\n",
       "  'Subdioecious',\n",
       "  'Subgynoecious',\n",
       "  'Synoecious',\n",
       "  'Trimonoecious',\n",
       "  'Trioecious',\n",
       "  'Unisexual',\n",
       "  'Hermaphroditic'],\n",
       " 'profile_position': ['profile_position',\n",
       "  'Backslope',\n",
       "  'Footslope',\n",
       "  'Shoulder',\n",
       "  'Summit',\n",
       "  'Toeslope'],\n",
       " 'samp_capt_status': ['active surveillance in response to an outbreak',\n",
       "  'active surveillance not initiated by an outbreak',\n",
       "  'farm sample',\n",
       "  'market sample',\n",
       "  'other'],\n",
       " 'samp_dis_stage': ['dissemination',\n",
       "  'growth and reproduction',\n",
       "  'infection',\n",
       "  'inoculation',\n",
       "  'penetration',\n",
       "  'other'],\n",
       " 'sediment_type': ['biogenous', 'cosmogenous', 'hydrogenous', 'lithogenous'],\n",
       " 'tidal_stage': ['low tide', 'ebb tide', 'flood tide', 'high tide'],\n",
       " 'tillage': ['Chisel',\n",
       "  'Cutting Disc',\n",
       "  'Disc Plough',\n",
       "  'Drill',\n",
       "  'Mouldboard',\n",
       "  'Ridge Till',\n",
       "  'Strip Tillage',\n",
       "  'Tined',\n",
       "  'Zonal Tillage'],\n",
       " 'trophic_level': ['autotroph',\n",
       "  'carboxydotroph',\n",
       "  'chemoautotroph',\n",
       "  'chemoheterotroph',\n",
       "  'chemolithoautotroph',\n",
       "  'chemolithotroph',\n",
       "  'chemoorganoheterotroph',\n",
       "  'chemoorganotroph',\n",
       "  'chemosynthetic',\n",
       "  'chemotroph',\n",
       "  'copiotroph',\n",
       "  'diazotroph',\n",
       "  'facultative',\n",
       "  'autotroph',\n",
       "  'heterotroph',\n",
       "  'lithoautotroph',\n",
       "  'lithoheterotroph',\n",
       "  'lithotroph',\n",
       "  'methanotroph',\n",
       "  'methylotroph',\n",
       "  'mixotroph',\n",
       "  'obligate',\n",
       "  'chemoautolithotroph',\n",
       "  'oligotroph',\n",
       "  'organoheterotroph',\n",
       "  'organotroph',\n",
       "  'photoautotroph',\n",
       "  'photoheterotroph',\n",
       "  'photolithoautotroph',\n",
       "  'photolithotroph',\n",
       "  'photosynthetic',\n",
       "  'phototroph']}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct_dol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "79d7c0c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['biol_stat',\n",
       " 'biotic_relationship',\n",
       " 'cur_land_use',\n",
       " 'drainage_class',\n",
       " 'fao_class',\n",
       " 'growth_facil',\n",
       " 'growth_habit',\n",
       " 'growth_medium',\n",
       " 'horizon',\n",
       " 'investigation_type',\n",
       " 'oxy_stat_samp',\n",
       " 'plant_sex',\n",
       " 'plant_struc',\n",
       " 'profile_position',\n",
       " 'samp_capt_status',\n",
       " 'samp_dis_stage',\n",
       " 'samp_store_temp',\n",
       " 'sediment_type',\n",
       " 'store_cond',\n",
       " 'tidal_stage',\n",
       " 'tillage',\n",
       " 'trophic_level']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct_keys = list(ct_dol.keys())\n",
    "ct_keys.sort()\n",
    "ct_keys"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "519682cf",
   "metadata": {},
   "source": [
    "## Compare to MIxS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "23d917eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for which is the \"column header\" different from the display name?\n",
    "# for which is the \"definition\" different from ???\n",
    "# guidance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "63e917c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mixs_yaml_view = SchemaView(mixs_yaml_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c444ab88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# which \"name\" values aren't MiXS slots at all?\n",
    "# could be minor spelling differences etc.\n",
    "\n",
    "all_mixs_slots_dict = mixs_yaml_view.all_slots()\n",
    "ams_keys = list(all_mixs_slots_dict.keys())\n",
    "ams_keys.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c19dce0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['?? Incubation collection date',\n",
       " '?? Incubation start date',\n",
       " 'extreme_salinity',\n",
       " 'host_infra_spec_name',\n",
       " 'host_infra_spec_rank',\n",
       " 'investigation_type',\n",
       " 'isotope_exposure',\n",
       " 'other',\n",
       " 'package',\n",
       " 'sample_size',\n",
       " 'source_mat_ID',\n",
       " 'tot_nitro_content_meth',\n",
       " 'unique_ID',\n",
       " 'water_content_meth']"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_iot_slots = iot_glossary[\"name\"]\n",
    "\n",
    "iot_slots_only = list(set(all_iot_slots) - set(ams_keys))\n",
    "\n",
    "iot_slots_only.sort()\n",
    "\n",
    "iot_slots_only"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e477f452",
   "metadata": {},
   "source": [
    "```\n",
    " '?? Incubation collection date',\n",
    " '?? Incubation start date',\n",
    " 'extreme_salinity',\n",
    " 'host_infra_spec_name', host_infra_specific_name\n",
    " 'host_infra_spec_rank', host_infra_specific_rank\n",
    " 'investigation_type',\n",
    " 'isotope_exposure',\n",
    " 'other',\n",
    " 'package',\n",
    " 'sample_size', samp_size\n",
    " 'source_mat_ID', source_mat_id\n",
    " 'tot_nitro_content_meth', tot_n_meth\n",
    " 'unique_ID',\n",
    " 'water_content_meth' water_cont_soil_meth\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "dac4a34f",
   "metadata": {},
   "outputs": [],
   "source": [
    "iot_undefined_slots_followup = iot_glossary.loc[\n",
    "    iot_glossary[\"name\"].isin(iot_slots_only)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "eb61690c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column Header</th>\n",
       "      <th>name</th>\n",
       "      <th>mixs_6_slot_name</th>\n",
       "      <th>different?</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Guidance</th>\n",
       "      <th>Expected Value</th>\n",
       "      <th>syntax</th>\n",
       "      <th>Category</th>\n",
       "      <th>Associated Packages</th>\n",
       "      <th>Origin</th>\n",
       "      <th>Notes</th>\n",
       "      <th>GitHub Ticket</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Unique ID</td>\n",
       "      <td>unique_ID</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>A globally unique identifier assigned to the b...</td>\n",
       "      <td>Field REQUIRED for ALL sample submission. Opti...</td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>all</td>\n",
       "      <td></td>\n",
       "      <td>v6 has unique ID and sample name synonymous. B...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Analysis/Data Type</td>\n",
       "      <td>investigation_type</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td>This field is constrained to contain only a se...</td>\n",
       "      <td></td>\n",
       "      <td>drop down selection list</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>all</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>enumeration values are EMSL origin, MIxS has d...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sample Type</td>\n",
       "      <td>package</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td>MIxS Package</td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>all</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>This is assumed by the package, but I think sh...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Source Material ID</td>\n",
       "      <td>source_mat_ID</td>\n",
       "      <td>source_mat_id</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>A unique identifier assigned to a material sam...</td>\n",
       "      <td>A unique identifier assigned to an original ma...</td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>all</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>New v6 definition is better to indicate bio sa...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td></td>\n",
       "      <td>?? Incubation collection date</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Need to come up with a term for this or how we...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td></td>\n",
       "      <td>?? Incubation start date</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Sample size</td>\n",
       "      <td>sample_size</td>\n",
       "      <td>samp_size</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>The total amount or size (volume (ml), mass (g...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{value}{text}</td>\n",
       "      <td>required</td>\n",
       "      <td>all</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>I need how much sample is sent to EMSL. This i...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>isotope exposure/addition treatment</td>\n",
       "      <td>isotope_exposure</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>required where applicable</td>\n",
       "      <td>soil; sediment; pore_water; plant_associated; ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://github.com/GenomicsStandardsConsortium...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>treatment-other details</td>\n",
       "      <td>other</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>required where applicable</td>\n",
       "      <td>soil; sediment; pore_water; plant_associated; ...</td>\n",
       "      <td></td>\n",
       "      <td>link github ticket</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Other</td>\n",
       "      <td>other</td>\n",
       "      <td>other</td>\n",
       "      <td>TRUE</td>\n",
       "      <td></td>\n",
       "      <td>Other details about your sample that you feel ...</td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>required where applicable</td>\n",
       "      <td>all</td>\n",
       "      <td></td>\n",
       "      <td>I would like somewhere to  caputre any informa...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>extreme_unusual_properties/salinity</td>\n",
       "      <td>extreme_salinity</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>Measured salinity</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{float} {unit}</td>\n",
       "      <td></td>\n",
       "      <td>soil; water_extract_soil</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>host infra-specific name</td>\n",
       "      <td>host_infra_spec_name</td>\n",
       "      <td>host_infra_specific_name</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>Taxonomic information about the host below sub...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td></td>\n",
       "      <td>plant_associated</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>host infra-specific rank</td>\n",
       "      <td>host_infra_spec_rank</td>\n",
       "      <td>host_infra_specific_rank</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>Taxonomic rank information about the host belo...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td></td>\n",
       "      <td>plant_associated</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>total nitrogen content method</td>\n",
       "      <td>tot_nitro_content_meth</td>\n",
       "      <td>tot_n_meth</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>Reference or method used in determining the to...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{PMID}|{DOI}|{URL}</td>\n",
       "      <td></td>\n",
       "      <td>soil; sediment; water_extract_soil</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>water content method</td>\n",
       "      <td>water_content_meth</td>\n",
       "      <td>water_cont_soil_meth</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>Reference or method used in determining the wa...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{PMID}|{DOI}|{URL}</td>\n",
       "      <td></td>\n",
       "      <td>soil; sediment</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Column Header                           name  \\\n",
       "1                              Unique ID                      unique_ID   \n",
       "3                     Analysis/Data Type             investigation_type   \n",
       "4                            Sample Type                        package   \n",
       "5                     Source Material ID                  source_mat_ID   \n",
       "8                                         ?? Incubation collection date   \n",
       "9                                              ?? Incubation start date   \n",
       "13                           Sample size                    sample_size   \n",
       "22   isotope exposure/addition treatment               isotope_exposure   \n",
       "27               treatment-other details                          other   \n",
       "39                                 Other                          other   \n",
       "91   extreme_unusual_properties/salinity               extreme_salinity   \n",
       "114             host infra-specific name           host_infra_spec_name   \n",
       "115             host infra-specific rank           host_infra_spec_rank   \n",
       "236        total nitrogen content method         tot_nitro_content_meth   \n",
       "246                 water content method             water_content_meth   \n",
       "\n",
       "             mixs_6_slot_name different?  \\\n",
       "1                                  FALSE   \n",
       "3                                  FALSE   \n",
       "4                                  FALSE   \n",
       "5               source_mat_id       TRUE   \n",
       "8                                  FALSE   \n",
       "9                                  FALSE   \n",
       "13                  samp_size      FALSE   \n",
       "22                                 FALSE   \n",
       "27                                 FALSE   \n",
       "39                      other       TRUE   \n",
       "91                                 FALSE   \n",
       "114  host_infra_specific_name      FALSE   \n",
       "115  host_infra_specific_rank      FALSE   \n",
       "236                tot_n_meth      FALSE   \n",
       "246      water_cont_soil_meth      FALSE   \n",
       "\n",
       "                                            Definition  \\\n",
       "1    A globally unique identifier assigned to the b...   \n",
       "3                                                        \n",
       "4                                                        \n",
       "5    A unique identifier assigned to a material sam...   \n",
       "8                                                        \n",
       "9                                                 None   \n",
       "13   The total amount or size (volume (ml), mass (g...   \n",
       "22                                                       \n",
       "27                                                       \n",
       "39                                                       \n",
       "91                                   Measured salinity   \n",
       "114  Taxonomic information about the host below sub...   \n",
       "115  Taxonomic rank information about the host belo...   \n",
       "236  Reference or method used in determining the to...   \n",
       "246  Reference or method used in determining the wa...   \n",
       "\n",
       "                                              Guidance Expected Value  \\\n",
       "1    Field REQUIRED for ALL sample submission. Opti...                  \n",
       "3    This field is constrained to contain only a se...                  \n",
       "4                                         MIxS Package                  \n",
       "5    A unique identifier assigned to an original ma...                  \n",
       "8                                                                       \n",
       "9                                                 None           None   \n",
       "13                                                                      \n",
       "22                                                                      \n",
       "27                                                                      \n",
       "39   Other details about your sample that you feel ...                  \n",
       "91                                                                      \n",
       "114                                                                     \n",
       "115                                                                     \n",
       "236                                                                     \n",
       "246                                                                     \n",
       "\n",
       "                       syntax                   Category  \\\n",
       "1                      {text}      sample identification   \n",
       "3    drop down selection list      sample identification   \n",
       "4                      {text}      sample identification   \n",
       "5                      {text}      sample identification   \n",
       "8                                                          \n",
       "9                        None                       None   \n",
       "13              {value}{text}                   required   \n",
       "22                     {text}  required where applicable   \n",
       "27                     {text}  required where applicable   \n",
       "39                     {text}  required where applicable   \n",
       "91             {float} {unit}                              \n",
       "114                    {text}                              \n",
       "115                    {text}                              \n",
       "236        {PMID}|{DOI}|{URL}                              \n",
       "246        {PMID}|{DOI}|{URL}                              \n",
       "\n",
       "                                   Associated Packages Origin  \\\n",
       "1                                                  all          \n",
       "3                                                  all   MIxS   \n",
       "4                                                  all   MIxS   \n",
       "5                                                  all   MIxS   \n",
       "8                                                               \n",
       "9                                                 None   None   \n",
       "13                                                 all   MIxS   \n",
       "22   soil; sediment; pore_water; plant_associated; ...          \n",
       "27   soil; sediment; pore_water; plant_associated; ...          \n",
       "39                                                 all          \n",
       "91                            soil; water_extract_soil   MIxS   \n",
       "114                                   plant_associated   MIxS   \n",
       "115                                   plant_associated   MIxS   \n",
       "236                 soil; sediment; water_extract_soil   MIxS   \n",
       "246                                     soil; sediment   MIxS   \n",
       "\n",
       "                                                 Notes  \\\n",
       "1    v6 has unique ID and sample name synonymous. B...   \n",
       "3    enumeration values are EMSL origin, MIxS has d...   \n",
       "4    This is assumed by the package, but I think sh...   \n",
       "5    New v6 definition is better to indicate bio sa...   \n",
       "8    Need to come up with a term for this or how we...   \n",
       "9                                                 None   \n",
       "13   I need how much sample is sent to EMSL. This i...   \n",
       "22                                                       \n",
       "27                                  link github ticket   \n",
       "39   I would like somewhere to  caputre any informa...   \n",
       "91                                                None   \n",
       "114                                               None   \n",
       "115                                               None   \n",
       "236                                               None   \n",
       "246                                               None   \n",
       "\n",
       "                                         GitHub Ticket  \n",
       "1                                                 None  \n",
       "3                                                 None  \n",
       "4                                                 None  \n",
       "5                                                 None  \n",
       "8                                                 None  \n",
       "9                                                 None  \n",
       "13                                                None  \n",
       "22   https://github.com/GenomicsStandardsConsortium...  \n",
       "27                                                None  \n",
       "39                                                None  \n",
       "91                                                None  \n",
       "114                                               None  \n",
       "115                                               None  \n",
       "236                                               None  \n",
       "246                                               None  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iot_undefined_slots_followup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "85b62364",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_check = iot_glossary[[\"name\", \"mixs_6_slot_name\"]]\n",
    "slot_check = slot_check.loc[slot_check[\"name\"].ne(slot_check[\"mixs_6_slot_name\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9b98abe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_check.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "562abae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "iot_package_slots = iot_glossary[[\"mixs_6_slot_name\", \"Associated Packages\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1f8333d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "iot_package_slots = iot_package_slots.loc[iot_package_slots[\"mixs_6_slot_name\"].ne(\"\")]\n",
    "iot_package_slots = iot_package_slots.loc[\n",
    "    ~iot_package_slots[\"mixs_6_slot_name\"].isnull()\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c49139c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "iot_package_slots.to_clipboard()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "44b86308",
   "metadata": {},
   "outputs": [],
   "source": [
    "iot_package_slots = iot_package_slots.explode(\"Associated Packages\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d08cf28d",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_package_slots = list(\n",
    "    iot_package_slots[\"mixs_6_slot_name\"].loc[\n",
    "        iot_package_slots[\"Associated Packages\"].eq(\"all\")\n",
    "    ]\n",
    ")\n",
    "all_package_slots.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "41e14b5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['collection_date',\n",
       " 'growth_facil',\n",
       " 'other',\n",
       " 'samp_collect_device',\n",
       " 'samp_mat_process',\n",
       " 'samp_name',\n",
       " 'samp_size',\n",
       " 'samp_store_temp',\n",
       " 'source_mat_id',\n",
       " 'store_cond']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_package_slots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a7a2adf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'collection_date;growth_facil;other;samp_collect_device;samp_mat_process;samp_name;samp_size;samp_store_temp;source_mat_id;store_cond'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_package_slots_string = \";\".join(all_package_slots)\n",
    "all_package_slots_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "215abecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "selective_package_slots = iot_package_slots.loc[\n",
    "    iot_package_slots[\"Associated Packages\"].ne(\"all\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "870bb70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = (\n",
    "    selective_package_slots[\"Associated Packages\"]\n",
    "    .str.split(\" *; *\", expand=False)\n",
    "    .copy()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "82257d69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c8dd654e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mark/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/pandas/core/indexing.py:1667: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = value\n"
     ]
    }
   ],
   "source": [
    "selective_package_slots.loc[:, \"splitted\"] = temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "0dcf9b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "selective_package_slots = selective_package_slots[[\"mixs_6_slot_name\", \"splitted\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2437dc77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mixs_6_slot_name</th>\n",
       "      <th>splitted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>geo_loc_name</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>geo_loc_name</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>geo_loc_name</td>\n",
       "      <td>pore_water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>geo_loc_name</td>\n",
       "      <td>plant_associated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>geo_loc_name</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>water_cont_soil_meth</td>\n",
       "      <td>soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>water_cont_soil_meth</td>\n",
       "      <td>sediment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>water_current</td>\n",
       "      <td>pore_water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>water_current</td>\n",
       "      <td>water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>water_temp_regm</td>\n",
       "      <td>plant_associated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>596 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         mixs_6_slot_name          splitted\n",
       "14           geo_loc_name              soil\n",
       "14           geo_loc_name          sediment\n",
       "14           geo_loc_name        pore_water\n",
       "14           geo_loc_name  plant_associated\n",
       "14           geo_loc_name             water\n",
       "..                    ...               ...\n",
       "246  water_cont_soil_meth              soil\n",
       "246  water_cont_soil_meth          sediment\n",
       "247         water_current        pore_water\n",
       "247         water_current             water\n",
       "248       water_temp_regm  plant_associated\n",
       "\n",
       "[596 rows x 2 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selective_package_slots = selective_package_slots.explode(\"splitted\")\n",
    "selective_package_slots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "4736d62a",
   "metadata": {},
   "outputs": [],
   "source": [
    "packagewise = (\n",
    "    selective_package_slots.astype(str)\n",
    "    .groupby(\"splitted\", as_index=False)\n",
    "    .agg(\";\".join)[selective_package_slots.columns]\n",
    ")\n",
    "packagewise = packagewise[[\"splitted\", \"mixs_6_slot_name\"]]\n",
    "packagewise.columns = [\"package\", \"mixs_6_slot_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "cd441667",
   "metadata": {},
   "outputs": [],
   "source": [
    "packagewise[\"mixs_6_slot_name\"] = (\n",
    "    packagewise[\"mixs_6_slot_name\"].astype(str) + \";\" + all_package_slots_string\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "8d5b0bd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>mixs_6_slot_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>culture_environmental</td>\n",
       "      <td>geo_loc_name;lat_lon;elev;chem_administration;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mixed_culture</td>\n",
       "      <td>chem_administration;watering_regm;air_temp_reg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>geo_loc_name;lat_lon;elev;chem_administration;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pore_water</td>\n",
       "      <td>geo_loc_name;lat_lon;elev;depth;chem_administr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pure_culture</td>\n",
       "      <td>chem_administration;watering_regm;air_temp_reg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sediment</td>\n",
       "      <td>geo_loc_name;lat_lon;elev;depth;chem_administr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>soil</td>\n",
       "      <td>geo_loc_name;lat_lon;elev;depth;chem_administr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>water</td>\n",
       "      <td>geo_loc_name;lat_lon;elev;depth;chem_administr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>water_extract_biosolid</td>\n",
       "      <td>geo_loc_name;lat_lon;elev;chem_administration;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>water_extract_soil</td>\n",
       "      <td>geo_loc_name;lat_lon;elev;depth;chem_administr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  package                                   mixs_6_slot_name\n",
       "0   culture_environmental  geo_loc_name;lat_lon;elev;chem_administration;...\n",
       "1           mixed_culture  chem_administration;watering_regm;air_temp_reg...\n",
       "2        plant_associated  geo_loc_name;lat_lon;elev;chem_administration;...\n",
       "3              pore_water  geo_loc_name;lat_lon;elev;depth;chem_administr...\n",
       "4            pure_culture  chem_administration;watering_regm;air_temp_reg...\n",
       "5                sediment  geo_loc_name;lat_lon;elev;depth;chem_administr...\n",
       "6                    soil  geo_loc_name;lat_lon;elev;depth;chem_administr...\n",
       "7                   water  geo_loc_name;lat_lon;elev;depth;chem_administr...\n",
       "8  water_extract_biosolid  geo_loc_name;lat_lon;elev;chem_administration;...\n",
       "9      water_extract_soil  geo_loc_name;lat_lon;elev;depth;chem_administr..."
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "packagewise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "acc91bab",
   "metadata": {},
   "outputs": [],
   "source": [
    "iot_packages = list(packagewise[\"package\"])\n",
    "iot_packages.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f7b471ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['culture_environmental',\n",
       " 'mixed_culture',\n",
       " 'plant_associated',\n",
       " 'pore_water',\n",
       " 'pure_culture',\n",
       " 'sediment',\n",
       " 'soil',\n",
       " 'water',\n",
       " 'water_extract_biosolid',\n",
       " 'water_extract_soil']"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iot_packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "22e35a91",
   "metadata": {},
   "outputs": [],
   "source": [
    "mixs_all_classes = mixs_yaml_view.all_classes()\n",
    "mixs_all_classes = list(mixs_all_classes.keys())\n",
    "mixs_all_classes.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "d0fe9367",
   "metadata": {},
   "outputs": [],
   "source": [
    "mixs_root_classes = mixs_yaml_view.class_roots()\n",
    "mixs_root_classes.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c282738a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MIGS bacteria',\n",
       " 'MIGS eukaryote',\n",
       " 'MIGS org',\n",
       " 'MIGS plant',\n",
       " 'MIGS virus',\n",
       " 'MIMAG',\n",
       " 'MIMARKS specimen',\n",
       " 'MIMARKS survey',\n",
       " 'MIMS',\n",
       " 'MISAG',\n",
       " 'MIUVIG']"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mixers = []\n",
    "for i in mixs_all_classes:\n",
    "    #     print(i)\n",
    "    checkee = mixs_yaml_view.get_class(i)\n",
    "    mixers = mixers + checkee.mixins\n",
    "\n",
    "mixers = list(set(mixers))\n",
    "mixers.sort()\n",
    "mixers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2dbb1cae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['agriculture',\n",
       " 'air',\n",
       " 'built environment',\n",
       " 'core',\n",
       " 'food-animal and animal feed',\n",
       " 'food-farm environment',\n",
       " 'food-food production facility',\n",
       " 'food-human foods',\n",
       " 'host-associated',\n",
       " 'human-associated',\n",
       " 'human-gut',\n",
       " 'human-oral',\n",
       " 'human-skin',\n",
       " 'human-vaginal',\n",
       " 'hydrocarbon resources-cores',\n",
       " 'hydrocarbon resources-fluids_swabs',\n",
       " 'microbial mat_biofilm',\n",
       " 'miscellaneous natural or artificial environment',\n",
       " 'plant-associated',\n",
       " 'quantity value',\n",
       " 'sediment',\n",
       " 'soil',\n",
       " 'symbiont-associated',\n",
       " 'wastewater_sludge',\n",
       " 'water']"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mixs_unmixed_roots = list(set(mixs_root_classes) - set(mixers))\n",
    "mixs_unmixed_roots.sort()\n",
    "mixs_unmixed_roots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00f7e5ea",
   "metadata": {},
   "source": [
    "could exclude core package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5b3ae1a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['culture_environmental',\n",
       " 'mixed_culture',\n",
       " 'plant_associated',\n",
       " 'pore_water',\n",
       " 'pure_culture',\n",
       " 'water_extract_biosolid',\n",
       " 'water_extract_soil']"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iot_only_packages = list(set(iot_packages) - set(mixs_unmixed_roots))\n",
    "iot_only_packages.sort()\n",
    "iot_only_packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "92863e5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "shared_packages = list(set(iot_packages).intersection(set(mixs_unmixed_roots)))\n",
    "shared_packages.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "3d74b9c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sediment', 'soil', 'water']"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared_packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "3fe9f3c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sediment\n",
      "soil\n",
      "water\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'sediment': {'IoT only': ['air_temp_regm',\n",
       "   'alkalinity_method',\n",
       "   'biotic_regm',\n",
       "   'biotic_relationship',\n",
       "   'climate_environment',\n",
       "   'extreme_event',\n",
       "   'fire',\n",
       "   'flooding',\n",
       "   'gaseous_environment',\n",
       "   'growth_facil',\n",
       "   'humidity_regm',\n",
       "   'light_regm',\n",
       "   'microbial_biomass',\n",
       "   'microbial_biomass_meth',\n",
       "   'other',\n",
       "   'ph_meth',\n",
       "   'samp_collect_device',\n",
       "   'samp_mat_process',\n",
       "   'samp_name',\n",
       "   'samp_size',\n",
       "   'sieving',\n",
       "   'source_mat_id',\n",
       "   'store_cond',\n",
       "   'tot_n_meth',\n",
       "   'tot_org_c_meth',\n",
       "   'water_cont_soil_meth',\n",
       "   'watering_regm'],\n",
       "  'MIxS 6 only': ['env_broad_scale',\n",
       "   'env_local_scale',\n",
       "   'env_medium',\n",
       "   'samp_store_dur',\n",
       "   'samp_store_loc']},\n",
       " 'soil': {'IoT only': ['air_temp_regm',\n",
       "   'biotic_regm',\n",
       "   'biotic_relationship',\n",
       "   'chem_administration',\n",
       "   'climate_environment',\n",
       "   'gaseous_environment',\n",
       "   'growth_facil',\n",
       "   'horizon',\n",
       "   'humidity_regm',\n",
       "   'light_regm',\n",
       "   'microbial_biomass_meth',\n",
       "   'other',\n",
       "   'oxy_stat_samp',\n",
       "   'previous_land_use_meth',\n",
       "   'samp_collect_device',\n",
       "   'samp_mat_process',\n",
       "   'samp_name',\n",
       "   'samp_size',\n",
       "   'samp_store_temp',\n",
       "   'source_mat_id',\n",
       "   'texture',\n",
       "   'texture_meth',\n",
       "   'tot_n_meth',\n",
       "   'watering_regm'],\n",
       "  'MIxS 6 only': ['env_broad_scale',\n",
       "   'env_local_scale',\n",
       "   'env_medium',\n",
       "   'micro_biomass_meth',\n",
       "   'pool_dna_extracts',\n",
       "   'prev_land_use_meth',\n",
       "   'soil_horizon',\n",
       "   'soil_text_measure',\n",
       "   'soil_texture_meth',\n",
       "   'temp',\n",
       "   'tot_nitro_cont_meth']},\n",
       " 'water': {'IoT only': ['air_temp_regm',\n",
       "   'biotic_regm',\n",
       "   'climate_environment',\n",
       "   'gaseous_environment',\n",
       "   'growth_facil',\n",
       "   'humidity_regm',\n",
       "   'light_regm',\n",
       "   'other',\n",
       "   'ph_meth',\n",
       "   'samp_collect_device',\n",
       "   'samp_mat_process',\n",
       "   'samp_name',\n",
       "   'samp_size',\n",
       "   'size_frac',\n",
       "   'source_mat_id',\n",
       "   'store_cond',\n",
       "   'watering_regm'],\n",
       "  'MIxS 6 only': ['env_broad_scale',\n",
       "   'env_local_scale',\n",
       "   'env_medium',\n",
       "   'samp_store_dur',\n",
       "   'samp_store_loc']}}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slot_usage_differences = {}\n",
    "for i in shared_packages:\n",
    "    print(i)\n",
    "    # assume one matching row\n",
    "    ii = packagewise[\"mixs_6_slot_name\"].loc[packagewise[\"package\"].eq(i)]\n",
    "    ii = ii.iloc[0]\n",
    "    ii = re.split(\" *; *\", ii)\n",
    "    ii.sort()\n",
    "    im = mixs_yaml_view.class_induced_slots(i)\n",
    "    im = [x.name for x in im]\n",
    "    im.sort()\n",
    "    i_only = list(set(ii) - set(im))\n",
    "    i_only.sort()\n",
    "    m_only = list(set(im) - set(ii))\n",
    "    m_only.sort()\n",
    "    slot_usage_differences[i] = {\"IoT only\": i_only, \"MIxS 6 only\": m_only}\n",
    "\n",
    "slot_usage_differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2bc3e750",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    247\n",
       "Name: Expected Value, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iot_glossary[\"Expected Value\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "bdcd1c57",
   "metadata": {},
   "outputs": [],
   "source": [
    "syntax_frame = (\n",
    "    iot_glossary[\"syntax\"]\n",
    "    .value_counts()\n",
    "    .rename_axis(\"unique_values\")\n",
    "    .reset_index(name=\"counts\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "576978be",
   "metadata": {},
   "outputs": [],
   "source": [
    "syntax_frame.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "918b47d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "category_frame = (\n",
    "    iot_glossary[\"Category\"]\n",
    "    .value_counts()\n",
    "    .rename_axis(\"unique_values\")\n",
    "    .reset_index(name=\"counts\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "14cc28a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_values</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>required where applicable</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>required</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample identification</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               unique_values  counts\n",
       "0                                210\n",
       "1  required where applicable      26\n",
       "2                   required       6\n",
       "3      sample identification       5"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "category_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "a0f68bdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "origin_frame = (\n",
    "    iot_glossary[\"Origin\"]\n",
    "    .value_counts()\n",
    "    .rename_axis(\"unique_values\")\n",
    "    .reset_index(name=\"counts\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "cef6cd56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_values</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIxS</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  unique_values  counts\n",
       "0          MIxS     242\n",
       "1                     5"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "origin_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3797ccce",
   "metadata": {},
   "source": [
    "## Start building DataHarmonizer templates\n",
    "\n",
    "### take all packages and slots, but repair typos in slot names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "eb913211",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_to_pack_4_dh = iot_glossary[[\"name\", \"mixs_6_slot_name\", \"Associated Packages\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "2ebb273d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mark/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/pandas/core/indexing.py:1667: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = value\n"
     ]
    }
   ],
   "source": [
    "temp = slot_to_pack_4_dh[\"Associated Packages\"].str.split(\" *; *\", expand=False).copy()\n",
    "slot_to_pack_4_dh.loc[:, \"ap_list\"] = temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "8989a1c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>mixs_6_slot_name</th>\n",
       "      <th>Associated Packages</th>\n",
       "      <th>ap_list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>unique_ID</td>\n",
       "      <td></td>\n",
       "      <td>all</td>\n",
       "      <td>[all]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>samp_name</td>\n",
       "      <td>samp_name</td>\n",
       "      <td>all</td>\n",
       "      <td>[all]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>investigation_type</td>\n",
       "      <td></td>\n",
       "      <td>all</td>\n",
       "      <td>[all]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>package</td>\n",
       "      <td></td>\n",
       "      <td>all</td>\n",
       "      <td>[all]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>source_mat_ID</td>\n",
       "      <td>source_mat_id</td>\n",
       "      <td>all</td>\n",
       "      <td>[all]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>wastewater_type</td>\n",
       "      <td>wastewater_type</td>\n",
       "      <td>water_extract_biosolid</td>\n",
       "      <td>[water_extract_biosolid]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>water_content</td>\n",
       "      <td>water_content</td>\n",
       "      <td>soil; sediment; water_extract_soil</td>\n",
       "      <td>[soil, sediment, water_extract_soil]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>water_content_meth</td>\n",
       "      <td>water_cont_soil_meth</td>\n",
       "      <td>soil; sediment</td>\n",
       "      <td>[soil, sediment]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>water_current</td>\n",
       "      <td>water_current</td>\n",
       "      <td>pore_water; water</td>\n",
       "      <td>[pore_water, water]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>water_temp_regm</td>\n",
       "      <td>water_temp_regm</td>\n",
       "      <td>plant_associated</td>\n",
       "      <td>[plant_associated]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>248 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   name      mixs_6_slot_name  \\\n",
       "1             unique_ID                         \n",
       "2             samp_name             samp_name   \n",
       "3    investigation_type                         \n",
       "4               package                         \n",
       "5         source_mat_ID         source_mat_id   \n",
       "..                  ...                   ...   \n",
       "244     wastewater_type       wastewater_type   \n",
       "245       water_content         water_content   \n",
       "246  water_content_meth  water_cont_soil_meth   \n",
       "247       water_current         water_current   \n",
       "248     water_temp_regm       water_temp_regm   \n",
       "\n",
       "                    Associated Packages                               ap_list  \n",
       "1                                   all                                 [all]  \n",
       "2                                   all                                 [all]  \n",
       "3                                   all                                 [all]  \n",
       "4                                   all                                 [all]  \n",
       "5                                   all                                 [all]  \n",
       "..                                  ...                                   ...  \n",
       "244              water_extract_biosolid              [water_extract_biosolid]  \n",
       "245  soil; sediment; water_extract_soil  [soil, sediment, water_extract_soil]  \n",
       "246                      soil; sediment                      [soil, sediment]  \n",
       "247                   pore_water; water                   [pore_water, water]  \n",
       "248                    plant_associated                    [plant_associated]  \n",
       "\n",
       "[248 rows x 4 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slot_to_pack_4_dh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e8aa8b91",
   "metadata": {},
   "outputs": [],
   "source": [
    "iot_packages = list(slot_to_pack_4_dh[\"ap_list\"])\n",
    "iot_packages = [i for i in iot_packages if i]\n",
    "iot_packages = sum(iot_packages, [])\n",
    "iot_packages = list(set(iot_packages))\n",
    "iot_packages.sort()\n",
    "iot_packages.remove(\"\")\n",
    "iot_packages.remove(\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "1004280c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_372797/4018325788.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  slot_to_pack_4_dh[\"repaired_name\"] = slot_to_pack_4_dh[\"mixs_6_slot_name\"]\n",
      "/home/mark/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_block(indexer, value, name)\n"
     ]
    }
   ],
   "source": [
    "slot_to_pack_4_dh[\"repaired_name\"] = slot_to_pack_4_dh[\"mixs_6_slot_name\"]\n",
    "\n",
    "slot_to_pack_4_dh[\"repaired_name\"].loc[\n",
    "    slot_to_pack_4_dh[\"repaired_name\"] == \"\"\n",
    "] = slot_to_pack_4_dh[\"name\"].loc[slot_to_pack_4_dh[\"repaired_name\"] == \"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "e7689e1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# slot_to_pack_4_dh[\"ap_list\"].loc[\n",
    "#     slot_to_pack_4_dh[\"Associated Packages\"] == \"all\"\n",
    "# ] = iot_packages\n",
    "\n",
    "slot_to_pack_4_dh[\"ap_list\"].loc[\n",
    "    slot_to_pack_4_dh[\"Associated Packages\"] == \"all\"\n",
    "] = slot_to_pack_4_dh[\"Associated Packages\"].apply(lambda _: iot_packages)\n",
    "\n",
    "# slot_to_pack_4_dh[\"Associated Packages\"].apply(lambda _: iot_packages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "84119fa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_to_pack_4_dh = slot_to_pack_4_dh[[\"repaired_name\", \"ap_list\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "5f770b1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_to_pack_4_dh = slot_to_pack_4_dh.explode(\"ap_list\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "70714c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_to_pack_4_dh = (\n",
    "    slot_to_pack_4_dh.astype(str)\n",
    "    .groupby(\"ap_list\", as_index=False)\n",
    "    .agg(\";\".join)[slot_to_pack_4_dh.columns]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "3fec31ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_to_pack_4_dh = slot_to_pack_4_dh.loc[slot_to_pack_4_dh[\"ap_list\"].ne(\"\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "bc62b7b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_to_pack_4_dh = slot_to_pack_4_dh.loc[slot_to_pack_4_dh[\"ap_list\"].ne(\"None\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "543fba73",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_to_pack_4_dh = slot_to_pack_4_dh.loc[~slot_to_pack_4_dh[\"ap_list\"].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "121dff3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_to_pack_4_dh = slot_to_pack_4_dh[[\"ap_list\", \"repaired_name\"]]\n",
    "slot_to_pack_4_dh.columns = [\"package\", \"slots\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "aa248f33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>package</th>\n",
       "      <th>slots</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>culture_environmental</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mixed_culture</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pore_water</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pure_culture</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sediment</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>soil</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>water</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>water_extract_biosolid</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>water_extract_soil</td>\n",
       "      <td>unique_ID;samp_name;investigation_type;package...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   package                                              slots\n",
       "2    culture_environmental  unique_ID;samp_name;investigation_type;package...\n",
       "3            mixed_culture  unique_ID;samp_name;investigation_type;package...\n",
       "4         plant_associated  unique_ID;samp_name;investigation_type;package...\n",
       "5               pore_water  unique_ID;samp_name;investigation_type;package...\n",
       "6             pure_culture  unique_ID;samp_name;investigation_type;package...\n",
       "7                 sediment  unique_ID;samp_name;investigation_type;package...\n",
       "8                     soil  unique_ID;samp_name;investigation_type;package...\n",
       "9                    water  unique_ID;samp_name;investigation_type;package...\n",
       "10  water_extract_biosolid  unique_ID;samp_name;investigation_type;package...\n",
       "11      water_extract_soil  unique_ID;samp_name;investigation_type;package..."
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slot_to_pack_4_dh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3f8e2865",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_details_4_dh = iot_glossary.copy()\n",
    "slot_details_4_dh.drop(\n",
    "    labels=[\"Associated Packages\", \"Expected Value\"], axis=1, inplace=True\n",
    ")\n",
    "\n",
    "slot_details_4_dh[\"repaired_name\"] = slot_details_4_dh[\"mixs_6_slot_name\"]\n",
    "\n",
    "slot_details_4_dh[\"repaired_name\"].loc[\n",
    "    slot_details_4_dh[\"repaired_name\"] == \"\"\n",
    "] = slot_details_4_dh[\"name\"].loc[slot_details_4_dh[\"repaired_name\"] == \"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "17b42fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# slot_details_4_dh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "f9a526df",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_details_4_dh[\"Category\"].loc[slot_details_4_dh[\"Category\"] == \"\"] = \"optional\"\n",
    "slot_details_4_dh[\"Category\"].loc[slot_details_4_dh[\"Category\"].isnull()] = \"optional\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ec61c7c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column Header</th>\n",
       "      <th>name</th>\n",
       "      <th>mixs_6_slot_name</th>\n",
       "      <th>different?</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Guidance</th>\n",
       "      <th>syntax</th>\n",
       "      <th>Category</th>\n",
       "      <th>Origin</th>\n",
       "      <th>Notes</th>\n",
       "      <th>GitHub Ticket</th>\n",
       "      <th>repaired_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Unique ID</td>\n",
       "      <td>unique_ID</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td>A globally unique identifier assigned to the b...</td>\n",
       "      <td>Field REQUIRED for ALL sample submission. Opti...</td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td></td>\n",
       "      <td>v6 has unique ID and sample name synonymous. B...</td>\n",
       "      <td>None</td>\n",
       "      <td>unique_ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sample Name</td>\n",
       "      <td>samp_name</td>\n",
       "      <td>samp_name</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>A local identifier or name that for the materi...</td>\n",
       "      <td>ID that is present on the shipped sample. Huma...</td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>v6</td>\n",
       "      <td>None</td>\n",
       "      <td>samp_name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Analysis/Data Type</td>\n",
       "      <td>investigation_type</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td>This field is constrained to contain only a se...</td>\n",
       "      <td>drop down selection list</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>enumeration values are EMSL origin, MIxS has d...</td>\n",
       "      <td>None</td>\n",
       "      <td>investigation_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sample Type</td>\n",
       "      <td>package</td>\n",
       "      <td></td>\n",
       "      <td>FALSE</td>\n",
       "      <td></td>\n",
       "      <td>MIxS Package</td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>This is assumed by the package, but I think sh...</td>\n",
       "      <td>None</td>\n",
       "      <td>package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Source Material ID</td>\n",
       "      <td>source_mat_ID</td>\n",
       "      <td>source_mat_id</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>A unique identifier assigned to a material sam...</td>\n",
       "      <td>A unique identifier assigned to an original ma...</td>\n",
       "      <td>{text}</td>\n",
       "      <td>sample identification</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>New v6 definition is better to indicate bio sa...</td>\n",
       "      <td>None</td>\n",
       "      <td>source_mat_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>wastewater type</td>\n",
       "      <td>wastewater_type</td>\n",
       "      <td>wastewater_type</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>the origin of wastewater such as human waste, ...</td>\n",
       "      <td></td>\n",
       "      <td>{text}</td>\n",
       "      <td>optional</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>wastewater_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>water content</td>\n",
       "      <td>water_content</td>\n",
       "      <td>water_content</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>Water content measurement</td>\n",
       "      <td></td>\n",
       "      <td>{float}</td>\n",
       "      <td>optional</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>water_content</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>water content method</td>\n",
       "      <td>water_content_meth</td>\n",
       "      <td>water_cont_soil_meth</td>\n",
       "      <td>FALSE</td>\n",
       "      <td>Reference or method used in determining the wa...</td>\n",
       "      <td></td>\n",
       "      <td>{PMID}|{DOI}|{URL}</td>\n",
       "      <td>optional</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>water_cont_soil_meth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>water current</td>\n",
       "      <td>water_current</td>\n",
       "      <td>water_current</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>Measurement of magnitude and direction of flow...</td>\n",
       "      <td></td>\n",
       "      <td>{float} {unit}</td>\n",
       "      <td>optional</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>water_current</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>water temperature regimen</td>\n",
       "      <td>water_temp_regm</td>\n",
       "      <td>water_temp_regm</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>Information about treatment involving an expos...</td>\n",
       "      <td></td>\n",
       "      <td>measurement value {value} {unit}; treatment in...</td>\n",
       "      <td>optional</td>\n",
       "      <td>MIxS</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>water_temp_regm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>248 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Column Header                name      mixs_6_slot_name  \\\n",
       "1                    Unique ID           unique_ID                         \n",
       "2                  Sample Name           samp_name             samp_name   \n",
       "3           Analysis/Data Type  investigation_type                         \n",
       "4                  Sample Type             package                         \n",
       "5           Source Material ID       source_mat_ID         source_mat_id   \n",
       "..                         ...                 ...                   ...   \n",
       "244            wastewater type     wastewater_type       wastewater_type   \n",
       "245              water content       water_content         water_content   \n",
       "246       water content method  water_content_meth  water_cont_soil_meth   \n",
       "247              water current       water_current         water_current   \n",
       "248  water temperature regimen     water_temp_regm       water_temp_regm   \n",
       "\n",
       "    different?                                         Definition  \\\n",
       "1        FALSE  A globally unique identifier assigned to the b...   \n",
       "2         TRUE  A local identifier or name that for the materi...   \n",
       "3        FALSE                                                      \n",
       "4        FALSE                                                      \n",
       "5         TRUE  A unique identifier assigned to a material sam...   \n",
       "..         ...                                                ...   \n",
       "244       TRUE  the origin of wastewater such as human waste, ...   \n",
       "245       TRUE                          Water content measurement   \n",
       "246      FALSE  Reference or method used in determining the wa...   \n",
       "247       TRUE  Measurement of magnitude and direction of flow...   \n",
       "248       TRUE  Information about treatment involving an expos...   \n",
       "\n",
       "                                              Guidance  \\\n",
       "1    Field REQUIRED for ALL sample submission. Opti...   \n",
       "2    ID that is present on the shipped sample. Huma...   \n",
       "3    This field is constrained to contain only a se...   \n",
       "4                                         MIxS Package   \n",
       "5    A unique identifier assigned to an original ma...   \n",
       "..                                                 ...   \n",
       "244                                                      \n",
       "245                                                      \n",
       "246                                                      \n",
       "247                                                      \n",
       "248                                                      \n",
       "\n",
       "                                                syntax               Category  \\\n",
       "1                                               {text}  sample identification   \n",
       "2                                               {text}  sample identification   \n",
       "3                             drop down selection list  sample identification   \n",
       "4                                               {text}  sample identification   \n",
       "5                                               {text}  sample identification   \n",
       "..                                                 ...                    ...   \n",
       "244                                             {text}               optional   \n",
       "245                                            {float}               optional   \n",
       "246                                 {PMID}|{DOI}|{URL}               optional   \n",
       "247                                     {float} {unit}               optional   \n",
       "248  measurement value {value} {unit}; treatment in...               optional   \n",
       "\n",
       "    Origin                                              Notes GitHub Ticket  \\\n",
       "1           v6 has unique ID and sample name synonymous. B...          None   \n",
       "2     MIxS                                                 v6          None   \n",
       "3     MIxS  enumeration values are EMSL origin, MIxS has d...          None   \n",
       "4     MIxS  This is assumed by the package, but I think sh...          None   \n",
       "5     MIxS  New v6 definition is better to indicate bio sa...          None   \n",
       "..     ...                                                ...           ...   \n",
       "244   MIxS                                               None          None   \n",
       "245   MIxS                                               None          None   \n",
       "246   MIxS                                               None          None   \n",
       "247   MIxS                                               None          None   \n",
       "248   MIxS                                               None          None   \n",
       "\n",
       "            repaired_name  \n",
       "1               unique_ID  \n",
       "2               samp_name  \n",
       "3      investigation_type  \n",
       "4                 package  \n",
       "5           source_mat_id  \n",
       "..                    ...  \n",
       "244       wastewater_type  \n",
       "245         water_content  \n",
       "246  water_cont_soil_meth  \n",
       "247         water_current  \n",
       "248       water_temp_regm  \n",
       "\n",
       "[248 rows x 12 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slot_details_4_dh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "85df0dee",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_categories = list(set(list(slot_details_4_dh[\"Category\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "47c5d861",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sample identification', 'required', 'required where applicable', 'optional']"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slot_categories"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d575bae8",
   "metadata": {},
   "source": [
    "# Reconcile these\n",
    "\n",
    "## Index of Terms\n",
    "\n",
    "- Column Header\n",
    "- name -> repaired_name\n",
    "- Definition\n",
    "- Guidance\n",
    "- Expected Value\n",
    "- syntax\n",
    "- Category\n",
    "- Associated Packages\n",
    "- Origin\n",
    "- Notes\n",
    "- GitHub Ticket\n",
    "\n",
    "\n",
    "## DataHarmonizer\n",
    "- Ontology ID\n",
    "- parent class (section)\n",
    "- label\n",
    "- datatype\n",
    "- ~source~\n",
    "- data status (semicolon separated list of options like \"not collected\", \"missing\", not applicable\")\n",
    "- requirement (\"\", \"required\" or \"recommended\")\n",
    "- min value\n",
    "- max value\n",
    "- capitalize\n",
    "- pattern\n",
    "- description\n",
    "- guidance\n",
    "- examples\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b3973d8",
   "metadata": {},
   "source": [
    "## skip unless creating DH templates\n",
    "i.e. skip for creating YAML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e2454fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def template_package(\n",
    "    current_package,\n",
    "    slot_to_package_df,\n",
    "    slot_details_df,\n",
    "    enums_dict,\n",
    "    template_prefix,\n",
    "    template_suffix,\n",
    "):\n",
    "    print(current_package)\n",
    "\n",
    "    main_row_list = []\n",
    "    enum_row_list = []\n",
    "\n",
    "    for i in slot_categories:\n",
    "        print(i)\n",
    "        current_row = blank_row.copy()\n",
    "        current_row[\"label\"] = i\n",
    "        main_row_list.append(current_row)\n",
    "\n",
    "    package_slots = slot_to_pack_4_dh[\"slots\"].loc[\n",
    "        slot_to_pack_4_dh[\"package\"] == current_package\n",
    "    ]\n",
    "    package_slots = package_slots.iloc[0]\n",
    "    package_slots = package_slots.split(\";\")\n",
    "\n",
    "    for i in package_slots:\n",
    "        print(i)\n",
    "        current_details = slot_details_4_dh.loc[slot_details_4_dh[\"repaired_name\"] == i]\n",
    "        current_row = blank_row.copy()\n",
    "        #     \"Ontology ID\"\n",
    "        current_row[\"parent class\"] = current_details[\"Category\"].iloc[0]\n",
    "        current_row[\"label\"] = i\n",
    "        current_row[\"datatype\"] = \"xs:token\"\n",
    "        # day resolution may not be specific enough\n",
    "        if current_details[\"syntax\"].iloc[0] == \"{timestamp}\":\n",
    "            current_row[\"datatype\"] = \"xs:date\"\n",
    "        if current_details[\"syntax\"].iloc[0] == \"{float}\":\n",
    "            current_row[\"datatype\"] = \"xs:decimal\"\n",
    "        if current_details[\"syntax\"].iloc[0] == \"{value}\":\n",
    "            current_row[\"datatype\"] = \"xs:decimal\"\n",
    "        # {integer} doesn't actually = xs:nonNegativeInteger\n",
    "        if current_details[\"syntax\"].iloc[0] == \"{integer}\":\n",
    "            current_row[\"datatype\"] = \"xs:nonNegativeInteger\"\n",
    "        if i == \"unique_ID\":\n",
    "            current_row[\"datatype\"] = \"xs:unique\"\n",
    "        if current_details[\"syntax\"].iloc[0] == \"{float} {unit}\":\n",
    "            current_row[\"pattern\"] = \"^[+-]?([0-9]*[.])?[0-9]+ \\S+$\"\n",
    "        #     \"source\": \"\",\n",
    "        #     \"data status\": \"\",\n",
    "        if current_details[\"Category\"].iloc[0] in required_sections:\n",
    "            current_row[\"requirement\"] = \"required\"\n",
    "        #     \"min value\": \"\",\n",
    "        #     \"max value\": \"\",\n",
    "        #     \"capitalize\": \"\",\n",
    "        current_row[\"description\"] = current_details[\"Column Header\"].iloc[0]\n",
    "        current_row[\"guidance\"] = current_details[\"Guidance\"].iloc[0]\n",
    "        current_row[\"examples\"] = current_details[\"syntax\"].iloc[0]\n",
    "        if i in ct_keys:\n",
    "            current_row[\"datatype\"] = \"select\"\n",
    "            # map?\n",
    "            # indent?\n",
    "            current_enums = ct_dol[i]\n",
    "            current_enums.sort()\n",
    "            for j in current_enums:\n",
    "                print(\"    \" + j)\n",
    "                current_enum_row = blank_row.copy()\n",
    "                current_enum_row[\"label\"] = j\n",
    "                current_enum_row[\"parent class\"] = i\n",
    "                enum_row_list.append(current_enum_row)\n",
    "        main_row_list.append(current_row)\n",
    "    print(\"\\n\")\n",
    "\n",
    "    current_frame = pd.DataFrame(main_row_list)\n",
    "    enum_frame = pd.DataFrame(enum_row_list)\n",
    "    assembled_frame = pd.concat([current_frame, enum_frame])\n",
    "\n",
    "    # create directory if necessary\n",
    "    required_directory = dh_template_prefix + current_package\n",
    "    # print(required_directory)\n",
    "    os.makedirs(required_directory, exist_ok=True)\n",
    "    # copy from soil directory to new required_directory?\n",
    "    ref_temp_src = dh_template_root + ref_temp_filename\n",
    "    ref_temp_dest = required_directory + \"/\" + ref_temp_filename\n",
    "    copyfile(ref_temp_src, ref_temp_dest)\n",
    "\n",
    "    ref_temp_src = dh_template_root + \"export.js\"\n",
    "    ref_temp_dest = required_directory + \"/\" + \"export.js\"\n",
    "    copyfile(ref_temp_src, ref_temp_dest)\n",
    "\n",
    "    # should escape characters that break filenames like whitespaces\n",
    "    current_template_file = required_directory + dh_template_suffix\n",
    "    assembled_frame.to_csv(current_template_file, index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94549a14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# template_package(\n",
    "#     \"soil\",\n",
    "#     slot_to_package_df=slot_to_pack_4_dh,\n",
    "#     slot_details_df=slot_details_4_dh,\n",
    "#     enums_dict=ct_dol,\n",
    "#     template_prefix=dh_template_prefix,\n",
    "#     template_suffix=dh_template_suffix,\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d2445f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for current_package in iot_packages:\n",
    "    #     print(current_package)\n",
    "    template_package(\n",
    "        current_package,\n",
    "        slot_to_package_df=slot_to_pack_4_dh,\n",
    "        slot_details_df=slot_details_4_dh,\n",
    "        enums_dict=ct_dol,\n",
    "        template_prefix=dh_template_prefix,\n",
    "        template_suffix=dh_template_suffix,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "596db81f",
   "metadata": {},
   "source": [
    "Add to `TEMPLATES` constant at top of `DataHarmonizer/main/script.js`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd82d0de",
   "metadata": {},
   "outputs": [],
   "source": [
    "for current_package in iot_packages:\n",
    "    print(\n",
    "        \"'Index of Terms \"\n",
    "        + current_package\n",
    "        + \"':    {'folder': 'IoT_\"\n",
    "        + current_package\n",
    "        + \"', 'status': 'published'},\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecf6e1b6",
   "metadata": {},
   "source": [
    "- Don't forget to run `DataHarmonizer/template/make_all.sh`\n",
    "- note that the refernce templates (`reference_template.html`) are very generic\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf94ab6a",
   "metadata": {},
   "source": [
    "## Pick back up here to make LinkiML model of IoT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "7d4b91ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "slot_list = list(slot_details_4_dh[\"repaired_name\"])\n",
    "\n",
    "class_list = list(slot_to_pack_4_dh[\"package\"])\n",
    "\n",
    "IoT_yaml_file_name = \"IoT.yaml\"\n",
    "\n",
    "minimal_model_file = \"minimal.yaml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "59de7510",
   "metadata": {},
   "outputs": [],
   "source": [
    "minimal_model = SchemaView(minimal_model_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "de2d07ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in slot_list:\n",
    "    #     print(i)\n",
    "    minimal_model.add_slot(SlotDefinition(name=i))\n",
    "\n",
    "for i in class_list:\n",
    "    #     print(i)\n",
    "    minimal_model.add_class(ClassDefinition(name=i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "e708c985",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ClassDefinition(name='culture_environmental', id_prefixes=[], definition_uri=None, aliases=[], local_names={}, conforms_to=None, mappings=[], exact_mappings=[], close_mappings=[], related_mappings=[], narrow_mappings=[], broad_mappings=[], extensions={}, annotations={}, description=None, alt_descriptions={}, title=None, deprecated=None, todos=[], notes=[], comments=[], examples=[], in_subset=[], from_schema='https://w3id.org/minimal', imported_from=None, see_also=[], deprecated_element_has_exact_replacement=None, deprecated_element_has_possible_replacement=None, is_a=None, abstract=None, mixin=None, mixins=[], apply_to=[], values_from=[], created_by=None, created_on=None, last_updated_on=None, modified_by=None, status=None, string_serialization=None, slots=[], slot_usage={}, attributes={}, class_uri=None, subclass_of=None, union_of=[], defining_slots=[], tree_root=None, unique_keys={}, rules=[], classification_rules=[], any_of=[], exactly_one_of=[], none_of=[], all_of=[], slot_conditions={})"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# magically avoids bogus(?) \"class culture_environmental has no from_schema\" error\n",
    "minimal_model.get_class(\"culture_environmental\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "c64f6e79",
   "metadata": {},
   "outputs": [],
   "source": [
    "schema_not_view = minimal_model.schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "a2b1355b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type(schema_not_view)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "342779c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "generated_yaml = YAMLGenerator(schema_not_view).serialize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "0a29b4cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generated_yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "528efc64",
   "metadata": {},
   "outputs": [],
   "source": [
    "IoT_yaml_file = open(IoT_yaml_file_name, \"w\")\n",
    "n = IoT_yaml_file.write(generated_yaml)\n",
    "IoT_yaml_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b27eee7d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
