{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a9590e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# use openpyxl for reading xlsx spreadsheet\n",
    "# https://github.com/ontodev/cogs for google sheets? may assume single tab\n",
    "import json\n",
    "from math import isnan\n",
    "import re\n",
    "from collections import Counter\n",
    "import uuid\n",
    "\n",
    "from Bio import Entrez\n",
    "import xmltodict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "06044849",
   "metadata": {},
   "outputs": [],
   "source": [
    "# could use entrez to retrieve addtional project details based on NCBI IDs\n",
    "# initialize some default parameters\n",
    "Entrez.email = \"MAM@lbl.gov\"  # provide your email address\n",
    "\n",
    "# db = 'bioproject'\n",
    "db = \"dbvar\"\n",
    "\n",
    "paramEutils = {\"usehistory\": \"Y\"}  # Use Entrez search history to cache results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "17c974f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list Entrez databases\n",
    "record = Entrez.read(Entrez.einfo())\n",
    "# record"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df713137",
   "metadata": {},
   "source": [
    "## Provenance\n",
    "\n",
    "saved https://docs.google.com/spreadsheets/d/1GZayIFIrY2jdoxRIpk9KDTBLiE71VVtb7YAd5ZSYGR0/edit#gid=0\n",
    "to `input/Example-Soil_NMDC_SampleMetadata.xlsx` on 2021-07-13\n",
    "\n",
    "Also \n",
    "- `Example-Soil_NMDC_SampleMetadata_202107201146.xlsx`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d8d777f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "example_file = \"../input/Example-Soil_NMDC_SampleMetadata_202107201146.xlsx\"\n",
    "mixs_spreadsheet_url = \"https://github.com/GenomicsStandardsConsortium/mixs-legacy/blob/master/mixs5/mixs_v5.xlsx?raw=true\"\n",
    "\n",
    "# could also get MIxS knoweldge from NMDC schema... have been menaing to parse that anyway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a5723557",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Data Validation extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['ReadMe', 'ProjectInformation', 'AssociatedResearchers', 'Metadata', 'EnvironmentalMetadata', 'EcosystemTerms', 'MenuTerms'])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_frame_structure = pd.read_excel(example_file, sheet_name=None)\n",
    "example_sheet_names = example_frame_structure.keys()\n",
    "example_sheet_names"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52b779bd",
   "metadata": {},
   "source": [
    "The \"Metadata\" sheet in the NMDC Sample Metadata template has four header rows and color coded blocks of columns. The green and red sections have fixed columns. Here, the second though fourth rows provide elaboration on the expected content. _Remember, Pandas interprets the very first row as a dataframe's headers by default._\n",
    "\n",
    "These tabs will have to be parsed, too. `EnvironmentalMetadata` isn't populated in the example file yet.\n",
    "- ProjectInformation\n",
    "- AssociatedResearchers\n",
    "- EnvironmentalMetadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0e44f1bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "header_rows = 4\n",
    "\n",
    "# column/color block realtions for `Metadata` tab\n",
    "\n",
    "# \"Sample Identification\"\n",
    "green_id_col = [\"ID\"]\n",
    "additional_green_col_names = [\n",
    "    \"sample_name\",\n",
    "    \"investigation_type\",\n",
    "    \"package\",\n",
    "    \"source_mat_ID\",\n",
    "]\n",
    "green_cols = green_id_col + additional_green_col_names\n",
    "\n",
    "# \"Required for all samples\"\n",
    "red_col_names = [\n",
    "    \"growth_facil\",\n",
    "    \"geo_loc_name\",\n",
    "    \"lat_lon\",\n",
    "    \"collection_date\",\n",
    "    \"samp_mat_process\",\n",
    "    \"store_cond\",\n",
    "    \"samp_store_temp\",\n",
    "    \"samp_size\",\n",
    "    \"env_broad_scale\",\n",
    "    \"env_local_scale\",\n",
    "    \"env_medium\",\n",
    "    \"gold_ecosystem\",\n",
    "    \"microbiome_taxonomy\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f72f3623",
   "metadata": {},
   "source": [
    "## ProjectInformation\n",
    "\n",
    "The project data are are organized in a single column\n",
    "\n",
    "Transforming them here to a single-row dataframe, based on prior experience with metadata tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "42012f18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Study Metadata</th>\n",
       "      <th>Project/Study Name</th>\n",
       "      <th>Principal Investigator Name</th>\n",
       "      <th>Principal Investigator Email</th>\n",
       "      <th>Principal Investigator ORCiD</th>\n",
       "      <th>LinkOut Webpage</th>\n",
       "      <th>Project/Study Description</th>\n",
       "      <th>Associated Publications</th>\n",
       "      <th>Dataset DOIs</th>\n",
       "      <th>Type of samples</th>\n",
       "      <th>Alternative Names</th>\n",
       "      <th>EMSL Proposal/Study Number</th>\n",
       "      <th>GOLD Study ID</th>\n",
       "      <th>JGI Proposal ID</th>\n",
       "      <th>Umbrella Bio Project Name</th>\n",
       "      <th>Umbrella Bio Project ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>\"Soil microbial response to elevated temperatu...</td>\n",
       "      <td>Montana Smith</td>\n",
       "      <td>montana.smith@pnnl.gov</td>\n",
       "      <td>0000-0002-8683-0050</td>\n",
       "      <td>https://microbiomedata.org/ | https://github.c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://doi.org/10.1016/j.soilbio.2019.107561 ...</td>\n",
       "      <td>https://doi.org/10.25585/1487765</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EMSL:29728</td>\n",
       "      <td>GOLD:Gs01971387</td>\n",
       "      <td>JGI:1781</td>\n",
       "      <td>NCBI Accession: PRJNA594403</td>\n",
       "      <td>NCBI ID: 594403</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Study Metadata                                 Project/Study Name  \\\n",
       "1               \"Soil microbial response to elevated temperatu...   \n",
       "\n",
       "Study Metadata Principal Investigator Name Principal Investigator Email  \\\n",
       "1                            Montana Smith       montana.smith@pnnl.gov   \n",
       "\n",
       "Study Metadata Principal Investigator ORCiD  \\\n",
       "1                       0000-0002-8683-0050   \n",
       "\n",
       "Study Metadata                                    LinkOut Webpage  \\\n",
       "1               https://microbiomedata.org/ | https://github.c...   \n",
       "\n",
       "Study Metadata Project/Study Description  \\\n",
       "1                                    NaN   \n",
       "\n",
       "Study Metadata                            Associated Publications  \\\n",
       "1               https://doi.org/10.1016/j.soilbio.2019.107561 ...   \n",
       "\n",
       "Study Metadata                      Dataset DOIs Type of samples  \\\n",
       "1               https://doi.org/10.25585/1487765            soil   \n",
       "\n",
       "Study Metadata Alternative Names EMSL Proposal/Study Number    GOLD Study ID  \\\n",
       "1                            NaN                 EMSL:29728  GOLD:Gs01971387   \n",
       "\n",
       "Study Metadata JGI Proposal ID    Umbrella Bio Project Name  \\\n",
       "1                     JGI:1781  NCBI Accession: PRJNA594403   \n",
       "\n",
       "Study Metadata Umbrella Bio Project ID  \n",
       "1                      NCBI ID: 594403  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_projinf_frame = pd.read_excel(example_file, sheet_name=\"ProjectInformation\")\n",
    "example_projinf_frame = example_projinf_frame.transpose()\n",
    "example_projinf_frame.columns = example_projinf_frame.iloc[0]\n",
    "example_projinf_frame = example_projinf_frame.reset_index(drop=True)\n",
    "\n",
    "# drop field names once they've become column names\n",
    "example_projinf_frame = example_projinf_frame.drop(0)\n",
    "\n",
    "# # drop blank left-most column\n",
    "column_numbers = [\n",
    "    x for x in range(example_projinf_frame.shape[1])\n",
    "]  # list of columns' integer indices\n",
    "column_numbers.remove(0)  # removing column integer index 0\n",
    "example_projinf_frame = example_projinf_frame.iloc[\n",
    "    :, column_numbers\n",
    "]  # return all columns except the 0th column\n",
    "\n",
    "example_projinf_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d37b958e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Project/Study Name': '\"Soil microbial response to elevated temperatures and increased carbon availability\"',\n",
       " 'Principal Investigator Name': 'Montana Smith',\n",
       " 'Principal Investigator Email': 'montana.smith@pnnl.gov',\n",
       " 'Principal Investigator ORCiD': '0000-0002-8683-0050',\n",
       " 'LinkOut Webpage': 'https://microbiomedata.org/ | https://github.com/microbiomedata',\n",
       " 'Project/Study Description': nan,\n",
       " 'Associated Publications': 'https://doi.org/10.1016/j.soilbio.2019.107561 | Keiser, Ashley D., et al. \"Peatland microbial community response to altered climate tempered by nutrient availability.\" Soil Biology and Biochemistry 137 (2019): 107561.',\n",
       " 'Dataset DOIs': 'https://doi.org/10.25585/1487765',\n",
       " 'Type of samples': 'soil',\n",
       " 'Alternative Names': nan,\n",
       " 'EMSL Proposal/Study Number': 'EMSL:29728',\n",
       " 'GOLD Study ID': 'GOLD:Gs01971387',\n",
       " 'JGI Proposal ID': 'JGI:1781',\n",
       " 'Umbrella Bio Project Name': 'NCBI Accession: PRJNA594403',\n",
       " 'Umbrella Bio Project ID': 'NCBI ID: 594403'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proj_row_list = example_projinf_frame.to_dict(\"records\")\n",
    "proj_row_list\n",
    "proj_obj = proj_row_list[0]\n",
    "proj_obj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "76e06144",
   "metadata": {},
   "outputs": [],
   "source": [
    "# value of having both BioProject ID and accession \"name\"\n",
    "proj_bioproj_id = proj_obj[\"Umbrella Bio Project ID\"]\n",
    "handle = Entrez.esummary(db=\"bioproject\", id=\"594403\")\n",
    "record = Entrez.read(handle)\n",
    "handle.close()\n",
    "# record\n",
    "\n",
    "bioproj_record_json = json.dumps(record, indent=4, sort_keys=False, default=str)\n",
    "# print(bioproj_record_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e3e040cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 'f505e81e3d3b4e66bb5541ada3af77fa',\n",
       " 'name': '\"Soil microbial response to elevated temperatures and increased carbon availability\"',\n",
       " 'description': nan,\n",
       " 'principal_investigator': {'has raw value': 'Montana Smith',\n",
       "  'orcid': '0000-0002-8683-0050'},\n",
       " 'type': 'Study',\n",
       " 'websites': ['https://microbiomedata.org/',\n",
       "  'https://github.com/microbiomedata'],\n",
       " 'publications': ['https://doi.org/10.1016/j.soilbio.2019.107561',\n",
       "  'Keiser, Ashley D., et al. \"Peatland microbial community response to altered climate tempered by nutrient availability.\" Soil Biology and Biochemistry 137 (2019): 107561.'],\n",
       " 'doi': {'has raw value': 'https://doi.org/10.25585/1487765'}}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proj_uuid = uuid.uuid4().hex\n",
    "proj_nmdc = {\n",
    "    # which ID to use?\n",
    "    # using a UUID as a placeholder\n",
    "    \"id\": proj_uuid,\n",
    "    # name, titel, dewcription, etc.?\n",
    "    \"name\": proj_obj[\"Project/Study Name\"],\n",
    "    \"description\": proj_obj[\"Project/Study Description\"],\n",
    "    \"principal_investigator\": {\n",
    "        \"has raw value\": proj_obj[\"Principal Investigator Name\"],\n",
    "        \"orcid\": proj_obj[\"Principal Investigator ORCiD\"],\n",
    "    },\n",
    "    \"type\": \"Study\",\n",
    "    \"websites\": re.split(\" *\\| *\", proj_obj[\"LinkOut Webpage\"]),\n",
    "    \"publications\": re.split(\" *\\| *\", proj_obj[\"Associated Publications\"]),\n",
    "    # does the template expect multiple DOIs?\n",
    "    \"doi\": {\n",
    "        \"has raw value\": proj_obj[\"Dataset DOIs\"],\n",
    "    },\n",
    "}\n",
    "proj_nmdc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00dbb6da",
   "metadata": {},
   "source": [
    "### Project fields in template\n",
    "\n",
    "- Project/Study Name\n",
    "- Project/Study Description\n",
    "- Alternative Names\n",
    "- Principal Investigator Name\n",
    "- Principal Investigator Email\n",
    "- Principal Investigator ORCiD\n",
    "- LinkOut Webpage\n",
    "- Associated Publications\n",
    "- Dataset DOIs\n",
    "- EMSL Proposal/Study Number\n",
    "- GOLD Study ID\n",
    "- JGI Proposal ID\n",
    "- Umbrella Bio Project Name\n",
    "- Umbrella Bio Project ID\n",
    "- Type of samples\n",
    "\n",
    "Might have to use APIs like from Entrez to look up details for some of those IDs?\n",
    "\n",
    "### Sample data from mongodb\n",
    "\n",
    "```JSON\n",
    "{\n",
    "  \"_id\": {\n",
    "    \"$oid\": \"60e840cbe9822b255ad93845\"\n",
    "  },\n",
    "  \"id\": \"gold:Gs0103573\",\n",
    "  \"name\": \"Populus root and rhizosphere microbial communities from Tennessee, USA\",\n",
    "  \"description\": \"This study is part of the Plant-Microbe Interfaces Science Focus Area, which aims to gain a deeper understanding of the diversity and functioning of mutually beneficial interactions between plants and microbes in the rhizosphere. Ongoing efforts focus on characterizing and interpreting such interfaces using systems comprising plants and microbes, in particular the poplar tree (Populus) and its microbial community in the context of favorable plant microbe interactions.\",\n",
    "  \"ecosystem\": \"Host-associated\",\n",
    "  \"ecosystem_category\": \"Plants\",\n",
    "  \"ecosystem_type\": \"Unclassified\",\n",
    "  \"ecosystem_subtype\": \"Unclassified\",\n",
    "  \"specific_ecosystem\": \"Unclassified\",\n",
    "  \"principal_investigator\": {\n",
    "    \"has_raw_value\": \"Dale Pelletier\"\n",
    "  },\n",
    "  \"type\": \"nmdc:Study\",\n",
    "  \"websites\": [\n",
    "    \"https://pmiweb.ornl.gov/pmi-project-aims/\"\n",
    "  ],\n",
    "  \"title\": \"Defining the functional diversity of the Populus root microbiome\",\n",
    "  \"doi\": {\n",
    "    \"has_raw_value\": \"https://doi.org/10.25585/1488096\"\n",
    "  },\n",
    "  \"publications\": [\n",
    "    \"https://doi.org/10.1128/mSystems.00045-18\"\n",
    "  ]\n",
    "}\n",
    "```\n",
    "\n",
    "### Alignment with schema\n",
    "\n",
    "https://microbiomedata.github.io/nmdc-schema/Study.html\n",
    "\n",
    "- which proposal, study, project ID to use in `id` slot?\n",
    "- Project/Study Name -> `title`\n",
    "- Project/Study Description -> `description`\n",
    "- ? -> `name`\n",
    "- `principal_investigator`\n",
    "    - Principal Investigator Name -> `has_raw_value`\n",
    "    - Principal Investigator ORCiD -> `orcid`\n",
    "    - https://microbiomedata.github.io/nmdc-schema/PersonValue.html\n",
    "- all `ecosystem...` fields missing from template?\n",
    "- \"Biosample\" -> `type`\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b2e3fba",
   "metadata": {},
   "source": [
    "## AssociatedResearchers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "65ed2a86",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Data Validation extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>email</th>\n",
       "      <th>ORCiD</th>\n",
       "      <th>Role- CRediT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pajau Vangay</td>\n",
       "      <td>pvangay@lbl.gov</td>\n",
       "      <td>0000-0002-9231-0692</td>\n",
       "      <td>Project administration</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Name            email                ORCiD            Role- CRediT\n",
       "0  Pajau Vangay  pvangay@lbl.gov  0000-0002-9231-0692  Project administration"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_assocs_frame = pd.read_excel(example_file, sheet_name=\"AssociatedResearchers\")\n",
    "example_assocs_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9795efce",
   "metadata": {},
   "source": [
    "> /Users/MAM/Documents/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Data Validation extension is not supported and will be removed\n",
    "  warn(msg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "185b84b4",
   "metadata": {},
   "source": [
    "## `Metadata` tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "689fd14b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# example_sheet_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "725a8aa8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Data Validation extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "example_metadata_frame = pd.read_excel(example_file, sheet_name=\"Metadata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1faee61b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>investigation_type</th>\n",
       "      <th>package</th>\n",
       "      <th>source_mat_ID</th>\n",
       "      <th>growth_facil</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>...</th>\n",
       "      <th>treatment.3</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>size_fract</th>\n",
       "      <th>select</th>\n",
       "      <th>select.1</th>\n",
       "      <th>select.2</th>\n",
       "      <th>select.3</th>\n",
       "      <th>select.4</th>\n",
       "      <th>select.5</th>\n",
       "      <th>select.6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Universally unique ID (ex: IGSN, UUID)</td>\n",
       "      <td>Sample name/Laboratory ID</td>\n",
       "      <td>Analysis/ Data Type</td>\n",
       "      <td>Sample Type</td>\n",
       "      <td>Source Material ID</td>\n",
       "      <td>Growth Facility</td>\n",
       "      <td>Geographical Location Name</td>\n",
       "      <td>latitude;longitude</td>\n",
       "      <td>Collection Date and Time</td>\n",
       "      <td>sample material processing</td>\n",
       "      <td>...</td>\n",
       "      <td>watering_regm</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "      <td>sieving</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Field REQUIRED for ALL sample submission. Opti...</td>\n",
       "      <td>Human readable ID. This is the ID that will be...</td>\n",
       "      <td>This field is constrained to contain only a se...</td>\n",
       "      <td>This field contains the type of sample type as...</td>\n",
       "      <td>A unique identifier assigned to an original ma...</td>\n",
       "      <td>Type of facility where the sample was collecte...</td>\n",
       "      <td>Detailed geographic location of sampling site</td>\n",
       "      <td>The geographical origin of the sample as defin...</td>\n",
       "      <td>The time of sampling, either as an instance (s...</td>\n",
       "      <td>Any processing applied to the sample during or...</td>\n",
       "      <td>...</td>\n",
       "      <td>treatment- watering regimen/schedule</td>\n",
       "      <td>sample collection device-slide hammer corer</td>\n",
       "      <td>size fraction-sieve size</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{text}</td>\n",
       "      <td>{text}</td>\n",
       "      <td>{text};{text};{text}</td>\n",
       "      <td>drop down selection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>drop down selection</td>\n",
       "      <td>{text}</td>\n",
       "      <td>{float};{float}</td>\n",
       "      <td>{YYYY-MM-DDTHH:MM:SS} or {YYYY-MM-DD}</td>\n",
       "      <td>{text}</td>\n",
       "      <td>...</td>\n",
       "      <td>{float} {unit};{Rn/start_time/end_time/duration}</td>\n",
       "      <td>{text}</td>\n",
       "      <td>{value}{units}</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  ID  \\\n",
       "0             Universally unique ID (ex: IGSN, UUID)   \n",
       "1  Field REQUIRED for ALL sample submission. Opti...   \n",
       "2                                             {text}   \n",
       "\n",
       "                                         sample_name  \\\n",
       "0                          Sample name/Laboratory ID   \n",
       "1  Human readable ID. This is the ID that will be...   \n",
       "2                                             {text}   \n",
       "\n",
       "                                  investigation_type  \\\n",
       "0                                Analysis/ Data Type   \n",
       "1  This field is constrained to contain only a se...   \n",
       "2                               {text};{text};{text}   \n",
       "\n",
       "                                             package  \\\n",
       "0                                        Sample Type   \n",
       "1  This field contains the type of sample type as...   \n",
       "2                                drop down selection   \n",
       "\n",
       "                                       source_mat_ID  \\\n",
       "0                                 Source Material ID   \n",
       "1  A unique identifier assigned to an original ma...   \n",
       "2                                                NaN   \n",
       "\n",
       "                                        growth_facil  \\\n",
       "0                                    Growth Facility   \n",
       "1  Type of facility where the sample was collecte...   \n",
       "2                                drop down selection   \n",
       "\n",
       "                                    geo_loc_name  \\\n",
       "0                     Geographical Location Name   \n",
       "1  Detailed geographic location of sampling site   \n",
       "2                                         {text}   \n",
       "\n",
       "                                             lat_lon  \\\n",
       "0                                 latitude;longitude   \n",
       "1  The geographical origin of the sample as defin...   \n",
       "2                                    {float};{float}   \n",
       "\n",
       "                                     collection_date  \\\n",
       "0                           Collection Date and Time   \n",
       "1  The time of sampling, either as an instance (s...   \n",
       "2              {YYYY-MM-DDTHH:MM:SS} or {YYYY-MM-DD}   \n",
       "\n",
       "                                    samp_mat_process  ...  \\\n",
       "0                         sample material processing  ...   \n",
       "1  Any processing applied to the sample during or...  ...   \n",
       "2                                             {text}  ...   \n",
       "\n",
       "                                        treatment.3  \\\n",
       "0                                     watering_regm   \n",
       "1              treatment- watering regimen/schedule   \n",
       "2  {float} {unit};{Rn/start_time/end_time/duration}   \n",
       "\n",
       "                           samp_collect_device                size_fract  \\\n",
       "0                            slide_hammer_core                   sieving   \n",
       "1  sample collection device-slide hammer corer  size fraction-sieve size   \n",
       "2                                       {text}            {value}{units}   \n",
       "\n",
       "   select select.1 select.2 select.3 select.4 select.5 select.6  \n",
       "0  select   select   select   select   select   select   select  \n",
       "1                                                                \n",
       "2                                                                \n",
       "\n",
       "[3 rows x 34 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_metadata_headers = example_metadata_frame.loc[\n",
    "    0 : (header_rows - 2),\n",
    "]\n",
    "example_metadata_headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75682032",
   "metadata": {},
   "outputs": [],
   "source": [
    "example_metadata_headers_cols = pd.Series(example_metadata_headers.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d1437d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# discard any column whose Pandas header starts with \"select\" possibly followed by a period and some digits\n",
    "# these go up to column AH now (#34 when starting at 1)\n",
    "# could that increase?\n",
    "\n",
    "select_flag = example_metadata_headers_cols.str.match(\n",
    "    \"^select(\\\\.\\\\d+)?$\", case=True, flags=0, na=None\n",
    ")\n",
    "select_cols = example_metadata_headers_cols[select_flag]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8c4b136e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a22c72c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# all column names except for those that start with \"select...\"\n",
    "keeper_cols = example_metadata_headers_cols[~select_flag]\n",
    "# keeper_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "510235d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>investigation_type</th>\n",
       "      <th>package</th>\n",
       "      <th>source_mat_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "      <td>P4_-0-10_4C</td>\n",
       "      <td>genome; metabolome</td>\n",
       "      <td>plant_associated</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "      <td>L10_-20-30_13C-Gluc</td>\n",
       "      <td>transcriptome; metabolome</td>\n",
       "      <td>soil</td>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "      <td>6_J2_75-WHC</td>\n",
       "      <td>organic matter</td>\n",
       "      <td>soil</td>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "      <td>T4-35</td>\n",
       "      <td>lipidome; 16S-Amplicon; ITS-Amplicon</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "      <td>2020-05-21_S19</td>\n",
       "      <td>proteome; genome</td>\n",
       "      <td>soil</td>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ID          sample_name  \\\n",
       "3  UUID:472894-473947-847398          P4_-0-10_4C   \n",
       "4  UUID:684267-410686-971057  L10_-20-30_13C-Gluc   \n",
       "5  UUID:472894-473947-847396          6_J2_75-WHC   \n",
       "6  UUID:516394-970067-847398                T4-35   \n",
       "7  UUID:472894-473947-642384       2020-05-21_S19   \n",
       "\n",
       "                     investigation_type           package  \\\n",
       "3                    genome; metabolome  plant_associated   \n",
       "4             transcriptome; metabolome              soil   \n",
       "5                        organic matter              soil   \n",
       "6  lipidome; 16S-Amplicon; ITS-Amplicon              soil   \n",
       "7                      proteome; genome              soil   \n",
       "\n",
       "               source_mat_ID  \n",
       "3                        NaN  \n",
       "4  UUID:472894-473947-847398  \n",
       "5  UUID:472894-473947-847398  \n",
       "6                        NaN  \n",
       "7  UUID:516394-970067-847398  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_metadata_content = example_metadata_frame.loc[\n",
    "    3:,\n",
    "]\n",
    "example_metadata_content = example_metadata_content[list(keeper_cols)]\n",
    "\n",
    "green_content = example_metadata_content[green_cols]\n",
    "\n",
    "green_content"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b414940",
   "metadata": {},
   "source": [
    "### Assumptions:\n",
    "- lowercased column header of select.* means no data?\n",
    "- **also check for non-NAs?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c12d6668",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>select</th>\n",
       "      <th>select.1</th>\n",
       "      <th>select.2</th>\n",
       "      <th>select.3</th>\n",
       "      <th>select.4</th>\n",
       "      <th>select.5</th>\n",
       "      <th>select.6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   select select.1 select.2 select.3 select.4 select.5 select.6\n",
       "0  select   select   select   select   select   select   select\n",
       "1                                                              \n",
       "2                                                              \n",
       "3     NaN      NaN      NaN      NaN      NaN      NaN      NaN\n",
       "4     NaN      NaN      NaN      NaN      NaN      NaN      NaN\n",
       "5     NaN      NaN      NaN      NaN      NaN      NaN      NaN\n",
       "6     NaN      NaN      NaN      NaN      NaN      NaN      NaN\n",
       "7     NaN      NaN      NaN      NaN      NaN      NaN      NaN"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make sure no sample metadata slipped in here\n",
    "\n",
    "select_content = example_metadata_frame[select_cols]\n",
    "select_content"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ba4429d",
   "metadata": {},
   "source": [
    "### Be prepared to check template assumptions against MIxS expectations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1ae2eae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mixs_spreadsheet = pd.read_excel(\n",
    "    mixs_spreadsheet_url, sheet_name=\"MIxS\", engine=\"openpyxl\"\n",
    ")\n",
    "mixs_spreadsheet\n",
    "\n",
    "investigation_type_row = mixs_spreadsheet[\n",
    "    mixs_spreadsheet[\"Structured comment name\"] == \"investigation_type\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "33f8cf3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['eukaryote, bacteria_archaea, plasmid, virus, organelle, metagenome,mimarks-survey, mimarks-specimen, metatranscriptome, single amplified genome, metagenome-assembled genome, or uncultivated viral genomes'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "investigation_type_row[\"Expected value\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c63b75ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['[eukaryote|bacteria_archaea|plasmid|virus|organelle|metagenome|metatranscriptome|mimarks-survey|mimarks-specimen|misag|mimag|miuvig]'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "investigation_type_row[\"Value syntax\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0a9666b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1], dtype=object)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# array?!\n",
    "investigation_type_row[\"Occurence\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e39e5f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "green_row_list = green_content.to_dict(\"records\")\n",
    "\n",
    "# I made the column/slot associations by hand (search feature in nmdc schema docs web page)\n",
    "#   would it been possible to do it by programtaticaly examining the schema?\n",
    "#   or at least account for which populated columns hadn't been mapped\n",
    "#   I'd like to open nmdc.yaml as linkedml and automatically include all imports\n",
    "green_nmdc = [\n",
    "    {\n",
    "        \"id\": item[\"ID\"],\n",
    "        \"name\": item[\"sample_name\"],\n",
    "        \"env_package\": item[\"package\"],\n",
    "        \"source_mat_id\": item[\"source_mat_ID\"],\n",
    "        # investigation_type is not multi-valued according to mixs yet\n",
    "        #   https://github.com/GenomicsStandardsConsortium/mixs-legacy/blob/master/mixs5/mixs_v5.xlsx\n",
    "        #   see Occurence column\n",
    "        #   eukaryote, bacteria_archaea, plasmid, virus, organelle, metagenome,mimarks-survey,\n",
    "        #   mimarks-specimen, metatranscriptome, single amplified genome, metagenome-assembled genome, or uncultivated viral genomes\n",
    "        # and has a smaller set of permissibles that are found in Montana's example data\n",
    "        # todo: replace \"; ?\" with \"|\" but otherwise leave as a flast string\n",
    "        # \"investigation_type\": re.split(\"; ?\", item[\"investigation_type\"]),\n",
    "        # choices according to the template\n",
    "        #   investigation_type\n",
    "        #   Select Analyses\n",
    "        #   genome\n",
    "        #   transcriptome\n",
    "        #   16S-Amplicon\n",
    "        #   ITS-Amplicon\n",
    "        #   18S-Amplicon\n",
    "        #   proteome\n",
    "        #   metabolome\n",
    "        #   lipidome\n",
    "        #   organic matter\n",
    "        #   imaging- light\n",
    "        #   imaging- electron\n",
    "        #   imaging- ion\n",
    "        #   chemical speciation/mapping\n",
    "        #   molecular structure\n",
    "        \"investigation_type\": re.sub(r\"; *\", \"|\", item[\"investigation_type\"]),\n",
    "        \"type\": \"Biosample\",\n",
    "        \"part_of\": proj_uuid,\n",
    "    }\n",
    "    for item in green_row_list\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "47cf1991",
   "metadata": {},
   "outputs": [],
   "source": [
    "red_content = example_metadata_content[green_id_col + red_col_names]\n",
    "\n",
    "# red_content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "94b0d66c",
   "metadata": {},
   "outputs": [],
   "source": [
    "red_row_list = red_content.to_dict(\"records\")\n",
    "\n",
    "red_nmdc = [\n",
    "    {\n",
    "        \"id\": item[\"ID\"],\n",
    "        # template's use of a nmdc/mixs recognized name\n",
    "        #   doesn't mean that the data won't need any tidying\n",
    "        \"growth_facil\": item[\"growth_facil\"],\n",
    "        \"geo_loc_name\": item[\"geo_loc_name\"],\n",
    "        \"samp_mat_process\": item[\"samp_mat_process\"],\n",
    "        \"store_cond\": item[\"store_cond\"],\n",
    "        \"samp_store_temp\": item[\"samp_store_temp\"],\n",
    "        \"samp_size\": item[\"samp_size\"],\n",
    "        \"env_broad_scale\": item[\"env_broad_scale\"],\n",
    "        \"env_local_scale\": item[\"env_local_scale\"],\n",
    "        \"env_medium\": item[\"env_medium\"],\n",
    "        \"lat_lon\": item[\"lat_lon\"],\n",
    "        # gold_ecosystem -> gold_path_field ?\n",
    "        \"gold_path_field\": item[\"gold_ecosystem\"],\n",
    "        \"collection_date\": item[\"collection_date\"],\n",
    "        \"ncbi_taxonomy_name\": item[\"microbiome_taxonomy\"],\n",
    "    }\n",
    "    for item in red_row_list\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cb93c9e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "green_red = green_nmdc.copy()\n",
    "for i in range(len(green_nmdc)):\n",
    "    # get rid of NaNs\n",
    "    green_red[i] = {k: v for k, v in green_red[i].items() if v == v}\n",
    "    red_temp = red_nmdc[i]\n",
    "    clean_dict = {k: v for k, v in red_temp.items() if v == v}\n",
    "    green_red[i].update(clean_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3d653d9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# default=str for Object of type datetime is not JSON serializable on collection_date\n",
    "green_red_json = json.dumps(green_red, indent=4, sort_keys=False, default=str)\n",
    "\n",
    "# print(green_red_json)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9db06f7b",
   "metadata": {},
   "source": [
    "The columns and semantics of the blue \"Required where applicable\" section are variable. Users are supposed to fill in data required by the environmental packages describing their samples? All four of the header rows may need to be parsed/interpreted in order to convert the data rows into NMDC JSON data objects and record corresponding units.\n",
    "\n",
    "The blue section is made up of all non \"select\" columns minus the green and red cols, or all columns at postion S (20, starting from 1) or greater?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ded8f63e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>treatment.1</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>elev</th>\n",
       "      <th>depth.1</th>\n",
       "      <th>depth</th>\n",
       "      <th>treatment.2</th>\n",
       "      <th>treatment</th>\n",
       "      <th>treatment.3</th>\n",
       "      <th>size_fract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4mm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "      <td>1000µg of C (glucose)/g soil</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0-1</td>\n",
       "      <td>13C Glucose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2mm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "      <td>1000µg of C (glucose)/g soil</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75% water holding capacity; moisture maintaine...</td>\n",
       "      <td>2mm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>1000</td>\n",
       "      <td>-20-30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5-.25mm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>1000</td>\n",
       "      <td>-20-30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1-.05mm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ID                   treatment.1  \\\n",
       "3  UUID:472894-473947-847398                           NaN   \n",
       "4  UUID:684267-410686-971057  1000µg of C (glucose)/g soil   \n",
       "5  UUID:472894-473947-847396  1000µg of C (glucose)/g soil   \n",
       "6  UUID:516394-970067-847398                           NaN   \n",
       "7  UUID:472894-473947-642384                           NaN   \n",
       "\n",
       "          samp_collect_device  elev depth.1 depth  treatment.2 treatment  \\\n",
       "3  11/2' x 6\"; stainless stee   500     NaN   0-1          NaN       NaN   \n",
       "4  11/2' x 6\"; stainless stee   500     NaN   0-1  13C Glucose       NaN   \n",
       "5  11/2' x 6\"; stainless stee   500     NaN   0-1          NaN       NaN   \n",
       "6  11/2' x 6\"; stainless stee  1000  -20-30   NaN          NaN       35C   \n",
       "7  11/2' x 6\"; stainless stee  1000  -20-30   NaN          NaN       NaN   \n",
       "\n",
       "                                         treatment.3 size_fract  \n",
       "3                                                NaN        4mm  \n",
       "4                                                NaN        2mm  \n",
       "5  75% water holding capacity; moisture maintaine...        2mm  \n",
       "6                                                NaN  0.5-.25mm  \n",
       "7                                                NaN    1-.05mm  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blue_cols = list((set(keeper_cols) - set(green_cols)) - set(red_col_names))\n",
    "blue_content = example_metadata_content[green_id_col + blue_cols]\n",
    "blue_content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "01535d53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>treatment.1</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>elev</th>\n",
       "      <th>depth.1</th>\n",
       "      <th>depth</th>\n",
       "      <th>treatment.2</th>\n",
       "      <th>treatment</th>\n",
       "      <th>treatment.3</th>\n",
       "      <th>size_fract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chem_administration</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "      <td>m</td>\n",
       "      <td>m</td>\n",
       "      <td>cm</td>\n",
       "      <td>isotope_exposure</td>\n",
       "      <td>air_temp_regm</td>\n",
       "      <td>watering_regm</td>\n",
       "      <td>sieving</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>treatment-chemical administration/addition</td>\n",
       "      <td>sample collection device-slide hammer corer</td>\n",
       "      <td>elevation-meters</td>\n",
       "      <td>depth-meters</td>\n",
       "      <td>depth-centimeters</td>\n",
       "      <td>treatment-isotope exposure/addition</td>\n",
       "      <td>treatment-air temperature regimen</td>\n",
       "      <td>treatment- watering regimen/schedule</td>\n",
       "      <td>size fraction-sieve size</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{termLabel} {[termID]}; {timestamp}</td>\n",
       "      <td>{text}</td>\n",
       "      <td>{value} meters</td>\n",
       "      <td>{value} meters</td>\n",
       "      <td>{value} centimeters</td>\n",
       "      <td>{text}</td>\n",
       "      <td>{float} {unit};{Rn/start_time/end_time/duration}</td>\n",
       "      <td>{float} {unit};{Rn/start_time/end_time/duration}</td>\n",
       "      <td>{value}{units}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  treatment.1  \\\n",
       "0                         chem_administration   \n",
       "1  treatment-chemical administration/addition   \n",
       "2         {termLabel} {[termID]}; {timestamp}   \n",
       "\n",
       "                           samp_collect_device              elev  \\\n",
       "0                            slide_hammer_core                 m   \n",
       "1  sample collection device-slide hammer corer  elevation-meters   \n",
       "2                                       {text}    {value} meters   \n",
       "\n",
       "          depth.1                depth                          treatment.2  \\\n",
       "0               m                   cm                     isotope_exposure   \n",
       "1    depth-meters    depth-centimeters  treatment-isotope exposure/addition   \n",
       "2  {value} meters  {value} centimeters                               {text}   \n",
       "\n",
       "                                          treatment  \\\n",
       "0                                     air_temp_regm   \n",
       "1                 treatment-air temperature regimen   \n",
       "2  {float} {unit};{Rn/start_time/end_time/duration}   \n",
       "\n",
       "                                        treatment.3                size_fract  \n",
       "0                                     watering_regm                   sieving  \n",
       "1              treatment- watering regimen/schedule  size fraction-sieve size  \n",
       "2  {float} {unit};{Rn/start_time/end_time/duration}            {value}{units}  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blue_headers = example_metadata_headers[blue_cols]\n",
    "blue_headers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8f6a06b",
   "metadata": {},
   "source": [
    "## template's `size_fract` = schema's `size_frac`\n",
    "\n",
    "For domain and range, NMDC schema says `None →` _OPT_ `QuantityValue`\n",
    "\n",
    "mongodb query `{ \"size_frac\": { $exists: true }}` against the `biosample_set` collection returns 0 documents\n",
    "\n",
    "for reference, `{ \"depth\": { $exists: true }}` **does** return hits\n",
    "\n",
    "The template says that size_fract can? must? be used with\n",
    "- sediment\n",
    "- soil\n",
    "- water\n",
    "\n",
    "The size_fract details must? be one of \n",
    "- filter_size\t{value}{units}\n",
    "- filter_type\t{text}\n",
    "- sieving\t{value}{units}\n",
    "\n",
    "The following columns are also allowed? by the template for water package samples\n",
    "\n",
    "```\n",
    "size_frac_low\tsize-fraction lower threshold\t{float} {unit}\n",
    "size_frac_up\tsize-fraction upper threshold\t{float} {unit}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "441b8c2e",
   "metadata": {},
   "source": [
    "## `treatment`\n",
    "\n",
    "There can be multiple treatment columns. Google Sheets allows multiple column headers with the same value, but Pandas adds numeric suffixes to make them unique.\n",
    "\n",
    "The schema doesn't have a treatment class or slot. Rather, schema defined terms can be found in row 0 of the headers frame.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29eb5612",
   "metadata": {},
   "source": [
    "## `elev`\n",
    "\n",
    "`alt` also allowed? Intended usage? Neither appears in mongodb yet\n",
    "\n",
    "template specifies meters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bc59f82",
   "metadata": {},
   "source": [
    "## `samp_collect_device`\n",
    "\n",
    "range = (optionally language-typed) TextValue\n",
    "\n",
    "No mongodb hits\n",
    "\n",
    "- sediment\n",
    "- soil\n",
    "- water\n",
    "\n",
    "```\n",
    "shovel\t{text}\n",
    "slide_hammer_core\t{text}\n",
    "deep_corer\t{text}\n",
    "russian_corer\t{text}\n",
    "sipper\t{text}\n",
    "other-samp_collect_device\t{text}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "77a90f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# todo: illustrate use of annotator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2e00a6d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# what if there are multiple depth columns with different units?\n",
    "\n",
    "# which slots can be multivalued and therefore take multiple columns?\n",
    "# they may or may not have suffixes like treatment, treatment.1, treatment.2\n",
    "\n",
    "blue_row_list = blue_content.to_dict(\"records\")\n",
    "\n",
    "# any one of these columns could be missing\n",
    "# how to avoid KeyErrors?\n",
    "\n",
    "blue_nmdc = [\n",
    "    {\n",
    "        \"id\": item[\"ID\"],\n",
    "        ## size_frac, not size_fract\n",
    "        # \"size_frac\": item[\"size_fract\"],\n",
    "        # \"samp_collect_device\": item[\"samp_collect_device\"],\n",
    "        ## plant_struc, host_taxid, host_common_name are valid for XXX package/investigation type\n",
    "        ## but doesn't appear in this soil sample data\n",
    "        # \"plant_struc\": item[\"plant_struc\"],\n",
    "        # \"host_taxid\": item[\"host_taxid\"],\n",
    "        # \"host_common_name\": item[\"host_common_name\"],\n",
    "        # https://microbiomedata.github.io/nmdc-schema/QuantityValue.html\n",
    "        \"depth\": {\n",
    "            \"has_raw_value\": item[\"depth\"],\n",
    "            \"has_unit\": example_metadata_headers.loc[0, \"depth\"],\n",
    "        },\n",
    "        \"elev\": {\n",
    "            \"has_raw_value\": item[\"elev\"],\n",
    "            \"has_unit\": example_metadata_headers.loc[0, \"elev\"],\n",
    "        },\n",
    "    }\n",
    "    for item in blue_row_list\n",
    "]\n",
    "\n",
    "# blue_nmdc\n",
    "\n",
    "# blue_json = json.dumps(blue_nmdc, indent=4, sort_keys=False, default=str)\n",
    "\n",
    "# print(blue_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2978c3bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#  REFACTOR\n",
    "\n",
    "green_red_blue = green_red.copy()\n",
    "for i in range(len(green_nmdc)):\n",
    "    ## NaNs should have already been removed\n",
    "    ##   but only for atomic values. Didn't do any deep search.\n",
    "    # green_red[i] = {k: v for k, v in green_red[i].items() if v == v}\n",
    "    blue_temp = blue_nmdc[i]\n",
    "    clean_dict = {k: v for k, v in blue_temp.items() if v == v}\n",
    "    green_red_blue[i].update(clean_dict)\n",
    "\n",
    "green_red_blue_json = json.dumps(green_red_blue, indent=4, sort_keys=False, default=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5e9d5b59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"id\": \"f505e81e3d3b4e66bb5541ada3af77fa\",\n",
      "    \"name\": \"\\\"Soil microbial response to elevated temperatures and increased carbon availability\\\"\",\n",
      "    \"description\": NaN,\n",
      "    \"principal_investigator\": {\n",
      "        \"has raw value\": \"Montana Smith\",\n",
      "        \"orcid\": \"0000-0002-8683-0050\"\n",
      "    },\n",
      "    \"type\": \"Study\",\n",
      "    \"websites\": [\n",
      "        \"https://microbiomedata.org/\",\n",
      "        \"https://github.com/microbiomedata\"\n",
      "    ],\n",
      "    \"publications\": [\n",
      "        \"https://doi.org/10.1016/j.soilbio.2019.107561\",\n",
      "        \"Keiser, Ashley D., et al. \\\"Peatland microbial community response to altered climate tempered by nutrient availability.\\\" Soil Biology and Biochemistry 137 (2019): 107561.\"\n",
      "    ],\n",
      "    \"doi\": {\n",
      "        \"has raw value\": \"https://doi.org/10.25585/1487765\"\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "proj_json = json.dumps(proj_nmdc, indent=4, sort_keys=False, default=str)\n",
    "print(proj_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "121b1050",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "    {\n",
      "        \"id\": \"UUID:472894-473947-847398\",\n",
      "        \"name\": \"P4_-0-10_4C\",\n",
      "        \"env_package\": \"plant_associated\",\n",
      "        \"investigation_type\": \"genome|metabolome\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"f505e81e3d3b4e66bb5541ada3af77fa\",\n",
      "        \"growth_facil\": \"field\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"snap freeze in liquid N\",\n",
      "        \"store_cond\": \"fresh\",\n",
      "        \"samp_store_temp\": \"4 degree Celsius\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251709, -119.728663\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2020-05-21T12:00:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": \"0-1\",\n",
      "            \"has_unit\": \"cm\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 500,\n",
      "            \"has_unit\": \"m\"\n",
      "        }\n",
      "    },\n",
      "    {\n",
      "        \"id\": \"UUID:684267-410686-971057\",\n",
      "        \"name\": \"L10_-20-30_13C-Gluc\",\n",
      "        \"env_package\": \"soil\",\n",
      "        \"source_mat_id\": \"UUID:472894-473947-847398\",\n",
      "        \"investigation_type\": \"transcriptome|metabolome\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"f505e81e3d3b4e66bb5541ada3af77fa\",\n",
      "        \"growth_facil\": \"field_incubation\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"2micron sieve, wet to 25% gravimetric water capacity\",\n",
      "        \"store_cond\": \"frozen\",\n",
      "        \"samp_store_temp\": \"-80 degree Celsius\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251709, -119.728663\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2020-05-21T13:30:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": \"0-1\",\n",
      "            \"has_unit\": \"cm\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 500,\n",
      "            \"has_unit\": \"m\"\n",
      "        }\n",
      "    },\n",
      "    {\n",
      "        \"id\": \"UUID:472894-473947-847396\",\n",
      "        \"name\": \"6_J2_75-WHC\",\n",
      "        \"env_package\": \"soil\",\n",
      "        \"source_mat_id\": \"UUID:472894-473947-847398\",\n",
      "        \"investigation_type\": \"organic matter\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"f505e81e3d3b4e66bb5541ada3af77fa\",\n",
      "        \"growth_facil\": \"field\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"aggregate separation at optimal moisture, dried at 100C\",\n",
      "        \"store_cond\": \"other\",\n",
      "        \"samp_store_temp\": \"room temperature\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251709, -119.728663\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2020-05-21 00:00:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": \"0-1\",\n",
      "            \"has_unit\": \"cm\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 500,\n",
      "            \"has_unit\": \"m\"\n",
      "        }\n",
      "    },\n",
      "    {\n",
      "        \"id\": \"UUID:516394-970067-847398\",\n",
      "        \"name\": \"T4-35\",\n",
      "        \"env_package\": \"soil\",\n",
      "        \"investigation_type\": \"lipidome|16S-Amplicon|ITS-Amplicon\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"f505e81e3d3b4e66bb5541ada3af77fa\",\n",
      "        \"growth_facil\": \"field\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"crushed via morter and pestle, temperature acclimation to 35C\",\n",
      "        \"store_cond\": \"lyophilized\",\n",
      "        \"samp_store_temp\": \"-80 degree Celsius\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251500, -119.728483\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2021-01-05T15:05:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": NaN,\n",
      "            \"has_unit\": \"cm\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 1000,\n",
      "            \"has_unit\": \"m\"\n",
      "        }\n",
      "    },\n",
      "    {\n",
      "        \"id\": \"UUID:472894-473947-642384\",\n",
      "        \"name\": \"2020-05-21_S19\",\n",
      "        \"env_package\": \"soil\",\n",
      "        \"source_mat_id\": \"UUID:516394-970067-847398\",\n",
      "        \"investigation_type\": \"proteome|genome\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"f505e81e3d3b4e66bb5541ada3af77fa\",\n",
      "        \"growth_facil\": \"field\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"ground via omni bead beater\",\n",
      "        \"store_cond\": \"frozen\",\n",
      "        \"samp_store_temp\": \"-20 degree Celsius\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251500, -119.728483\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2021-03-11T09:25:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": NaN,\n",
      "            \"has_unit\": \"cm\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 1000,\n",
      "            \"has_unit\": \"m\"\n",
      "        }\n",
      "    }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "print(green_red_blue_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08a0bbf1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
