{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7fbced8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "import xmltodict\n",
    "import yaml\n",
    "from pymongo import MongoClient\n",
    "import warnings\n",
    "import pandas as pd\n",
    "from pprint import pprint\n",
    "import re\n",
    "from linkml.utils.schemaloader import SchemaLoader\n",
    "from linkml_runtime.dumpers import yaml_dumper\n",
    "import uuid\n",
    "import math\n",
    "from sklearn.cluster import OPTICS\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fc1c7ec",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "`pip install -r requirements_issue_375.txt`\n",
    "\n",
    "- Copy `config/nmdc_mongo_db_template.yaml` to `config/nmdc_mongo_db.yaml` and replace the `<SECRET>`s with the host address and your password for the NMDC mongodb server.\n",
    "\n",
    "```\n",
    "mongo_nmdc_password:\n",
    "  \"<SECRET>\"\n",
    "mongo_nmdc_host:\n",
    "  \"<SECRET>\"\n",
    "```\n",
    "\n",
    "- Clone the nmdc-schema GitHub repo from git@github.com:microbiomedata/nmdc-schema.git\n",
    "- Change the location of `local_nmdc_linkml_file` if necessary below\n",
    "- Clone the mixs-source GitHub repo from git@github.com:cmungall/mixs-source.git\n",
    "- Change the location of `local_mixs_linkml_file` if necessary below\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9795efce",
   "metadata": {},
   "source": [
    "# Notes\n",
    "\n",
    "See also\n",
    "- https://github.com/microbiomedata/nmdc-metadata/issues/375\n",
    "- https://docs.google.com/spreadsheets/d/1GZayIFIrY2jdoxRIpk9KDTBLiE71VVtb7YAd5ZSYGR0/edit#gid=0\n",
    "\n",
    "Did CJM say that `samp_collec_device` has the range of ontology term? \n",
    "\n",
    "```\n",
    "name: samp_collec_device\n",
    "...\n",
    "range: string\n",
    "...\n",
    "description: The device used to collect an environmental sample. This field accepts\n",
    "  terms listed under environmental sampling device (http://purl.obolibrary.org/obo/ENVO).\n",
    "  This field also accepts terms listed under specimen collection device (http://purl.obolibrary.org/obo/GENEPIO_0002094).\n",
    "```\n",
    "\n",
    "When opening Excel spreadsheets:\n",
    "\n",
    "> /Users/MAM/Documents/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Data Validation extension is not supported and will be removed\n",
    "  warn(msg)\n",
    "  \n",
    "Addressed with https://docs.python.org/3/library/warnings.html\n",
    "\n",
    "Still getting some `value is trying to be set on a copy of a slice` warnings\n",
    "\n",
    "Still getting some `Unrecognized prefix` warnings on `SchemaLoader()`\n",
    "\n",
    "Slots available for https://cmungall.github.io/mixs-source/QuantityValue/: \n",
    "\n",
    "- `has_raw_value`\n",
    "- `has unit`\n",
    "\n",
    "and \n",
    "\n",
    "```\n",
    "âžžhas numeric value OPT\n",
    "range: Double\n",
    "```\n",
    "\n",
    "Need to process `EnvironmentalMetadata` tab\n",
    "\n",
    "No plant associated data in the example template yet, no parsing example. Should follow established patterns\n",
    "\n",
    "```\n",
    "13\tplant_associated\thost_common_name\n",
    "14\tplant_associated\thost_taxid\n",
    "11\tplant_associated\tplant_struc\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e86b59a",
   "metadata": {},
   "source": [
    "Is bmt useful for winnowing down the relevant MIxS slots?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ceabce5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import bmt\n",
    "\n",
    "# tk = bmt.Toolkit(\"../../nmdc-schema/src/schema/nmdc.yaml\")\n",
    "\n",
    "# attrib_descendants = tk.get_descendants(name='attribute')\n",
    "# attrib_descendants.sort()\n",
    "\n",
    "# print(len(attrib_descendants))\n",
    "\n",
    "# attrib_descendants\n",
    "\n",
    "# # 614 out of the 699 slots in mixs are descendants of 'attribute'\n",
    "##  filtering doens't seem worth it here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df713137",
   "metadata": {},
   "source": [
    "# Provenance\n",
    "\n",
    "\n",
    "\n",
    "saved https://docs.google.com/spreadsheets/d/1GZayIFIrY2jdoxRIpk9KDTBLiE71VVtb7YAd5ZSYGR0/edit#gid=0\n",
    "to `input/Example-Soil_NMDC_SampleMetadata.xlsx` on 2021-07-13\n",
    "\n",
    "Also \n",
    "- `Example-Soil_NMDC_SampleMetadata_20210802.xlsx`\n",
    "- `Example-Soil_NMDC_SampleMetadata_202107201146.xlsx`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d81e1a4",
   "metadata": {},
   "source": [
    "The \"Metadata\" sheet in the NMDC Sample Metadata template has four header rows and color coded blocks of columns. The green and red sections have fixed columns. Here, the second though fourth rows provide elaboration on the expected content. _Remember, Pandas interprets the very first row as a dataframe's headers by default._\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0e44f1bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_file = \"../input/Example-Soil_NMDC_SampleMetadata_20210802.xlsx\"\n",
    "header_rows = 4\n",
    "\n",
    "# column/color block realtions for `Metadata` tab\n",
    "# \"Sample Identification\"\n",
    "green_id_col = [\"ID\"]\n",
    "additional_green_col_names = [\n",
    "    \"sample_name\",\n",
    "    \"investigation_type\",\n",
    "    \"package\",\n",
    "    \"source_mat_ID\",\n",
    "]\n",
    "green_cols = green_id_col + additional_green_col_names\n",
    "\n",
    "# \"Required for all samples\"\n",
    "red_col_names = [\n",
    "    \"growth_facil\",\n",
    "    \"geo_loc_name\",\n",
    "    \"lat_lon\",\n",
    "    \"collection_date\",\n",
    "    \"samp_mat_process\",\n",
    "    \"store_cond\",\n",
    "    \"samp_store_temp\",\n",
    "    \"samp_size\",\n",
    "    \"env_broad_scale\",\n",
    "    \"env_local_scale\",\n",
    "    \"env_medium\",\n",
    "    \"gold_ecosystem\",\n",
    "    \"microbiome_taxonomy\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9251d324",
   "metadata": {},
   "outputs": [],
   "source": [
    "local_nmdc_linkml_file = \"../../nmdc-schema/src/schema/nmdc.yaml\"\n",
    "local_mixs_linkml_file = \"../../mixs-source/model/schema/mixs.yaml\"\n",
    "\n",
    "# Donny suggested some other way of conencting to mongodb\n",
    "mongodb_conf_file = \"../config/nmdc_mongo_db.yaml\"\n",
    "mongodb_user = \"mam\"\n",
    "mongodb_host = \"mongo-ext.nmdc-runtime-dev.polyneme.xyz\"\n",
    "\n",
    "# could use entrez to retrieve addtional project details based on NCBI IDs\n",
    "# initialize some default parameters\n",
    "Entrez.email = \"MAM@lbl.gov\"  # provide your email address\n",
    "paramEutils = {\"usehistory\": \"Y\"}  # Use Entrez search history to cache results\n",
    "\n",
    "# mixs_spreadsheet_url = \"https://github.com/GenomicsStandardsConsortium/mixs-legacy/blob/master/mixs5/mixs_v5.xlsx?raw=true\"\n",
    "# # could also get MIxS knoweldge from NMDC schema... have been menaing to parse that anyway\n",
    "# # CJM encourages use of his MiXS 6 yaml file (local_mixs_linkml_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6c5749e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(mongodb_conf_file) as conf_file:\n",
    "    conf_data = yaml.load(conf_file, Loader=yaml.Loader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0a17f69c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mongodb://[username:password@]host1[:port1][,...hostN[:portN]][/[defaultauthdb][?options]]\n",
    "# mongo_conn_str = \"mongodb://localhost\"\n",
    "\n",
    "mongo_conn_str = (\n",
    "    \"mongodb://\"\n",
    "    + mongodb_user\n",
    "    + \":\"\n",
    "    + conf_data[\"mongo_nmdc_password\"]\n",
    "    + \"@\"\n",
    "    + conf_data[\"mongo_nmdc_host\"]\n",
    ")\n",
    "\n",
    "\n",
    "client = MongoClient(mongo_conn_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dba8f2bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_nans_from_dict(inp_dict):\n",
    "    clean_dict = {k: v for k, v in inp_dict.items() if v == v}\n",
    "    return clean_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "124fe4cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_j_on_index(left_j, right_j):\n",
    "    left_copy = left_j.copy()\n",
    "    for i in range(len(green_nmdc)):\n",
    "        left_copy[i] = remove_nans_from_dict(left_copy[i])\n",
    "        right_temp = right_j[i]\n",
    "        right_temp = remove_nans_from_dict(right_temp)\n",
    "        left_copy[i].update(right_temp)\n",
    "    return left_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "516cefab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# should we check that each ID is only used once in each dict?\n",
    "def merge_j_on_id(left_j, right_j):\n",
    "    # the zipping and unzipping will be repeated unneccessarily?\n",
    "    left_ids = [a_dict[\"id\"] for a_dict in left_j]\n",
    "    left_zip = dict(zip(left_ids, left_j))\n",
    "    right_ids = [a_dict[\"id\"] for a_dict in right_j]\n",
    "    right_zip = dict(zip(right_ids, right_j))\n",
    "    shared_ids = set(left_ids).intersection(set(right_ids))\n",
    "    for i in shared_ids:\n",
    "        # this will get repeated each time this function is called. refactor?\n",
    "        left_zip[i] = remove_nans_from_dict(left_zip[i])\n",
    "        right_zip[i] = remove_nans_from_dict(right_zip[i])\n",
    "        left_zip[i].update(right_zip[i])\n",
    "    # unzip\n",
    "    left_values_only = list(left_zip.values())\n",
    "    return left_values_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "02746b67",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_searchable_sheet(df):\n",
    "    zipped_tmts = dict(zip(df.columns, range(len(df.columns))))\n",
    "    searchable = df.rename(columns=zipped_tmts).stack()\n",
    "    return searchable\n",
    "\n",
    "\n",
    "def search_get_coords(searchable, search_string):\n",
    "    match_coords_list = searchable[(searchable == search_string)].index.tolist()\n",
    "    return match_coords_list\n",
    "\n",
    "\n",
    "def euc_pair_dist(point1, point2):\n",
    "    x_dist = point2[0] - point1[0]\n",
    "    xds = x_dist ** 2\n",
    "    y_dist = point2[1] - point1[1]\n",
    "    yds = y_dist ** 2\n",
    "    dist_sum = xds + yds\n",
    "    dist_root = math.sqrt(dist_sum)\n",
    "    return dist_root\n",
    "\n",
    "\n",
    "def find_first_closest(xy_tuple_list, origin):\n",
    "    dists = []\n",
    "    for xy_tuple in xy_tuple_list:\n",
    "        current_dist = euc_pair_dist(origin, xy_tuple)\n",
    "        dists.append(current_dist)\n",
    "    val, idx = min((val, idx) for (idx, val) in enumerate(dists))\n",
    "    return xy_tuple_list[idx]\n",
    "\n",
    "\n",
    "def get_from_find_to_end(df, start, colcount):\n",
    "    df_row_count = len(df.index)\n",
    "    found_to_end = df.iloc[\n",
    "        start[0] : (df_row_count - 1), start[1] : (start[1] + (colcount))\n",
    "    ]\n",
    "    found_to_end = found_to_end.reset_index(drop=True)\n",
    "    return found_to_end\n",
    "\n",
    "\n",
    "# assuming we should search for nas in first column\n",
    "def get_from_start_to_na(df):\n",
    "    df_isna = df.isna()\n",
    "    df_isna_loc = df_isna.idxmax().values[0]\n",
    "    # could this ever be larger than one element?\n",
    "    df_to_na = df[0:df_isna_loc]\n",
    "    return df_to_na\n",
    "\n",
    "\n",
    "def colnames_from_rowvals(df, row):\n",
    "    df.columns = df.iloc[row].values\n",
    "    return df\n",
    "\n",
    "\n",
    "def colnames_from_rowvals(df, row):\n",
    "    df.columns = df.iloc[row].values\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8ce97e32",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_rows_matching_list(\n",
    "    df, column, exclusion_list, remove_nans=True, remove_header_val=True\n",
    "):\n",
    "    if remove_header_val:\n",
    "        col_header = df.columns[column]\n",
    "        exclusion_list.append(col_header)\n",
    "    extract = df.iloc[:, column]\n",
    "    goodflag = ~extract.isin(exclusion_list)\n",
    "    keepers = df[goodflag]\n",
    "    if remove_nans:\n",
    "        keepers.dropna(axis=0, how=\"any\", inplace=True)\n",
    "    keepers.reset_index(drop=True, inplace=True)\n",
    "    return keepers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a53ffae8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_optics_coord_frame(df, current_search_string):\n",
    "    # may have already been assembled elsewhere ... factor back out?\n",
    "    searchable_menu = make_searchable_sheet(df)\n",
    "    tx_coords = search_get_coords(searchable_menu, current_search_string)\n",
    "    tx_coord_frame = pd.DataFrame(tx_coords, columns=[\"y\", \"x\"])\n",
    "    tx_custers = OPTICS(min_samples=2).fit(tx_coords)\n",
    "    tx_per_coord_labels = tx_custers.labels_\n",
    "    tx_coord_frame[\"cluster\"] = tx_per_coord_labels\n",
    "    return tx_coord_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "06859d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_colors = {\n",
    "    -1: \"gray\",\n",
    "    0: \"red\",\n",
    "    1: \"black\",\n",
    "    2: \"green\",\n",
    "    3: \"orange\",\n",
    "    4: \"purple\",\n",
    "    5: \"brown\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6484f3fa",
   "metadata": {},
   "source": [
    "### `get_optics_coord_frame`\n",
    "\n",
    "- Shouldn't assume that there will be one largest cluster\n",
    "- Shouldn't assume number of clusters/cluster_colors\n",
    "- These cluster_colors don't have anything to do with the color blocks int the Metadata tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c52873a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_data_from_extended_cluster_coords(\n",
    "    data_df, cluster_coords_df, largest_index, additional_columns=0\n",
    "):\n",
    "    temp = cluster_coords_df[cluster_coords_df[\"cluster\"] == largest_index]\n",
    "    xmin = temp[\"x\"].min()\n",
    "    xmax = temp[\"x\"].max()\n",
    "    ymin = temp[\"y\"].min()\n",
    "    ymax = temp[\"y\"].max()\n",
    "\n",
    "    additional_columns = 3\n",
    "    extended_x_max = xmax + additional_columns\n",
    "    data_from_extended_cluster_coords = data_df.loc[ymin:ymax, xmin:extended_x_max]\n",
    "    return data_from_extended_cluster_coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d994af0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pivot_multi_columns(colname_base):\n",
    "    temp_match_cols = [col for col in blue_cols if re.search(r\"^\" + colname_base, col)]\n",
    "    temp_cols = green_id_col + temp_match_cols\n",
    "    temp_content = blue_content[temp_cols]\n",
    "    temp_content_piv = pd.melt(temp_content, id_vars=\"ID\")\n",
    "    temp_content_piv\n",
    "    nan_flag = temp_content_piv[\"value\"].isna()\n",
    "    temp_content_piv = temp_content_piv[~nan_flag]\n",
    "    temp_headers = blue_headers[temp_match_cols]\n",
    "    temp_headers = temp_headers.iloc[0:1, :]\n",
    "    temp_headers_piv = pd.melt(temp_headers)\n",
    "    temp_piv_merged = temp_content_piv.merge(\n",
    "        right=temp_headers_piv, how=\"left\", left_on=\"variable\", right_on=\"variable\"\n",
    "    )\n",
    "    return temp_piv_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6a406879",
   "metadata": {},
   "outputs": [],
   "source": [
    "def concatenate_melted_multi(\n",
    "    melted_multi_df,\n",
    "    data_col_name,\n",
    "    source_id_col_name=\"ID\",\n",
    "    x_drop=\"value_x\",\n",
    "    y_drop=\"value_y\",\n",
    "):\n",
    "    temp_multi = melted_multi_df\n",
    "    cols_for_concat = [x_drop, y_drop]\n",
    "    temp_multi[data_col_name] = melted_multi_df[cols_for_concat].apply(\n",
    "        lambda row: \" \".join(row.values.astype(str)), axis=1\n",
    "    )\n",
    "    temp_multi = temp_multi.drop([x_drop, y_drop], axis=1)\n",
    "    temp_multi_pivot = temp_multi.pivot(\n",
    "        index=source_id_col_name, columns=\"variable\", values=data_col_name\n",
    "    )\n",
    "    temp_multi_pivot[\"id\"] = temp_multi_pivot.index\n",
    "    temp_multi_pivot = temp_multi_pivot.reset_index(drop=True)\n",
    "    return temp_multi_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c28f61e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_melted_multi(\n",
    "    melted_multi_df,\n",
    "    source_id_col_name=\"ID\",\n",
    "    x_drop=\"value_x\",\n",
    "    y_drop=\"value_y\",\n",
    "):\n",
    "    temp_recon = melted_multi_df.pivot(\n",
    "        index=source_id_col_name, columns=y_drop, values=x_drop\n",
    "    )\n",
    "    temp_recon[\"id\"] = temp_recon.index\n",
    "    temp_recon = temp_recon.reset_index(drop=True)\n",
    "    return temp_recon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "973ab05c",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings(\"ignore\", category=UserWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6d39614",
   "metadata": {},
   "source": [
    "Should we want to double check anything with NCBI..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3beba855",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DbList': ['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'ncbisearch', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'biosystems', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list Entrez databases\n",
    "entrez_databases = Entrez.read(Entrez.einfo())\n",
    "entrez_databases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "926a63c9",
   "metadata": {},
   "source": [
    "# List the tabs (sheets) in the NMDC Biosample Metadata Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a5723557",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ReadMe',\n",
       " 'ProjectInformation',\n",
       " 'AssociatedResearchers',\n",
       " 'Metadata',\n",
       " 'EnvironmentalMetadata',\n",
       " 'EcosystemTerms',\n",
       " 'MenuTerms']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_frame_structure = pd.read_excel(template_file, sheet_name=None)\n",
    "template_sheet_names = template_frame_structure.keys()\n",
    "list(template_sheet_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d70fc3c",
   "metadata": {},
   "source": [
    "# `ProjectInformation` tab\n",
    "\n",
    "The project data are are organized in a single column\n",
    "\n",
    "They correspond to MIxS `study` modeling\n",
    "\n",
    "Transforming them here to a single-row dataframe, based on prior experience with metadata tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "42012f18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Study Metadata</th>\n",
       "      <th>Project/Study Name</th>\n",
       "      <th>Principal Investigator Name</th>\n",
       "      <th>Principal Investigator Email</th>\n",
       "      <th>Principal Investigator ORCiD</th>\n",
       "      <th>LinkOut Webpage</th>\n",
       "      <th>Project/Study Description</th>\n",
       "      <th>Associated Publications</th>\n",
       "      <th>Dataset DOIs</th>\n",
       "      <th>Type of samples</th>\n",
       "      <th>Alternative Names</th>\n",
       "      <th>EMSL Proposal/Study Number</th>\n",
       "      <th>GOLD Study ID</th>\n",
       "      <th>JGI Proposal ID</th>\n",
       "      <th>Umbrella Bio Project Name</th>\n",
       "      <th>Umbrella Bio Project ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>\"Soil microbial response to elevated temperatu...</td>\n",
       "      <td>Montana Smith</td>\n",
       "      <td>montana.smith@pnnl.gov</td>\n",
       "      <td>0000-0002-8683-0050</td>\n",
       "      <td>https://microbiomedata.org/ | https://github.c...</td>\n",
       "      <td>Northern latitude peatlands contain large rese...</td>\n",
       "      <td>https://doi.org/10.1016/j.soilbio.2019.107561 ...</td>\n",
       "      <td>https://doi.org/10.25585/1487765</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EMSL:29728</td>\n",
       "      <td>GOLD:Gs01971387</td>\n",
       "      <td>JGI:1781</td>\n",
       "      <td>NCBI Accession: PRJNA594403</td>\n",
       "      <td>NCBI ID: 594403</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Study Metadata                                 Project/Study Name  \\\n",
       "1               \"Soil microbial response to elevated temperatu...   \n",
       "\n",
       "Study Metadata Principal Investigator Name Principal Investigator Email  \\\n",
       "1                            Montana Smith       montana.smith@pnnl.gov   \n",
       "\n",
       "Study Metadata Principal Investigator ORCiD  \\\n",
       "1                       0000-0002-8683-0050   \n",
       "\n",
       "Study Metadata                                    LinkOut Webpage  \\\n",
       "1               https://microbiomedata.org/ | https://github.c...   \n",
       "\n",
       "Study Metadata                          Project/Study Description  \\\n",
       "1               Northern latitude peatlands contain large rese...   \n",
       "\n",
       "Study Metadata                            Associated Publications  \\\n",
       "1               https://doi.org/10.1016/j.soilbio.2019.107561 ...   \n",
       "\n",
       "Study Metadata                      Dataset DOIs Type of samples  \\\n",
       "1               https://doi.org/10.25585/1487765            soil   \n",
       "\n",
       "Study Metadata Alternative Names EMSL Proposal/Study Number    GOLD Study ID  \\\n",
       "1                            NaN                 EMSL:29728  GOLD:Gs01971387   \n",
       "\n",
       "Study Metadata JGI Proposal ID    Umbrella Bio Project Name  \\\n",
       "1                     JGI:1781  NCBI Accession: PRJNA594403   \n",
       "\n",
       "Study Metadata Umbrella Bio Project ID  \n",
       "1                      NCBI ID: 594403  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_projinf_frame = pd.read_excel(template_file, sheet_name=\"ProjectInformation\")\n",
    "template_projinf_frame = template_projinf_frame.transpose()\n",
    "template_projinf_frame.columns = template_projinf_frame.iloc[0]\n",
    "template_projinf_frame = template_projinf_frame.reset_index(drop=True)\n",
    "\n",
    "# drop field names once they've become column names\n",
    "template_projinf_frame = template_projinf_frame.drop(0)\n",
    "\n",
    "# # drop blank left-most column\n",
    "column_numbers = [\n",
    "    x for x in range(template_projinf_frame.shape[1])\n",
    "]  # list of columns' integer indices\n",
    "column_numbers.remove(0)  # removing column integer index 0\n",
    "template_projinf_frame = template_projinf_frame.iloc[\n",
    "    :, column_numbers\n",
    "]  # return all columns except the 0th column\n",
    "\n",
    "template_projinf_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d37b958e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Alternative Names': nan,\n",
      " 'Associated Publications': 'https://doi.org/10.1016/j.soilbio.2019.107561 | '\n",
      "                            'Keiser, Ashley D., et al. \"Peatland microbial '\n",
      "                            'community response to altered climate tempered by '\n",
      "                            'nutrient availability.\" Soil Biology and '\n",
      "                            'Biochemistry 137 (2019): 107561.',\n",
      " 'Dataset DOIs': 'https://doi.org/10.25585/1487765',\n",
      " 'EMSL Proposal/Study Number': 'EMSL:29728',\n",
      " 'GOLD Study ID': 'GOLD:Gs01971387',\n",
      " 'JGI Proposal ID': 'JGI:1781',\n",
      " 'LinkOut Webpage': 'https://microbiomedata.org/ | '\n",
      "                    'https://github.com/microbiomedata',\n",
      " 'Principal Investigator Email': 'montana.smith@pnnl.gov',\n",
      " 'Principal Investigator Name': 'Montana Smith',\n",
      " 'Principal Investigator ORCiD': '0000-0002-8683-0050',\n",
      " 'Project/Study Description': 'Northern latitude peatlands contain large '\n",
      "                              'reserves of soil carbon (C) due in part to '\n",
      "                              'climatic constraints on decomposition, '\n",
      "                              'including low temperatures and water inundated '\n",
      "                              'soils.... As expected, the two LMWCC substrate '\n",
      "                              'additions increased C respiration....',\n",
      " 'Project/Study Name': '\"Soil microbial response to elevated temperatures and '\n",
      "                       'increased carbon availability\"',\n",
      " 'Type of samples': 'soil',\n",
      " 'Umbrella Bio Project ID': 'NCBI ID: 594403',\n",
      " 'Umbrella Bio Project Name': 'NCBI Accession: PRJNA594403'}\n"
     ]
    }
   ],
   "source": [
    "proj_row_list = template_projinf_frame.to_dict(\"records\")\n",
    "proj_row_list\n",
    "proj_obj = proj_row_list[0]\n",
    "pprint(proj_obj)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35243f52",
   "metadata": {},
   "source": [
    "## Example of obtaining project details from NCBI with Entrez\n",
    "\n",
    "Reddy: no API for GOLD and JGI?\n",
    "But the values may be in NCBI?\n",
    "What about EMSL or"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1239d6f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# value of having both BioProject ID and accession \"name\"?\n",
    "proj_bioproj_id = proj_obj[\"Umbrella Bio Project ID\"]\n",
    "bioproj_rhs = re.sub(\"^.*: *\", \"\", proj_bioproj_id)\n",
    "# bioproj_rhs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "83b0ad0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = Entrez.esummary(db=\"bioproject\", id=bioproj_rhs)\n",
    "record = Entrez.read(handle)\n",
    "handle.close()\n",
    "# record\n",
    "\n",
    "bioproj_record_json = json.dumps(record, indent=4, sort_keys=False, default=str)\n",
    "# print(bioproj_record_json)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "956688c7",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Get NMDC classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "85fab034",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Unrecognized prefix: rdfs\n",
      "Unrecognized prefix: insdc.srs\n",
      "Unrecognized prefix: mgnify\n",
      "Unrecognized prefix: UniProtKB\n",
      "Unrecognized prefix: gtpo\n",
      "Unrecognized prefix: COG\n",
      "Unrecognized prefix: MetaCyc\n",
      "Unrecognized prefix: EC\n",
      "Unrecognized prefix: MetaNetX\n",
      "Unrecognized prefix: RetroRules\n"
     ]
    }
   ],
   "source": [
    "# nmdc_schema_resolved = load_raw_schema(local_nmdc_linkml_file)\n",
    "\n",
    "nmdc_schema_resolved = SchemaLoader(local_nmdc_linkml_file).resolve()\n",
    "\n",
    "nmdc_resolved_classes = list(nmdc_schema_resolved.classes.keys())\n",
    "nmdc_resolved_classes.sort()\n",
    "\n",
    "# nmdc_resolved_classes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94d73209",
   "metadata": {},
   "source": [
    "## Get NMDC schema `study` slots\n",
    "_Included inherited slots_\n",
    "\n",
    "This and more will all be reconciled below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5195401d",
   "metadata": {},
   "outputs": [],
   "source": [
    "study_slots = nmdc_schema_resolved.classes[\"study\"].slots\n",
    "study_slots.sort()\n",
    "# study_slots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00dbb6da",
   "metadata": {},
   "source": [
    "## Get `ProjectInformation` fields from `Example-Soil_NMDC_SampleMetadata`\n",
    "\n",
    "Might have to use APIs like from Entrez to look up details for some of those IDs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d8cf548c",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_fields = list(template_projinf_frame.columns)\n",
    "project_fields.sort()\n",
    "# project_fields"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0989dfbb",
   "metadata": {},
   "source": [
    "## Get `study` data from mongodb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "dadaaf92",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'_id': ObjectId('60e840cbe9822b255ad93845'),\n",
      " 'description': 'This study is part of the Plant-Microbe Interfaces Science '\n",
      "                'Focus Area, which aims to gain a deeper understanding of the '\n",
      "                'diversity and functioning of mutually beneficial interactions '\n",
      "                'between plants and microbes in the rhizosphere. Ongoing '\n",
      "                'efforts focus on characterizing and interpreting such '\n",
      "                'interfaces using systems comprising plants and microbes, in '\n",
      "                'particular the poplar tree (Populus) and its microbial '\n",
      "                'community in the context of favorable plant microbe '\n",
      "                'interactions.',\n",
      " 'doi': {'has_raw_value': 'https://doi.org/10.25585/1488096'},\n",
      " 'ecosystem': 'Host-associated',\n",
      " 'ecosystem_category': 'Plants',\n",
      " 'ecosystem_subtype': 'Unclassified',\n",
      " 'ecosystem_type': 'Unclassified',\n",
      " 'id': 'gold:Gs0103573',\n",
      " 'name': 'Populus root and rhizosphere microbial communities from Tennessee, '\n",
      "         'USA',\n",
      " 'principal_investigator': {'has_raw_value': 'Dale Pelletier'},\n",
      " 'publications': ['https://doi.org/10.1128/mSystems.00045-18'],\n",
      " 'specific_ecosystem': 'Unclassified',\n",
      " 'title': 'Defining the functional diversity of the Populus root microbiome',\n",
      " 'type': 'nmdc:Study',\n",
      " 'websites': ['https://pmiweb.ornl.gov/pmi-project-aims/']}\n"
     ]
    }
   ],
   "source": [
    "db = client.nmdc\n",
    "coll = db[\"study_set\"]\n",
    "\n",
    "any_study = coll.find_one({\"type\": \"nmdc:Study\"})\n",
    "pprint(any_study)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d27b605e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>ecosystem_category</th>\n",
       "      <th>ecosystem_type</th>\n",
       "      <th>ecosystem_subtype</th>\n",
       "      <th>specific_ecosystem</th>\n",
       "      <th>principal_investigator</th>\n",
       "      <th>type</th>\n",
       "      <th>websites</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>publications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>60e840cbe9822b255ad93845</td>\n",
       "      <td>gold:Gs0103573</td>\n",
       "      <td>Populus root and rhizosphere microbial communi...</td>\n",
       "      <td>This study is part of the Plant-Microbe Interf...</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Plants</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>{'has_raw_value': 'Dale Pelletier'}</td>\n",
       "      <td>nmdc:Study</td>\n",
       "      <td>[https://pmiweb.ornl.gov/pmi-project-aims/]</td>\n",
       "      <td>Defining the functional diversity of the Popul...</td>\n",
       "      <td>{'has_raw_value': 'https://doi.org/10.25585/14...</td>\n",
       "      <td>[https://doi.org/10.1128/mSystems.00045-18]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>60e840cbe9822b255ad93846</td>\n",
       "      <td>gold:Gs0110138</td>\n",
       "      <td>Peatland microbial communities from Minnesota,...</td>\n",
       "      <td>This study is part of the Spruce and Peatland ...</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>Wetlands</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>{'has_raw_value': 'Erik Lilleskov'}</td>\n",
       "      <td>nmdc:Study</td>\n",
       "      <td>[https://mnspruce.ornl.gov/project/overview]</td>\n",
       "      <td>Fungal, bacterial, and archaeal communities me...</td>\n",
       "      <td>{'has_raw_value': 'https://doi.org/10.25585/14...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>60e840cbe9822b255ad93847</td>\n",
       "      <td>gold:Gs0114663</td>\n",
       "      <td>Riverbed sediment microbial communities from t...</td>\n",
       "      <td>This research project, led by James Stegen at ...</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Aquatic</td>\n",
       "      <td>Freshwater</td>\n",
       "      <td>River</td>\n",
       "      <td>Sediment</td>\n",
       "      <td>{'has_raw_value': 'James Stegen', 'profile_ima...</td>\n",
       "      <td>nmdc:Study</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'has_raw_value': 'https://doi.org/10.25585/14...</td>\n",
       "      <td>[https://doi.org/10.1371/journal.pone.0228165]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>60e840cbe9822b255ad93848</td>\n",
       "      <td>gold:Gs0114675</td>\n",
       "      <td>Deep subsurface shale carbon reservoir microbi...</td>\n",
       "      <td>This project aims to improve the understanding...</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Deep subsurface</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>{'has_raw_value': 'Kelly Wrighton', 'profile_i...</td>\n",
       "      <td>nmdc:Study</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'has_raw_value': 'https://doi.org/10.25585/14...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>60e840cbe9822b255ad93849</td>\n",
       "      <td>gold:Gs0128850</td>\n",
       "      <td>Lab enrichment of tropical soil microbial comm...</td>\n",
       "      <td>This study is part of the Microbes Persist: Sy...</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Forest Soil</td>\n",
       "      <td>{'has_raw_value': 'Jennifer Pett-Ridge', 'prof...</td>\n",
       "      <td>nmdc:Study</td>\n",
       "      <td>[https://sc-programs.llnl.gov/biological-and-e...</td>\n",
       "      <td>Microbial Carbon Transformations in Wet Tropic...</td>\n",
       "      <td>{'has_raw_value': 'https://doi.org/10.25585/14...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>60e840cbe9822b255ad9384a</td>\n",
       "      <td>gold:Gs0135149</td>\n",
       "      <td>Bulk soil microbial communities from the East ...</td>\n",
       "      <td>This research project aimed to understand how ...</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Meadow</td>\n",
       "      <td>Bulk soil</td>\n",
       "      <td>{'has_raw_value': 'Eoin Brodie', 'profile_imag...</td>\n",
       "      <td>nmdc:Study</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'has_raw_value': 'https://doi.org/10.25585/14...</td>\n",
       "      <td>[https://doi.org/10.21952/WTR/1573029]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>60e840cbe9822b255ad9384c</td>\n",
       "      <td>gold:Gs0154044</td>\n",
       "      <td>Bulk and rhizosphere soil microbial communitie...</td>\n",
       "      <td>The goal of this Bio-Scales Pilot Project stud...</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>{'has_raw_value': 'Mitchel J. Doktycz', 'profi...</td>\n",
       "      <td>nmdc:Study</td>\n",
       "      <td>[https://www.ornl.gov/staff-profile/mitchel-j-...</td>\n",
       "      <td>Bio-Scales: Defining plant gene function and i...</td>\n",
       "      <td>{'has_raw_value': 'https://doi.org/10.46936/10...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        _id              id  \\\n",
       "0  60e840cbe9822b255ad93845  gold:Gs0103573   \n",
       "1  60e840cbe9822b255ad93846  gold:Gs0110138   \n",
       "2  60e840cbe9822b255ad93847  gold:Gs0114663   \n",
       "3  60e840cbe9822b255ad93848  gold:Gs0114675   \n",
       "4  60e840cbe9822b255ad93849  gold:Gs0128850   \n",
       "5  60e840cbe9822b255ad9384a  gold:Gs0135149   \n",
       "6  60e840cbe9822b255ad9384c  gold:Gs0154044   \n",
       "\n",
       "                                                name  \\\n",
       "0  Populus root and rhizosphere microbial communi...   \n",
       "1  Peatland microbial communities from Minnesota,...   \n",
       "2  Riverbed sediment microbial communities from t...   \n",
       "3  Deep subsurface shale carbon reservoir microbi...   \n",
       "4  Lab enrichment of tropical soil microbial comm...   \n",
       "5  Bulk soil microbial communities from the East ...   \n",
       "6  Bulk and rhizosphere soil microbial communitie...   \n",
       "\n",
       "                                         description        ecosystem  \\\n",
       "0  This study is part of the Plant-Microbe Interf...  Host-associated   \n",
       "1  This study is part of the Spruce and Peatland ...    Environmental   \n",
       "2  This research project, led by James Stegen at ...    Environmental   \n",
       "3  This project aims to improve the understanding...    Environmental   \n",
       "4  This study is part of the Microbes Persist: Sy...    Environmental   \n",
       "5  This research project aimed to understand how ...    Environmental   \n",
       "6  The goal of this Bio-Scales Pilot Project stud...    Environmental   \n",
       "\n",
       "  ecosystem_category   ecosystem_type ecosystem_subtype specific_ecosystem  \\\n",
       "0             Plants     Unclassified      Unclassified       Unclassified   \n",
       "1            Aquatic       Freshwater          Wetlands       Unclassified   \n",
       "2            Aquatic       Freshwater             River           Sediment   \n",
       "3        Terrestrial  Deep subsurface      Unclassified       Unclassified   \n",
       "4        Terrestrial             Soil      Unclassified        Forest Soil   \n",
       "5        Terrestrial             Soil            Meadow          Bulk soil   \n",
       "6        Terrestrial             Soil      Unclassified       Unclassified   \n",
       "\n",
       "                              principal_investigator        type  \\\n",
       "0                {'has_raw_value': 'Dale Pelletier'}  nmdc:Study   \n",
       "1                {'has_raw_value': 'Erik Lilleskov'}  nmdc:Study   \n",
       "2  {'has_raw_value': 'James Stegen', 'profile_ima...  nmdc:Study   \n",
       "3  {'has_raw_value': 'Kelly Wrighton', 'profile_i...  nmdc:Study   \n",
       "4  {'has_raw_value': 'Jennifer Pett-Ridge', 'prof...  nmdc:Study   \n",
       "5  {'has_raw_value': 'Eoin Brodie', 'profile_imag...  nmdc:Study   \n",
       "6  {'has_raw_value': 'Mitchel J. Doktycz', 'profi...  nmdc:Study   \n",
       "\n",
       "                                            websites  \\\n",
       "0        [https://pmiweb.ornl.gov/pmi-project-aims/]   \n",
       "1       [https://mnspruce.ornl.gov/project/overview]   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  [https://sc-programs.llnl.gov/biological-and-e...   \n",
       "5                                                NaN   \n",
       "6  [https://www.ornl.gov/staff-profile/mitchel-j-...   \n",
       "\n",
       "                                               title  \\\n",
       "0  Defining the functional diversity of the Popul...   \n",
       "1  Fungal, bacterial, and archaeal communities me...   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  Microbial Carbon Transformations in Wet Tropic...   \n",
       "5                                                NaN   \n",
       "6  Bio-Scales: Defining plant gene function and i...   \n",
       "\n",
       "                                                 doi  \\\n",
       "0  {'has_raw_value': 'https://doi.org/10.25585/14...   \n",
       "1  {'has_raw_value': 'https://doi.org/10.25585/14...   \n",
       "2  {'has_raw_value': 'https://doi.org/10.25585/14...   \n",
       "3  {'has_raw_value': 'https://doi.org/10.25585/14...   \n",
       "4  {'has_raw_value': 'https://doi.org/10.25585/14...   \n",
       "5  {'has_raw_value': 'https://doi.org/10.25585/14...   \n",
       "6  {'has_raw_value': 'https://doi.org/10.46936/10...   \n",
       "\n",
       "                                     publications  \n",
       "0     [https://doi.org/10.1128/mSystems.00045-18]  \n",
       "1                                             NaN  \n",
       "2  [https://doi.org/10.1371/journal.pone.0228165]  \n",
       "3                                             NaN  \n",
       "4                                             NaN  \n",
       "5          [https://doi.org/10.21952/WTR/1573029]  \n",
       "6                                             NaN  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_studies = list(coll.find({}))\n",
    "\n",
    "study_frame = pd.DataFrame(all_studies)\n",
    "\n",
    "study_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "859ed3f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['_id',\n",
       " 'description',\n",
       " 'doi',\n",
       " 'ecosystem',\n",
       " 'ecosystem_category',\n",
       " 'ecosystem_subtype',\n",
       " 'ecosystem_type',\n",
       " 'id',\n",
       " 'name',\n",
       " 'principal_investigator',\n",
       " 'publications',\n",
       " 'specific_ecosystem',\n",
       " 'title',\n",
       " 'type',\n",
       " 'websites']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_cols = list(study_frame.columns)\n",
    "study_cols.sort()\n",
    "study_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8029ca36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "_id                       0\n",
       "id                        0\n",
       "name                      0\n",
       "description               0\n",
       "ecosystem                 0\n",
       "ecosystem_category        0\n",
       "ecosystem_type            0\n",
       "ecosystem_subtype         0\n",
       "specific_ecosystem        0\n",
       "principal_investigator    0\n",
       "type                      0\n",
       "websites                  3\n",
       "title                     3\n",
       "doi                       0\n",
       "publications              4\n",
       "dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_frame.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c2480f9",
   "metadata": {},
   "source": [
    "## Curated matches between schema and template\n",
    "\n",
    "| NMDC schema study slots  | ProjectInformation fields from Example-Soil_NMDC_SampleMetadata   |\n",
    "|:-------------------------|:------------------------------------------------------------------|\n",
    "| alternative names        | Alternative Names                                                 |\n",
    "| description              | Project/Study Description                                         |\n",
    "| doi                      | Dataset DOIs                                                      |\n",
    "| title                    | Project/Study Name                                                |\n",
    "| principal investigator   | Principal Investigator ORCiD, Principal Investigator ORCiD        |\n",
    "| publications             | Associated Publications                                           |\n",
    "| type                     | constant 'study'                                                  |\n",
    "| websites                 | LinkOut Webpage                                                   |\n",
    "\n",
    "\n",
    "## How many of these ID-like columns do we expect to be populated? \n",
    "\n",
    "0? 1? Multiple? All?\n",
    "\n",
    "**How to prioritize picking just one?**\n",
    "\n",
    "Putting all of them in `alternate identifiers` for the time being\n",
    "\n",
    "- GOLD Study ID\n",
    "    - CJM says prioritize this.\n",
    "    - Then what?\n",
    "- EMSL Proposal/Study Number\n",
    "- JGI Proposal ID\n",
    "- Umbrella Bio Project ID\n",
    "    - what is the relation between this and the bioproject accession?\n",
    "\n",
    "<!-- **Prioritizing GOLD. Need additional rules.** -->\n",
    "\n",
    "## Mongodb utilized but unaccounted-for slots\n",
    "Ie. these slots are defined in the NMDC schema and have at least some data in the current mongodb. However, there don't seem to be any corresponding ProjectInformation columns in the template \n",
    "\n",
    "- id (see above)\n",
    "- All study-wide ecosystem terms\n",
    "    - ecosystem\n",
    "    - ecosystem_category\n",
    "    - ecosystem_subtype\n",
    "    - ecosystem_type\n",
    "    - specific_ecosystem\n",
    "    \n",
    "```\n",
    "title OPT\n",
    "Description: A name given to the entity that differs from the name/label programatically assigned to it. For example, when extracting study information for GOLD, the GOLD system has assigned a name/label. However, for display purposes, we may also wish the capture the title of the proposal that was used to fund the study.\n",
    "Range: String\n",
    "```\n",
    "    \n",
    "## Unaccounted-for template columns\n",
    "\n",
    "- Principal Investigator Email\n",
    "    - see https://microbiomedata.github.io/nmdc-schema/PersonValue.html\n",
    "    - see https://github.com/microbiomedata/nmdc-schema/pull/105\n",
    "- Type of samples\n",
    "- Umbrella Bio Project Name\n",
    "    - that's really the bioproject accession\n",
    "\n",
    "## Non-utilized and unaccounted-for slots\n",
    "\n",
    "Acting on these is probably the lowest priority\n",
    "    \n",
    "- abstract\n",
    "- alternate identifiers\n",
    "- alternative descriptions\n",
    "- alternative titles\n",
    "- objective\n",
    "- name etc. (Inherited from named thing)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e3e040cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'GOLD study identifiers': 'GOLD:Gs01971387',\n",
      " 'INSDC bioproject identifiers': 'NCBI Accession: PRJNA594403',\n",
      " 'alternate identifiers': ['EMSL:29728', 'JGI:1781', 'NCBI ID: 594403'],\n",
      " 'description': 'Northern latitude peatlands contain large reserves of soil '\n",
      "                'carbon (C) due in part to climatic constraints on '\n",
      "                'decomposition, including low temperatures and water inundated '\n",
      "                'soils.... As expected, the two LMWCC substrate additions '\n",
      "                'increased C respiration....',\n",
      " 'doi': {'has raw value': 'https://doi.org/10.25585/1487765'},\n",
      " 'id': 'GOLD:Gs01971387',\n",
      " 'principal_investigator': {'email': 'montana.smith@pnnl.gov',\n",
      "                            'has raw value': 'Montana Smith',\n",
      "                            'orcid': '0000-0002-8683-0050'},\n",
      " 'publications': ['https://doi.org/10.1016/j.soilbio.2019.107561',\n",
      "                  'Keiser, Ashley D., et al. \"Peatland microbial community '\n",
      "                  'response to altered climate tempered by nutrient '\n",
      "                  'availability.\" Soil Biology and Biochemistry 137 (2019): '\n",
      "                  '107561.'],\n",
      " 'title': '\"Soil microbial response to elevated temperatures and increased '\n",
      "          'carbon availability\"',\n",
      " 'type': 'Study',\n",
      " 'websites': ['https://microbiomedata.org/',\n",
      "              'https://github.com/microbiomedata']}\n"
     ]
    }
   ],
   "source": [
    "proj_uuid = uuid.uuid4().hex\n",
    "proj_obj_keys = list(proj_obj.keys())\n",
    "proj_nmdc = {\n",
    "    # CJS says to preferrentially use GOLD id\n",
    "    # we will need rules for prioritizing others\n",
    "    # checking for length > 0 is presumably not sifficient here\n",
    "    # although we have taked about doing validataion outside of this parsing script\n",
    "    # using a UUID as a fallback\n",
    "    \"id\": proj_obj[\"GOLD Study ID\"]\n",
    "    if len(proj_obj[\"GOLD Study ID\"]) > 0\n",
    "    else proj_obj[\"Umbrella Bio Project Name\"]\n",
    "    if len(proj_obj[\"Umbrella Bio Project Name\"]) > 0\n",
    "    else proj_uuid,\n",
    "    # scrub quotation marks from name etc?\n",
    "    \"title\": proj_obj[\"Project/Study Name\"],\n",
    "    \"description\": proj_obj[\"Project/Study Description\"].replace('\"', \"\")\n",
    "    if (\n",
    "        \"Project/Study Description\" in proj_obj_keys\n",
    "        and len(proj_obj[\"Project/Study Description\"]) > 0\n",
    "    )\n",
    "    else \"\",\n",
    "    # email slot in https://github.com/microbiomedata/nmdc-schema/pull/105/files\n",
    "    \"principal_investigator\": {\n",
    "        \"has raw value\": proj_obj[\"Principal Investigator Name\"],\n",
    "        \"orcid\": proj_obj[\"Principal Investigator ORCiD\"],\n",
    "        \"email\": proj_obj[\"Principal Investigator Email\"],\n",
    "    },\n",
    "    \"type\": \"Study\",\n",
    "    \"websites\": re.split(\" *\\| *\", proj_obj[\"LinkOut Webpage\"]),\n",
    "    \"publications\": re.split(\" *\\| *\", proj_obj[\"Associated Publications\"]),\n",
    "    # does the template expect multiple DOIs?\n",
    "    \"doi\": {\n",
    "        \"has raw value\": proj_obj[\"Dataset DOIs\"],\n",
    "    },\n",
    "    #     \"title\": proj_obj[\"title\"]\n",
    "    #     if (\"title\" in proj_obj_keys and len(proj_obj[\"title\"]) > 0)\n",
    "    #     else \"\",\n",
    "    # these should use specific slots and or include recognized prefixes\n",
    "    \"alternate identifiers\": [\n",
    "        proj_obj[\"EMSL Proposal/Study Number\"],\n",
    "        proj_obj[\"JGI Proposal ID\"],\n",
    "        proj_obj[\"Umbrella Bio Project ID\"],\n",
    "    ],\n",
    "    # should we expect vertical pipes and split on them for multi values?\n",
    "    \"INSDC bioproject identifiers\": proj_obj[\"Umbrella Bio Project Name\"],\n",
    "    \"GOLD study identifiers\": proj_obj[\"GOLD Study ID\"],\n",
    "}\n",
    "pprint(proj_nmdc)\n",
    "# print(proj_nmdc)\n",
    "# proj_nmdc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "33d5b874",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Project/Study Name',\n",
       " 'Principal Investigator Name',\n",
       " 'Principal Investigator Email',\n",
       " 'Principal Investigator ORCiD',\n",
       " 'LinkOut Webpage',\n",
       " 'Project/Study Description',\n",
       " 'Associated Publications',\n",
       " 'Dataset DOIs',\n",
       " 'Type of samples',\n",
       " 'Alternative Names',\n",
       " 'EMSL Proposal/Study Number',\n",
       " 'GOLD Study ID',\n",
       " 'JGI Proposal ID',\n",
       " 'Umbrella Bio Project Name',\n",
       " 'Umbrella Bio Project ID']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(proj_obj.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dfb1fb8",
   "metadata": {},
   "source": [
    "# `AssociatedResearchers` tab\n",
    "\n",
    "How should this be modeled?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "65ed2a86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>email</th>\n",
       "      <th>ORCiD</th>\n",
       "      <th>Role- CRediT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pajau Vangay</td>\n",
       "      <td>pvangay@lbl.gov</td>\n",
       "      <td>0000-0002-9231-0692</td>\n",
       "      <td>Project administration</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Name            email                ORCiD            Role- CRediT\n",
       "0  Pajau Vangay  pvangay@lbl.gov  0000-0002-9231-0692  Project administration"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_assocs_frame = pd.read_excel(template_file, sheet_name=\"AssociatedResearchers\")\n",
    "template_assocs_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e72b88e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_assocs_row_list = template_assocs_frame.to_dict(\"records\")\n",
    "# template_assocs_row_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "40800160",
   "metadata": {},
   "outputs": [],
   "source": [
    "# these sematnics haven't been finalized\n",
    "template_assocs_nmdc = [\n",
    "    {\n",
    "        \"type\": \"credit record\",\n",
    "        \"has contributor role\": item[\"Role- CRediT\"],\n",
    "        \"applies to person\": {\n",
    "            \"has_raw_value\": item[\"Name\"],\n",
    "            \"orcid\": item[\"ORCiD\"],\n",
    "            \"email\": item[\"email\"],\n",
    "        },\n",
    "    }\n",
    "    for item in template_assocs_row_list\n",
    "]\n",
    "\n",
    "# template_assocs_nmdc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1d1b0f25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 'GOLD:Gs01971387',\n",
       " 'title': '\"Soil microbial response to elevated temperatures and increased carbon availability\"',\n",
       " 'description': 'Northern latitude peatlands contain large reserves of soil carbon (C) due in part to climatic constraints on decomposition, including low temperatures and water inundated soils.... As expected, the two LMWCC substrate additions increased C respiration....',\n",
       " 'principal_investigator': {'has raw value': 'Montana Smith',\n",
       "  'orcid': '0000-0002-8683-0050',\n",
       "  'email': 'montana.smith@pnnl.gov'},\n",
       " 'type': 'Study',\n",
       " 'websites': ['https://microbiomedata.org/',\n",
       "  'https://github.com/microbiomedata'],\n",
       " 'publications': ['https://doi.org/10.1016/j.soilbio.2019.107561',\n",
       "  'Keiser, Ashley D., et al. \"Peatland microbial community response to altered climate tempered by nutrient availability.\" Soil Biology and Biochemistry 137 (2019): 107561.'],\n",
       " 'doi': {'has raw value': 'https://doi.org/10.25585/1487765'},\n",
       " 'alternate identifiers': ['EMSL:29728', 'JGI:1781', 'NCBI ID: 594403'],\n",
       " 'INSDC bioproject identifiers': 'NCBI Accession: PRJNA594403',\n",
       " 'GOLD study identifiers': 'GOLD:Gs01971387',\n",
       " 'has credit record': [{'type': 'credit record',\n",
       "   'has contributor role': 'Project administration',\n",
       "   'applies to person': {'has_raw_value': 'Pajau Vangay',\n",
       "    'orcid': '0000-0002-9231-0692',\n",
       "    'email': 'pvangay@lbl.gov'}}]}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proj_nmdc[\"has credit record\"] = template_assocs_nmdc\n",
    "# pprint(proj_nmdc)\n",
    "proj_nmdc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2350335",
   "metadata": {},
   "source": [
    "# `Metadata` tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "689fd14b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# template_sheet_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "725a8aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_metadata_frame = pd.read_excel(template_file, sheet_name=\"Metadata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1faee61b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>investigation_type</th>\n",
       "      <th>package</th>\n",
       "      <th>source_mat_ID</th>\n",
       "      <th>growth_facil</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>samp_mat_process</th>\n",
       "      <th>...</th>\n",
       "      <th>treatment.3</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>size_fract</th>\n",
       "      <th>select</th>\n",
       "      <th>select.1</th>\n",
       "      <th>select.2</th>\n",
       "      <th>select.3</th>\n",
       "      <th>select.4</th>\n",
       "      <th>select.5</th>\n",
       "      <th>select.6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Universally unique ID (ex: IGSN, UUID)</td>\n",
       "      <td>Sample name/Laboratory ID</td>\n",
       "      <td>Analysis/ Data Type</td>\n",
       "      <td>Sample Type</td>\n",
       "      <td>Source Material ID</td>\n",
       "      <td>Growth Facility</td>\n",
       "      <td>Geographical Location Name</td>\n",
       "      <td>latitude;longitude</td>\n",
       "      <td>Collection Date and Time</td>\n",
       "      <td>sample material processing</td>\n",
       "      <td>...</td>\n",
       "      <td>watering_regm</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "      <td>sieving</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Field REQUIRED for ALL sample submission. Opti...</td>\n",
       "      <td>Human readable ID. This is the ID that will be...</td>\n",
       "      <td>This field is constrained to contain only a se...</td>\n",
       "      <td>This field contains the type of sample type as...</td>\n",
       "      <td>A unique identifier assigned to an original ma...</td>\n",
       "      <td>Type of facility where the sample was collecte...</td>\n",
       "      <td>Detailed geographic location of sampling site</td>\n",
       "      <td>The geographical origin of the sample as defin...</td>\n",
       "      <td>The time of sampling, either as an instance (s...</td>\n",
       "      <td>Any processing applied to the sample during or...</td>\n",
       "      <td>...</td>\n",
       "      <td>treatment- watering regimen/schedule</td>\n",
       "      <td>sample collection device-slide hammer corer</td>\n",
       "      <td>size fraction-sieve size</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{text}</td>\n",
       "      <td>{text}</td>\n",
       "      <td>{text};{text};{text}</td>\n",
       "      <td>drop down selection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>drop down selection</td>\n",
       "      <td>{text}</td>\n",
       "      <td>{float};{float}</td>\n",
       "      <td>{YYYY-MM-DDTHH:MM:SS} or {YYYY-MM-DD}</td>\n",
       "      <td>{text}</td>\n",
       "      <td>...</td>\n",
       "      <td>{float} {unit};{Rn/start_time/end_time/duration}</td>\n",
       "      <td>{text}</td>\n",
       "      <td>{value}{units}</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  ID  \\\n",
       "0             Universally unique ID (ex: IGSN, UUID)   \n",
       "1  Field REQUIRED for ALL sample submission. Opti...   \n",
       "2                                             {text}   \n",
       "\n",
       "                                         sample_name  \\\n",
       "0                          Sample name/Laboratory ID   \n",
       "1  Human readable ID. This is the ID that will be...   \n",
       "2                                             {text}   \n",
       "\n",
       "                                  investigation_type  \\\n",
       "0                                Analysis/ Data Type   \n",
       "1  This field is constrained to contain only a se...   \n",
       "2                               {text};{text};{text}   \n",
       "\n",
       "                                             package  \\\n",
       "0                                        Sample Type   \n",
       "1  This field contains the type of sample type as...   \n",
       "2                                drop down selection   \n",
       "\n",
       "                                       source_mat_ID  \\\n",
       "0                                 Source Material ID   \n",
       "1  A unique identifier assigned to an original ma...   \n",
       "2                                                NaN   \n",
       "\n",
       "                                        growth_facil  \\\n",
       "0                                    Growth Facility   \n",
       "1  Type of facility where the sample was collecte...   \n",
       "2                                drop down selection   \n",
       "\n",
       "                                    geo_loc_name  \\\n",
       "0                     Geographical Location Name   \n",
       "1  Detailed geographic location of sampling site   \n",
       "2                                         {text}   \n",
       "\n",
       "                                             lat_lon  \\\n",
       "0                                 latitude;longitude   \n",
       "1  The geographical origin of the sample as defin...   \n",
       "2                                    {float};{float}   \n",
       "\n",
       "                                     collection_date  \\\n",
       "0                           Collection Date and Time   \n",
       "1  The time of sampling, either as an instance (s...   \n",
       "2              {YYYY-MM-DDTHH:MM:SS} or {YYYY-MM-DD}   \n",
       "\n",
       "                                    samp_mat_process  ...  \\\n",
       "0                         sample material processing  ...   \n",
       "1  Any processing applied to the sample during or...  ...   \n",
       "2                                             {text}  ...   \n",
       "\n",
       "                                        treatment.3  \\\n",
       "0                                     watering_regm   \n",
       "1              treatment- watering regimen/schedule   \n",
       "2  {float} {unit};{Rn/start_time/end_time/duration}   \n",
       "\n",
       "                           samp_collect_device                size_fract  \\\n",
       "0                            slide_hammer_core                   sieving   \n",
       "1  sample collection device-slide hammer corer  size fraction-sieve size   \n",
       "2                                       {text}            {value}{units}   \n",
       "\n",
       "   select select.1 select.2 select.3 select.4 select.5 select.6  \n",
       "0  select   select   select   select   select   select   select  \n",
       "1                                                                \n",
       "2                                                                \n",
       "\n",
       "[3 rows x 34 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_metadata_headers = template_metadata_frame.loc[\n",
    "    0 : (header_rows - 2),\n",
    "]\n",
    "template_metadata_headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "75682032",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_metadata_headers_cols = pd.Series(template_metadata_headers.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d1437d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# discard any column whose Pandas header starts with \"select\" possibly followed by a period and some digits\n",
    "# these go up to column AH now (#34 when starting at 1)\n",
    "# could that increase?\n",
    "\n",
    "select_flag = template_metadata_headers_cols.str.match(\n",
    "    \"^select(\\\\.\\\\d+)?$\", case=True, flags=0, na=None\n",
    ")\n",
    "select_cols = template_metadata_headers_cols[select_flag]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8c4b136e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a22c72c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# all column names except for those that start with \"select...\"\n",
    "keeper_cols = template_metadata_headers_cols[~select_flag]\n",
    "# keeper_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "89fbe0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_metadata_content = template_metadata_frame.loc[\n",
    "    3:,\n",
    "]\n",
    "template_metadata_content = template_metadata_content[list(keeper_cols)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e756768c",
   "metadata": {},
   "source": [
    "## Assumption:\n",
    "- lowercased column header of select.* means no data?\n",
    "- **also check for non-NAs?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "c12d6668",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>select</th>\n",
       "      <th>select.1</th>\n",
       "      <th>select.2</th>\n",
       "      <th>select.3</th>\n",
       "      <th>select.4</th>\n",
       "      <th>select.5</th>\n",
       "      <th>select.6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "      <td>select</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   select select.1 select.2 select.3 select.4 select.5 select.6\n",
       "0  select   select   select   select   select   select   select\n",
       "1                                                              \n",
       "2                                                              \n",
       "3     NaN      NaN      NaN      NaN      NaN      NaN      NaN\n",
       "4     NaN      NaN      NaN      NaN      NaN      NaN      NaN\n",
       "5     NaN      NaN      NaN      NaN      NaN      NaN      NaN\n",
       "6     NaN      NaN      NaN      NaN      NaN      NaN      NaN\n",
       "7     NaN      NaN      NaN      NaN      NaN      NaN      NaN"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make sure no sample metadata slipped in here\n",
    "\n",
    "select_content = template_metadata_frame[select_cols]\n",
    "select_content"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9aabecd",
   "metadata": {},
   "source": [
    "## Be prepared to check template assumptions against MIxS expectations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "03c0832e",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_menu_terms = pd.read_excel(\n",
    "    template_file, sheet_name=\"MenuTerms\", nrows=1000, header=None\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "379e2984",
   "metadata": {},
   "outputs": [],
   "source": [
    "# template_menu_terms.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "34f00eb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the spot in the MenuTerms tab where \"investigation_type\" is mentioned\n",
    "searchable_menu = make_searchable_sheet(template_menu_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "36211cb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_search_string = \"Look Up Options\"\n",
    "loo_coords = search_get_coords(searchable_menu, current_search_string)\n",
    "closest_found = find_first_closest(loo_coords, (0, 0))\n",
    "found_to_end = get_from_find_to_end(template_menu_terms, closest_found, colcount=2)\n",
    "first_loo_found_to_na = get_from_start_to_na(found_to_end)\n",
    "# first_loo_found_to_na\n",
    "\n",
    "# the source sheet for this extract was intentionally parsed with header=None\n",
    "# is there any row that could serve as the header?\n",
    "# remove rows where column \"0\" is \"Look Up Options\", \"Sample_category\", \"select\"\n",
    "# and column \"1\" is NaN, \"Variable\", \"select\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "aa602c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "package_to_blue_col = colnames_from_rowvals(first_loo_found_to_na, 1)\n",
    "# package_to_blue_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "e9763c0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return func(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "package_to_blue_col = remove_rows_matching_list(\n",
    "    package_to_blue_col, 1, [\"Variable\", \"select\"]\n",
    ")\n",
    "# A value is trying to be set on a copy of a slice from a DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6f1d393",
   "metadata": {},
   "source": [
    "## Which columns (variables) go with which packages (Sample_categories)?\n",
    "_According to the template_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "a35b42f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_category</th>\n",
       "      <th>Variable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>elev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>host_common_name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>host_taxid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>plant_struc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sediment</td>\n",
       "      <td>depth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sediment</td>\n",
       "      <td>elev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sediment</td>\n",
       "      <td>samp_collect_device</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sediment</td>\n",
       "      <td>size_fract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sediment</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>soil</td>\n",
       "      <td>depth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>elev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>soil</td>\n",
       "      <td>samp_collect_device</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>soil</td>\n",
       "      <td>size_fract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>soil</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>water</td>\n",
       "      <td>depth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>water</td>\n",
       "      <td>elev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>water</td>\n",
       "      <td>samp_collect_device</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>water</td>\n",
       "      <td>size_fract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>water</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample_category             Variable\n",
       "10  plant_associated                 elev\n",
       "13  plant_associated     host_common_name\n",
       "14  plant_associated           host_taxid\n",
       "11  plant_associated          plant_struc\n",
       "12  plant_associated            treatment\n",
       "6           sediment                depth\n",
       "5           sediment                 elev\n",
       "8           sediment  samp_collect_device\n",
       "9           sediment           size_fract\n",
       "7           sediment            treatment\n",
       "1               soil                depth\n",
       "0               soil                 elev\n",
       "3               soil  samp_collect_device\n",
       "4               soil           size_fract\n",
       "2               soil            treatment\n",
       "16             water                depth\n",
       "15             water                 elev\n",
       "18             water  samp_collect_device\n",
       "19             water           size_fract\n",
       "17             water            treatment"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "package_to_blue_col.sort_values(by=[\"Sample_category\", \"Variable\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "6a544a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_search_string = \"investigation_type\"\n",
    "inv_type_coords = search_get_coords(searchable_menu, current_search_string)\n",
    "closest_found = find_first_closest(inv_type_coords, (0, 0))\n",
    "# closest_found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "70f00991",
   "metadata": {},
   "outputs": [],
   "source": [
    "found_to_end = get_from_find_to_end(template_menu_terms, closest_found, colcount=1)\n",
    "investigation_types = get_from_start_to_na(found_to_end)\n",
    "# type(investigation_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "05de0079",
   "metadata": {},
   "outputs": [],
   "source": [
    "investigation_types = colnames_from_rowvals(investigation_types, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "3a02c17c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/MAM/Documents/gitrepos/nmdc-metadata/venv/lib/python3.9/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  return func(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "investigation_types = remove_rows_matching_list(\n",
    "    investigation_types, 0, [\"Select Analyses\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b81cf66",
   "metadata": {},
   "source": [
    "## What are the permitted investigation types\n",
    "_According to the template_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "5fec6452",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>investigation_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>transcriptome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16S-Amplicon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ITS-Amplicon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18S-Amplicon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>proteome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>metabolome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>lipidome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>organic matter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>imaging- light</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>imaging- electron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>imaging- ion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chemical speciation/mapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>molecular structure</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             investigation_type\n",
       "0                        genome\n",
       "1                 transcriptome\n",
       "2                  16S-Amplicon\n",
       "3                  ITS-Amplicon\n",
       "4                  18S-Amplicon\n",
       "5                      proteome\n",
       "6                    metabolome\n",
       "7                      lipidome\n",
       "8                organic matter\n",
       "9                imaging- light\n",
       "10            imaging- electron\n",
       "11                 imaging- ion\n",
       "12  chemical speciation/mapping\n",
       "13          molecular structure"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "investigation_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd683c17",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Now getting MIxS knowledge from CJM's YAML, not NMDC's\n",
    "Or from GSC's spreadsheet\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "cfb7c377",
   "metadata": {},
   "outputs": [],
   "source": [
    "mixs_schema_resolved = SchemaLoader(local_mixs_linkml_file).resolve()\n",
    "mixs_slots = list(mixs_schema_resolved.slots.keys())\n",
    "mixs_slots.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "b3a34547",
   "metadata": {},
   "outputs": [],
   "source": [
    "mixs_schema_inv_type = mixs_schema_resolved.slots[\"investigation_type\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "ced4f972",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['bacteria_archaea',\n",
       " 'eukaryote',\n",
       " 'metagenome',\n",
       " 'metatranscriptome',\n",
       " 'mimag',\n",
       " 'mimarks-specimen',\n",
       " 'mimarks-survey',\n",
       " 'misag',\n",
       " 'miuvig',\n",
       " 'organelle',\n",
       " 'plasmid',\n",
       " 'virus']"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mixs_schema_inv_type_pattern = mixs_schema_inv_type.pattern\n",
    "temp = re.sub(\"\\[|\\]\", \"\", mixs_schema_inv_type_pattern)\n",
    "mixs_schema_inv_type_pattern = re.split(\" *\\| *\", temp)\n",
    "mixs_schema_inv_type_pattern.sort()\n",
    "mixs_schema_inv_type_pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "26c6b9c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "mixs_schema_inv_type_multivalued = mixs_schema_inv_type.multivalued\n",
    "\n",
    "print(mixs_schema_inv_type_multivalued)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84e7748f",
   "metadata": {},
   "source": [
    "`Investigation_type` is not asserted as multi-valued "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba3af1e5",
   "metadata": {},
   "source": [
    "## Metadata tab, green block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "510235d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>investigation_type</th>\n",
       "      <th>package</th>\n",
       "      <th>source_mat_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "      <td>P4_-0-10_4C</td>\n",
       "      <td>genome; metabolome</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "      <td>L10_-20-30_13C-Gluc</td>\n",
       "      <td>transcriptome; metabolome</td>\n",
       "      <td>soil</td>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "      <td>6_J2_75-WHC</td>\n",
       "      <td>organic matter</td>\n",
       "      <td>soil</td>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "      <td>T4-35</td>\n",
       "      <td>lipidome; 16S-Amplicon; ITS-Amplicon</td>\n",
       "      <td>soil</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "      <td>2020-05-21_S19</td>\n",
       "      <td>proteome; genome</td>\n",
       "      <td>soil</td>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ID          sample_name  \\\n",
       "3  UUID:472894-473947-847398          P4_-0-10_4C   \n",
       "4  UUID:684267-410686-971057  L10_-20-30_13C-Gluc   \n",
       "5  UUID:472894-473947-847396          6_J2_75-WHC   \n",
       "6  UUID:516394-970067-847398                T4-35   \n",
       "7  UUID:472894-473947-642384       2020-05-21_S19   \n",
       "\n",
       "                     investigation_type package              source_mat_ID  \n",
       "3                    genome; metabolome    soil                        NaN  \n",
       "4             transcriptome; metabolome    soil  UUID:472894-473947-847398  \n",
       "5                        organic matter    soil  UUID:472894-473947-847398  \n",
       "6  lipidome; 16S-Amplicon; ITS-Amplicon    soil                        NaN  \n",
       "7                      proteome; genome    soil  UUID:516394-970067-847398  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "green_content = template_metadata_content[green_cols]\n",
    "\n",
    "green_content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "e39e5f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "green_row_list = green_content.to_dict(\"records\")\n",
    "\n",
    "# I made the column/slot associations by hand (search feature in nmdc schema docs web page)\n",
    "#   would it been possible to do it by programtaticaly examining the schema?\n",
    "#   or at least account for which populated columns hadn't been mapped\n",
    "#   I'd like to open nmdc.yaml as linkedml and automatically include all imports\n",
    "green_nmdc = [\n",
    "    {\n",
    "        \"id\": item[\"ID\"],\n",
    "        \"name\": item[\"sample_name\"],\n",
    "        \"env_package\": item[\"package\"],\n",
    "        \"source_mat_id\": item[\"source_mat_ID\"],\n",
    "        # \"investigation_type\": re.split(\"; *\", item[\"investigation_type\"]),\n",
    "        \"investigation_type\": re.sub(r\"; *\", \"|\", item[\"investigation_type\"]),\n",
    "        \"type\": \"Biosample\",\n",
    "        \"part_of\": proj_uuid,\n",
    "    }\n",
    "    for item in green_row_list\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "202cbce3",
   "metadata": {},
   "source": [
    "### Metadata tab, red block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "47cf1991",
   "metadata": {},
   "outputs": [],
   "source": [
    "red_content = template_metadata_content[green_id_col + red_col_names]\n",
    "\n",
    "# red_content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "94b0d66c",
   "metadata": {},
   "outputs": [],
   "source": [
    "red_row_list = red_content.to_dict(\"records\")\n",
    "\n",
    "red_nmdc = [\n",
    "    {\n",
    "        \"id\": item[\"ID\"],\n",
    "        # template's use of a nmdc/mixs recognized name\n",
    "        #   doesn't mean that the data won't need any tidying\n",
    "        \"growth_facil\": item[\"growth_facil\"],\n",
    "        \"geo_loc_name\": item[\"geo_loc_name\"],\n",
    "        \"samp_mat_process\": item[\"samp_mat_process\"],\n",
    "        \"store_cond\": item[\"store_cond\"],\n",
    "        \"samp_store_temp\": item[\"samp_store_temp\"],\n",
    "        \"samp_size\": item[\"samp_size\"],\n",
    "        \"env_broad_scale\": item[\"env_broad_scale\"],\n",
    "        \"env_local_scale\": item[\"env_local_scale\"],\n",
    "        \"env_medium\": item[\"env_medium\"],\n",
    "        \"lat_lon\": item[\"lat_lon\"],\n",
    "        # gold_ecosystem -> gold_path_field ?\n",
    "        \"gold_path_field\": item[\"gold_ecosystem\"],\n",
    "        \"collection_date\": item[\"collection_date\"],\n",
    "        \"ncbi_taxonomy_name\": item[\"microbiome_taxonomy\"],\n",
    "    }\n",
    "    for item in red_row_list\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "63476b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cumulative_json_merge = green_nmdc\n",
    "cumulative_json_merge = merge_j_on_id(cumulative_json_merge, red_nmdc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "05283a38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cumulative_json_merge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fa03488",
   "metadata": {},
   "source": [
    "## Metadata tab, blue block"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73d9b165",
   "metadata": {},
   "source": [
    "The columns and semantics of the blue \"Required where applicable\" section are variable. Users are supposed to fill in data required by the environmental packages describing their samples? All four of the header rows may need to be parsed/interpreted in order to convert the data rows into NMDC JSON data objects and record corresponding units.\n",
    "\n",
    "The blue section is made up of all non \"select\" columns minus the green and red cols, or all columns at postion S (20, starting from 1) or greater?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e90bbea7",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "5746c25d",
   "metadata": {},
   "outputs": [],
   "source": [
    "blue_cols = list((set(keeper_cols) - set(green_cols)) - set(red_col_names))\n",
    "blue_cols = [\"ID\"] + blue_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ded8f63e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>depth.1</th>\n",
       "      <th>treatment.1</th>\n",
       "      <th>treatment.2</th>\n",
       "      <th>size_fract</th>\n",
       "      <th>depth</th>\n",
       "      <th>treatment.3</th>\n",
       "      <th>treatment</th>\n",
       "      <th>elev</th>\n",
       "      <th>samp_collect_device</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4mm</td>\n",
       "      <td>0-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>500.0</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000Âµg of C (glucose)/g soil</td>\n",
       "      <td>13C Glucose</td>\n",
       "      <td>2mm</td>\n",
       "      <td>0-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>500.0</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000Âµg of C (glucose)/g soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2mm</td>\n",
       "      <td>0-1</td>\n",
       "      <td>75% water holding capacity; moisture maintaine...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>500.0</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "      <td>-20-30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5-.25mm</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35C</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "      <td>-20-30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1-.05mm</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ID depth.1                   treatment.1  \\\n",
       "3  UUID:472894-473947-847398     NaN                           NaN   \n",
       "4  UUID:684267-410686-971057     NaN  1000Âµg of C (glucose)/g soil   \n",
       "5  UUID:472894-473947-847396     NaN  1000Âµg of C (glucose)/g soil   \n",
       "6  UUID:516394-970067-847398  -20-30                           NaN   \n",
       "7  UUID:472894-473947-642384  -20-30                           NaN   \n",
       "\n",
       "   treatment.2 size_fract depth  \\\n",
       "3          NaN        4mm   0-1   \n",
       "4  13C Glucose        2mm   0-1   \n",
       "5          NaN        2mm   0-1   \n",
       "6          NaN  0.5-.25mm   NaN   \n",
       "7          NaN    1-.05mm   NaN   \n",
       "\n",
       "                                         treatment.3 treatment    elev  \\\n",
       "3                                                NaN       NaN   500.0   \n",
       "4                                                NaN       NaN   500.0   \n",
       "5  75% water holding capacity; moisture maintaine...       NaN   500.0   \n",
       "6                                                NaN       35C  1000.0   \n",
       "7                                                NaN       NaN  1000.0   \n",
       "\n",
       "          samp_collect_device  \n",
       "3  11/2' x 6\"; stainless stee  \n",
       "4  11/2' x 6\"; stainless stee  \n",
       "5  11/2' x 6\"; stainless stee  \n",
       "6  11/2' x 6\"; stainless stee  \n",
       "7  11/2' x 6\"; stainless stee  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blue_content = template_metadata_content[blue_cols]\n",
    "blue_content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "01535d53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>depth.1</th>\n",
       "      <th>treatment.1</th>\n",
       "      <th>treatment.2</th>\n",
       "      <th>size_fract</th>\n",
       "      <th>depth</th>\n",
       "      <th>treatment.3</th>\n",
       "      <th>treatment</th>\n",
       "      <th>elev</th>\n",
       "      <th>samp_collect_device</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Universally unique ID (ex: IGSN, UUID)</td>\n",
       "      <td>cm</td>\n",
       "      <td>chem_administration</td>\n",
       "      <td>isotope_exposure</td>\n",
       "      <td>sieving</td>\n",
       "      <td>m</td>\n",
       "      <td>watering_regm</td>\n",
       "      <td>air_temp_regm</td>\n",
       "      <td>m</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Field REQUIRED for ALL sample submission. Opti...</td>\n",
       "      <td>depth-centimeters</td>\n",
       "      <td>treatment-chemical administration/addition</td>\n",
       "      <td>treatment-isotope exposure/addition</td>\n",
       "      <td>size fraction-sieve size</td>\n",
       "      <td>depth-meters</td>\n",
       "      <td>treatment- watering regimen/schedule</td>\n",
       "      <td>treatment-air temperature regimen</td>\n",
       "      <td>elevation-meters</td>\n",
       "      <td>sample collection device-slide hammer corer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{text}</td>\n",
       "      <td>{value} centimeters</td>\n",
       "      <td>{termLabel} {[termID]}; {timestamp}</td>\n",
       "      <td>{text}</td>\n",
       "      <td>{value}{units}</td>\n",
       "      <td>{value} meters</td>\n",
       "      <td>{float} {unit};{Rn/start_time/end_time/duration}</td>\n",
       "      <td>{float} {unit};{Rn/start_time/end_time/duration}</td>\n",
       "      <td>{value} meters</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  ID              depth.1  \\\n",
       "0             Universally unique ID (ex: IGSN, UUID)                   cm   \n",
       "1  Field REQUIRED for ALL sample submission. Opti...    depth-centimeters   \n",
       "2                                             {text}  {value} centimeters   \n",
       "\n",
       "                                  treatment.1  \\\n",
       "0                         chem_administration   \n",
       "1  treatment-chemical administration/addition   \n",
       "2         {termLabel} {[termID]}; {timestamp}   \n",
       "\n",
       "                           treatment.2                size_fract  \\\n",
       "0                     isotope_exposure                   sieving   \n",
       "1  treatment-isotope exposure/addition  size fraction-sieve size   \n",
       "2                               {text}            {value}{units}   \n",
       "\n",
       "            depth                                       treatment.3  \\\n",
       "0               m                                     watering_regm   \n",
       "1    depth-meters              treatment- watering regimen/schedule   \n",
       "2  {value} meters  {float} {unit};{Rn/start_time/end_time/duration}   \n",
       "\n",
       "                                          treatment              elev  \\\n",
       "0                                     air_temp_regm                 m   \n",
       "1                 treatment-air temperature regimen  elevation-meters   \n",
       "2  {float} {unit};{Rn/start_time/end_time/duration}    {value} meters   \n",
       "\n",
       "                           samp_collect_device  \n",
       "0                            slide_hammer_core  \n",
       "1  sample collection device-slide hammer corer  \n",
       "2                                       {text}  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blue_headers = template_metadata_headers[blue_cols]\n",
    "blue_headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "d6277ad0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_category</th>\n",
       "      <th>Variable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sediment</td>\n",
       "      <td>depth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>soil</td>\n",
       "      <td>depth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>water</td>\n",
       "      <td>depth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>elev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sediment</td>\n",
       "      <td>elev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>soil</td>\n",
       "      <td>elev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>water</td>\n",
       "      <td>elev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>host_common_name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>host_taxid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>plant_struc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sediment</td>\n",
       "      <td>samp_collect_device</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>soil</td>\n",
       "      <td>samp_collect_device</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>water</td>\n",
       "      <td>samp_collect_device</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sediment</td>\n",
       "      <td>size_fract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>soil</td>\n",
       "      <td>size_fract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>water</td>\n",
       "      <td>size_fract</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>plant_associated</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sediment</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>soil</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>water</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample_category             Variable\n",
       "6           sediment                depth\n",
       "1               soil                depth\n",
       "16             water                depth\n",
       "10  plant_associated                 elev\n",
       "5           sediment                 elev\n",
       "0               soil                 elev\n",
       "15             water                 elev\n",
       "13  plant_associated     host_common_name\n",
       "14  plant_associated           host_taxid\n",
       "11  plant_associated          plant_struc\n",
       "8           sediment  samp_collect_device\n",
       "3               soil  samp_collect_device\n",
       "18             water  samp_collect_device\n",
       "9           sediment           size_fract\n",
       "4               soil           size_fract\n",
       "19             water           size_fract\n",
       "12  plant_associated            treatment\n",
       "7           sediment            treatment\n",
       "2               soil            treatment\n",
       "17             water            treatment"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "package_to_blue_col.sort_values(by=[\"Variable\", \"Sample_category\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b9ae073",
   "metadata": {},
   "source": [
    "How to describe any patterns in the use of columns from the blue section? How to assess whether they are aligned with usage allowed by the MIxS schema?\n",
    "\n",
    "Which columns can appear more than once? Certainly `treatment`, because a sample could be exposed to temperature, moisture, chemical treatments, etc. `treatment` is not defined by MIxS, but MIxS slots do appear further down in the headers frame.\n",
    "\n",
    "There are *scalar* measurements with units like `depth` and `elevation`. Values appear per-row in the content frame. Units are constant per column and appear in the headers frame. It makes sense that multiple `depth` columns are allowed as they could take multiple different units. It appears that `elev` is also allowed to appear multiple times, although it's units are constrained to `m`\n",
    "\n",
    "It appears that the header rows for the `samp_collect_device` and `size_fract` columns are intended to capture something like subproperties, although those are not defined in MIxS. If this is the case, the row values would be string values for those subproperties. It it unclear whether those string values provide a level of detail that would be utilized in user searches. It does not appear that `samp_collect_device` and `size_fract` are allowed to appear multiple times.\n",
    "\n",
    "\n",
    "All of these conclusions are based on examination of the NMDC Sample Metadata Template's `MenuTerms` tab. This notebook provides methods, like `get_optics_coord_frame`, for finding relevant sections.\n",
    "\n",
    "Note that column titles can appear identically multiple times in the Google Sheet, but will be suffixed with `\\.\\d+` in a Pandas Dataframe.\n",
    "\n",
    "Not yet characterized. It looks like these can appear multiple times.\n",
    "\n",
    "- host_common_name\n",
    "- host_taxid\n",
    "- plant_struc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a78fb541",
   "metadata": {},
   "source": [
    "**Assuming all columns could be repeated**\n",
    "\n",
    "Generalized methods:\n",
    "- pivot_multi_columns\n",
    "- either\n",
    "    1. rename_melted_multi\n",
    "    1. concatenate_melted_multi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8983a333",
   "metadata": {},
   "source": [
    "### Which columns are duplicated in the current example template?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "6985d156",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['depth', 'treatment']"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dupe_blues = [col for col in blue_cols if \".\" in col]\n",
    "base_dupes = [re.sub(\"\\.\\d+$\", \"\", colname) for colname in dupe_blues]\n",
    "unique_dupes = list(set(base_dupes))\n",
    "\n",
    "# it's almost certain that there will be multiple teratmetn columns.\n",
    "# handle them seperately\n",
    "# unique_dupes.remove(\"treatment\")\n",
    "unique_dupes.sort()\n",
    "unique_dupes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cad4532",
   "metadata": {},
   "source": [
    "### `depth`\n",
    "\n",
    "`depth` and `elev` are scalars: elements with a value and a unit. I don' think they need or benefit from any special processing, but they will need validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "10a4485a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>variable</th>\n",
       "      <th>depth</th>\n",
       "      <th>units</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "      <td>depth.1</td>\n",
       "      <td>-20-30</td>\n",
       "      <td>cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "      <td>depth.1</td>\n",
       "      <td>-20-30</td>\n",
       "      <td>cm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "      <td>depth</td>\n",
       "      <td>0-1</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "      <td>depth</td>\n",
       "      <td>0-1</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "      <td>depth</td>\n",
       "      <td>0-1</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ID variable   depth units\n",
       "0  UUID:516394-970067-847398  depth.1  -20-30    cm\n",
       "1  UUID:472894-473947-642384  depth.1  -20-30    cm\n",
       "2  UUID:472894-473947-847398    depth     0-1     m\n",
       "3  UUID:684267-410686-971057    depth     0-1     m\n",
       "4  UUID:472894-473947-847396    depth     0-1     m"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_multi = pivot_multi_columns(\"depth\")\n",
    "depth_multi = depth_multi.rename(columns={\"value_x\": \"depth\", \"value_y\": \"units\"})\n",
    "depth_multi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "a05785c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check for duplicate IDs ?\n",
    "depth_row_list = depth_multi.to_dict(\"records\")\n",
    "depth_nmdc = [\n",
    "    {\n",
    "        \"id\": item[\"ID\"],\n",
    "        \"depth\": {\"has_raw_value\": item[\"depth\"], \"has_unit\": item[\"units\"]},\n",
    "    }\n",
    "    for item in depth_row_list\n",
    "]\n",
    "cumulative_json_merge = merge_j_on_id(cumulative_json_merge, depth_nmdc)\n",
    "# cumulative_json_merge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d39afe2",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f57722f9",
   "metadata": {},
   "source": [
    "### `elev`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "0885cb63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>variable</th>\n",
       "      <th>elev</th>\n",
       "      <th>units</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "      <td>elev</td>\n",
       "      <td>500.0</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "      <td>elev</td>\n",
       "      <td>500.0</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "      <td>elev</td>\n",
       "      <td>500.0</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "      <td>elev</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "      <td>elev</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ID variable    elev units\n",
       "0  UUID:472894-473947-847398     elev   500.0     m\n",
       "1  UUID:684267-410686-971057     elev   500.0     m\n",
       "2  UUID:472894-473947-847396     elev   500.0     m\n",
       "3  UUID:516394-970067-847398     elev  1000.0     m\n",
       "4  UUID:472894-473947-642384     elev  1000.0     m"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "elev_multi = pivot_multi_columns(\"elev\")\n",
    "elev_multi = elev_multi.rename(columns={\"value_x\": \"elev\", \"value_y\": \"units\"})\n",
    "elev_multi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "f40ae571",
   "metadata": {},
   "outputs": [],
   "source": [
    "elev_row_list = elev_multi.to_dict(\"records\")\n",
    "elev_nmdc = [\n",
    "    {\n",
    "        \"id\": item[\"ID\"],\n",
    "        \"elev\": {\"has_raw_value\": item[\"elev\"], \"has_unit\": item[\"units\"]},\n",
    "    }\n",
    "    for item in elev_row_list\n",
    "]\n",
    "# elev_nmdc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "16e92002",
   "metadata": {},
   "outputs": [],
   "source": [
    "cumulative_json_merge = merge_j_on_id(cumulative_json_merge, elev_nmdc)\n",
    "# cumulative_json_merge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e61d5530",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bc59f82",
   "metadata": {},
   "source": [
    "### `samp_collect_device`\n",
    "\n",
    "Double check spelling. See for example https://cmungall.github.io/mixs-source/samp_collec_device/#slot-samp_collec_device"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df26e5fe",
   "metadata": {},
   "source": [
    "No illustrative mongodb hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "fe3dc3df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "coll = db[\"biosample_set\"]\n",
    "any_study = coll.find_one({\"samp_collec_device\": {\"$exists\": True}})\n",
    "pprint(any_study)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd854169",
   "metadata": {},
   "source": [
    "_Counter example of a successful search_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "bcda2f85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'_id': ObjectId('60e840cde9822b255ad942ee'),\n",
      " 'add_date': '2015-05-28',\n",
      " 'collection_date': {'has_raw_value': '2014-09-23'},\n",
      " 'community': 'microbial communities',\n",
      " 'depth': {'has_numeric_value': 0.5,\n",
      "           'has_raw_value': '0.5',\n",
      "           'has_unit': 'meter'},\n",
      " 'depth2': {'has_numeric_value': 1.0,\n",
      "            'has_raw_value': '1.0',\n",
      "            'has_unit': 'meter'},\n",
      " 'description': 'Sterilized sand packs were incubated back in the ground and '\n",
      "                'collected at time point T2.',\n",
      " 'ecosystem': 'Engineered',\n",
      " 'ecosystem_category': 'Artificial ecosystem',\n",
      " 'ecosystem_subtype': 'Unclassified',\n",
      " 'ecosystem_type': 'Sand microcosm',\n",
      " 'env_broad_scale': {'has_raw_value': 'ENVO:01000253'},\n",
      " 'env_local_scale': {'has_raw_value': 'ENVO:01000621'},\n",
      " 'env_medium': {'has_raw_value': 'ENVO:01000017'},\n",
      " 'geo_loc_name': {'has_raw_value': 'USA: Columbia River, Washington'},\n",
      " 'habitat': 'sand microcosm',\n",
      " 'id': 'gold:Gb0115217',\n",
      " 'identifier': 'GW-RW T2_23-Sept-14',\n",
      " 'lat_lon': {'has_raw_value': '46.37228379 -119.2717467',\n",
      "             'latitude': 46.37228379,\n",
      "             'longitude': -119.2717467},\n",
      " 'location': 'groundwater-surface water interaction zone in Washington, USA',\n",
      " 'mod_date': '2021-06-17',\n",
      " 'name': 'Sand microcosm microbial communities from a hyporheic zone in '\n",
      "         'Columbia River, Washington, USA - GW-RW T2_23-Sept-14',\n",
      " 'ncbi_taxonomy_name': 'sediment metagenome',\n",
      " 'part_of': ['gold:Gs0114663'],\n",
      " 'sample_collection_site': 'sand microcosm',\n",
      " 'specific_ecosystem': 'Unclassified',\n",
      " 'type': 'nmdc:Biosample'}\n"
     ]
    }
   ],
   "source": [
    "coll = db[\"biosample_set\"]\n",
    "any_study = coll.find_one({\"depth\": {\"$exists\": True}})\n",
    "pprint(any_study)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70ab5733",
   "metadata": {},
   "source": [
    "#### How does MIxS define `samp_collec_device`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "329b0e59",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "name: samp_collec_device\n",
      "definition_uri: https://w3id.org/mixs/vocab/samp_collec_device\n",
      "aliases:\n",
      "- sample collection device\n",
      "mappings:\n",
      "- MIXS:0000002\n",
      "description: The device used to collect an environmental sample. This field accepts\n",
      "  terms listed under environmental sampling device (http://purl.obolibrary.org/obo/ENVO).\n",
      "  This field also accepts terms listed under specimen collection device (http://purl.obolibrary.org/obo/GENEPIO_0002094).\n",
      "comments:\n",
      "- 'Expected value: device name'\n",
      "- 'Position: 31.0'\n",
      "examples:\n",
      "- value: swab, biopsy, niskin bottle, push core, drag swab [GENEPIO:0002713]\n",
      "from_schema: http://w3id.org/mixs/terms\n",
      "exact_mappings:\n",
      "- 'MIGS:'\n",
      "is_a: nucleic acid sequence source field\n",
      "range: string\n",
      "slot_uri: MIXS:0000002\n",
      "owner: core\n",
      "domain_of:\n",
      "- MIGS eukaryote\n",
      "- MIGS bacteria\n",
      "- MIGS plant\n",
      "- MIGS virus\n",
      "- MIGS org\n",
      "- ME\n",
      "- MIMARKS specimen\n",
      "- MIMARKS survey\n",
      "- MISAG\n",
      "- MIMAG\n",
      "- MIUVIG\n",
      "- core\n",
      "pattern: '{termLabel} {[termID]}|{text}'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml_dumper.dumps(mixs_schema_resolved.slots[\"samp_collec_device\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d52e69da",
   "metadata": {},
   "source": [
    "#### What do the `samp_collect_device` MenuTerms in the template look like\n",
    "\n",
    "This approach is elaborated under treatments below\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "d071532c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>sample collection device-shovel</td>\n",
       "      <td>shovel</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>sample collection device-slide hammer corer</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>sample collection device- deep corer</td>\n",
       "      <td>deep_corer</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>sample collection device-russian corer</td>\n",
       "      <td>russian_corer</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>sample collection device-sipper</td>\n",
       "      <td>sipper</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>sample collection device-other</td>\n",
       "      <td>other-samp_collect_device</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      0                                            1  \\\n",
       "37  samp_collect_device              sample collection device-shovel   \n",
       "38  samp_collect_device  sample collection device-slide hammer corer   \n",
       "39  samp_collect_device         sample collection device- deep corer   \n",
       "40  samp_collect_device       sample collection device-russian corer   \n",
       "41  samp_collect_device              sample collection device-sipper   \n",
       "42  samp_collect_device               sample collection device-other   \n",
       "\n",
       "                            2       3  \n",
       "37                     shovel  {text}  \n",
       "38          slide_hammer_core  {text}  \n",
       "39                 deep_corer  {text}  \n",
       "40              russian_corer  {text}  \n",
       "41                     sipper  {text}  \n",
       "42  other-samp_collect_device  {text}  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scd_coord_frame = get_optics_coord_frame(template_menu_terms, \"samp_collect_device\")\n",
    "scd_cluster_sizes = scd_coord_frame[\"cluster\"].value_counts()\n",
    "scd_largest_index = scd_cluster_sizes.index[0]\n",
    "template_scd_menu_terms = get_data_from_extended_cluster_coords(\n",
    "    template_menu_terms, scd_coord_frame, scd_largest_index, additional_columns=3\n",
    ")\n",
    "\n",
    "template_scd_menu_terms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c667a13a",
   "metadata": {},
   "source": [
    "####  Still not clear on whether to expect multi-columns for  `samp_collect_device`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "06398872",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>variable</th>\n",
       "      <th>value_x</th>\n",
       "      <th>value_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>11/2' x 6\"; stainless stee</td>\n",
       "      <td>slide_hammer_core</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ID             variable                     value_x  \\\n",
       "0  UUID:472894-473947-847398  samp_collect_device  11/2' x 6\"; stainless stee   \n",
       "1  UUID:684267-410686-971057  samp_collect_device  11/2' x 6\"; stainless stee   \n",
       "2  UUID:472894-473947-847396  samp_collect_device  11/2' x 6\"; stainless stee   \n",
       "3  UUID:516394-970067-847398  samp_collect_device  11/2' x 6\"; stainless stee   \n",
       "4  UUID:472894-473947-642384  samp_collect_device  11/2' x 6\"; stainless stee   \n",
       "\n",
       "             value_y  \n",
       "0  slide_hammer_core  \n",
       "1  slide_hammer_core  \n",
       "2  slide_hammer_core  \n",
       "3  slide_hammer_core  \n",
       "4  slide_hammer_core  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scd_multi = pivot_multi_columns(\"samp_collect_device\")\n",
    "scd_multi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "b5212519",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>variable</th>\n",
       "      <th>samp_collect_device</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11/2' x 6\"; stainless stee slide_hammer_core</td>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11/2' x 6\"; stainless stee slide_hammer_core</td>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11/2' x 6\"; stainless stee slide_hammer_core</td>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11/2' x 6\"; stainless stee slide_hammer_core</td>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11/2' x 6\"; stainless stee slide_hammer_core</td>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "variable                           samp_collect_device  \\\n",
       "0         11/2' x 6\"; stainless stee slide_hammer_core   \n",
       "1         11/2' x 6\"; stainless stee slide_hammer_core   \n",
       "2         11/2' x 6\"; stainless stee slide_hammer_core   \n",
       "3         11/2' x 6\"; stainless stee slide_hammer_core   \n",
       "4         11/2' x 6\"; stainless stee slide_hammer_core   \n",
       "\n",
       "variable                         id  \n",
       "0         UUID:472894-473947-642384  \n",
       "1         UUID:472894-473947-847396  \n",
       "2         UUID:472894-473947-847398  \n",
       "3         UUID:516394-970067-847398  \n",
       "4         UUID:684267-410686-971057  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scd_multi_pivot = concatenate_melted_multi(scd_multi, \"samp_collect_device\")\n",
    "scd_multi_pivot.rename(columns={\"samp_collect_device\": \"samp_collec_device\"})\n",
    "scd_multi_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "bedc5429",
   "metadata": {},
   "outputs": [],
   "source": [
    "scd_row_list = scd_multi_pivot.to_dict(\"records\")\n",
    "# pprint(scd_row_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "258945e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cumulative_json_merge = merge_j_on_id(cumulative_json_merge, scd_row_list)\n",
    "# cumulative_json_merge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24d37121",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "### `size_fract`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8f6a06b",
   "metadata": {},
   "source": [
    "see https://cmungall.github.io/mixs-source/size_frac/\n",
    "\n",
    "No mongodb hits\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "df5736e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "coll = db[\"biosample_set\"]\n",
    "any_study = coll.find_one({\"size_frac\": {\"$exists\": True}})\n",
    "pprint(any_study)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "876222a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sf_coord_frame = get_optics_coord_frame(template_menu_terms, \"size_fract\")\n",
    "# sf_coord_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "c174dff3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       " 0    4\n",
       " 1    3\n",
       "-1    1\n",
       "Name: cluster, dtype: int64"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sf_cluster_sizes = sf_coord_frame[\"cluster\"].value_counts()\n",
    "sf_cluster_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "12f5c500",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "sf_largest_index = sf_cluster_sizes.index[0]\n",
    "print(sf_largest_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "d470dc22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>size_fract</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>size_fract</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>select</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>elev</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>depth</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>treatment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>size_fract</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>select</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>elev</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>plant_struc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>treatment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>host_common_name</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>host_taxid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>select</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>elev</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>depth</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>treatment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>samp_collect_device</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>size_fract</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      1    2    3           4    5    6    7\n",
       "7   samp_collect_device  NaN  NaN  size_fract  NaN  NaN  NaN\n",
       "8            size_fract  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "9                select  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "10                 elev  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "11                depth  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "12            treatment  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "13  samp_collect_device  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "14           size_fract  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "15               select  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "16                 elev  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "17          plant_struc  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "18            treatment  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "19     host_common_name  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "20           host_taxid  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "21               select  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "22                 elev  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "23                depth  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "24            treatment  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "25  samp_collect_device  NaN  NaN         NaN  NaN  NaN  NaN\n",
       "26           size_fract  NaN  NaN         NaN  NaN  NaN  NaN"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_sf_menu_terms = get_data_from_extended_cluster_coords(\n",
    "    template_menu_terms, sf_coord_frame, sf_largest_index\n",
    ")\n",
    "template_sf_menu_terms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9800691e",
   "metadata": {},
   "source": [
    "\"size_frac\" does appear several times in that cluster, but  the rest of the cells don't do much to elaborate on \"size_frac\". Try the next largest cluster, number 1.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "d9d66aca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>size_fract</td>\n",
       "      <td>size fraction-filter size</td>\n",
       "      <td>filter_size</td>\n",
       "      <td>{value}{units}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>size_fract</td>\n",
       "      <td>size fraction-filter type</td>\n",
       "      <td>filter_type</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>size_fract</td>\n",
       "      <td>size fraction-sieve size</td>\n",
       "      <td>sieving</td>\n",
       "      <td>{value}{units}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0                          1            2               3\n",
       "52  size_fract  size fraction-filter size  filter_size  {value}{units}\n",
       "53  size_fract  size fraction-filter type  filter_type          {text}\n",
       "54  size_fract   size fraction-sieve size      sieving  {value}{units}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_sf_menu_terms = get_data_from_extended_cluster_coords(\n",
    "    template_menu_terms, sf_coord_frame, 1, additional_columns=3\n",
    ")\n",
    "\n",
    "template_sf_menu_terms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3017ec7",
   "metadata": {},
   "source": [
    "#### Of the values in column 2, plus \"size_frac\", which are defined as MIxS slots?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "7d99a268",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ME_adapters',\n",
       " 'ME_annot',\n",
       " 'ME_assembly_name',\n",
       " 'ME_assembly_qual',\n",
       " 'ME_assembly_software',\n",
       " 'ME_associated resource',\n",
       " 'ME_collection_date',\n",
       " 'ME_env_broad_scale',\n",
       " 'ME_env_local_scale',\n",
       " 'ME_env_medium',\n",
       " 'ME_env_package',\n",
       " 'ME_experimental_factor',\n",
       " 'ME_feat_pred',\n",
       " 'ME_geo_loc_name',\n",
       " 'ME_investigation_type',\n",
       " 'ME_lat_lon',\n",
       " 'ME_lib_layout',\n",
       " 'ME_lib_reads_seqd',\n",
       " 'ME_lib_screen',\n",
       " 'ME_lib_size',\n",
       " 'ME_lib_vector',\n",
       " 'ME_mid',\n",
       " 'ME_nucl_acid_amp',\n",
       " 'ME_nucl_acid_ext',\n",
       " 'ME_number_contig',\n",
       " 'ME_project_name',\n",
       " 'ME_ref_biomaterial',\n",
       " 'ME_ref_db',\n",
       " 'ME_rel_to_oxygen',\n",
       " 'ME_samp_collec_device',\n",
       " 'ME_samp_collec_method',\n",
       " 'ME_samp_mat_process',\n",
       " 'ME_samp_name',\n",
       " 'ME_samp_size',\n",
       " 'ME_samp_taxon_id',\n",
       " 'ME_samp_vol_we_dna_ext',\n",
       " 'ME_seq_meth',\n",
       " 'ME_sim_search_meth',\n",
       " 'ME_size_frac',\n",
       " 'ME_sop',\n",
       " 'ME_source_mat_id',\n",
       " 'ME_submitted_to_insdc',\n",
       " 'ME_tax_class',\n",
       " 'MIGS bacteria_adapters',\n",
       " 'MIGS bacteria_annot',\n",
       " 'MIGS bacteria_assembly_name',\n",
       " 'MIGS bacteria_assembly_qual',\n",
       " 'MIGS bacteria_assembly_software',\n",
       " 'MIGS bacteria_associated resource',\n",
       " 'MIGS bacteria_biotic_relationship',\n",
       " 'MIGS bacteria_collection_date',\n",
       " 'MIGS bacteria_compl_score',\n",
       " 'MIGS bacteria_compl_software',\n",
       " 'MIGS bacteria_encoded_traits',\n",
       " 'MIGS bacteria_env_broad_scale',\n",
       " 'MIGS bacteria_env_local_scale',\n",
       " 'MIGS bacteria_env_medium',\n",
       " 'MIGS bacteria_env_package',\n",
       " 'MIGS bacteria_estimated_size',\n",
       " 'MIGS bacteria_experimental_factor',\n",
       " 'MIGS bacteria_extrachrom_elements',\n",
       " 'MIGS bacteria_feat_pred',\n",
       " 'MIGS bacteria_geo_loc_name',\n",
       " 'MIGS bacteria_health_disease_stat',\n",
       " 'MIGS bacteria_host_disease_stat',\n",
       " 'MIGS bacteria_host_spec_range',\n",
       " 'MIGS bacteria_investigation_type',\n",
       " 'MIGS bacteria_isol_growth_condt',\n",
       " 'MIGS bacteria_lat_lon',\n",
       " 'MIGS bacteria_lib_layout',\n",
       " 'MIGS bacteria_lib_reads_seqd',\n",
       " 'MIGS bacteria_lib_screen',\n",
       " 'MIGS bacteria_lib_size',\n",
       " 'MIGS bacteria_lib_vector',\n",
       " 'MIGS bacteria_nucl_acid_amp',\n",
       " 'MIGS bacteria_nucl_acid_ext',\n",
       " 'MIGS bacteria_num_replicons',\n",
       " 'MIGS bacteria_number_contig',\n",
       " 'MIGS bacteria_pathogenicity',\n",
       " 'MIGS bacteria_project_name',\n",
       " 'MIGS bacteria_ref_biomaterial',\n",
       " 'MIGS bacteria_ref_db',\n",
       " 'MIGS bacteria_rel_to_oxygen',\n",
       " 'MIGS bacteria_samp_collec_device',\n",
       " 'MIGS bacteria_samp_collec_method',\n",
       " 'MIGS bacteria_samp_mat_process',\n",
       " 'MIGS bacteria_samp_name',\n",
       " 'MIGS bacteria_samp_size',\n",
       " 'MIGS bacteria_samp_taxon_id',\n",
       " 'MIGS bacteria_samp_vol_we_dna_ext',\n",
       " 'MIGS bacteria_seq_meth',\n",
       " 'MIGS bacteria_sim_search_meth',\n",
       " 'MIGS bacteria_sop',\n",
       " 'MIGS bacteria_source_mat_id',\n",
       " 'MIGS bacteria_specific_host',\n",
       " 'MIGS bacteria_submitted_to_insdc',\n",
       " 'MIGS bacteria_subspecf_gen_lin',\n",
       " 'MIGS bacteria_tax_class',\n",
       " 'MIGS bacteria_tax_ident',\n",
       " 'MIGS bacteria_trophic_level',\n",
       " 'MIGS eukaryote_adapters',\n",
       " 'MIGS eukaryote_annot',\n",
       " 'MIGS eukaryote_assembly_name',\n",
       " 'MIGS eukaryote_assembly_qual',\n",
       " 'MIGS eukaryote_assembly_software',\n",
       " 'MIGS eukaryote_associated resource',\n",
       " 'MIGS eukaryote_biotic_relationship',\n",
       " 'MIGS eukaryote_collection_date',\n",
       " 'MIGS eukaryote_compl_score',\n",
       " 'MIGS eukaryote_compl_software',\n",
       " 'MIGS eukaryote_env_broad_scale',\n",
       " 'MIGS eukaryote_env_local_scale',\n",
       " 'MIGS eukaryote_env_medium',\n",
       " 'MIGS eukaryote_env_package',\n",
       " 'MIGS eukaryote_estimated_size',\n",
       " 'MIGS eukaryote_experimental_factor',\n",
       " 'MIGS eukaryote_extrachrom_elements',\n",
       " 'MIGS eukaryote_feat_pred',\n",
       " 'MIGS eukaryote_geo_loc_name',\n",
       " 'MIGS eukaryote_health_disease_stat',\n",
       " 'MIGS eukaryote_host_disease_stat',\n",
       " 'MIGS eukaryote_host_spec_range',\n",
       " 'MIGS eukaryote_investigation_type',\n",
       " 'MIGS eukaryote_isol_growth_condt',\n",
       " 'MIGS eukaryote_lat_lon',\n",
       " 'MIGS eukaryote_lib_layout',\n",
       " 'MIGS eukaryote_lib_reads_seqd',\n",
       " 'MIGS eukaryote_lib_screen',\n",
       " 'MIGS eukaryote_lib_size',\n",
       " 'MIGS eukaryote_lib_vector',\n",
       " 'MIGS eukaryote_nucl_acid_amp',\n",
       " 'MIGS eukaryote_nucl_acid_ext',\n",
       " 'MIGS eukaryote_num_replicons',\n",
       " 'MIGS eukaryote_number_contig',\n",
       " 'MIGS eukaryote_pathogenicity',\n",
       " 'MIGS eukaryote_ploidy',\n",
       " 'MIGS eukaryote_project_name',\n",
       " 'MIGS eukaryote_propagation',\n",
       " 'MIGS eukaryote_ref_biomaterial',\n",
       " 'MIGS eukaryote_ref_db',\n",
       " 'MIGS eukaryote_samp_collec_device',\n",
       " 'MIGS eukaryote_samp_collec_method',\n",
       " 'MIGS eukaryote_samp_mat_process',\n",
       " 'MIGS eukaryote_samp_name',\n",
       " 'MIGS eukaryote_samp_size',\n",
       " 'MIGS eukaryote_samp_taxon_id',\n",
       " 'MIGS eukaryote_samp_vol_we_dna_ext',\n",
       " 'MIGS eukaryote_seq_meth',\n",
       " 'MIGS eukaryote_sim_search_meth',\n",
       " 'MIGS eukaryote_sop',\n",
       " 'MIGS eukaryote_source_mat_id',\n",
       " 'MIGS eukaryote_specific_host',\n",
       " 'MIGS eukaryote_submitted_to_insdc',\n",
       " 'MIGS eukaryote_subspecf_gen_lin',\n",
       " 'MIGS eukaryote_tax_class',\n",
       " 'MIGS eukaryote_tax_ident',\n",
       " 'MIGS eukaryote_trophic_level',\n",
       " 'MIGS org_adapters',\n",
       " 'MIGS org_annot',\n",
       " 'MIGS org_assembly_name',\n",
       " 'MIGS org_assembly_qual',\n",
       " 'MIGS org_assembly_software',\n",
       " 'MIGS org_associated resource',\n",
       " 'MIGS org_collection_date',\n",
       " 'MIGS org_compl_score',\n",
       " 'MIGS org_compl_software',\n",
       " 'MIGS org_env_broad_scale',\n",
       " 'MIGS org_env_local_scale',\n",
       " 'MIGS org_env_medium',\n",
       " 'MIGS org_env_package',\n",
       " 'MIGS org_estimated_size',\n",
       " 'MIGS org_experimental_factor',\n",
       " 'MIGS org_extrachrom_elements',\n",
       " 'MIGS org_feat_pred',\n",
       " 'MIGS org_geo_loc_name',\n",
       " 'MIGS org_investigation_type',\n",
       " 'MIGS org_isol_growth_condt',\n",
       " 'MIGS org_lat_lon',\n",
       " 'MIGS org_lib_layout',\n",
       " 'MIGS org_lib_reads_seqd',\n",
       " 'MIGS org_lib_screen',\n",
       " 'MIGS org_lib_size',\n",
       " 'MIGS org_lib_vector',\n",
       " 'MIGS org_nucl_acid_amp',\n",
       " 'MIGS org_nucl_acid_ext',\n",
       " 'MIGS org_number_contig',\n",
       " 'MIGS org_project_name',\n",
       " 'MIGS org_ref_biomaterial',\n",
       " 'MIGS org_ref_db',\n",
       " 'MIGS org_samp_collec_device',\n",
       " 'MIGS org_samp_collec_method',\n",
       " 'MIGS org_samp_mat_process',\n",
       " 'MIGS org_samp_name',\n",
       " 'MIGS org_samp_size',\n",
       " 'MIGS org_samp_taxon_id',\n",
       " 'MIGS org_samp_vol_we_dna_ext',\n",
       " 'MIGS org_seq_meth',\n",
       " 'MIGS org_sim_search_meth',\n",
       " 'MIGS org_sop',\n",
       " 'MIGS org_source_mat_id',\n",
       " 'MIGS org_submitted_to_insdc',\n",
       " 'MIGS org_subspecf_gen_lin',\n",
       " 'MIGS org_tax_class',\n",
       " 'MIGS org_tax_ident',\n",
       " 'MIGS plant_adapters',\n",
       " 'MIGS plant_annot',\n",
       " 'MIGS plant_assembly_name',\n",
       " 'MIGS plant_assembly_qual',\n",
       " 'MIGS plant_assembly_software',\n",
       " 'MIGS plant_associated resource',\n",
       " 'MIGS plant_collection_date',\n",
       " 'MIGS plant_compl_score',\n",
       " 'MIGS plant_compl_software',\n",
       " 'MIGS plant_encoded_traits',\n",
       " 'MIGS plant_env_broad_scale',\n",
       " 'MIGS plant_env_local_scale',\n",
       " 'MIGS plant_env_medium',\n",
       " 'MIGS plant_env_package',\n",
       " 'MIGS plant_estimated_size',\n",
       " 'MIGS plant_experimental_factor',\n",
       " 'MIGS plant_feat_pred',\n",
       " 'MIGS plant_geo_loc_name',\n",
       " 'MIGS plant_host_spec_range',\n",
       " 'MIGS plant_investigation_type',\n",
       " 'MIGS plant_isol_growth_condt',\n",
       " 'MIGS plant_lat_lon',\n",
       " 'MIGS plant_lib_layout',\n",
       " 'MIGS plant_lib_reads_seqd',\n",
       " 'MIGS plant_lib_screen',\n",
       " 'MIGS plant_lib_size',\n",
       " 'MIGS plant_lib_vector',\n",
       " 'MIGS plant_nucl_acid_amp',\n",
       " 'MIGS plant_nucl_acid_ext',\n",
       " 'MIGS plant_number_contig',\n",
       " 'MIGS plant_project_name',\n",
       " 'MIGS plant_propagation',\n",
       " 'MIGS plant_ref_biomaterial',\n",
       " 'MIGS plant_ref_db',\n",
       " 'MIGS plant_samp_collec_device',\n",
       " 'MIGS plant_samp_collec_method',\n",
       " 'MIGS plant_samp_mat_process',\n",
       " 'MIGS plant_samp_name',\n",
       " 'MIGS plant_samp_size',\n",
       " 'MIGS plant_samp_taxon_id',\n",
       " 'MIGS plant_samp_vol_we_dna_ext',\n",
       " 'MIGS plant_seq_meth',\n",
       " 'MIGS plant_sim_search_meth',\n",
       " 'MIGS plant_sop',\n",
       " 'MIGS plant_source_mat_id',\n",
       " 'MIGS plant_specific_host',\n",
       " 'MIGS plant_submitted_to_insdc',\n",
       " 'MIGS plant_subspecf_gen_lin',\n",
       " 'MIGS plant_tax_class',\n",
       " 'MIGS plant_tax_ident',\n",
       " 'MIGS virus_adapters',\n",
       " 'MIGS virus_annot',\n",
       " 'MIGS virus_assembly_name',\n",
       " 'MIGS virus_assembly_qual',\n",
       " 'MIGS virus_assembly_software',\n",
       " 'MIGS virus_associated resource',\n",
       " 'MIGS virus_biotic_relationship',\n",
       " 'MIGS virus_collection_date',\n",
       " 'MIGS virus_compl_score',\n",
       " 'MIGS virus_compl_software',\n",
       " 'MIGS virus_encoded_traits',\n",
       " 'MIGS virus_env_broad_scale',\n",
       " 'MIGS virus_env_local_scale',\n",
       " 'MIGS virus_env_medium',\n",
       " 'MIGS virus_env_package',\n",
       " 'MIGS virus_estimated_size',\n",
       " 'MIGS virus_experimental_factor',\n",
       " 'MIGS virus_feat_pred',\n",
       " 'MIGS virus_geo_loc_name',\n",
       " 'MIGS virus_health_disease_stat',\n",
       " 'MIGS virus_host_disease_stat',\n",
       " 'MIGS virus_host_spec_range',\n",
       " 'MIGS virus_investigation_type',\n",
       " 'MIGS virus_isol_growth_condt',\n",
       " 'MIGS virus_lat_lon',\n",
       " 'MIGS virus_lib_layout',\n",
       " 'MIGS virus_lib_reads_seqd',\n",
       " 'MIGS virus_lib_screen',\n",
       " 'MIGS virus_lib_size',\n",
       " 'MIGS virus_lib_vector',\n",
       " 'MIGS virus_nucl_acid_amp',\n",
       " 'MIGS virus_nucl_acid_ext',\n",
       " 'MIGS virus_num_replicons',\n",
       " 'MIGS virus_number_contig',\n",
       " 'MIGS virus_pathogenicity',\n",
       " 'MIGS virus_project_name',\n",
       " 'MIGS virus_propagation',\n",
       " 'MIGS virus_ref_biomaterial',\n",
       " 'MIGS virus_ref_db',\n",
       " 'MIGS virus_samp_collec_device',\n",
       " 'MIGS virus_samp_collec_method',\n",
       " 'MIGS virus_samp_mat_process',\n",
       " 'MIGS virus_samp_name',\n",
       " 'MIGS virus_samp_size',\n",
       " 'MIGS virus_samp_taxon_id',\n",
       " 'MIGS virus_samp_vol_we_dna_ext',\n",
       " 'MIGS virus_seq_meth',\n",
       " 'MIGS virus_sim_search_meth',\n",
       " 'MIGS virus_sop',\n",
       " 'MIGS virus_source_mat_id',\n",
       " 'MIGS virus_specific_host',\n",
       " 'MIGS virus_submitted_to_insdc',\n",
       " 'MIGS virus_subspecf_gen_lin',\n",
       " 'MIGS virus_tax_class',\n",
       " 'MIGS virus_tax_ident',\n",
       " 'MIGS virus_virus_enrich_appr',\n",
       " 'MIMAG_adapters',\n",
       " 'MIMAG_annot',\n",
       " 'MIMAG_assembly_name',\n",
       " 'MIMAG_assembly_qual',\n",
       " 'MIMAG_assembly_software',\n",
       " 'MIMAG_associated resource',\n",
       " 'MIMAG_bin_param',\n",
       " 'MIMAG_bin_software',\n",
       " 'MIMAG_collection_date',\n",
       " 'MIMAG_compl_appr',\n",
       " 'MIMAG_compl_score',\n",
       " 'MIMAG_compl_software',\n",
       " 'MIMAG_contam_score',\n",
       " 'MIMAG_contam_screen_input',\n",
       " 'MIMAG_contam_screen_param',\n",
       " 'MIMAG_decontam_software',\n",
       " 'MIMAG_env_broad_scale',\n",
       " 'MIMAG_env_local_scale',\n",
       " 'MIMAG_env_medium',\n",
       " 'MIMAG_env_package',\n",
       " 'MIMAG_experimental_factor',\n",
       " 'MIMAG_feat_pred',\n",
       " 'MIMAG_geo_loc_name',\n",
       " 'MIMAG_investigation_type',\n",
       " 'MIMAG_lat_lon',\n",
       " 'MIMAG_lib_layout',\n",
       " 'MIMAG_lib_reads_seqd',\n",
       " 'MIMAG_lib_screen',\n",
       " 'MIMAG_lib_size',\n",
       " 'MIMAG_lib_vector',\n",
       " 'MIMAG_mag_cov_software',\n",
       " 'MIMAG_mid',\n",
       " 'MIMAG_nucl_acid_amp',\n",
       " 'MIMAG_nucl_acid_ext',\n",
       " 'MIMAG_number_contig',\n",
       " 'MIMAG_project_name',\n",
       " 'MIMAG_reassembly_bin',\n",
       " 'MIMAG_ref_biomaterial',\n",
       " 'MIMAG_ref_db',\n",
       " 'MIMAG_rel_to_oxygen',\n",
       " 'MIMAG_samp_collec_device',\n",
       " 'MIMAG_samp_collec_method',\n",
       " 'MIMAG_samp_mat_process',\n",
       " 'MIMAG_samp_name',\n",
       " 'MIMAG_samp_size',\n",
       " 'MIMAG_samp_taxon_id',\n",
       " 'MIMAG_samp_vol_we_dna_ext',\n",
       " 'MIMAG_seq_meth',\n",
       " 'MIMAG_sim_search_meth',\n",
       " 'MIMAG_size_frac',\n",
       " 'MIMAG_sop',\n",
       " 'MIMAG_source_mat_id',\n",
       " 'MIMAG_submitted_to_insdc',\n",
       " 'MIMAG_tax_class',\n",
       " 'MIMAG_tax_ident',\n",
       " 'MIMAG_trna_ext_software',\n",
       " 'MIMAG_trnas',\n",
       " 'MIMAG_x_16s_recover',\n",
       " 'MIMARKS specimen_adapters',\n",
       " 'MIMARKS specimen_assembly_software',\n",
       " 'MIMARKS specimen_associated resource',\n",
       " 'MIMARKS specimen_chimera_check',\n",
       " 'MIMARKS specimen_collection_date',\n",
       " 'MIMARKS specimen_env_broad_scale',\n",
       " 'MIMARKS specimen_env_local_scale',\n",
       " 'MIMARKS specimen_env_medium',\n",
       " 'MIMARKS specimen_env_package',\n",
       " 'MIMARKS specimen_experimental_factor',\n",
       " 'MIMARKS specimen_geo_loc_name',\n",
       " 'MIMARKS specimen_investigation_type',\n",
       " 'MIMARKS specimen_lat_lon',\n",
       " 'MIMARKS specimen_lib_layout',\n",
       " 'MIMARKS specimen_lib_reads_seqd',\n",
       " 'MIMARKS specimen_lib_screen',\n",
       " 'MIMARKS specimen_lib_size',\n",
       " 'MIMARKS specimen_lib_vector',\n",
       " 'MIMARKS specimen_mid',\n",
       " 'MIMARKS specimen_nucl_acid_amp',\n",
       " 'MIMARKS specimen_nucl_acid_ext',\n",
       " 'MIMARKS specimen_pcr_cond',\n",
       " 'MIMARKS specimen_pcr_primers',\n",
       " 'MIMARKS specimen_project_name',\n",
       " 'MIMARKS specimen_rel_to_oxygen',\n",
       " 'MIMARKS specimen_samp_collec_device',\n",
       " 'MIMARKS specimen_samp_collec_method',\n",
       " 'MIMARKS specimen_samp_mat_process',\n",
       " 'MIMARKS specimen_samp_name',\n",
       " 'MIMARKS specimen_samp_size',\n",
       " 'MIMARKS specimen_samp_taxon_id',\n",
       " 'MIMARKS specimen_samp_vol_we_dna_ext',\n",
       " 'MIMARKS specimen_seq_meth',\n",
       " 'MIMARKS specimen_seq_quality_check',\n",
       " 'MIMARKS specimen_size_frac',\n",
       " 'MIMARKS specimen_sop',\n",
       " 'MIMARKS specimen_source_mat_id',\n",
       " 'MIMARKS specimen_submitted_to_insdc',\n",
       " 'MIMARKS specimen_target_gene',\n",
       " 'MIMARKS specimen_target_subfragment',\n",
       " 'MIMARKS survey_associated resource',\n",
       " 'MIMARKS survey_biotic_relationship',\n",
       " 'MIMARKS survey_chimera_check',\n",
       " 'MIMARKS survey_collection_date',\n",
       " 'MIMARKS survey_env_broad_scale',\n",
       " 'MIMARKS survey_env_local_scale',\n",
       " 'MIMARKS survey_env_medium',\n",
       " 'MIMARKS survey_env_package',\n",
       " 'MIMARKS survey_experimental_factor',\n",
       " 'MIMARKS survey_extrachrom_elements',\n",
       " 'MIMARKS survey_geo_loc_name',\n",
       " 'MIMARKS survey_investigation_type',\n",
       " 'MIMARKS survey_isol_growth_condt',\n",
       " 'MIMARKS survey_lat_lon',\n",
       " 'MIMARKS survey_nucl_acid_amp',\n",
       " 'MIMARKS survey_nucl_acid_ext',\n",
       " 'MIMARKS survey_pcr_cond',\n",
       " 'MIMARKS survey_pcr_primers',\n",
       " 'MIMARKS survey_project_name',\n",
       " 'MIMARKS survey_rel_to_oxygen',\n",
       " 'MIMARKS survey_samp_collec_device',\n",
       " 'MIMARKS survey_samp_collec_method',\n",
       " 'MIMARKS survey_samp_mat_process',\n",
       " 'MIMARKS survey_samp_name',\n",
       " 'MIMARKS survey_samp_size',\n",
       " 'MIMARKS survey_samp_taxon_id',\n",
       " 'MIMARKS survey_samp_vol_we_dna_ext',\n",
       " 'MIMARKS survey_seq_meth',\n",
       " 'MIMARKS survey_seq_quality_check',\n",
       " 'MIMARKS survey_sop',\n",
       " 'MIMARKS survey_source_mat_id',\n",
       " 'MIMARKS survey_submitted_to_insdc',\n",
       " 'MIMARKS survey_subspecf_gen_lin',\n",
       " 'MIMARKS survey_target_gene',\n",
       " 'MIMARKS survey_target_subfragment',\n",
       " 'MIMARKS survey_trophic_level',\n",
       " 'MISAG_adapters',\n",
       " 'MISAG_annot',\n",
       " 'MISAG_assembly_name',\n",
       " 'MISAG_assembly_qual',\n",
       " 'MISAG_assembly_software',\n",
       " 'MISAG_associated resource',\n",
       " 'MISAG_collection_date',\n",
       " 'MISAG_compl_appr',\n",
       " 'MISAG_compl_score',\n",
       " 'MISAG_compl_software',\n",
       " 'MISAG_contam_score',\n",
       " 'MISAG_contam_screen_input',\n",
       " 'MISAG_contam_screen_param',\n",
       " 'MISAG_decontam_software',\n",
       " 'MISAG_env_broad_scale',\n",
       " 'MISAG_env_local_scale',\n",
       " 'MISAG_env_medium',\n",
       " 'MISAG_env_package',\n",
       " 'MISAG_experimental_factor',\n",
       " 'MISAG_feat_pred',\n",
       " 'MISAG_geo_loc_name',\n",
       " 'MISAG_investigation_type',\n",
       " 'MISAG_lat_lon',\n",
       " 'MISAG_lib_layout',\n",
       " 'MISAG_lib_reads_seqd',\n",
       " 'MISAG_lib_screen',\n",
       " 'MISAG_lib_size',\n",
       " 'MISAG_lib_vector',\n",
       " 'MISAG_mid',\n",
       " 'MISAG_nucl_acid_amp',\n",
       " 'MISAG_nucl_acid_ext',\n",
       " 'MISAG_number_contig',\n",
       " 'MISAG_project_name',\n",
       " 'MISAG_ref_biomaterial',\n",
       " 'MISAG_ref_db',\n",
       " 'MISAG_rel_to_oxygen',\n",
       " 'MISAG_samp_collec_device',\n",
       " 'MISAG_samp_collec_method',\n",
       " 'MISAG_samp_mat_process',\n",
       " 'MISAG_samp_name',\n",
       " 'MISAG_samp_size',\n",
       " 'MISAG_samp_taxon_id',\n",
       " 'MISAG_samp_vol_we_dna_ext',\n",
       " 'MISAG_seq_meth',\n",
       " 'MISAG_sim_search_meth',\n",
       " 'MISAG_single_cell_lysis_appr',\n",
       " 'MISAG_single_cell_lysis_prot',\n",
       " 'MISAG_size_frac',\n",
       " 'MISAG_sop',\n",
       " 'MISAG_sort_tech',\n",
       " 'MISAG_source_mat_id',\n",
       " 'MISAG_submitted_to_insdc',\n",
       " 'MISAG_tax_class',\n",
       " 'MISAG_tax_ident',\n",
       " 'MISAG_trna_ext_software',\n",
       " 'MISAG_trnas',\n",
       " 'MISAG_wga_amp_appr',\n",
       " 'MISAG_wga_amp_kit',\n",
       " 'MISAG_x_16s_recover',\n",
       " 'MIUVIG_adapters',\n",
       " 'MIUVIG_annot',\n",
       " 'MIUVIG_assembly_name',\n",
       " 'MIUVIG_assembly_qual',\n",
       " 'MIUVIG_assembly_software',\n",
       " 'MIUVIG_associated resource',\n",
       " 'MIUVIG_bin_param',\n",
       " 'MIUVIG_bin_software',\n",
       " 'MIUVIG_biotic_relationship',\n",
       " 'MIUVIG_collection_date',\n",
       " 'MIUVIG_compl_appr',\n",
       " 'MIUVIG_compl_score',\n",
       " 'MIUVIG_compl_software',\n",
       " 'MIUVIG_detec_type',\n",
       " 'MIUVIG_env_broad_scale',\n",
       " 'MIUVIG_env_local_scale',\n",
       " 'MIUVIG_env_medium',\n",
       " 'MIUVIG_env_package',\n",
       " 'MIUVIG_estimated_size',\n",
       " 'MIUVIG_experimental_factor',\n",
       " 'MIUVIG_feat_pred',\n",
       " 'MIUVIG_geo_loc_name',\n",
       " 'MIUVIG_host_disease_stat',\n",
       " 'MIUVIG_host_pred_appr',\n",
       " 'MIUVIG_host_pred_est_acc',\n",
       " 'MIUVIG_host_spec_range',\n",
       " 'MIUVIG_investigation_type',\n",
       " 'MIUVIG_lat_lon',\n",
       " 'MIUVIG_lib_layout',\n",
       " 'MIUVIG_lib_reads_seqd',\n",
       " 'MIUVIG_lib_screen',\n",
       " 'MIUVIG_lib_size',\n",
       " 'MIUVIG_lib_vector',\n",
       " 'MIUVIG_mag_cov_software',\n",
       " 'MIUVIG_mid',\n",
       " 'MIUVIG_nucl_acid_amp',\n",
       " 'MIUVIG_nucl_acid_ext',\n",
       " 'MIUVIG_number_contig',\n",
       " 'MIUVIG_otu_class_appr',\n",
       " 'MIUVIG_otu_db',\n",
       " 'MIUVIG_otu_seq_comp_appr',\n",
       " 'MIUVIG_pathogenicity',\n",
       " 'MIUVIG_pred_genome_struc',\n",
       " 'MIUVIG_pred_genome_type',\n",
       " 'MIUVIG_project_name',\n",
       " 'MIUVIG_reassembly_bin',\n",
       " 'MIUVIG_ref_biomaterial',\n",
       " 'MIUVIG_ref_db',\n",
       " 'MIUVIG_samp_collec_device',\n",
       " 'MIUVIG_samp_collec_method',\n",
       " 'MIUVIG_samp_mat_process',\n",
       " 'MIUVIG_samp_name',\n",
       " 'MIUVIG_samp_size',\n",
       " 'MIUVIG_samp_taxon_id',\n",
       " 'MIUVIG_samp_vol_we_dna_ext',\n",
       " 'MIUVIG_seq_meth',\n",
       " 'MIUVIG_sim_search_meth',\n",
       " 'MIUVIG_single_cell_lysis_appr',\n",
       " 'MIUVIG_single_cell_lysis_prot',\n",
       " 'MIUVIG_size_frac',\n",
       " 'MIUVIG_sop',\n",
       " 'MIUVIG_sort_tech',\n",
       " 'MIUVIG_source_mat_id',\n",
       " 'MIUVIG_source_uvig',\n",
       " 'MIUVIG_specific_host',\n",
       " 'MIUVIG_submitted_to_insdc',\n",
       " 'MIUVIG_tax_class',\n",
       " 'MIUVIG_tax_ident',\n",
       " 'MIUVIG_trna_ext_software',\n",
       " 'MIUVIG_trnas',\n",
       " 'MIUVIG_vir_ident_software',\n",
       " 'MIUVIG_virus_enrich_appr',\n",
       " 'MIUVIG_wga_amp_appr',\n",
       " 'MIUVIG_wga_amp_kit',\n",
       " 'abs_air_humidity',\n",
       " 'adapters',\n",
       " 'add_recov_method',\n",
       " 'additional_info',\n",
       " 'address',\n",
       " 'adj_room',\n",
       " 'aero_struc',\n",
       " 'agrochem_addition',\n",
       " 'air_alt',\n",
       " 'air_barometric_press',\n",
       " 'air_carb_dioxide',\n",
       " 'air_carb_monoxide',\n",
       " 'air_chem_administration',\n",
       " 'air_elev',\n",
       " 'air_humidity',\n",
       " 'air_methane',\n",
       " 'air_misc_param',\n",
       " 'air_organism_count',\n",
       " 'air_oxy_stat_samp',\n",
       " 'air_oxygen',\n",
       " 'air_perturbation',\n",
       " 'air_pollutants',\n",
       " 'air_project_name',\n",
       " 'air_resp_part_matter',\n",
       " 'air_salinity',\n",
       " 'air_samp_name',\n",
       " 'air_samp_salinity',\n",
       " 'air_samp_store_dur',\n",
       " 'air_samp_store_loc',\n",
       " 'air_samp_store_temp',\n",
       " 'air_samp_vol_we_dna_ext',\n",
       " 'air_solar_irradiance',\n",
       " 'air_temp',\n",
       " 'air_temp_regm',\n",
       " 'air_ventilation_rate',\n",
       " 'air_ventilation_type',\n",
       " 'air_volatile_org_comp',\n",
       " 'air_wind_direction',\n",
       " 'air_wind_speed',\n",
       " 'al_sat',\n",
       " 'al_sat_meth',\n",
       " 'alkalinity',\n",
       " 'alkalinity_method',\n",
       " 'alkyl_diethers',\n",
       " 'alt',\n",
       " 'aminopept_act',\n",
       " 'ammonium',\n",
       " 'amniotic_fluid_color',\n",
       " 'amount_light',\n",
       " 'ances_data',\n",
       " 'annot',\n",
       " 'annual_precpt',\n",
       " 'annual_temp',\n",
       " 'antibiotic_regm',\n",
       " 'api',\n",
       " 'arch_struc',\n",
       " 'aromatics_pc',\n",
       " 'asphaltenes_pc',\n",
       " 'assembly_name',\n",
       " 'assembly_qual',\n",
       " 'assembly_software',\n",
       " 'associated resource',\n",
       " 'atmospheric_data',\n",
       " 'avg_dew_point',\n",
       " 'avg_occup',\n",
       " 'avg_temp',\n",
       " 'bac_prod',\n",
       " 'bac_resp',\n",
       " 'bacteria_carb_prod',\n",
       " 'barometric_press',\n",
       " 'basin',\n",
       " 'bathroom_count',\n",
       " 'bedroom_count',\n",
       " 'benzene',\n",
       " 'bin_param',\n",
       " 'bin_software',\n",
       " 'biochem_oxygen_dem',\n",
       " 'biocide',\n",
       " 'biocide_admin_method',\n",
       " 'biol_stat',\n",
       " 'biomass',\n",
       " 'biotic_regm',\n",
       " 'biotic_relationship',\n",
       " 'birth_control',\n",
       " 'bishomohopanol',\n",
       " 'blood_blood_disord',\n",
       " 'blood_press_diast',\n",
       " 'blood_press_syst',\n",
       " 'bromide',\n",
       " 'build_docs',\n",
       " 'build_occup_type',\n",
       " 'building_setting',\n",
       " 'built environment_abs_air_humidity',\n",
       " 'built environment_address',\n",
       " 'built environment_adj_room',\n",
       " 'built environment_aero_struc',\n",
       " 'built environment_air_temp',\n",
       " 'built environment_amount_light',\n",
       " 'built environment_arch_struc',\n",
       " 'built environment_avg_dew_point',\n",
       " 'built environment_avg_occup',\n",
       " 'built environment_avg_temp',\n",
       " 'built environment_bathroom_count',\n",
       " 'built environment_bedroom_count',\n",
       " 'built environment_build_docs',\n",
       " 'built environment_build_occup_type',\n",
       " 'built environment_building_setting',\n",
       " 'built environment_built_struc_age',\n",
       " 'built environment_built_struc_set',\n",
       " 'built environment_built_struc_type',\n",
       " 'built environment_carb_dioxide',\n",
       " 'built environment_ceil_area',\n",
       " 'built environment_ceil_cond',\n",
       " 'built environment_ceil_finish_mat',\n",
       " 'built environment_ceil_struc',\n",
       " 'built environment_ceil_texture',\n",
       " 'built environment_ceil_thermal_mass',\n",
       " 'built environment_ceil_type',\n",
       " 'built environment_ceil_water_mold',\n",
       " 'built environment_cool_syst_id',\n",
       " 'built environment_date_last_rain',\n",
       " 'built environment_dew_point',\n",
       " 'built environment_door_comp_type',\n",
       " 'built environment_door_cond',\n",
       " 'built environment_door_direct',\n",
       " 'built environment_door_loc',\n",
       " 'built environment_door_mat',\n",
       " 'built environment_door_move',\n",
       " 'built environment_door_size',\n",
       " 'built environment_door_type',\n",
       " 'built environment_door_type_metal',\n",
       " 'built environment_door_type_wood',\n",
       " 'built environment_door_water_mold',\n",
       " 'built environment_drawings',\n",
       " 'built environment_elevator',\n",
       " 'built environment_escalator',\n",
       " 'built environment_exp_duct',\n",
       " 'built environment_exp_pipe',\n",
       " 'built environment_ext_door',\n",
       " 'built environment_ext_wall_orient',\n",
       " 'built environment_ext_window_orient',\n",
       " 'built environment_filter_type',\n",
       " 'built environment_fireplace_type',\n",
       " 'built environment_floor_age',\n",
       " 'built environment_floor_area',\n",
       " 'built environment_floor_cond',\n",
       " 'built environment_floor_count',\n",
       " 'built environment_floor_finish_mat',\n",
       " 'built environment_floor_struc',\n",
       " 'built environment_floor_thermal_mass',\n",
       " 'built environment_floor_water_mold',\n",
       " 'built environment_freq_clean',\n",
       " 'built environment_freq_cook',\n",
       " 'built environment_furniture',\n",
       " 'built environment_gender_restroom',\n",
       " 'built environment_hall_count',\n",
       " 'built environment_handidness',\n",
       " 'built environment_heat_cool_type',\n",
       " 'built environment_heat_deliv_loc',\n",
       " 'built environment_heat_sys_deliv_meth',\n",
       " 'built environment_heat_system_id',\n",
       " 'built environment_height_carper_fiber',\n",
       " 'built environment_indoor_space',\n",
       " 'built environment_indoor_surf',\n",
       " 'built environment_inside_lux',\n",
       " 'built environment_int_wall_cond',\n",
       " 'built environment_last_clean',\n",
       " 'built environment_light_type',\n",
       " 'built environment_max_occup',\n",
       " 'built environment_mech_struc',\n",
       " 'built environment_number_pets',\n",
       " 'built environment_number_plants',\n",
       " 'built environment_number_resident',\n",
       " 'built environment_occup_density_samp',\n",
       " 'built environment_occup_document',\n",
       " 'built environment_occup_samp',\n",
       " 'built environment_organism_count',\n",
       " 'built environment_pres_animal_insect',\n",
       " 'built environment_project_name',\n",
       " 'built environment_quad_pos',\n",
       " 'built environment_rel_air_humidity',\n",
       " 'built environment_rel_humidity_out',\n",
       " 'built environment_rel_samp_loc',\n",
       " 'built environment_room_air_exch_rate',\n",
       " 'built environment_room_architec_elem',\n",
       " 'built environment_room_condt',\n",
       " 'built environment_room_connected',\n",
       " 'built environment_room_count',\n",
       " 'built environment_room_dim',\n",
       " 'built environment_room_door_dist',\n",
       " 'built environment_room_door_share',\n",
       " 'built environment_room_hallway',\n",
       " 'built environment_room_loc',\n",
       " 'built environment_room_moist_dam_hist',\n",
       " 'built environment_room_net_area',\n",
       " 'built environment_room_occup',\n",
       " 'built environment_room_samp_pos',\n",
       " 'built environment_room_type',\n",
       " 'built environment_room_vol',\n",
       " 'built environment_room_wall_share',\n",
       " 'built environment_room_window_count',\n",
       " 'built environment_samp_floor',\n",
       " 'built environment_samp_name',\n",
       " 'built environment_samp_room_id',\n",
       " 'built environment_samp_sort_meth',\n",
       " 'built environment_samp_time_out',\n",
       " 'built environment_samp_weather',\n",
       " 'built environment_season',\n",
       " 'built environment_season_use',\n",
       " 'built environment_shad_dev_water_mold',\n",
       " 'built environment_shading_device_cond',\n",
       " 'built environment_shading_device_loc',\n",
       " 'built environment_shading_device_mat',\n",
       " 'built environment_shading_device_type',\n",
       " 'built environment_space_typ_state',\n",
       " 'built environment_specific',\n",
       " 'built environment_specific_humidity',\n",
       " 'built environment_substructure_type',\n",
       " 'built environment_surf_air_cont',\n",
       " 'built environment_surf_humidity',\n",
       " 'built environment_surf_material',\n",
       " 'built environment_surf_moisture',\n",
       " 'built environment_surf_moisture_ph',\n",
       " 'built environment_surf_temp',\n",
       " 'built environment_temp_out',\n",
       " 'built environment_train_line',\n",
       " 'built environment_train_stat_loc',\n",
       " 'built environment_train_stop_loc',\n",
       " 'built environment_typ_occup_density',\n",
       " 'built environment_ventilation_type',\n",
       " 'built environment_vis_media',\n",
       " 'built environment_wall_area',\n",
       " 'built environment_wall_const_type',\n",
       " 'built environment_wall_finish_mat',\n",
       " 'built environment_wall_height',\n",
       " 'built environment_wall_loc',\n",
       " 'built environment_wall_surf_treatment',\n",
       " 'built environment_wall_texture',\n",
       " 'built environment_wall_thermal_mass',\n",
       " 'built environment_wall_water_mold',\n",
       " 'built environment_water_feat_size',\n",
       " 'built environment_water_feat_type',\n",
       " 'built environment_weekday',\n",
       " 'built environment_window_cond',\n",
       " 'built environment_window_cover',\n",
       " 'built environment_window_horiz_pos',\n",
       " 'built environment_window_loc',\n",
       " 'built environment_window_mat',\n",
       " 'built environment_window_open_freq',\n",
       " 'built environment_window_size',\n",
       " 'built environment_window_status',\n",
       " 'built environment_window_type',\n",
       " 'built environment_window_vert_pos',\n",
       " 'built environment_window_water_mold',\n",
       " 'built_struc_age',\n",
       " 'built_struc_set',\n",
       " 'built_struc_type',\n",
       " 'calcium',\n",
       " 'carb_dioxide',\n",
       " 'carb_monoxide',\n",
       " 'carb_nitro_ratio',\n",
       " 'ceil_area',\n",
       " 'ceil_cond',\n",
       " 'ceil_finish_mat',\n",
       " 'ceil_struc',\n",
       " 'ceil_texture',\n",
       " 'ceil_thermal_mass',\n",
       " 'ceil_type',\n",
       " 'ceil_water_mold',\n",
       " 'chem_administration',\n",
       " 'chem_mutagen',\n",
       " 'chem_oxygen_dem',\n",
       " 'chem_treat_method',\n",
       " 'chem_treatment',\n",
       " 'chimera_check',\n",
       " 'chloride',\n",
       " 'chlorophyll',\n",
       " 'climate_environment',\n",
       " 'collection_date',\n",
       " 'compl_appr',\n",
       " 'compl_score',\n",
       " 'compl_software',\n",
       " 'conduc',\n",
       " 'contam_score',\n",
       " 'contam_screen_input',\n",
       " 'contam_screen_param',\n",
       " 'cool_syst_id',\n",
       " 'core field',\n",
       " 'crop_rotation',\n",
       " 'cult_root_med',\n",
       " 'cur_land_use',\n",
       " 'cur_vegetation',\n",
       " 'cur_vegetation_meth',\n",
       " 'date_last_rain',\n",
       " 'decontam_software',\n",
       " 'density',\n",
       " 'depos_env',\n",
       " 'depth',\n",
       " 'dermatology_disord',\n",
       " 'detec_type',\n",
       " 'dew_point',\n",
       " 'diet_last_six_month',\n",
       " 'diether_lipids',\n",
       " 'diss_carb_dioxide',\n",
       " 'diss_hydrogen',\n",
       " 'diss_inorg_carb',\n",
       " 'diss_inorg_nitro',\n",
       " 'diss_inorg_phosp',\n",
       " 'diss_iron',\n",
       " 'diss_org_carb',\n",
       " 'diss_org_nitro',\n",
       " 'diss_oxygen',\n",
       " 'diss_oxygen_fluid',\n",
       " 'dominant_hand',\n",
       " 'door_comp_type',\n",
       " 'door_cond',\n",
       " 'door_direct',\n",
       " 'door_loc',\n",
       " 'door_mat',\n",
       " 'door_move',\n",
       " 'door_size',\n",
       " 'door_type',\n",
       " 'door_type_metal',\n",
       " 'door_type_wood',\n",
       " 'door_water_mold',\n",
       " 'douche',\n",
       " 'down_par',\n",
       " 'drainage_class',\n",
       " 'drawings',\n",
       " 'drug_usage',\n",
       " 'efficiency_percent',\n",
       " 'elev',\n",
       " 'elevator',\n",
       " 'emulsions',\n",
       " 'encoded_traits',\n",
       " 'env_broad_scale',\n",
       " 'env_local_scale',\n",
       " 'env_medium',\n",
       " 'env_package',\n",
       " 'environment field',\n",
       " 'escalator',\n",
       " 'estimated_size',\n",
       " 'ethnicity',\n",
       " 'ethylbenzene',\n",
       " 'exp_duct',\n",
       " 'exp_pipe',\n",
       " 'experimental_factor',\n",
       " 'ext_door',\n",
       " 'ext_wall_orient',\n",
       " 'ext_window_orient',\n",
       " 'extrachrom_elements',\n",
       " 'extreme_event',\n",
       " 'extreme_salinity',\n",
       " 'fao_class',\n",
       " 'feat_pred',\n",
       " 'fertilizer_regm',\n",
       " 'field',\n",
       " 'filter_type',\n",
       " 'fire',\n",
       " 'fireplace_type',\n",
       " 'flooding',\n",
       " 'floor_age',\n",
       " 'floor_area',\n",
       " 'floor_cond',\n",
       " 'floor_count',\n",
       " 'floor_finish_mat',\n",
       " 'floor_struc',\n",
       " 'floor_thermal_mass',\n",
       " 'floor_water_mold',\n",
       " 'fluor',\n",
       " 'foetal_health_stat',\n",
       " 'freq_clean',\n",
       " 'freq_cook',\n",
       " 'fungicide_regm',\n",
       " 'furniture',\n",
       " 'gaseous_environment',\n",
       " 'gaseous_substances',\n",
       " 'gastrointest_disord',\n",
       " 'gender_restroom',\n",
       " 'genetic_mod',\n",
       " 'geo_loc_name',\n",
       " 'gestation_state',\n",
       " 'glucosidase_act',\n",
       " 'gravidity',\n",
       " 'gravity',\n",
       " 'growth_facil',\n",
       " 'growth_habit',\n",
       " 'growth_hormone_regm',\n",
       " 'gynecologic_disord',\n",
       " 'hall_count',\n",
       " 'handidness',\n",
       " 'hc_produced',\n",
       " 'hcr',\n",
       " 'hcr_fw_salinity',\n",
       " 'hcr_geol_age',\n",
       " 'hcr_pressure',\n",
       " 'hcr_temp',\n",
       " 'health_disease_stat',\n",
       " 'heat_cool_type',\n",
       " 'heat_deliv_loc',\n",
       " 'heat_sys_deliv_meth',\n",
       " 'heat_system_id',\n",
       " 'heavy_metals',\n",
       " 'heavy_metals_meth',\n",
       " 'height_carper_fiber',\n",
       " 'herbicide_regm',\n",
       " 'horizon',\n",
       " 'horizon_meth',\n",
       " 'host-associated_alt',\n",
       " 'host-associated_ances_data',\n",
       " 'host-associated_biol_stat',\n",
       " 'host-associated_blood_press_diast',\n",
       " 'host-associated_blood_press_syst',\n",
       " 'host-associated_chem_administration',\n",
       " 'host-associated_depth',\n",
       " 'host-associated_elev',\n",
       " 'host-associated_genetic_mod',\n",
       " 'host-associated_gravidity',\n",
       " 'host-associated_host_age',\n",
       " 'host-associated_host_body_habitat',\n",
       " 'host-associated_host_body_product',\n",
       " 'host-associated_host_body_site',\n",
       " 'host-associated_host_body_temp',\n",
       " ...]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mixs_slots.sort()\n",
    "mixs_slots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "b7418c46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term</th>\n",
       "      <th>MIxS defined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>filter_size</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>filter_type</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sieving</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>size_frac</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          term  MIxS defined\n",
       "0  filter_size         False\n",
       "1  filter_type          True\n",
       "2      sieving          True\n",
       "3    size_frac          True"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "size_fract_template_values = list(template_sf_menu_terms[2].values)\n",
    "size_fract_template_values.append(\"size_frac\")\n",
    "\n",
    "sf_term_assessments = []\n",
    "for sfterm in size_fract_template_values:\n",
    "    value_is_slot = sfterm in mixs_slots\n",
    "    temp_dict = {\"term\": sfterm, \"MIxS defined\": value_is_slot}\n",
    "    sf_term_assessments.append(temp_dict)\n",
    "\n",
    "sf_term_assessment_frame = pd.DataFrame(sf_term_assessments)\n",
    "sf_term_assessment_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dd8f434",
   "metadata": {},
   "source": [
    "#### Not `filter_size`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "9c61846d",
   "metadata": {},
   "outputs": [],
   "source": [
    "defined = list(\n",
    "    sf_term_assessment_frame[\"term\"].loc[sf_term_assessment_frame[\"MIxS defined\"]]\n",
    ")\n",
    "defined.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "eb3a5a3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filter_type\n",
      "name: filter_type\n",
      "definition_uri: https://w3id.org/mixs/vocab/filter_type\n",
      "aliases:\n",
      "- filter type\n",
      "mappings:\n",
      "- MIXS:0000765\n",
      "description: A device which removes solid particulates or airborne molecular contaminants\n",
      "comments:\n",
      "- 'Expected value: enumeration'\n",
      "- 'Occurrence: m'\n",
      "- 'Position: 17.0'\n",
      "- 'This field is used uniquely in: built environment'\n",
      "examples:\n",
      "- value: HEPA\n",
      "from_schema: http://w3id.org/mixs/terms\n",
      "is_a: core field\n",
      "range: filter_type_enum\n",
      "slot_uri: MIXS:0000765\n",
      "owner: built environment\n",
      "domain_of:\n",
      "- built environment\n",
      "pattern: '[particulate air filter|chemical air filter|low-MERV pleated media|HEPA|electrostatic|gas-phase\n",
      "  or ultraviolet air treatments]'\n",
      "\n",
      "\n",
      "sieving\n",
      "name: sieving\n",
      "definition_uri: https://w3id.org/mixs/vocab/sieving\n",
      "aliases:\n",
      "- composite design/sieving\n",
      "mappings:\n",
      "- MIXS:0000322\n",
      "description: Collection design of pooled samples and/or sieve size and amount of sample\n",
      "  sieved\n",
      "comments:\n",
      "- 'Expected value: design name and/or size;amount'\n",
      "- 'Occurrence: 1'\n",
      "- 'Position: 14.0'\n",
      "- 'This field is used uniquely in: soil'\n",
      "examples:\n",
      "- value: ''\n",
      "from_schema: http://w3id.org/mixs/terms\n",
      "is_a: core field\n",
      "range: string\n",
      "slot_uri: MIXS:0000322\n",
      "owner: soil\n",
      "domain_of:\n",
      "- soil\n",
      "pattern: '{{text}|{float} {unit}};{float} {unit}'\n",
      "\n",
      "\n",
      "size_frac\n",
      "name: size_frac\n",
      "definition_uri: https://w3id.org/mixs/vocab/size_frac\n",
      "aliases:\n",
      "- size fraction selected\n",
      "mappings:\n",
      "- MIXS:0000017\n",
      "description: Filtering pore size used in sample preparation\n",
      "comments:\n",
      "- 'Expected value: filter size value range'\n",
      "- 'Position: 33.0'\n",
      "examples:\n",
      "- value: 0-0.22 micrometer\n",
      "from_schema: http://w3id.org/mixs/terms\n",
      "exact_mappings:\n",
      "- 'MIGS:'\n",
      "is_a: nucleic acid sequence source field\n",
      "range: string\n",
      "slot_uri: MIXS:0000017\n",
      "owner: core\n",
      "domain_of:\n",
      "- ME\n",
      "- MIMARKS specimen\n",
      "- MISAG\n",
      "- MIMAG\n",
      "- MIUVIG\n",
      "- core\n",
      "pattern: '{float}-{float} {unit}'\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for one_defined in defined:\n",
    "    print(one_defined)\n",
    "    print(yaml_dumper.dumps(mixs_schema_resolved.slots[one_defined]))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "989fa830",
   "metadata": {},
   "source": [
    "#### see also `size_frac_low` and `size_frac_up`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "78c20c35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>value_y</th>\n",
       "      <th>sieving</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-.05mm</td>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2mm</td>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4mm</td>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.5-.25mm</td>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2mm</td>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "value_y    sieving                         id\n",
       "0          1-.05mm  UUID:472894-473947-642384\n",
       "1              2mm  UUID:472894-473947-847396\n",
       "2              4mm  UUID:472894-473947-847398\n",
       "3        0.5-.25mm  UUID:516394-970067-847398\n",
       "4              2mm  UUID:684267-410686-971057"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sf_multi = pivot_multi_columns(\"size_fract\")\n",
    "sf_multi = rename_melted_multi(sf_multi)\n",
    "sf_multi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "8d39ca5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'sieving': '1-.05mm', 'id': 'UUID:472894-473947-642384'},\n",
       " {'sieving': '2mm', 'id': 'UUID:472894-473947-847396'},\n",
       " {'sieving': '4mm', 'id': 'UUID:472894-473947-847398'},\n",
       " {'sieving': '0.5-.25mm', 'id': 'UUID:516394-970067-847398'},\n",
       " {'sieving': '2mm', 'id': 'UUID:684267-410686-971057'}]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sf_row_list = sf_multi.to_dict(\"records\")\n",
    "sf_row_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "4070b9dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cumulative_json_merge = merge_j_on_id(cumulative_json_merge, sf_row_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "465a6cba",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "### Treatments\n",
    "\n",
    "Where does the `MenuTerms` tab store it's beliefs about treatment modeling?\n",
    "\n",
    "If \"treatment\" appears in multiple locations on that sheet, we can \n",
    "- prioritize finding mentions that are closest to some specified coordinate, like the origin \n",
    "- prioritize finding a large cluster of mentions\n",
    "\n",
    "There can be multiple treatment columns. Google Sheets allows multiple column headers with the same value, but Pandas adds numeric suffixes to make them unique.\n",
    "\n",
    "The mixs schema doesn't have a treatment class or slot. Rather, schema defined terms can be found in row 0 of the headers frame."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36f426c0",
   "metadata": {},
   "source": [
    "#### Where is \"treatment\" mentioned in the MenuTerms?\n",
    "\n",
    "Search and cluster the results. Focus on the largest densest cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "44ecacd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "tx_coord_frame = get_optics_coord_frame(template_menu_terms, \"treatment\")\n",
    "# tx_coord_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "5593573f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2    9\n",
       "0    5\n",
       "1    4\n",
       "Name: cluster, dtype: int64"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what is the largest cluster?\n",
    "# could be multiple largest\n",
    "# take first ranked by default\n",
    "# densest cluster might be interesting too\n",
    "\n",
    "tx_cluster_sizes = tx_coord_frame[\"cluster\"].value_counts()\n",
    "tx_cluster_sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "55650109",
   "metadata": {},
   "outputs": [],
   "source": [
    "tx_largest_index = tx_cluster_sizes.index[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "de4efeb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(53.3, 2.6999999999999997)"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEGCAYAAABsLkJ6AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAOoElEQVR4nO3db4xldX3H8ffHXQgGZwLIdLNhoUuV1BhTsZkSraaxWA1tjdDGEE1b14Rk+0AbjE0VTRO1sfVPWrWPbKiYbhMVNqCF+KCRIMaaNsgsf+TPtgUJpGwWdqySGR5UBb59cA/tsMydnZ3dc8/e+3u/kpt7zu+cO/f7y8nez57f+ZeqQpLUnpcMXYAkaRgGgCQ1ygCQpEYZAJLUKANAkhq1fegCNuPcc8+t3bt3D12GJE2VAwcO/KiqFsYtn4oA2L17N0tLS0OXIUlTJcljGy13CEiSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1ygAY2vw8JP//mp8fuiJJjTAAhra6uvG8JPXEAJCkRhkAktQoA+DoMfhJj8PPzW08L0k9mYp7AfVqvTH3SY7Dr6xM7rskaQ33ACSpUQaAJDXKAJCkRhkA6x109UCspAZ4ENiDsJIa5R6AJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1ygCQpEb1FgBJvpzkSJL717Sdk+TWJA9172f39f2SpI31uQfwD8BlR7VdA9xWVRcBt3XzkqQB9BYAVfVd4MdHNV8O7Oum9wFX9PX9kqSNTfoYwI6qOtxNPwHsGLdikr1JlpIsLS8vT6Y6SWrIYAeBq6qA2mD5tVW1WFWLCwsLE6xMktow6QB4MslOgO79yIS/X5LUmXQA3ALs6ab3ADdP+PslSZ0+TwP9GvBvwC8neTzJVcCngbcmeQj4rW5ekjSA3h4JWVXvHrPoLX19pyRp87wSeH4ekhe+5ueHrkqSemcArK5urk2SZowBIEmNMgAkqVEGgCQ1ygCYm9tcmyTNmN5OA50aKytDVyBJg3APQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1ygCQpEYZAJLUKANAkhplAEhSowwASWqUATA/D8kLX/PzQ1clSb0zAFZXN9cmSTPGAJCkRhkAktQoA2BubnNtkjRjtg9dwOBWVoauQJIG4R6AJDXKAJCkRvUWAEnOT3J7kgeTPJDk6q79nCS3Jnmoez+7rxokSeP1uQfwDPCnVfVq4PXA+5K8GrgGuK2qLgJu6+YlSRPWWwBU1eGququbXgUOAucBlwP7utX2AVf0VYMkabyJHANIsht4HXAHsKOqDneLngB2jPnM3iRLSZaWl5cnUaYkNaX3AEjyMuAm4ANV9YJzLquqgFrvc1V1bVUtVtXiwsJC32VKUnN6DYAkpzH68f9KVX29a34yyc5u+U7gSJ81SJLW1+dZQAGuAw5W1efWLLoF2NNN7wFu7qsGSdJ4fV4J/Ebgj4D7ktzTtX0U+DSwP8lVwGPAlT3WIEkao7cAqKrvARmz+C19fa8kaXO8EliSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1ygCQpEYZAJLUKANAkhplAMzPQ/LC1/z80FVJUu8MgNXVzbVJ0owxACSpUQaAJDXKAJib21ybJM2YPp8HMB1WVo69jiTNIPcAJKlRBoAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktSoYwZAkj9JcvYkipEkTc5m9gB2AHcm2Z/ksiTpuyhJUv+OGQBV9efARcB1wHuBh5L8VZJX9FybJKlHmzoGUFUFPNG9ngHOBm5M8tkea5Mk9eiYD4RJcjXwHuBHwJeAP6uqnyd5CfAQ8KF+S5Qk9WEzTwQ7B/j9qnpsbWNVPZfk7f2UJUnq2zEDoKo+tsGygye3HEnSpPR2HUCSM5J8P8m9SR5I8omu/cIkdyR5OMkNSU7vqwZJ0nh9Xgj2U+DSqnotcDFwWZLXA58BPl9VrwR+AlzVYw2SpDF6C4AaebqbPa17FXApcGPXvg+4oq8aJEnj9XoriCTbktwDHAFuBX4IPFVVz3SrPA6cN+aze5MsJVlaXl7us0xJalKvAVBVz1bVxcAu4BLgVcfx2WurarGqFhcWFvoqUZKaNZGbwVXVU8DtwBuAs5I8f/bRLuDQJGqQptH8/DxJXvCan58fuqwXsc6Ta1J19nkW0EKSs7rplwJvBQ4yCoJ3dqvtAW7uqwZp2q2urm6qbWjWeXJNqs7NXAi2VTuBfUm2MQqa/VX1zSQPAtcn+SRwN6N7DEmSJqy3AKiqHwCvW6f9EUbHAyRJA/KBMNIpbG5ublNtQ7POk2tSdfY5BCTpBK2srAxdwqZY58k1qTrdA5CkRhkAktQoA0CSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1ygCQpEYZAJLUKANAkhplAEhSowwASWqUASBJjTIAJKlRBoAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmNMgAkqVG9B0CSbUnuTvLNbv7CJHckeTjJDUlO77sGSdKLTWIP4Grg4Jr5zwCfr6pXAj8BrppADZKko/QaAEl2Ab8LfKmbD3ApcGO3yj7gij5rkCStr+89gC8AHwKe6+ZfDjxVVc90848D5/VcgyRpHb0FQJK3A0eq6sAWP783yVKSpeXl5ZNcnSSpzz2ANwLvSPIocD2joZ+/Bc5Ksr1bZxdwaL0PV9W1VbVYVYsLCws9lilJbeotAKrqI1W1q6p2A+8Cvl1VfwDcDryzW20PcHNfNUiSxhviOoAPAx9M8jCjYwLXDVCDJDVv+7FXOXFV9R3gO930I8Alk/heSdJ4XgksSY0yACSpUQaAJDXKAJCkRhkAktSoiZwFNIT5T82z+rPV/5ufO32OlY+sDFiRJJ1aZnYPYO2P/3rzktS6mQ0ASdLGDABJapQBIEmNMgAkqVEGgCQ1ygCQpEYZAJLUKANAkhplAEhSowwASWqUASBJjTIAJKlRBoAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmN2t7nH0/yKLAKPAs8U1WLSc4BbgB2A48CV1bVT/qsQ5L0YpPYA/jNqrq4qha7+WuA26rqIuC2bl6SNGFDDAFdDuzrpvcBVwxQgyQ1r+8AKOBbSQ4k2du17aiqw930E8COnmuQJK2j12MAwJuq6lCSXwBuTfLvaxdWVSWp9T7YBcZegAsuuKDnMiWpPb3uAVTVoe79CPAN4BLgySQ7Abr3I2M+e21VLVbV4sLCQp9lSlKTeguAJGcmmXt+GngbcD9wC7CnW20PcHNfNUiSxutzCGgH8I0kz3/PV6vqn5PcCexPchXwGHBljzVIksboLQCq6hHgteu0/zfwlr6+V5K0OV4JLEmNmtkAmDt9bsN5SWpd36eBDmblIytDlyBJp7SZ3QOQJG3MAJCkRhkAktQoA0CSGmUASFKjDABJalSq1r0Z5yklyTKj20ZsxbnAj05iOaeCWeuT/Tn1zVqfZq0/sH6ffrGqxt5NcyoC4EQkWVrzNLKZMGt9sj+nvlnr06z1B7bWJ4eAJKlRBoAkNaqFALh26AJ6MGt9sj+nvlnr06z1B7bQp5k/BiBJWl8LewCSpHUYAJLUqJkOgCSXJfmPJA8nuWboek5UkkeT3JfkniRLQ9ezFUm+nORIkvvXtJ2T5NYkD3XvZw9Z4/EY05+PJznUbad7kvzOkDUejyTnJ7k9yYNJHkhyddc+zdtoXJ+mcjslOSPJ95Pc2/XnE137hUnu6H7vbkhy+jH/1qweA0iyDfhP4K3A48CdwLur6sFBCzsBSR4FFqtqai9gSfIbwNPAP1bVa7q2zwI/rqpPd0F9dlV9eMg6N2tMfz4OPF1Vfz1kbVuRZCews6ruSjIHHACuAN7L9G6jcX26kincThk9aP3Mqno6yWnA94CrgQ8CX6+q65P8HXBvVX1xo781y3sAlwAPV9UjVfUz4Hrg8oFral5VfRf48VHNlwP7uul9jP5xToUx/ZlaVXW4qu7qpleBg8B5TPc2GtenqVQjT3ezp3WvAi4FbuzaN7WNZjkAzgP+a83840zxRu8U8K0kB5LsHbqYk2hHVR3upp8AdgxZzEny/iQ/6IaIpma4ZK0ku4HXAXcwI9voqD7BlG6nJNuS3AMcAW4Ffgg8VVXPdKts6vdulgNgFr2pqn4V+G3gfd3ww0yp0ZjktI9LfhF4BXAxcBj4m0Gr2YIkLwNuAj5QVS94vuq0bqN1+jS126mqnq2qi4FdjEY7XrWVvzPLAXAIOH/N/K6ubWpV1aHu/QjwDUYbfhY82Y3TPj9ee2Tgek5IVT3Z/QN9Dvh7pmw7dePKNwFfqaqvd81TvY3W69O0byeAqnoKuB14A3BWkuef876p37tZDoA7gYu6I+OnA+8Cbhm4pi1LcmZ3AIskZwJvA+7f+FNT4xZgTze9B7h5wFpO2PM/lJ3fY4q2U3eA8TrgYFV9bs2iqd1G4/o0rdspyUKSs7rplzI60eUgoyB4Z7faprbRzJ4FBNCd1vUFYBvw5ar6y2Er2rokv8Tof/0A24GvTmN/knwNeDOjW9c+CXwM+CdgP3ABo9t+X1lVU3FgdUx/3sxoWKGAR4E/XjN+fkpL8ibgX4D7gOe65o8yGjOf1m00rk/vZgq3U5JfYXSQdxuj/8Tvr6q/6H4jrgfOAe4G/rCqfrrh35rlAJAkjTfLQ0CSpA0YAJLUKANAkhplAEhSowwASWqUASBJjTIAJKlRBoC0BUl+rbuJ2BndVdoPJHnN0HVJx8MLwaQtSvJJ4AzgpcDjVfWpgUuSjosBIG1Rd4+pO4H/AX69qp4duCTpuDgEJG3dy4GXAXOM9gSkqeIegLRFSW5hdPOtCxk9cvD9A5ckHZftx15F0tGSvAf4eVV9tXv+9L8mubSqvj10bdJmuQcgSY3yGIAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY36Xyq461R2Jc68AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# ax = plt.subplots()\n",
    "\n",
    "tx_coord_frame.plot.scatter(\n",
    "    x=\"x\", y=\"y\", c=tx_coord_frame[\"cluster\"].map(cluster_colors), marker=\"s\"\n",
    ")\n",
    "\n",
    "# ax.legend()\n",
    "# plt.legend()\n",
    "\n",
    "plt.ylim(max(plt.ylim()), min(plt.ylim()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "730975da",
   "metadata": {},
   "source": [
    "In this case, there is a large dense cluster in column 0, row numbers in the 40s"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1539176e",
   "metadata": {},
   "source": [
    "#### Treatments acknowledged by the Sample Meatadata Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "4da4f2fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>treatment</td>\n",
       "      <td>treatment-chemical administration/addition</td>\n",
       "      <td>chem_administration</td>\n",
       "      <td>{termLabel} {[termID]}; {timestamp}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>treatment</td>\n",
       "      <td>treatment- watering regimen/schedule</td>\n",
       "      <td>watering_regm</td>\n",
       "      <td>{float} {unit};{Rn/start_time/end_time/duration}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>treatment</td>\n",
       "      <td>treatment-air temperature regimen</td>\n",
       "      <td>air_temp_regm</td>\n",
       "      <td>{float} {unit};{Rn/start_time/end_time/duration}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>treatment</td>\n",
       "      <td>treatment-gas evironment/exposure</td>\n",
       "      <td>gaseous_environment</td>\n",
       "      <td>{text};{float} {unit};{Rn/start_time/end_time/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>treatment</td>\n",
       "      <td>treatment-isotope exposure/addition</td>\n",
       "      <td>isotope_exposure</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>treatment</td>\n",
       "      <td>treatment-climate environment</td>\n",
       "      <td>climate_environment</td>\n",
       "      <td>{text};{Rn/start_time/end_time/duration}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>treatment</td>\n",
       "      <td>treatment-humidity regimen/adition</td>\n",
       "      <td>humidity_regm</td>\n",
       "      <td>{float} {unit};{Rn/start_time/end_time/duration}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>treatment</td>\n",
       "      <td>treatment-light regimen</td>\n",
       "      <td>light_regm</td>\n",
       "      <td>{text};{float} {unit};{float} {unit}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>treatment</td>\n",
       "      <td>treatment-biotic regimen/addition</td>\n",
       "      <td>biotic_regm</td>\n",
       "      <td>{text}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0                                           1  \\\n",
       "43  treatment  treatment-chemical administration/addition   \n",
       "44  treatment        treatment- watering regimen/schedule   \n",
       "45  treatment           treatment-air temperature regimen   \n",
       "46  treatment           treatment-gas evironment/exposure   \n",
       "47  treatment         treatment-isotope exposure/addition   \n",
       "48  treatment               treatment-climate environment   \n",
       "49  treatment          treatment-humidity regimen/adition   \n",
       "50  treatment                     treatment-light regimen   \n",
       "51  treatment           treatment-biotic regimen/addition   \n",
       "\n",
       "                      2                                                  3  \n",
       "43  chem_administration                {termLabel} {[termID]}; {timestamp}  \n",
       "44        watering_regm   {float} {unit};{Rn/start_time/end_time/duration}  \n",
       "45        air_temp_regm   {float} {unit};{Rn/start_time/end_time/duration}  \n",
       "46  gaseous_environment  {text};{float} {unit};{Rn/start_time/end_time/...  \n",
       "47     isotope_exposure                                             {text}  \n",
       "48  climate_environment           {text};{Rn/start_time/end_time/duration}  \n",
       "49        humidity_regm   {float} {unit};{Rn/start_time/end_time/duration}  \n",
       "50           light_regm               {text};{float} {unit};{float} {unit}  \n",
       "51          biotic_regm                                             {text}  "
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_treatments = get_data_from_extended_cluster_coords(\n",
    "    template_menu_terms, tx_coord_frame, tx_largest_index, additional_columns=3\n",
    ")\n",
    "\n",
    "template_treatments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f9c4f03",
   "metadata": {},
   "source": [
    "#### Are there any treatments in the template that aren't defined by MIxS ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "2f531d64",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'isotope_exposure'}"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(template_treatments[2]) - set(mixs_slots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "8c72ac99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>value_y</th>\n",
       "      <th>air_temp_regm</th>\n",
       "      <th>chem_administration</th>\n",
       "      <th>isotope_exposure</th>\n",
       "      <th>watering_regm</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1000Âµg of C (glucose)/g soil</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75% water holding capacity; moisture maintaine...</td>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>35C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1000Âµg of C (glucose)/g soil</td>\n",
       "      <td>13C Glucose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "value_y air_temp_regm           chem_administration isotope_exposure  \\\n",
       "0                 NaN  1000Âµg of C (glucose)/g soil              NaN   \n",
       "1                 35C                           NaN              NaN   \n",
       "2                 NaN  1000Âµg of C (glucose)/g soil      13C Glucose   \n",
       "\n",
       "value_y                                      watering_regm  \\\n",
       "0        75% water holding capacity; moisture maintaine...   \n",
       "1                                                      NaN   \n",
       "2                                                      NaN   \n",
       "\n",
       "value_y                         id  \n",
       "0        UUID:472894-473947-847396  \n",
       "1        UUID:516394-970067-847398  \n",
       "2        UUID:684267-410686-971057  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx_melt = pivot_multi_columns(\"treatment\")\n",
    "tx_renamed = rename_melted_multi(tx_melt)\n",
    "tx_renamed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "40314706",
   "metadata": {},
   "outputs": [],
   "source": [
    "treatment_row_list = tx_renamed.to_dict(\"records\")\n",
    "cumulative_json_merge = merge_j_on_id(cumulative_json_merge, treatment_row_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c80a1b93",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "5547df71",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_env_met_frame = pd.read_excel(\n",
    "    template_file, sheet_name=\"EnvironmentalMetadata\"\n",
    ")\n",
    "# template_env_met_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "fb580900",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_env_met_headers_cols = pd.Series(template_env_met_frame.columns)\n",
    "\n",
    "firstbad = template_env_met_headers_cols.str.contains(\"Unnamed|#N/A\").idxmax()\n",
    "\n",
    "goodcols = list(template_env_met_headers_cols[0:firstbad])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "1e90c8df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_ID</th>\n",
       "      <th>lab_ID</th>\n",
       "      <th>oxy_stat_samp</th>\n",
       "      <th>cur_land_use</th>\n",
       "      <th>drainage_class</th>\n",
       "      <th>fao_class</th>\n",
       "      <th>biotic_relationship</th>\n",
       "      <th>profile_position</th>\n",
       "      <th>horizon</th>\n",
       "      <th>tillage</th>\n",
       "      <th>...</th>\n",
       "      <th>slope_gradient</th>\n",
       "      <th>slope_aspect</th>\n",
       "      <th>link_addit_analys</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>salinity_meth</th>\n",
       "      <th>heavy_metals</th>\n",
       "      <th>heavy_metals_meth</th>\n",
       "      <th>al_sat</th>\n",
       "      <th>al_sat_meth</th>\n",
       "      <th>misc_param</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Universally unique ID (ex: IGSN, UUID)</td>\n",
       "      <td>Laboratory ID</td>\n",
       "      <td>oxygen sample status</td>\n",
       "      <td>current land use</td>\n",
       "      <td>drainage classification</td>\n",
       "      <td>soil_taxonomic/FAO classification</td>\n",
       "      <td>biotic relationship</td>\n",
       "      <td>profile position</td>\n",
       "      <td>horizon</td>\n",
       "      <td>history/tillage</td>\n",
       "      <td>...</td>\n",
       "      <td>slope gradient</td>\n",
       "      <td>slope aspect</td>\n",
       "      <td>links to additional analysis</td>\n",
       "      <td>extreme_unusual_properties/salinity</td>\n",
       "      <td>extreme_unusual_properties/salinity method</td>\n",
       "      <td>extreme_unusual_properties/heavy metals</td>\n",
       "      <td>extreme_unusual_properties/heavy metals method</td>\n",
       "      <td>extreme_unusual_properties/Al saturation</td>\n",
       "      <td>extreme_unusual_properties/Al saturation method</td>\n",
       "      <td>miscellaneous parameter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{text}</td>\n",
       "      <td>{text}</td>\n",
       "      <td>controlled term, see readme</td>\n",
       "      <td>controlled term, see readme</td>\n",
       "      <td>controlled term, see readme</td>\n",
       "      <td>controlled term, see readme</td>\n",
       "      <td>controlled term, see readme</td>\n",
       "      <td>controlled term, see readme</td>\n",
       "      <td>controlled term, see readme</td>\n",
       "      <td>controlled term, see readme</td>\n",
       "      <td>...</td>\n",
       "      <td>{float} {unit}</td>\n",
       "      <td>{float} {unit}</td>\n",
       "      <td>{PMID}|{DOI}|{URL}</td>\n",
       "      <td>{float} {unit}</td>\n",
       "      <td>{PMID}|{DOI}|{URL}</td>\n",
       "      <td>{text};{float} {unit}</td>\n",
       "      <td>{PMID}|{DOI}|{URL}</td>\n",
       "      <td>{float} {unit}</td>\n",
       "      <td>{PMID}|{DOI}|{URL}</td>\n",
       "      <td>{text};{float} {unit}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                unique_ID         lab_ID  \\\n",
       "0  Universally unique ID (ex: IGSN, UUID)  Laboratory ID   \n",
       "1                                  {text}         {text}   \n",
       "\n",
       "                 oxy_stat_samp                 cur_land_use  \\\n",
       "0         oxygen sample status             current land use   \n",
       "1  controlled term, see readme  controlled term, see readme   \n",
       "\n",
       "                drainage_class                          fao_class  \\\n",
       "0      drainage classification  soil_taxonomic/FAO classification   \n",
       "1  controlled term, see readme        controlled term, see readme   \n",
       "\n",
       "           biotic_relationship             profile_position  \\\n",
       "0          biotic relationship             profile position   \n",
       "1  controlled term, see readme  controlled term, see readme   \n",
       "\n",
       "                       horizon                      tillage  ...  \\\n",
       "0                      horizon              history/tillage  ...   \n",
       "1  controlled term, see readme  controlled term, see readme  ...   \n",
       "\n",
       "   slope_gradient    slope_aspect             link_addit_analys  \\\n",
       "0  slope gradient    slope aspect  links to additional analysis   \n",
       "1  {float} {unit}  {float} {unit}            {PMID}|{DOI}|{URL}   \n",
       "\n",
       "                      extreme_salinity  \\\n",
       "0  extreme_unusual_properties/salinity   \n",
       "1                       {float} {unit}   \n",
       "\n",
       "                                salinity_meth  \\\n",
       "0  extreme_unusual_properties/salinity method   \n",
       "1                          {PMID}|{DOI}|{URL}   \n",
       "\n",
       "                              heavy_metals  \\\n",
       "0  extreme_unusual_properties/heavy metals   \n",
       "1                    {text};{float} {unit}   \n",
       "\n",
       "                                heavy_metals_meth  \\\n",
       "0  extreme_unusual_properties/heavy metals method   \n",
       "1                              {PMID}|{DOI}|{URL}   \n",
       "\n",
       "                                     al_sat  \\\n",
       "0  extreme_unusual_properties/Al saturation   \n",
       "1                            {float} {unit}   \n",
       "\n",
       "                                       al_sat_meth               misc_param  \n",
       "0  extreme_unusual_properties/Al saturation method  miscellaneous parameter  \n",
       "1                               {PMID}|{DOI}|{URL}    {text};{float} {unit}  \n",
       "\n",
       "[2 rows x 52 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_env_met_headers = template_env_met_frame.loc[0 : (header_rows - 3), goodcols]\n",
    "template_env_met_headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "b551403f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_ID</th>\n",
       "      <th>lab_ID</th>\n",
       "      <th>oxy_stat_samp</th>\n",
       "      <th>cur_land_use</th>\n",
       "      <th>drainage_class</th>\n",
       "      <th>fao_class</th>\n",
       "      <th>biotic_relationship</th>\n",
       "      <th>profile_position</th>\n",
       "      <th>horizon</th>\n",
       "      <th>tillage</th>\n",
       "      <th>...</th>\n",
       "      <th>slope_gradient</th>\n",
       "      <th>slope_aspect</th>\n",
       "      <th>link_addit_analys</th>\n",
       "      <th>extreme_salinity</th>\n",
       "      <th>salinity_meth</th>\n",
       "      <th>heavy_metals</th>\n",
       "      <th>heavy_metals_meth</th>\n",
       "      <th>al_sat</th>\n",
       "      <th>al_sat_meth</th>\n",
       "      <th>misc_param</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UUID:472894-473947-847398</td>\n",
       "      <td>P4_-0-10_4C</td>\n",
       "      <td>aerobic</td>\n",
       "      <td>Horticultural Plants (e.g. tulips)</td>\n",
       "      <td>Moderately Well</td>\n",
       "      <td>Andosols</td>\n",
       "      <td>free living</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B Horizon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UUID:684267-410686-971057</td>\n",
       "      <td>L10_-20-30_13C-Gluc</td>\n",
       "      <td>aerobic</td>\n",
       "      <td>Horticultural Plants (e.g. tulips)</td>\n",
       "      <td>Moderately Well</td>\n",
       "      <td>Andosols</td>\n",
       "      <td>free living</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B Horizon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UUID:472894-473947-847396</td>\n",
       "      <td>6_J2_75-WHC</td>\n",
       "      <td>aerobic</td>\n",
       "      <td>Horticultural Plants (e.g. tulips)</td>\n",
       "      <td>Moderately Well</td>\n",
       "      <td>Andosols</td>\n",
       "      <td>free living</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B Horizon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UUID:516394-970067-847398</td>\n",
       "      <td>T4-35</td>\n",
       "      <td>aerobic</td>\n",
       "      <td>Horticultural Plants (e.g. tulips)</td>\n",
       "      <td>Moderately Well</td>\n",
       "      <td>Andosols</td>\n",
       "      <td>free living</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O Horizon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UUID:472894-473947-642384</td>\n",
       "      <td>2020-05-21_S19</td>\n",
       "      <td>aerobic</td>\n",
       "      <td>Horticultural Plants (e.g. tulips)</td>\n",
       "      <td>Moderately Well</td>\n",
       "      <td>Andosols</td>\n",
       "      <td>free living</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O Horizon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   unique_ID               lab_ID oxy_stat_samp  \\\n",
       "2  UUID:472894-473947-847398          P4_-0-10_4C       aerobic   \n",
       "3  UUID:684267-410686-971057  L10_-20-30_13C-Gluc       aerobic   \n",
       "4  UUID:472894-473947-847396          6_J2_75-WHC       aerobic   \n",
       "5  UUID:516394-970067-847398                T4-35       aerobic   \n",
       "6  UUID:472894-473947-642384       2020-05-21_S19       aerobic   \n",
       "\n",
       "                         cur_land_use   drainage_class fao_class  \\\n",
       "2  Horticultural Plants (e.g. tulips)  Moderately Well  Andosols   \n",
       "3  Horticultural Plants (e.g. tulips)  Moderately Well  Andosols   \n",
       "4  Horticultural Plants (e.g. tulips)  Moderately Well  Andosols   \n",
       "5  Horticultural Plants (e.g. tulips)  Moderately Well  Andosols   \n",
       "6  Horticultural Plants (e.g. tulips)  Moderately Well  Andosols   \n",
       "\n",
       "  biotic_relationship profile_position    horizon tillage  ... slope_gradient  \\\n",
       "2         free living              NaN  B Horizon     NaN  ...            NaN   \n",
       "3         free living              NaN  B Horizon     NaN  ...            NaN   \n",
       "4         free living              NaN  B Horizon     NaN  ...            NaN   \n",
       "5         free living              NaN  O Horizon     NaN  ...            NaN   \n",
       "6         free living              NaN  O Horizon     NaN  ...            NaN   \n",
       "\n",
       "  slope_aspect link_addit_analys extreme_salinity salinity_meth heavy_metals  \\\n",
       "2          NaN               NaN              NaN           NaN          NaN   \n",
       "3          NaN               NaN              NaN           NaN          NaN   \n",
       "4          NaN               NaN              NaN           NaN          NaN   \n",
       "5          NaN               NaN              NaN           NaN          NaN   \n",
       "6          NaN               NaN              NaN           NaN          NaN   \n",
       "\n",
       "  heavy_metals_meth al_sat al_sat_meth misc_param  \n",
       "2               NaN    NaN         NaN        NaN  \n",
       "3               NaN    NaN         NaN        NaN  \n",
       "4               NaN    NaN         NaN        NaN  \n",
       "5               NaN    NaN         NaN        NaN  \n",
       "6               NaN    NaN         NaN        NaN  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template_env_met_content = template_env_met_frame.loc[2:, goodcols]\n",
    "template_env_met_content"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16f32073",
   "metadata": {},
   "source": [
    "In some of the parsing, I have been mapping \"bad\" columns names inside of a list comprehension. Here I'm renaming the table because mapping isn't require. A plain `to_dict(\"records\")` is sufficient. SO column renaming will be required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "8a7c5c9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_env_met_mapping = {\"unique_ID\": \"id\", \"lab_ID\": \"name\"}\n",
    "template_env_met_content = template_env_met_content.rename(\n",
    "    columns=template_env_met_mapping\n",
    ")\n",
    "template_env_met_headers = template_env_met_headers.rename(\n",
    "    columns=template_env_met_mapping\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5244a13",
   "metadata": {},
   "source": [
    "May not actually have to make any use of `template_env_met_headers`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "28a55021",
   "metadata": {},
   "outputs": [],
   "source": [
    "template_env_met_row_list = template_env_met_content.to_dict(\"records\")\n",
    "cumulative_json_merge = merge_j_on_id(cumulative_json_merge, template_env_met_row_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f870fe39",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "# Interim reshaped study data\n",
    "\n",
    "Not complete, encoded or validated "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "82f9f84f",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_nmdc = remove_nans_from_dict(proj_nmdc)\n",
    "# pprint(proj_nmdc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "6f5acdb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"id\": \"GOLD:Gs01971387\",\n",
      "    \"title\": \"\\\"Soil microbial response to elevated temperatures and increased carbon availability\\\"\",\n",
      "    \"description\": \"Northern latitude peatlands contain large reserves of soil carbon (C) due in part to climatic constraints on decomposition, including low temperatures and water inundated soils.... As expected, the two LMWCC substrate additions increased C respiration....\",\n",
      "    \"principal_investigator\": {\n",
      "        \"has raw value\": \"Montana Smith\",\n",
      "        \"orcid\": \"0000-0002-8683-0050\",\n",
      "        \"email\": \"montana.smith@pnnl.gov\"\n",
      "    },\n",
      "    \"type\": \"Study\",\n",
      "    \"websites\": [\n",
      "        \"https://microbiomedata.org/\",\n",
      "        \"https://github.com/microbiomedata\"\n",
      "    ],\n",
      "    \"publications\": [\n",
      "        \"https://doi.org/10.1016/j.soilbio.2019.107561\",\n",
      "        \"Keiser, Ashley D., et al. \\\"Peatland microbial community response to altered climate tempered by nutrient availability.\\\" Soil Biology and Biochemistry 137 (2019): 107561.\"\n",
      "    ],\n",
      "    \"doi\": {\n",
      "        \"has raw value\": \"https://doi.org/10.25585/1487765\"\n",
      "    },\n",
      "    \"alternate identifiers\": [\n",
      "        \"EMSL:29728\",\n",
      "        \"JGI:1781\",\n",
      "        \"NCBI ID: 594403\"\n",
      "    ],\n",
      "    \"INSDC bioproject identifiers\": \"NCBI Accession: PRJNA594403\",\n",
      "    \"GOLD study identifiers\": \"GOLD:Gs01971387\",\n",
      "    \"has credit record\": [\n",
      "        {\n",
      "            \"type\": \"credit record\",\n",
      "            \"has contributor role\": \"Project administration\",\n",
      "            \"applies to person\": {\n",
      "                \"has_raw_value\": \"Pajau Vangay\",\n",
      "                \"orcid\": \"0000-0002-9231-0692\",\n",
      "                \"email\": \"pvangay@lbl.gov\"\n",
      "            }\n",
      "        }\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "proj_json = json.dumps(proj_nmdc, indent=4, sort_keys=False, default=str)\n",
    "print(proj_json)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09873eb0",
   "metadata": {},
   "source": [
    "# Interim reshaped sample data\n",
    "\n",
    "Not complete, encoded or validated "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "cf678933",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "    {\n",
      "        \"id\": \"UUID:472894-473947-847398\",\n",
      "        \"name\": \"P4_-0-10_4C\",\n",
      "        \"env_package\": \"soil\",\n",
      "        \"investigation_type\": \"genome|metabolome\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"d787487292564776a968fb4efb1002b5\",\n",
      "        \"growth_facil\": \"field\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"snap freeze in liquid N\",\n",
      "        \"store_cond\": \"fresh\",\n",
      "        \"samp_store_temp\": \"4 degree Celsius\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251709, -119.728663\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2020-05-21T12:00:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": \"0-1\",\n",
      "            \"has_unit\": \"m\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 500.0,\n",
      "            \"has_unit\": \"m\"\n",
      "        },\n",
      "        \"samp_collect_device\": \"11/2' x 6\\\"; stainless stee slide_hammer_core\",\n",
      "        \"sieving\": \"4mm\",\n",
      "        \"oxy_stat_samp\": \"aerobic\",\n",
      "        \"cur_land_use\": \"Horticultural Plants (e.g. tulips)\",\n",
      "        \"drainage_class\": \"Moderately Well\",\n",
      "        \"fao_class\": \"Andosols\",\n",
      "        \"biotic_relationship\": \"free living\",\n",
      "        \"horizon\": \"B Horizon\",\n",
      "        \"cur_vegetation\": \"tall wheat grass\",\n",
      "        \"ph\": 6.2,\n",
      "        \"ph_meth\": \"https://doi.org/10.2136/sssabookser5.3.c16\",\n",
      "        \"microbial_biomass\": \"0.031431756 mg of C/gram dry soil\",\n",
      "        \"microbial_biomass_meth\": \"https://doi.org/10.1016/0038-0717(87)90052-6 \",\n",
      "        \"water_content\": \"0.5 g water/g soil\",\n",
      "        \"water_content_meth\": \"https://doi.org/10.1016/0016-7061(86)90018-2 \",\n",
      "        \"annual_temp\": \"18.5 celsius\",\n",
      "        \"season_temp\": \"30.44 celsius\",\n",
      "        \"soil_type\": \"sandy loam\"\n",
      "    },\n",
      "    {\n",
      "        \"id\": \"UUID:684267-410686-971057\",\n",
      "        \"name\": \"L10_-20-30_13C-Gluc\",\n",
      "        \"env_package\": \"soil\",\n",
      "        \"source_mat_id\": \"UUID:472894-473947-847398\",\n",
      "        \"investigation_type\": \"transcriptome|metabolome\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"d787487292564776a968fb4efb1002b5\",\n",
      "        \"growth_facil\": \"field_incubation\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"2micron sieve, wet to 25% gravimetric water capacity\",\n",
      "        \"store_cond\": \"frozen\",\n",
      "        \"samp_store_temp\": \"-80 degree Celsius\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251709, -119.728663\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2020-05-21T13:30:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": \"0-1\",\n",
      "            \"has_unit\": \"m\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 500.0,\n",
      "            \"has_unit\": \"m\"\n",
      "        },\n",
      "        \"samp_collect_device\": \"11/2' x 6\\\"; stainless stee slide_hammer_core\",\n",
      "        \"sieving\": \"2mm\",\n",
      "        \"chem_administration\": \"1000\\u00b5g of C (glucose)/g soil\",\n",
      "        \"isotope_exposure\": \"13C Glucose\",\n",
      "        \"oxy_stat_samp\": \"aerobic\",\n",
      "        \"cur_land_use\": \"Horticultural Plants (e.g. tulips)\",\n",
      "        \"drainage_class\": \"Moderately Well\",\n",
      "        \"fao_class\": \"Andosols\",\n",
      "        \"biotic_relationship\": \"free living\",\n",
      "        \"horizon\": \"B Horizon\",\n",
      "        \"cur_vegetation\": \"tall wheat grass\",\n",
      "        \"ph\": 6.2,\n",
      "        \"ph_meth\": \"https://doi.org/10.2136/sssabookser5.3.c16\",\n",
      "        \"microbial_biomass\": \"0.029525225 mg of C/gram dry soil\",\n",
      "        \"microbial_biomass_meth\": \"https://doi.org/10.1016/0038-0717(87)90052-6 \",\n",
      "        \"water_content\": \"0.5 g water/g soil\",\n",
      "        \"water_content_meth\": \"https://doi.org/10.1016/0016-7061(86)90018-2 \",\n",
      "        \"annual_temp\": \"18.5 celsius\",\n",
      "        \"season_temp\": \"30.44 celsius\",\n",
      "        \"soil_type\": \"sandy loam\"\n",
      "    },\n",
      "    {\n",
      "        \"id\": \"UUID:472894-473947-847396\",\n",
      "        \"name\": \"6_J2_75-WHC\",\n",
      "        \"env_package\": \"soil\",\n",
      "        \"source_mat_id\": \"UUID:472894-473947-847398\",\n",
      "        \"investigation_type\": \"organic matter\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"d787487292564776a968fb4efb1002b5\",\n",
      "        \"growth_facil\": \"field\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"aggregate separation at optimal moisture, dried at 100C\",\n",
      "        \"store_cond\": \"other\",\n",
      "        \"samp_store_temp\": \"room temperature\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251709, -119.728663\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2020-05-21 00:00:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": \"0-1\",\n",
      "            \"has_unit\": \"m\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 500.0,\n",
      "            \"has_unit\": \"m\"\n",
      "        },\n",
      "        \"samp_collect_device\": \"11/2' x 6\\\"; stainless stee slide_hammer_core\",\n",
      "        \"sieving\": \"2mm\",\n",
      "        \"chem_administration\": \"1000\\u00b5g of C (glucose)/g soil\",\n",
      "        \"watering_regm\": \"75% water holding capacity; moisture maintained every 3 days\",\n",
      "        \"oxy_stat_samp\": \"aerobic\",\n",
      "        \"cur_land_use\": \"Horticultural Plants (e.g. tulips)\",\n",
      "        \"drainage_class\": \"Moderately Well\",\n",
      "        \"fao_class\": \"Andosols\",\n",
      "        \"biotic_relationship\": \"free living\",\n",
      "        \"horizon\": \"B Horizon\",\n",
      "        \"cur_vegetation\": \"tall wheat grass\",\n",
      "        \"ph\": 6.2,\n",
      "        \"ph_meth\": \"https://doi.org/10.2136/sssabookser5.3.c16\",\n",
      "        \"microbial_biomass\": \"0.026930154 mg of C/gram dry soil\",\n",
      "        \"microbial_biomass_meth\": \"https://doi.org/10.1016/0038-0717(87)90052-6 \",\n",
      "        \"water_content\": \"75% water holding capacity\",\n",
      "        \"water_content_meth\": \"https://doi.org/10.2136/sssabookser5.1.2ed.c13\",\n",
      "        \"annual_temp\": \"18.5 celsius\",\n",
      "        \"season_temp\": \"30.44 celsius\",\n",
      "        \"soil_type\": \"sandy loam\"\n",
      "    },\n",
      "    {\n",
      "        \"id\": \"UUID:516394-970067-847398\",\n",
      "        \"name\": \"T4-35\",\n",
      "        \"env_package\": \"soil\",\n",
      "        \"investigation_type\": \"lipidome|16S-Amplicon|ITS-Amplicon\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"d787487292564776a968fb4efb1002b5\",\n",
      "        \"growth_facil\": \"field\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"crushed via morter and pestle, temperature acclimation to 35C\",\n",
      "        \"store_cond\": \"lyophilized\",\n",
      "        \"samp_store_temp\": \"-80 degree Celsius\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251500, -119.728483\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2021-01-05T15:05:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": \"-20-30\",\n",
      "            \"has_unit\": \"cm\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 1000.0,\n",
      "            \"has_unit\": \"m\"\n",
      "        },\n",
      "        \"samp_collect_device\": \"11/2' x 6\\\"; stainless stee slide_hammer_core\",\n",
      "        \"sieving\": \"0.5-.25mm\",\n",
      "        \"air_temp_regm\": \"35C\",\n",
      "        \"oxy_stat_samp\": \"aerobic\",\n",
      "        \"cur_land_use\": \"Horticultural Plants (e.g. tulips)\",\n",
      "        \"drainage_class\": \"Moderately Well\",\n",
      "        \"fao_class\": \"Andosols\",\n",
      "        \"biotic_relationship\": \"free living\",\n",
      "        \"horizon\": \"O Horizon\",\n",
      "        \"cur_vegetation\": \"tall wheat grass\",\n",
      "        \"ph\": 6.2,\n",
      "        \"ph_meth\": \"https://doi.org/10.2136/sssabookser5.3.c16\",\n",
      "        \"microbial_biomass\": \"0.031331657 mg of C/gram dry soil\",\n",
      "        \"microbial_biomass_meth\": \"https://doi.org/10.1016/0038-0717(87)90052-6 \",\n",
      "        \"water_content\": \"0.5 g water/g soil\",\n",
      "        \"water_content_meth\": \"https://doi.org/10.1016/0016-7061(86)90018-2 \",\n",
      "        \"annual_temp\": \"18.5 celsius\",\n",
      "        \"season_temp\": \"30.44 celsius\",\n",
      "        \"soil_type\": \"sandy loam\"\n",
      "    },\n",
      "    {\n",
      "        \"id\": \"UUID:472894-473947-642384\",\n",
      "        \"name\": \"2020-05-21_S19\",\n",
      "        \"env_package\": \"soil\",\n",
      "        \"source_mat_id\": \"UUID:516394-970067-847398\",\n",
      "        \"investigation_type\": \"proteome|genome\",\n",
      "        \"type\": \"Biosample\",\n",
      "        \"part_of\": \"d787487292564776a968fb4efb1002b5\",\n",
      "        \"growth_facil\": \"field\",\n",
      "        \"geo_loc_name\": \"USA; Prosser, Washington; Washington State University-Irrigated Agriculture Research and Extension Center\",\n",
      "        \"samp_mat_process\": \"ground via omni bead beater\",\n",
      "        \"store_cond\": \"frozen\",\n",
      "        \"samp_store_temp\": \"-20 degree Celsius\",\n",
      "        \"env_broad_scale\": \"arid biome\",\n",
      "        \"env_local_scale\": \"agricultural field\",\n",
      "        \"env_medium\": \"agricultural soil\",\n",
      "        \"lat_lon\": \"46.251500, -119.728483\",\n",
      "        \"gold_path_field\": \"5424:Environmental:Terrestrial:Soil:Bulk soil:Agricultural land\",\n",
      "        \"collection_date\": \"2021-03-11T09:25:00\",\n",
      "        \"ncbi_taxonomy_name\": \"soil metagenome\",\n",
      "        \"depth\": {\n",
      "            \"has_raw_value\": \"-20-30\",\n",
      "            \"has_unit\": \"cm\"\n",
      "        },\n",
      "        \"elev\": {\n",
      "            \"has_raw_value\": 1000.0,\n",
      "            \"has_unit\": \"m\"\n",
      "        },\n",
      "        \"samp_collect_device\": \"11/2' x 6\\\"; stainless stee slide_hammer_core\",\n",
      "        \"sieving\": \"1-.05mm\",\n",
      "        \"oxy_stat_samp\": \"aerobic\",\n",
      "        \"cur_land_use\": \"Horticultural Plants (e.g. tulips)\",\n",
      "        \"drainage_class\": \"Moderately Well\",\n",
      "        \"fao_class\": \"Andosols\",\n",
      "        \"biotic_relationship\": \"free living\",\n",
      "        \"horizon\": \"O Horizon\",\n",
      "        \"cur_vegetation\": \"tall wheat grass\",\n",
      "        \"ph\": 6.2,\n",
      "        \"ph_meth\": \"https://doi.org/10.2136/sssabookser5.3.c16\",\n",
      "        \"microbial_biomass\": \"0.032336283 mg of C/gram dry soil\",\n",
      "        \"microbial_biomass_meth\": \"https://doi.org/10.1016/0038-0717(87)90052-6 \",\n",
      "        \"water_content\": \"0.5 g water/g soil\",\n",
      "        \"water_content_meth\": \"https://doi.org/10.1016/0016-7061(86)90018-2 \",\n",
      "        \"annual_temp\": \"18.5 celsius\",\n",
      "        \"season_temp\": \"30.44 celsius\",\n",
      "        \"soil_type\": \"sandy loam\"\n",
      "    }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "cumulative_json_merge_json = json.dumps(\n",
    "    cumulative_json_merge, indent=4, sort_keys=False, default=str\n",
    ")\n",
    "\n",
    "print(cumulative_json_merge_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce0f1dc3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
