{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import concurrent\n",
    "import os\n",
    "from pathlib import Path\n",
    "from time import time\n",
    "\n",
    "import jsonschema\n",
    "import requests\n",
    "from mongospawn.schema import dbschema_from_file, collschemas_for\n",
    "from pymongo import MongoClient, ReplaceOne\n",
    "from toolz import keyfilter\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from collections import defaultdict\n",
    "import csv\n",
    "from datetime import datetime\n",
    "from functools import partial, reduce\n",
    "import json\n",
    "import os\n",
    "import re\n",
    "from pprint import pprint\n",
    "from zipfile import ZipFile\n",
    "\n",
    "from dictdiffer import diff\n",
    "from pymongo import DeleteMany, DeleteOne, InsertOne, MongoClient, ReplaceOne, UpdateOne\n",
    "from toolz import assoc_in, compose, concat, dissoc, keyfilter, get_in, merge, merge_with\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "\n",
    "def reset_database(db):\n",
    "    for coll_name in collschemas:\n",
    "        db.drop_collection(coll_name)\n",
    "        db.create_collection(\n",
    "            coll_name, validator={\"$jsonSchema\": collschemas[coll_name]}\n",
    "        )\n",
    "        db[coll_name].create_index(\"id\", unique=True)\n",
    "\n",
    "\n",
    "def jsonschema_for(collection_name=None):\n",
    "    if collection_name not in set(dbschema[\"properties\"]):\n",
    "        raise ValueError(\n",
    "            f'collection_name must be one of {set(dbschema[\"properties\"])}'\n",
    "        )\n",
    "    defn = dbschema[\"properties\"][collection_name][\"items\"][\"$ref\"].split(\"/\")[-1]\n",
    "    return dbschema[\"definitions\"][defn]\n",
    "\n",
    "\n",
    "def validator_for(collection):\n",
    "    return collection.options()[\"validator\"][\"$jsonSchema\"]\n",
    "\n",
    "\n",
    "def pick(whitelist, d):\n",
    "    return keyfilter(lambda k: k in whitelist, d)\n",
    "\n",
    "\n",
    "def conform(doc, collection_name=None):\n",
    "    \"\"\"Provides limited, conservative conformance on a docments.\n",
    "\n",
    "    - If additionalProperties is False, omit any supplied.\n",
    "    - If a field must be a list of strings, and a lone string is supplied, wrap it in a list.\n",
    "\n",
    "    \"\"\"\n",
    "    if collection_name not in set(dbschema[\"properties\"]):\n",
    "        raise ValueError(\n",
    "            f'collection_name must be one of {set(dbschema[\"properties\"])}'\n",
    "        )\n",
    "    defn = dbschema[\"properties\"][collection_name][\"items\"][\"$ref\"].split(\"/\")[-1]\n",
    "    schema = dbschema[\"definitions\"][defn]\n",
    "    if schema.get(\"additionalProperties\") is False:\n",
    "        doc = pick(list(schema[\"properties\"]), doc)\n",
    "    for k in list(doc.keys()):\n",
    "        if (\n",
    "            isinstance(doc[k], str)\n",
    "            and schema[\"properties\"].get(k, {}).get(\"type\") == \"array\"\n",
    "            and schema[\"properties\"][k][\"items\"][\"type\"] == \"string\"\n",
    "            and not isinstance(doc[k], list)\n",
    "        ):\n",
    "            doc[k] = [doc[k]]\n",
    "    return doc\n",
    "\n",
    "\n",
    "def validate(doc, collection_name=None, conform_doc=False):\n",
    "    if collection_name not in set(dbschema[\"properties\"]):\n",
    "        raise ValueError(\n",
    "            f'collection_name must be one of {set(dbschema[\"properties\"])}'\n",
    "        )\n",
    "    if conform_doc:\n",
    "        doc = conform(doc, collection_name=collection_name)\n",
    "    jsonschema.validate({collection_name: [doc]}, schema=dbschema)\n",
    "    return doc\n",
    "\n",
    "\n",
    "def fetch_json(url):\n",
    "    return requests.get(url).json()\n",
    "\n",
    "\n",
    "def fetch_and_validate_json(resource, collection_name=None, conform_doc=False):\n",
    "    \"\"\"Takes a URL or the pre-fetched resource (list or dict)\"\"\"\n",
    "    payload = fetch_json(resource) if isinstance(resource, str) else resource\n",
    "    validated = []\n",
    "    if isinstance(payload, list):\n",
    "        for doc in tqdm(payload):\n",
    "            validated.append(\n",
    "                validate(doc, collection_name=collection_name, conform_doc=conform_doc)\n",
    "            )\n",
    "    elif isinstance(payload, dict):\n",
    "        if set(payload) & set(dbschema[\"properties\"]):\n",
    "            for collection_name, docs in payload.items():\n",
    "                for doc in tqdm(docs, desc=collection_name):\n",
    "                    validated.append(\n",
    "                        validate(\n",
    "                            doc,\n",
    "                            collection_name=collection_name,\n",
    "                            conform_doc=conform_doc,\n",
    "                        )\n",
    "                    )\n",
    "        else:\n",
    "            validated.append(\n",
    "                validate(\n",
    "                    payload, collection_name=collection_name, conform_doc=conform_doc\n",
    "                )\n",
    "            )\n",
    "    else:\n",
    "        raise ValueError(f\"Fetched JSON must be a JSON array or object\")\n",
    "    return validated\n",
    "\n",
    "\n",
    "def add_to_db(validated, db, collection_name=None):\n",
    "    if collection_name not in set(dbschema[\"properties\"]):\n",
    "        raise ValueError(\n",
    "            f'collection_name must be one of {set(dbschema[\"properties\"])}'\n",
    "        )\n",
    "    if isinstance(validated, list):\n",
    "        db[collection_name].bulk_write(\n",
    "            [ReplaceOne({\"id\": v[\"id\"]}, v, upsert=True) for v in validated]\n",
    "        )\n",
    "    elif isinstance(validated, dict):\n",
    "        if set(validated) & set(dbschema[\"properties\"]):\n",
    "            for collection_name, docs in validated.items():\n",
    "                db[collection_name].bulk_write(\n",
    "                    [ReplaceOne({\"id\": v[\"id\"]}, v, upsert=True) for v in docs]\n",
    "                )\n",
    "        else:\n",
    "            db[collection_name].bulk_write(\n",
    "                [ReplaceOne({\"id\": validated[\"id\"]}, validated, upsert=True)]\n",
    "            )\n",
    "    else:\n",
    "        raise ValueError(f\"payload must be a list or dict\")\n",
    "\n",
    "\n",
    "def fetch_conform_and_persist(spec, db):\n",
    "    url = spec[\"url\"]\n",
    "    collection_name = spec[\"type\"]\n",
    "    print(f\"fetching {url} ({collection_name})\")\n",
    "    payload = fetch_and_validate_json(url, collection_name, conform_doc=True)\n",
    "    add_to_db(payload, db, collection_name)\n",
    "\n",
    "\n",
    "def fetch_conform_and_persist_from_manifest(spec, db):\n",
    "    error_urls = []\n",
    "    url_manifest = spec[\"url_manifest\"]\n",
    "    collection_name = spec[\"type\"]\n",
    "    urls = fetch_json(url_manifest)\n",
    "\n",
    "    pbar = tqdm(total=len(urls))\n",
    "\n",
    "    with concurrent.futures.ThreadPoolExecutor() as executor:\n",
    "        future_to_url = {\n",
    "            executor.submit(\n",
    "                fetch_and_validate_json, url, collection_name, conform_doc=True\n",
    "            ): url\n",
    "            for url in urls\n",
    "        }\n",
    "        for future in concurrent.futures.as_completed(future_to_url):\n",
    "            pbar.update(1)\n",
    "            url = future_to_url[future]\n",
    "            try:\n",
    "                payload = future.result()\n",
    "            except Exception as e:\n",
    "                error_urls.append((url, str(e)))\n",
    "            else:\n",
    "                add_to_db(payload, db, collection_name)\n",
    "\n",
    "    pbar.close()\n",
    "    return error_urls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import tee\n",
    "\n",
    "def pairwise(iterable):\n",
    "    \"s -> (s0,s1), (s1,s2), (s2, s3), ...\"\n",
    "    a, b = tee(iterable)\n",
    "    next(b, None)\n",
    "    return zip(a, b)\n",
    "\n",
    "def coalesce_acronyms(set_name):\n",
    "    out = \"\"\n",
    "    for this_part, next_part in pairwise(set_name.split(\"_\")):\n",
    "        if len(this_part) == 1 and len(next_part) == 1:\n",
    "            out += this_part\n",
    "        elif next_part == \"set\":\n",
    "            out += this_part + \"_set\"\n",
    "        else:\n",
    "            out += this_part + \"_\"\n",
    "    return out\n",
    "\n",
    "def snake_case_set_name(object_name):\n",
    "    first_pass =  re.sub(r'(?<!^)(?=[A-Z])', '_', object_name).lower() + \"_set\"\n",
    "    return coalesce_acronyms(first_pass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "nmdc_schema_json_path = str(\n",
    "    Path.cwd().parent.parent.joinpath(\"schema\", \"nmdc.schema.json\")\n",
    ")\n",
    "dbschema = dbschema_from_file(nmdc_schema_json_path)\n",
    "\n",
    "###########################\n",
    "# Adjustments for GSP below\n",
    "###########################\n",
    "\n",
    "defined_object_names = set(dbschema[\"definitions\"])\n",
    "\n",
    "set_for_object_name = {\n",
    "    spec[\"items\"][\"$ref\"].split(\"#/definitions/\")[-1]: set_name\n",
    "    for set_name, spec in dbschema[\"properties\"].items()\n",
    "}\n",
    "\n",
    "existing_set_names = set(dbschema[\"properties\"])\n",
    "\n",
    "for object_without_set in (defined_object_names - set(set_for_object_name.keys())):\n",
    "    proposed_set_name = snake_case_set_name(object_without_set)\n",
    "    if proposed_set_name not in existing_set_names:\n",
    "        dbschema[\"properties\"][proposed_set_name] = {\n",
    "            \"description\": (f\"This property links a database object to the set of\"\n",
    "                            f\" {object_without_set} objects within it.\"),\n",
    "            \"items\": {\"$ref\": f\"#/definitions/{object_without_set}\"},\n",
    "            \"type\": \"array\",\n",
    "        }\n",
    "        \n",
    "del dbschema[\"definitions\"][\"OmicsProcessing\"][\"additionalProperties\"]\n",
    "del dbschema[\"definitions\"][\"Biosample\"][\"additionalProperties\"]\n",
    "del dbschema[\"definitions\"][\"ReadQCAnalysisActivity\"][\"additionalProperties\"]\n",
    "dbschema = assoc_in(dbschema, [\"definitions\", \"ControlledTermValue\", \"properties\", \"term\", \"type\"], \"string\")\n",
    "del dbschema[\"definitions\"][\"ControlledTermValue\"][\"properties\"][\"term\"][\"$ref\"]\n",
    "dbschema = assoc_in(dbschema, [\"definitions\", \"MetagenomeAssembly\", \"properties\", \"scaf_l_gt50k\", \"type\"], \"number\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "collschemas = collschemas_for(dbschema)\n",
    "\n",
    "# Reconstruct\n",
    "set_for_object_name = {\n",
    "    spec[\"items\"][\"$ref\"].split(\"#/definitions/\")[-1]: set_name\n",
    "    for set_name, spec in dbschema[\"properties\"].items()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(Re-)load existing NMDC DB from file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "with ZipFile('../src/data/nmdc_database.json.zip') as myzip:\n",
    "    with myzip.open('nmdc_database.json') as f:\n",
    "        nmdc_database = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['activity_set',\n",
       " 'agent_set',\n",
       " 'attribute_value_set',\n",
       " 'biosample_processing_set',\n",
       " 'biosample_set',\n",
       " 'boolean_value_set',\n",
       " 'chemical_entity_set',\n",
       " 'controlled_term_value_set',\n",
       " 'data_object_set',\n",
       " 'database_set',\n",
       " 'environmental_material_term_set',\n",
       " 'foo',\n",
       " 'functional_annotation_set',\n",
       " 'gene_product_set',\n",
       " 'genome_feature_set',\n",
       " 'geolocation_value_set',\n",
       " 'instrument_set',\n",
       " 'integer_value_set',\n",
       " 'mag_bin_set',\n",
       " 'mags_activity_set',\n",
       " 'metabolite_quantification_set',\n",
       " 'metabolomics_analysis_activity_set',\n",
       " 'metagenome_annotation_activity_set',\n",
       " 'metagenome_assembly_set',\n",
       " 'metaproteomics_analysis_activity_set',\n",
       " 'nom_analysis_activity_set',\n",
       " 'omics_processing_set',\n",
       " 'ontology_class_set',\n",
       " 'orthology_group_set',\n",
       " 'pathway_set',\n",
       " 'peptide_quantification_set',\n",
       " 'person_set',\n",
       " 'person_value_set',\n",
       " 'protein_quantification_set',\n",
       " 'quantity_value_set',\n",
       " 'reaction_participant_set',\n",
       " 'reaction_set',\n",
       " 'read_QC_analysis_activity_set',\n",
       " 'read_based_analysis_activity_set',\n",
       " 'study_set',\n",
       " 'text_value_set',\n",
       " 'timestamp_value_set',\n",
       " 'url_value_set']"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "client = MongoClient(\n",
    "    host=os.getenv(\"NMDC_MONGO_HOST\"),\n",
    "    username=\"dwinston_rw\",\n",
    "    password=os.getenv(\"NMDC_MONGO_RW_PWD\"))\n",
    "dbname = \"dwinston_scratch\"\n",
    "db = client[dbname]\n",
    "\n",
    "reset_database(db)\n",
    "#for collection in nmdc_database:\n",
    "#    db[collection].insert_many(nmdc_database[collection])\n",
    "sorted(db.list_collection_names())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "study_set {'Study'}\n",
      "omics_processing_set {'OmicsProcessing'}\n",
      "biosample_set {'Biosample'}\n",
      "some data_object_set docs have no type\n",
      "data_object_set {'DataObject'}\n",
      "activity_set {'MetaProteomicAnalysis', 'MetagenomeAssembly', 'ReadQCAnalysisActivity', 'MetagenomeAnnotation'}\n"
     ]
    }
   ],
   "source": [
    "target_collection = {\n",
    "    \"Study\": \"study_set\",\n",
    "    \"OmicsProcessing\": \"omics_processing_set\",\n",
    "    \"Biosample\": \"biosample_set\",\n",
    "    \"DataObject\": \"data_object_set\",\n",
    "    \"MetagenomeAssembly\": \"metagenome_assembly_set\",\n",
    "    \"MetaProteomicAnalysis\": \"metaproteomics_analysis_activity_set\",\n",
    "    \"MetagenomeAnnotation\": \"metagenome_annotation_activity_set\",\n",
    "    \"ReadQCAnalysisActivity\": \"read_QC_analysis_activity_set\",\n",
    "    \"\": \"data_object_set\",\n",
    "}\n",
    "for collection in nmdc_database:\n",
    "    docs = nmdc_database[collection]\n",
    "    object_types = {d.get(\"type\", \"nmdc:\")[5:] for d in docs} - {\"\"}\n",
    "    if any(d for d in docs if \"type\" not in d):\n",
    "        print(\"some\",collection,\"docs have no type\")\n",
    "    print(collection, object_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "study_set\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b2978575b4d848f8b68148c901c6bc93",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/13 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "omics_processing_set\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "281b0e10e4424c01a879944305df1b8f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/7454 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "for source_collection in nmdc_database:\n",
    "    docs = nmdc_database[source_collection]\n",
    "    docs_per_target = defaultdict(list)\n",
    "    for d in docs:\n",
    "        type_ = d.get(\"type\", \"nmdc:\")[5:]\n",
    "        d_new = dissoc(d, \"type\")\n",
    "        \n",
    "        if \"lat_lon\" in d_new:\n",
    "            d_new[\"lat_lon\"].pop(\"type\", None)\n",
    "        for k_float in (\n",
    "            \"asm_score\", \"ctg_logsum\", \"ctg_powsum\", \"gap_pct\", \"gc_avg\", \"gc_std\",\n",
    "            \"scaf_logsum\", \"scaf_powsum\"):\n",
    "            if k_float in d_new:\n",
    "                d_new[k_float] = float(d_new[k_float]) \n",
    "        keys_with_term_ids = [\n",
    "            k for k in d_new\n",
    "            if isinstance(d_new[k], dict)\n",
    "            and \"term\" in d_new[k]\n",
    "            and \"id\" in d_new[k][\"term\"]\n",
    "        ]\n",
    "        for k in keys_with_term_ids:\n",
    "            d_new = assoc_in(d_new, [k, \"term\"], d_new[k][\"term\"][\"id\"])\n",
    "            \n",
    "        docs_per_target[target_collection[type_]].append(d_new)\n",
    "                \n",
    "    for collection_name, docs in docs_per_target.items():\n",
    "        print(collection_name)\n",
    "        payload = fetch_and_validate_json(docs, collection_name=collection_name)\n",
    "        add_to_db(payload, db, collection_name=collection_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "study_set: `Additional properties are not allowed ('principal_investigator_name', 'type' were unexpected)`\n",
    "\n",
    "omics_processing_set: `Additional properties are not allowed ('principal_investigator_name', 'mod_date', 'add_date', 'processing_institution', 'ncbi_project_name' were unexpected)`\n",
    "\n",
    "biosample_set: \n",
    "\n",
    "-`Additional properties are not allowed ('location', 'mod_date', 'identifier', 'habitat', 'ncbi_taxonomy_name', 'add_date', 'community', 'sample_collection_site' were unexpected)`\n",
    "\n",
    "-`Additional properties are not allowed ('type' was unexpected)` for `lat_lon`\n",
    "\n",
    "data_object_set: OK\n",
    "\n",
    "activity_set:\n",
    "- `Additional properties are not allowed ('contigs', 'ctg_N50', 'scaf_max', 'contig_bp', 'scaffolds', 'gc_std', 'gap_pct', 'num_input_reads', 'scaf_N50', 'ctg_max', 'ctg_L90', 'ctg_powsum', 'gc_avg', 'scaf_powsum', 'scaf_L90', 'ctg_L50', 'scaf_l_gt50k', 'scaf_N90', 'ctg_N90', 'scaf_L50', 'asm_score', 'scaf_n_gt50K', 'scaf_pct_gt50K', 'num_aligned_reads', 'scaf_bp', 'ctg_logsum', 'scaf_logsum' were unexpected)` for type 'nmdc:MetagenomeAssembly'\n",
    "\n",
    "- `Additional properties are not allowed ('input_read_count', 'output_read_count', 'input_read_bases', 'output_read_bases' were unexpected)` for type 'nmdc:ReadQCAnalysisActivity'\n",
    "\n",
    "metagenome_assembly_set:\n",
    "`Additional properties are not allowed ('scaf_l_gt50k' was unexpected)`\n",
    "\n",
    "read_QC_analysis_activity_set:\n",
    "- `Additional properties are not allowed ('output_read_bases', 'input_read_bases' were unexpected)`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load FICUS Brodie spreadsheet and create gold-id-to-igsn map."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "GOLD_ID_IDX = 5\n",
    "IGSN_IDX = 2\n",
    "\n",
    "igsn_golds = defaultdict(list)\n",
    "\n",
    "gold_id_pattern = re.compile(r\"Gb\\d+\")\n",
    "\n",
    "with open('../src/data/FICUS_Soil_Gs0135149_Brodie-12-23-2020_PS.xlsx - Brodie_Gs0135149_Soil_Metadata.csv') as f:\n",
    "    reader = csv.reader(f)\n",
    "    for row in reader:\n",
    "        gold_id = row[GOLD_ID_IDX]\n",
    "        igsn = row[IGSN_IDX]\n",
    "        if gold_id_pattern.fullmatch(gold_id):\n",
    "            igsn_golds[igsn].append(gold_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare helper function to compare timestamps given in e.g. \"15-MAY-20 08.30.01.000000000 am\" format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "dt_pattern = re.compile(r\"\\d{2}-(?P<month>\\w+)-\\d{2} \\d{2}\\.\\d{2}\\.\\d{2}\\.(?P<ns>\\d+) [A|P]M\")\n",
    "dt_format = \"%d-%b-%y %I.%M.%S.%f %p\"\n",
    "\n",
    "def order_timestamps(timestamps):\n",
    "    if not all(isinstance(ts, str) for ts in timestamps):\n",
    "        raise Exception(f\"{timestamps} not strings\")\n",
    "    as_datetimes = []\n",
    "    for ts in timestamps:\n",
    "        match = dt_pattern.search(ts)\n",
    "        first, month, rest = ts.partition(match.group(\"month\"))\n",
    "        ts_new = first + month[0] + month[1:].lower() + rest\n",
    "        ts_new = ts_new.replace(match.group(\"ns\"), match.group(\"ns\")[:-3]) # truncate to microseconds\n",
    "        as_datetimes.append(datetime.strptime(ts_new, dt_format))\n",
    "    sorted_dts = sorted(as_datetimes)\n",
    "    return [dt.strftime(dt_format) for dt in sorted_dts]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare helper-function pipeline to unify biosample_set documents that should be considered equivalent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "er_xna_pattern = re.compile(r\"ER_[D|R]NA_\\d+$\")\n",
    "\n",
    "def rstrip_name_ER_ID(d):\n",
    "    s = get_in([\"name\"], d)\n",
    "    s_new = er_xna_pattern.split(s)[0] if er_xna_pattern.search(s) else s\n",
    "    return assoc_in(d, [\"name\"], s_new)\n",
    "\n",
    "def capitalize_location_raw_value(d):\n",
    "    s = get_in([\"location\", \"has_raw_value\"], d)\n",
    "    s_new = s[0].upper() + s[1:]\n",
    "    return assoc_in(d, [\"location\", \"has_raw_value\"], s_new)\n",
    "\n",
    "pipeline = compose(\n",
    "    capitalize_location_raw_value,\n",
    "    rstrip_name_ER_ID,\n",
    "    lambda d: dissoc(d, \"_id\", \"id\", \"add_date\", \"mod_date\", \"identifier\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Produce new biosample objects with ISGN ids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_biosample_docs = []\n",
    "\n",
    "for igsn, golds in igsn_golds.items():\n",
    "    igsn_curie = \"igsn:\"+igsn\n",
    "    to_change = list(db.biosample_set.find({\"id\": {\"$in\": [f\"gold:{g}\" for g in golds]}}))\n",
    "    \n",
    "    # No merge needed, just change of id.\n",
    "    if len(to_change) == 1:\n",
    "        merged = assoc_in(to_change[0], [\"id\"], igsn_curie)\n",
    "        merged = assoc_in(merged, [\"identifier\", \"has_raw_value\"], igsn_curie)\n",
    "        merged_biosample_docs.append(merged)\n",
    "        continue\n",
    "    elif len(to_change) == 0:\n",
    "        continue\n",
    "\n",
    "    # Ensure that unification pipeline is adequate to resolve differences.\n",
    "    distilled = list(map(pipeline, to_change))\n",
    "    result = list(diff(distilled[0], distilled[1]))\n",
    "    assert result == []\n",
    "    \n",
    "    # Produce a merged document\n",
    "    earlier_ts, _ = order_timestamps([get_in([\"add_date\", \"has_raw_value\"], d) for d in to_change])\n",
    "    merged = assoc_in(distilled[0], [\"add_date\", \"has_raw_value\"], earlier_ts)\n",
    "    _, later_ts = order_timestamps([get_in([\"mod_date\", \"has_raw_value\"], d) for d in to_change])\n",
    "    merged = assoc_in(merged, [\"mod_date\", \"has_raw_value\"], later_ts)\n",
    "    merged = assoc_in(merged, [\"id\"], igsn_curie)\n",
    "    merged = assoc_in(merged, [\"identifier\", \"has_raw_value\"], igsn_curie)\n",
    "    \n",
    "    merged_biosample_docs.append(merged)\n",
    "    merged = None # defense against accidental reuse during next iteration.\n",
    "\n",
    "assert len(merged_biosample_docs) == len(igsn_golds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Delete old biosample objects and insert new ones in one bulk-write operation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(93, 48)"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "requests = [DeleteMany({\"id\": {\"$in\": [\"gold:\"+g for g in concat(igsn_golds.values())]}})]\n",
    "requests.extend([InsertOne(d) for d in merged_biosample_docs])\n",
    "result = db.biosample_set.bulk_write(requests)\n",
    "result.deleted_count, result.inserted_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Update omics_processing_set references to biosample_set ids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "goldid_igsn = {}\n",
    "for igsn, gids in igsn_golds.items():\n",
    "    for gid in gids:\n",
    "        goldid_igsn[gid] = igsn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "requests = []\n",
    "to_replace = {\"gold:\"+k: \"igsn:\"+v for k, v in goldid_igsn.items()}\n",
    "\n",
    "for doc in db.omics_processing_set.find({\"has_input\": {\"$in\": list(to_replace)}}):\n",
    "    operations = {\"$set\": {\n",
    "        \"has_input\": [to_replace.get(i, i) for i in doc[\"has_input\"]],\n",
    "    }}\n",
    "    requests.append({\"filter\": {\"_id\": doc[\"_id\"]}, \"update\": operations})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "rv = db.omics_processing_set.bulk_write([UpdateOne(**r) for r in requests])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "93"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rv.modified_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Update omics_processing_set references from EMSL ids to IGSNs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "EMSL_IDS_IDX = 7\n",
    "IGSN_IDX = 2\n",
    "\n",
    "igsn_emsls = {}\n",
    "\n",
    "emsl_ids_pattern = re.compile(r\"\\d+\")\n",
    "\n",
    "with open('../src/data/FICUS_Soil_Gs0135149_Brodie-12-23-2020_PS.xlsx - Brodie_Gs0135149_Soil_Metadata.csv') as f:\n",
    "    reader = csv.reader(f)\n",
    "    for row in reader:\n",
    "        emsl_ids = row[EMSL_IDS_IDX]\n",
    "        igsn = row[IGSN_IDX]\n",
    "        ids = emsl_ids_pattern.findall(emsl_ids)\n",
    "        # XXX some rows have emsl ids but no IGSN, so igsn.strip() check here\n",
    "        if igsn.strip() and ids:\n",
    "            igsn_emsls[igsn] = ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "emslid_igsn = {}\n",
    "for igsn, eids in igsn_emsls.items():\n",
    "    for eid in eids:\n",
    "        emslid_igsn[eid] = igsn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_with_emsl_id = db.omics_processing_set.count_documents(\n",
    "    {\"id\": {\"$in\": [\"emsl:\"+i for i in emslid_igsn]}})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "requests = []\n",
    "to_replace = {\"emsl:\"+k: \"igsn:\"+v for k, v in emslid_igsn.items()}\n",
    "to_replace.update({\"emsl:output_\"+k: \"igsn:\"+v for k, v in emslid_igsn.items()})\n",
    "\n",
    "def omit(blacklist, d):\n",
    "    return keyfilter(lambda k: k not in blacklist, d)\n",
    "\n",
    "def sans_mongo_id(d):\n",
    "    return omit([\"_id\"], d)\n",
    "\n",
    "\n",
    "for doc in db.omics_processing_set.find({\"has_input\": {\"$in\": list(to_replace)}}):\n",
    "    operations = {\"$set\": {\n",
    "        \"has_input\": [to_replace.get(i, i) for i in doc[\"has_input\"]],\n",
    "    }}\n",
    "    requests.append({\"filter\": {\"_id\": doc[\"_id\"]}, \"update\": operations})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "if requests:\n",
    "    rv = db.omics_processing_set.bulk_write([UpdateOne(**r) for r in requests])\n",
    "    print(rv.modified_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'additionalProperties': False,\n",
       " 'description': 'A material sample. It may be environmental (encompassing many organisms) or isolate or tissue.   An environmental sample containing genetic material from multiple individuals is commonly referred to as a biosample.',\n",
       " 'properties': {'agrochem_addition': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'al_sat': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'al_sat_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'alt': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'alternate_identifiers': {'description': 'The same biosample may have distinct identifiers in different databases (e.g. GOLD and EMSL)',\n",
       "   'items': {'type': 'string'},\n",
       "   'type': 'array'},\n",
       "  'annual_precpt': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'annual_temp': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'chlorophyll': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'collection_date': {'additionalProperties': False,\n",
       "   'description': 'A value that is a timestamp. The range should be ISO-8601',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TimestampValue',\n",
       "   'type': 'object'},\n",
       "  'crop_rotation': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'cur_land_use': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'cur_vegetation': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'cur_vegetation_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'depth': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'description': {'description': 'a human-readable description of a thing',\n",
       "   'type': 'string'},\n",
       "  'diss_oxygen': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'drainage_class': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'ecosystem': {'additionalProperties': False,\n",
       "   'description': 'The value for any value of a attribute for a sample. This object can hold both the un-normalized atomic value and the structured value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'AttributeValue',\n",
       "   'type': 'object'},\n",
       "  'ecosystem_category': {'additionalProperties': False,\n",
       "   'description': 'The value for any value of a attribute for a sample. This object can hold both the un-normalized atomic value and the structured value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'AttributeValue',\n",
       "   'type': 'object'},\n",
       "  'ecosystem_subtype': {'additionalProperties': False,\n",
       "   'description': 'The value for any value of a attribute for a sample. This object can hold both the un-normalized atomic value and the structured value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'AttributeValue',\n",
       "   'type': 'object'},\n",
       "  'ecosystem_type': {'additionalProperties': False,\n",
       "   'description': 'The value for any value of a attribute for a sample. This object can hold both the un-normalized atomic value and the structured value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'AttributeValue',\n",
       "   'type': 'object'},\n",
       "  'elev': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'env_broad_scale': {'additionalProperties': False,\n",
       "   'description': 'A controlled term or class from an ontology',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'term': {'description': 'pointer to an ontology class', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'ControlledTermValue',\n",
       "   'type': 'object'},\n",
       "  'env_local_scale': {'additionalProperties': False,\n",
       "   'description': 'A controlled term or class from an ontology',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'term': {'description': 'pointer to an ontology class', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'ControlledTermValue',\n",
       "   'type': 'object'},\n",
       "  'env_medium': {'additionalProperties': False,\n",
       "   'description': 'A controlled term or class from an ontology',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'term': {'description': 'pointer to an ontology class', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'ControlledTermValue',\n",
       "   'type': 'object'},\n",
       "  'env_package': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'extreme_event': {'additionalProperties': False,\n",
       "   'description': 'A value that is a timestamp. The range should be ISO-8601',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TimestampValue',\n",
       "   'type': 'object'},\n",
       "  'fao_class': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'fire': {'additionalProperties': False,\n",
       "   'description': 'A value that is a timestamp. The range should be ISO-8601',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TimestampValue',\n",
       "   'type': 'object'},\n",
       "  'flooding': {'additionalProperties': False,\n",
       "   'description': 'A value that is a timestamp. The range should be ISO-8601',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TimestampValue',\n",
       "   'type': 'object'},\n",
       "  'geo_loc_name': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'heavy_metals': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'heavy_metals_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'horizon': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'horizon_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'id': {'description': 'The primary identifier for the biosample. Example: GOLD:Gb0205609',\n",
       "   'type': 'string'},\n",
       "  'lat_lon': {'additionalProperties': False,\n",
       "   'description': \"A normalized value for a location on the earth's surface\",\n",
       "   'properties': {'has_raw_value': {'description': 'The raw value for a  geolocation should follow {lat} {long}',\n",
       "     'type': 'string'},\n",
       "    'latitude': {'description': 'latitude', 'type': 'number'},\n",
       "    'longitude': {'description': 'longitude', 'type': 'number'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'GeolocationValue',\n",
       "   'type': 'object'},\n",
       "  'link_addit_analys': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'link_class_info': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'link_climate_info': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'local_class': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'local_class_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'microbial_biomass': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'microbial_biomass_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'misc_param': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'name': {'description': 'A human readable name or description of the biosample',\n",
       "   'type': 'string'},\n",
       "  'nitrate': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'ph_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'pool_dna_extracts': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'previous_land_use': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'previous_land_use_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'profile_position': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'salinity': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'salinity_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'samp_vol_we_dna_ext': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'season_precpt': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'season_temp': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'sieving': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'slope_aspect': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'slope_gradient': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'soil_type': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'soil_type_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'specific_ecosystem': {'additionalProperties': False,\n",
       "   'description': 'The value for any value of a attribute for a sample. This object can hold both the un-normalized atomic value and the structured value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'AttributeValue',\n",
       "   'type': 'object'},\n",
       "  'store_cond': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'texture': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'texture_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'tillage': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'tot_nitro_content': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'tot_nitro_content_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'tot_org_c_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  'tot_org_carb': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'water_content': {'additionalProperties': False,\n",
       "   'description': 'A simple quantity, e.g. 2cm',\n",
       "   'properties': {'has_numeric_value': {'description': 'The number part of the quantity',\n",
       "     'type': 'number'},\n",
       "    'has_raw_value': {'description': 'Unnormalized atomic string representation, should in syntax {number} {unit}',\n",
       "     'type': 'string'},\n",
       "    'has_unit': {'description': 'The unit of the quantity', 'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'QuantityValue',\n",
       "   'type': 'object'},\n",
       "  'water_content_soil_meth': {'additionalProperties': False,\n",
       "   'description': 'A basic string value',\n",
       "   'properties': {'has_raw_value': {'description': 'The value that was specified for an annotation in raw form, i.e. a string. E.g. \"2 cm\" or \"2-4 cm\"',\n",
       "     'type': 'string'},\n",
       "    'language': {'description': 'Should use ISO 639-1 code e.g. \"en\", \"fr\"',\n",
       "     'type': 'string'},\n",
       "    'was_generated_by': {'type': 'string'}},\n",
       "   'title': 'TextValue',\n",
       "   'type': 'object'},\n",
       "  '_id': {'bsonType': 'objectId'}},\n",
       " 'required': ['lat_lon',\n",
       "  'env_broad_scale',\n",
       "  'env_local_scale',\n",
       "  'env_medium',\n",
       "  'id'],\n",
       " 'title': 'Biosample',\n",
       " 'type': 'object'}"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validator_for(db_share.biosample_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "admin_client = MongoClient(\n",
    "    host=os.getenv(\"NMDC_MONGO_HOST\"),\n",
    "    username=\"nmdc-admin\",\n",
    "    password=os.getenv(\"NMDC_MONGO_ADMIN_PWD\")\n",
    ")\n",
    "admin_dwinston_share = admin_client[\"dwinston_share\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_collection_names = sorted(set(target_collection.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reset_database_schema(db):\n",
    "    for coll_name in target_collection_names:\n",
    "        if coll_name not in db.list_collection_names():\n",
    "            db.create_collection(\n",
    "                coll_name, validator={\"$jsonSchema\": collschemas[coll_name]}\n",
    "            )\n",
    "            db[coll_name].create_index(\"id\", unique=True)\n",
    "        else:\n",
    "            print(coll_name)\n",
    "            db.command(\"collMod\", coll_name, validator={\"$jsonSchema\": collschemas[coll_name]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biosample_set\n",
      "data_object_set\n",
      "metagenome_annotation_activity_set\n",
      "metagenome_assembly_set\n",
      "metaproteomics_analysis_activity_set\n",
      "omics_processing_set\n",
      "read_QC_analysis_activity_set\n",
      "study_set\n"
     ]
    }
   ],
   "source": [
    "reset_database_schema(admin_dwinston_share)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biosample_set\n",
      "data_object_set\n",
      "metagenome_annotation_activity_set\n",
      "metagenome_assembly_set\n",
      "metaproteomics_analysis_activity_set\n",
      "omics_processing_set\n",
      "read_QC_analysis_activity_set\n",
      "study_set\n"
     ]
    }
   ],
   "source": [
    "db_share = client[\"dwinston_share\"]\n",
    "target_collection_names = sorted(set(target_collection.values()))\n",
    "for name in target_collection_names:\n",
    "    docs = [sans_mongo_id(d) for d in db[name].find()]\n",
    "    print(name)\n",
    "    add_to_db(docs, db_share, collection_name=name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nmdc",
   "language": "python",
   "name": "nmdc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
