<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>NMDC Metadata Standards Documentation - NMDC Metadata Documentation</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "NMDC Metadata Standards Documentation";
    var mkdocs_page_input_path = "Metadata_Documentation_Overview.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> NMDC Metadata Documentation</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">NMDC Metadata Standards Documentation</a>
    <ul class="current">
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../schema-validation/">NMDC Schema Validation</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../identifiers/">Identifiers in NMDC</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NMDC_MIxS_Soil_documentation/">MIxS Soil Package</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">NMDC Metadata Documentation</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>NMDC Metadata Standards Documentation</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="the-nmdc-metadata-standards-documentation">The NMDC Metadata Standards Documentation</h1>
<h1 id="introduction">Introduction</h1>
<p>This documentation provides details on the National Microbiome Data
Collaborative’s (<a href="http://microbiomedata.org"><u>NMDC</u></a>) approach to
sample and data processing metadata. These are key features that drive
the data search and discovery aspect of the NMDC data portal
(<a href="https://microbiomedata.org/data/"><u>https://microbiomedata.org/data/</u></a>).
If you are unfamiliar with these types of metadata (Figure 1), we
recommend you begin with an <em>Introduction to Metadata and Ontologies:
Everything You Always Wanted to Know About Metadata and Ontologies (But
Were Afraid to Ask)</em>
(<a href="https://doi.org/10.25979/1607365"><u>https://doi.org/10.25979/1607365</u></a>).</p>
<div align="center">
    <img src="../images/NMDC_metadata_img1.png" style="width:80%"/>
</div>

<p>Figure 1: Microbiome metadata types: Information that contextualizes
sample including its geographic location and collection date, sample
preparation, data processing methods, and data products produced from a
biological sample (Luke et al., 2020. Introduction to Metadata and
Ontologies: Everything You Always Wanted to Know About Metadata and
Ontologies (But Were Afraid to Ask). DOI: 10.25979/1607365).</p>
<p>All data integrated into the NMDC data portal must adhere to existing
metadata standards for proper indexing and display, and to ensure
accurate search results are returned. This documentation outlines the
standards and ontologies that were included in the NMDC data schema, a
framework that defines how data were defined and linked. For the
2019-2022 pilot initiative, the NMDC Metadata Standards Team (see the
<a href="https://microbiomedata.org/team/"><u>NMDC Team page</u></a>) leveraged
existing community-driven standards developed by the <a href="https://gensc.org/"><u>Genomics
Standards Consortium</u></a> (GSC), the Joint Genome
Institute (JGI) <a href="https://gold.jgi.doe.gov/"><u>Genomes Online
Database</u></a> (GOLD), and OBO Foundry’s
<a href="http://www.obofoundry.org/ontology/envo.html"><u>Environmental
Ontology</u></a> (EnvO). In
collaboration with these organizations, the NMDC has created a framework
for mapping these standards into an interoperable framework that can be
expanded to include additional standards and ontologies in the future.</p>
<p>Additional information on the activities by the NMDC Metadata Standards
team can be found on the NMDC website at:
<a href="https://microbiomedata.org/metadata/"><u>https://microbiomedata.org/metadata/</u></a></p>
<h1 id="standards-and-ontologies-used-by-the-nmdc">Standards and Ontologies used by the NMDC</h1>
<h2 id="sample-metadata">Sample Metadata</h2>
<h3 id="gsc-minimum-information-about-any-x-sequence-mixs">GSC Minimum Information about any (x) Sequence (MIxS)</h3>
<p>The GSC has developed standards for describing genomic and metagenomic
sequences, including the “minimum information about a genome sequence”
(<a href="https://pubmed.ncbi.nlm.nih.gov/18464787/"><u>MIGS</u></a>), the “minimum
information about a metagenome sequence”
(<a href="https://pubmed.ncbi.nlm.nih.gov/18464787/"><u>MIMS</u></a>), and the
“minimum information about a marker gene sequence”
(<a href="https://pubmed.ncbi.nlm.nih.gov/21552244/"><u>MIMARKS</u></a>). To
complement this community-driven standard effort, the GSC has also
developed a system for describing the environment from which a
biological sample originates, as “environmental packages” and
established a unified standard set of checklists through the minimum
information about any (x) sequence (<u>MIxS</u>). MIxS provides a
standardized data dictionary of sample descriptors (e.g., location,
environment, elevation, altitude, depth, etc.) organized into different
packages for 17 different sample environments.</p>
<p>To standardize how physical samples are described (i.e., sample
metadata, Figure 1), the NMDC schema includes environmental descriptors
from the GSC MIxS standards.</p>
<p><em>Explore how to create a MIxS-compliant sample metadata spreadsheet</em></p>
<ul>
<li>
<p>Review our example spreadsheet with sample metadata that has been
    converted to be compliant with the MIxS Soil environment package.
    Note that not all non-mandatory terms from the MIxs Soil package
    were relevant for these example samples, and hence were omitted for
    clarity.</p>
<ul>
<li>
<p><a href="https://docs.google.com/spreadsheets/d/1i2w2CEEHiMJZesi984LyU-ayaHKNFOCCN0TcPmKFda0/edit?usp=sharing"><u>Basic sample
    spreadsheet</u></a>
    (Tab 1 - before conversion to MIxS)</p>
</li>
<li>
<p><a href="https://docs.google.com/spreadsheets/d/1i2w2CEEHiMJZesi984LyU-ayaHKNFOCCN0TcPmKFda0/edit?usp=sharing"><u>MIxS-compliant soil
    spreadsheet</u></a>
    (Tab 2 - converted to MIxS Soil)</p>
</li>
</ul>
</li>
<li>
<p>Explore the mandatory, unique, and shared descriptors from the
    <a href="https://docs.google.com/document/d/1oNlMNQySuCoEeqhf1Qou8D-BV5bE76TkjrJLya8Ehw4/edit"><u>MIxS Soil
    package</u></a></p>
</li>
<li>
<p>Searchable descriptors from <strong><em>all</em></strong> MIxS environmental packages <em>-
    coming soon!</em></p>
</li>
<li>
<p>Learn more about all of the <a href="https://gensc.org/mixs"><u>17 MIxS environmental
    packages</u></a></p>
</li>
</ul>
<h3 id="genomes-online-database-gold">Genomes Online Database (GOLD)</h3>
<p>The JGI <a href="https://gold.jgi.doe.gov/"><u>Genomes OnLine Database</u></a>
(GOLD, <a href="https://pubmed.ncbi.nlm.nih.gov/33152092/"><u>Mukherjee
2021</u></a>) is an open-access
repository of genome, metagenome, and metatranscriptome sequencing
projects with their associated metadata. GOLD data are organized based
on Study, Biosample/Organism, Sequencing Project and Analysis Project
(<a href="https://pubmed.ncbi.nlm.nih.gov/30357420/"><u>Mukherjee 2017</u></a>).
Biosamples (defined as the physical material collected from an
environment) are described using a five-level ecosystem classification
<a href="https://pubmed.ncbi.nlm.nih.gov/20653767/">path (Figure 2)</a>; the NMDC
schema also uses this ecosystem classification to describe sample
environments.</p>
<div align="center">
<img src="../images/NMDC_metadata_img2.png" style="width:100%"/>
</div>

<p>Figure 2. The GOLD five-level ecosystem classification paths
(<a href="https://pubmed.ncbi.nlm.nih.gov/33152092/"><u>Mukherjee 2019</u></a>).</p>
<p><em>Overview of the GOLD ecosystem paths</em></p>
<ul>
<li>
<p><strong><em>Ecosystem</em></strong> describes biosamples using three different broadest
    contexts, namely environmental, engineered, and host-associated.</p>
</li>
<li>
<p><strong><em>Ecosystem category</em></strong> subdivides the ecosystem into categories,
    such as aquatic or terrestrial.</p>
</li>
<li>
<p><strong><em>Ecosystem type</em></strong> classifies those categories into types, such as
    freshwater or marine, cave, desert, soil, etc.</p>
</li>
<li>
<p><strong><em>Ecosystem subtype</em></strong> allows for additional environmental context
    or boundaries.</p>
</li>
<li>
<p><strong><em>Specific ecosystem</em></strong> that describes the environment that
    directly influences the sample or the environmental material itself.</p>
</li>
</ul>
<p><em>Explore how to map sample environments using the GOLD ecosystem
classification</em></p>
<ul>
<li>
<p>Learn more about the GOLD ecosystem paths using an <a href="https://gold.jgi.doe.gov/ecosystemtree"><u>interactive
    visualization tool</u></a>.</p>
</li>
<li>
<p>Review a
    <a href="https://drive.google.com/file/d/1h-FVY26G_Q_OazkZrYmlTg4QhQUZTRFY/view?usp=sharing"><u>step-by-step</u></a>
    example of how to assign the GOLD ecosystem classification to a lake
    sediment sample.</p>
</li>
</ul>
<h3 id="environmental-ontology-envo">Environmental Ontology (EnvO)</h3>
<p>The Environment Ontology (<u>EnvO</u>, <a href="https://pubmed.ncbi.nlm.nih.gov/27664130/"><u>Buttigieg
2016</u></a>) is a community-led
ontology that represents environmental entities such as biomes,
environmental features, and environmental materials. Each EnvO term is
identified using a unique <em>resource identifier</em> (e.g.,
<a href="https://en.wikipedia.org/wiki/CURIE"><u>CURIE</u></a> or
<a href="https://en.wikipedia.org/wiki/Internationalized_Resource_Identifier"><u>IRI</u></a>)
that resolves in a web browser. This ensures that EnvO’s terms (and
their definitions) are easy to find, reference, and share amongst
collaborators. It also ensures that datasets described using EnvO terms
can be more easily integrated and analyzed in a reproducible manner. And
since the meanings of the terms are precisely defined and accessible,
humans and computers can easily connect EnvO terms across datasets.</p>
<p>EnvO terms are the recommended values for several of the mandatory terms
in the MIxS packages, often referred to as the "MIxS triad”.</p>
<ul>
<li>
<p><strong>MIxS: env_broad_scale</strong> (a.k.a. Biome): The major environmental
    system that the sample or specimen came from. Often, the value for
    this term comes from EnvO’s
    <a href="http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_00000428"><u>biome</u></a>
    hierarchy, and is similar to GOLD’s <em>Ecosystem category.</em></p>
<ul>
<li><u>Examples:</u> forest biome, tropical biome, and oceanic
    pelagic zone biome</li>
</ul>
</li>
<li>
<p><strong>MIxS: env_local_scale</strong> (a.k.a. Feature): A more direct
    expression of the sample or specimen’s local vicinity, which likely
    has a significant influence on the sample or specimen. Possible
    values are listed in EnvO’s <a href="http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_01000813"><u>astronomical body
    part</u></a>
    hierarchy, which is similar to GOLD’s <em>Ecosystem type/subtype.</em></p>
<ul>
<li><u>Examples</u>: mountain, pond, whale fall, and karst</li>
</ul>
</li>
<li>
<p><strong>MIxS: env_medium</strong> (a.k.a. material): The environmental
    material(s) immediately surrounding your sample or specimen prior to
    sampling. Examples of this are found in EnvO’s <a href="http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_00010483"><u>environmental
    material</u></a>
    hierarchy, and is similar to GOLD’s <em>Specific ecosystem.</em></p>
<ul>
<li><u>Examples:</u> sediment, soil, water, and air</li>
</ul>
</li>
</ul>
<p><em>Explore how to map sample environments using the EnvO ecosystem
classification</em></p>
<p>Review a step-by-step example of how to assign EnvO terms to an
oligotrophic lake sediment sample below.</p>
<table>
 <tbody>
    <tr class="odd">
     <td width="35%"  valign="top">
        <p><strong>env_broad_scale (Biome)</strong></p>
        <p>Using <a href="http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_00000428"><u>EnvO biome</u></a> categories, <em>aquatic</em> is appropriate. However, since the EnvO is a hierarchical system, the aquatic biome has two sub-categories: freshwater and marine biomes. The freshwater biome is further divided into freshwater lake biome and freshwater river biome. Therefore, for a lake sediment sample, <em>freshwater lake biome</em> is the appropriate EnvO biome category.</p>
     </td>
     <td>
        <img src="../images/NMDC_metadata_img3.png" style="width:100%" />
     </td>
    </tr>
    <tr class="even">
    <td valign="top">
     <p><strong>env_local_scale (Feature)</strong></p>
     <p>Next, we describe the local environmental feature in the vicinity of and likely having a strong causal influence on the sample. Using the <a href="http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_01000813"><u>EnvO astronomical body part</u></a> categories, we step through the relevant categories (see figure on the right) until we reach the EnvO term <em>oligotrophic lake</em>.</p></td>
    <td>
        <img src="../images/NMDC_metadata_img4.png" style="width:100%" />
    </td>
    </tr>
    <tr class="odd">
    <td valign="top">
     <p><strong>env_medium (Material)</strong></p>
     <p>Finally, since the sample is <em>oligotrophic lake sediment</em>, the <a href="http://www.ontobee.org/ontology/ENVO?iri=http://purl.obolibrary.org/obo/ENVO_00002007"><u>EnvO environmental material</u></a> could be assigned <em>sediment</em>. But because the EnvO hierarchy provides sub-categories within <em>sediment</em>, the environmenta material will be assigned <em>lake sediment</em>.</p></td>
     <td>
         <img src="../images/NMDC_metadata_img5.png" style="width:100%" />
     </td>
    </tr>
 </tbody>
</table>

<p>Therefore, the EnvO triad for <em>oligotrophic lake sediment</em> is:</p>
<blockquote>
<p><strong>Env_broad_scale</strong>: freshwater lake biome [ENVO_01000252]</p>
<p><strong>Env_local_scale</strong>: oligotrophic lake [ENVO_01000774]</p>
<p><strong>Env_medium</strong>: lake sediment [ENVO_00000546]</p>
</blockquote>
<h3 id="classifying-samples-with-gold-and-mixsenvo">Classifying samples with GOLD and MIxS/EnvO</h3>
<p>The five-level GOLD ecosystem classification path and EnvO triad each
have unique advantages in describing the environmental context of a
biosample. The NMDC leverages the strengths of both the GOLD ecosystem
classification path and MIxS/EnvO triad. The assignment of MIxS/EnvO
triad for the biosamples currently in the NMDC data portal was achieved
through a manual curation process using various metadata fields of GOLD
biosamples fields, such as name, description, habitat, sample collection
site, identifier, ecosystem classification path, and study description.
The NMDC team is currently working on exploring solutions for automated
mapping between GOLD and MIxS/EnvO.</p>
<div align="center">
    <img src="../images/NMDC_metadata_img6.png" style="width:80%" />
</div>

<p>Figure 3: Mapping between the MIxS/EnvO triad and the GOLD ecosystem
classification enables integration of sample environments defined with
GOLD and MIxS/EnvO.</p>
<h2 id="data-processing-metadata">Data Processing Metadata</h2>
<p>In addition, the NMDC is adopting the MIxS standards for sequence data
types (e.g., sequencing method, pcr primers and conditions, etc.), and
are building on previous efforts by the <a href="http://www.psidev.info/groups/mass-spectrometry"><u>Proteomics Standards
Initiative</u></a> and
<a href="https://github.com/MSI-Metabolomics-Standards-Initiative/CIMR"><u>Metabolomics Standards
Initiative</u></a>
to develop standards and controlled vocabularies for mass spectrometry
data types (e.g., ionization mode, mass resolution, scan rate, etc.).
<em>Additional details on the processing metadata are coming soon.</em></p>
<h1 id="overview-of-the-nmdc-data-schema">Overview of the NMDC Data Schema</h1>
<p>The NMDC has developed a normalized metadata
<a href="https://github.com/microbiomedata/nmdc-metadata"><u>schema</u></a>
(available in the NMDC GitHub) for representing studies, samples,
relationships between samples, and associated data objects. The schema
is organized into object classes, which act as nodes. Each class has
associated slots, which are fields that contain metadata that describe
the object. For more in-depth information, full documentation of the
NMDC schema can be found
<a href="https://microbiomedata.github.io/nmdc-metadata/#classes"><u>here</u></a>.</p>
<p>For the NMDC pilot, a python
<a href="https://github.com/microbiomedata/nmdc-metadata"><u>toolkit</u></a> for
generating NMDC-compliant JavaScript Object Notation (JSON) objects was
developed to create ETL (Extract-Transform-Load) software to ingest
metadata from the DOE User Facilities. Read more about the data in the
NMDC pilot <a href="https://microbiomedata.org/data/"><u>here</u></a>.</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../schema-validation/" class="btn btn-neutral float-right" title="NMDC Schema Validation">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href=".." class="btn btn-neutral" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href=".." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../schema-validation/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
