<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Class: MetagenomeAssembly - NMDC Metadata Documentation</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Class: MetagenomeAssembly";
    var mkdocs_page_input_path = "MetagenomeAssembly.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> NMDC Metadata Documentation</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../Metadata_Documentation_Overview/">NMDC Metadata Standards Documentation</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../schema-validation/">NMDC Schema Validation</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../identifiers/">Identifiers in NMDC</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NMDC_MIxS_Soil_documentation/">MIxS Soil Package</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">NMDC Metadata Documentation</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Class: MetagenomeAssembly</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="class-metagenomeassembly">Class: MetagenomeAssembly</h1>
<p>URI: <a href="https://microbiomedata/meta/MetagenomeAssembly">nmdc:MetagenomeAssembly</a></p>
<p><img alt="img" src="http://yuml.me/diagram/nofunky;dir:TB/class/[WorkflowExecutionActivity],[NamedThing],[Database]++-%20metagenome%20assembly%20set%200..*&gt;[MetagenomeAssembly&#124;asm_score:float%20%3F;scaffolds:float%20%3F;scaf_logsum:float%20%3F;scaf_powsum:float%20%3F;scaf_max:float%20%3F;scaf_bp:float%20%3F;scaf_N50:float%20%3F;scaf_N90:float%20%3F;scaf_L50:float%20%3F;scaf_L90:float%20%3F;scaf_n_gt50K:float%20%3F;scaf_l_gt50K:float%20%3F;scaf_pct_gt50K:float%20%3F;contigs:float%20%3F;contig_bp:float%20%3F;ctg_N50:float%20%3F;ctg_L50:float%20%3F;ctg_N90:float%20%3F;ctg_L90:float%20%3F;ctg_logsum:float%20%3F;ctg_powsum:float%20%3F;ctg_max:float%20%3F;gap_pct:float%20%3F;gc_std:float%20%3F;gc_avg:float%20%3F;num_input_reads:float%20%3F;num_aligned_reads:float%20%3F;execution_resource(i):string%20%3F;git_url(i):string%20%3F;id(i):string;name(i):string%20%3F;type(i):string%20%3F;started_at_time(i):string%20%3F;ended_at_time(i):string%20%3F;used(i):string%20%3F],[WorkflowExecutionActivity]^-[MetagenomeAssembly],[Database],[Activity]" /></p>
<h2 id="parents">Parents</h2>
<ul>
<li>is_a: <a href="../WorkflowExecutionActivity/">WorkflowExecutionActivity</a> - Represents an instance of an execution of a particular workflow</li>
</ul>
<h2 id="referenced-by-class">Referenced by class</h2>
<ul>
<li><strong><a href="../Database/">Database</a></strong> <em><a href="../metagenome_assembly_set/">metagenome assembly set</a></em>  <sub>0..<em></sub>  </em><em><a href="./">MetagenomeAssembly</a></em>*</li>
</ul>
<h2 id="attributes">Attributes</h2>
<h3 id="own">Own</h3>
<ul>
<li><a href="../asm_score/">asm_score</a>  <sub>OPT</sub><ul>
<li>Description: A score for comparing metagenomic assembly quality from same sample.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../contig_bp/">contig_bp</a>  <sub>OPT</sub><ul>
<li>Description: Total size in bp of all contigs.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../contigs/">contigs</a>  <sub>OPT</sub><ul>
<li>Description: The sum of the (length*log(length)) of all contigs, times some constant.  Increase the contiguity, the score will increase</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../ctg_L50/">ctg_L50</a>  <sub>OPT</sub><ul>
<li>Description: Given a set of contigs, the L50 is defined as the sequence length of the shortest contig at 50% of the total genome length.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../ctg_L90/">ctg_L90</a>  <sub>OPT</sub><ul>
<li>Description: The L90 statistic is less than or equal to the L50 statistic; it is the length for which the collection of all contigs of that length or longer contains at least 90% of the sum of the lengths of all contigs.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../ctg_N50/">ctg_N50</a>  <sub>OPT</sub><ul>
<li>Description: Given a set of contigs, each with its own length, the N50 count is defined as the smallest number of contigs whose length sum makes up half of genome size.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../ctg_N90/">ctg_N90</a>  <sub>OPT</sub><ul>
<li>Description: Given a set of contigs, each with its own length, the N90 count is defined as the smallest number of contigs whose length sum makes up 90% of genome size.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../ctg_logsum/">ctg_logsum</a>  <sub>OPT</sub><ul>
<li>Description: Maximum contig length.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../ctg_max/">ctg_max</a>  <sub>OPT</sub><ul>
<li>Description: Maximum contig length.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../ctg_powsum/">ctg_powsum</a>  <sub>OPT</sub><ul>
<li>Description: Powersum of all contigs is the same as logsum except that it uses the sum of (length*(length^P)) for some power P (default P=0.25).</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../gap_pct/">gap_pct</a>  <sub>OPT</sub><ul>
<li>Description: The gap size percentage of all scaffolds.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../gc_avg/">gc_avg</a>  <sub>OPT</sub><ul>
<li>Description: Average of GC content of all contigs.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../gc_std/">gc_std</a>  <sub>OPT</sub><ul>
<li>Description: Standard deviation of GC content of all contigs.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../num_aligned_reads/">num_aligned_reads</a>  <sub>OPT</sub><ul>
<li>Description: The sequence count number of input reads aligned to assembled contigs.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../num_input_reads/">num_input_reads</a>  <sub>OPT</sub><ul>
<li>Description: The sequence count number of input reads for assembly.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_L50/">scaf_L50</a>  <sub>OPT</sub><ul>
<li>Description: Given a set of scaffolds, the L50 is defined as the sequence length of the shortest scaffold at 50% of the total genome length.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_L90/">scaf_L90</a>  <sub>OPT</sub><ul>
<li>Description: The L90 statistic is less than or equal to the L50 statistic; it is the length for which the collection of all scaffolds of that length or longer contains at least 90% of the sum of the lengths of all scaffolds.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_N50/">scaf_N50</a>  <sub>OPT</sub><ul>
<li>Description: Given a set of scaffolds, each with its own length, the N50 count is defined as the smallest number of scaffolds whose length sum makes up half of genome size.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_N90/">scaf_N90</a>  <sub>OPT</sub><ul>
<li>Description: Given a set of scaffolds, each with its own length, the N90 count is defined as the smallest number of scaffolds whose length sum makes up 90% of genome size.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_bp/">scaf_bp</a>  <sub>OPT</sub><ul>
<li>Description: Total size in bp of all scaffolds.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_l_gt50K/">scaf_l_gt50K</a>  <sub>OPT</sub><ul>
<li>Description: Total size in bp of all scaffolds greater than 50 KB.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_logsum/">scaf_logsum</a>  <sub>OPT</sub><ul>
<li>Description: The sum of the (length*log(length)) of all scaffolds, times some constant.  Increase the contiguity, the score will increase</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_max/">scaf_max</a>  <sub>OPT</sub><ul>
<li>Description: Maximum scaffold length.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_n_gt50K/">scaf_n_gt50K</a>  <sub>OPT</sub><ul>
<li>Description: Total sequence count of scaffolds greater than 50 KB.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_pct_gt50K/">scaf_pct_gt50K</a>  <sub>OPT</sub><ul>
<li>Description: Total sequence size percentage of scaffolds greater than 50 KB.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaf_powsum/">scaf_powsum</a>  <sub>OPT</sub><ul>
<li>Description: Powersum of all scaffolds is the same as logsum except that it uses the sum of (length*(length^P)) for some power P (default P=0.25).</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
<li><a href="../scaffolds/">scaffolds</a>  <sub>OPT</sub><ul>
<li>Description: Total sequence count of all scaffolds.</li>
<li>range: <a href="../types/Float/">Float</a></li>
</ul>
</li>
</ul>
<h3 id="inherited-from-workflow-execution-activity">Inherited from workflow execution activity:</h3>
<ul>
<li><a href="../ended_at_time/">ended at time</a>  <sub>OPT</sub><ul>
<li>range: <a href="../types/String/">String</a></li>
</ul>
</li>
<li><a href="../execution_resource/">execution resource</a>  <sub>OPT</sub><ul>
<li>Description: Example: NERSC-Cori</li>
<li>range: <a href="../types/String/">String</a></li>
</ul>
</li>
<li><a href="../git_url/">git url</a>  <sub>OPT</sub><ul>
<li>Description: Example: https://github.com/microbiomedata/mg_annotation/releases/tag/0.1</li>
<li>range: <a href="../types/String/">String</a></li>
</ul>
</li>
<li><a href="../has_input/">has input</a>  <sub>0..*</sub><ul>
<li>Description: An input to a process.</li>
<li>range: <a href="../NamedThing/">NamedThing</a></li>
</ul>
</li>
<li><a href="../has_output/">has output</a>  <sub>0..*</sub><ul>
<li>Description: An output biosample to a processing step</li>
<li>range: <a href="../NamedThing/">NamedThing</a></li>
</ul>
</li>
<li><a href="../id/">id</a>  <sub>REQ</sub><ul>
<li>Description: A unique identifier for a thing. Must be either a CURIE shorthand for a URI or a complete URI</li>
<li>range: <a href="../types/String/">String</a></li>
</ul>
</li>
<li><a href="../name/">name</a>  <sub>OPT</sub><ul>
<li>Description: A human readable label for an entity</li>
<li>range: <a href="../types/String/">String</a></li>
</ul>
</li>
<li><a href="../started_at_time/">started at time</a>  <sub>OPT</sub><ul>
<li>range: <a href="../types/String/">String</a></li>
</ul>
</li>
<li><a href="../type/">type</a>  <sub>OPT</sub><ul>
<li>Description: An optional string that species the type object.  This is used to allow for searches for different kinds of objects.</li>
<li>range: <a href="../types/String/">String</a></li>
<li>Example: nmdc:Biosample None</li>
<li>Example: nmdc:Study None</li>
</ul>
</li>
<li><a href="../used/">used</a>  <sub>OPT</sub><ul>
<li>range: <a href="../types/String/">String</a></li>
</ul>
</li>
<li><a href="../was_informed_by/">was informed by</a>  <sub>OPT</sub><ul>
<li>range: <a href="../Activity/">Activity</a></li>
</ul>
</li>
<li><a href="../workflow_execution_activity_was_associated_with/">workflow execution activityâžžwas associated with</a>  <sub>OPT</sub><ul>
<li>Description: the agent/entity associated with the generation of the file</li>
<li>range: <a href="../WorkflowExecutionActivity/">WorkflowExecutionActivity</a></li>
</ul>
</li>
</ul>
<h2 id="other-properties">Other properties</h2>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>In Subsets:</strong></td>
<td></td>
<td>workflow subset</td>
</tr>
</tbody>
</table>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
