<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>MIxS Soil Package - NMDC Metadata Documentation</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "MIxS Soil Package";
    var mkdocs_page_input_path = "NMDC_MIxS_Soil_documentation.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> NMDC Metadata Documentation</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../Metadata_Documentation_Overview/">NMDC Metadata Standards Documentation</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../schema-validation/">NMDC Schema Validation</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../identifiers/">Identifiers in NMDC</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">MIxS Soil Package</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#mandatory-descriptors-of-mixs-soil-packages-are">Mandatory descriptors of MIxS Soil packages are:</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#unique-descriptors-46-in-mixs-soil-package">Unique descriptors (46) in MIxS Soil package</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#other-descriptors-non-mandatory-and-non-unique-descriptors-from-mixs-soil-package">Other descriptors (non mandatory and non-unique descriptors) from MIxS Soil package</a>
    </li>
    </ul>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">NMDC Metadata Documentation</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>MIxS Soil Package</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="mixs-soil-package">MIxS Soil Package</h1>
<p>The MIxS Soil Package contains a list of 145 descriptors to describe the
soil sample taken from various environments including soil from,
cropland, dryland, forest, grassland soil, coastal sand dune, permafrost
soil. These 145 descriptors have been provided in different sections
namely soil, nucleic acid sequence source, environment, sequencing,
investigation and MIxS extension. We have grouped these descriptors into
mandatory descriptors, unique descriptors and other descriptors (non
mandatory and non unique).</p>
<p>Some examples of biosamples described using MIxS-Soil package (v5) terms:</p>
<p><a href="https://www.ncbi.nlm.nih.gov/biosample/SAMN07125075"><u>https://www.ncbi.nlm.nih.gov/biosample/SAMN07125075</u></a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/biosample/SAMN08902834"><u>https://www.ncbi.nlm.nih.gov/biosample/SAMN08902834</u></a></p>
<h2 id="mandatory-descriptors-of-mixs-soil-packages-are"><strong>Mandatory descriptors of MIxS Soil packages are:</strong></h2>
<p>The MIxS soil package has 12 mandatory descriptors including <strong>'depth'</strong>
and <strong>'elevation'</strong>. These 12 mandatory descriptors with descriptor
name, definition, section of the MIxS package, expected value, value
syntax for all of the descriptors and preferred unit and example value
when available are listed below.</p>
<ul>
<li>
<p>investigation_type - Nucleic Acid Sequence Report is the root
    element of all MIGS/MIMS compliant reports as standardized by
    Genomic Standards Consortium. This field is either
    eukaryote,bacteria,virus,plasmid,organelle,
    metagenome,mimarks-survey, mimarks-specimen, metatranscriptome,
    single amplified genome, metagenome-assembled genome, or
    uncultivated viral genome.<br />
    Section : investigation<br />
    Expected value : eukaryote, bacteria_archaea, plasmid, virus,
    organelle, metagenome,mimarks-survey, mimarks-specimen,
    metatranscriptome, single amplified genome, metagenome-assembled
    genome, or uncultivated viral genomes<br />
    Value syntax :
    [eukaryote|bacteria_archaea|plasmid|virus|organelle|metagenome|metatranscriptome|mimarks-survey|mimarks-specimen|misag|mimag|miuvig]<br />
    Example : metagenome</p>
</li>
<li>
<p>project_name - Name of the project within which the sequencing was organized.</p>
<p>Section : investigation<br />
Expected value :<br />
Value syntax : {text}</p>
<p>The project name in the NMDC follows standardized metagenome naming
scheme as per the Genomes Online Database (GOLD) that can be accessed
from<a href="https://gold.jgi.doe.gov/resources/Standardized_Metagenome_Naming.pdf"><u>https://gold.jgi.doe.gov/resources/Standardized_Metagenome_Naming.pdf</u></a></p>
<p>The following four metadata are used in the naming of the project:</p>
<p>[Habitat] [Type of communities] [ Location, including the
country/ocean] – [Identifier]</p>
<p>For example, for the following metadata:</p>
<p>Habitat: Permafrost</p>
<p>COMMUNITY: microbial communities</p>
<p>GEOGRAPHIC_LOCATION: Sweden: Stordalen mire</p>
<p>Sample_Identifier: 20120800_S1X</p>
<p>Project name for <strong>metagenome</strong> would be:</p>
<p>Permafrost microbial communities from Stordalen mire, Sweden -
20120800_S1X.</p>
<p>Project name for <strong>Metatranscriptome</strong> would be:</p>
<p>Metatranscriptome of permafrost microbial communities from Stordalen
mire, Sweden - 20120800_S1X</p>
</li>
<li>
<p>lat_lon - The geographical origin of the sample as defined by
    latitude and longitude. The values should be reported in decimal
    degrees and in WGS84 system.<br />
    Section : environment<br />
    Expected value : decimal degrees<br />
    Value syntax : {float} {float}<br />
    Example : 50.586825 6.408977</p>
</li>
<li>
<p>geo_loc_name - The geographical origin of the sample as defined by
    the country or sea name followed by specific region name. Country
    or sea names should be chosen from the INSDC country list
    (<a href="http://insdc.org/country.html"><u>http://insdc.org/country.html</u></a>),
    or the GAZ ontology that can be accessed from
    <a href="http://www.ontobee.org/ontology/GAZ"><u>http://www.ontobee.org/ontology/GAZ</u></a>
    or
    <a href="http://purl.bioontology.org/ontology/GAZ"><u>http://purl.bioontology.org/ontology/GAZ</u></a>.<br />
    Section : environment<br />
    Expected value : country or sea name (INSDC or
    GAZ);region(GAZ);specific location name<br />
    Value syntax : {term};{term};{text}<br />
    Example : Germany;North Rhine-Westphalia;Eifel National Park</p>
</li>
<li>
<p>collection_date - The time of sampling, either as an instance
    (single point in time) or interval. In case no exact time is
    available, the date/time can be right truncated i.e. all of these
    are valid times: 2008-01-23T19:23:10+00:00; 2008-01-23T19:23:10;
    2008-01-23; 2008-01; 2008; Except: 2008-01; 2008 all are ISO8601
    compliant.<br />
    Section : environment<br />
    Expected value : date and time<br />
    Value syntax : {timestamp}<br />
    Example : 2018-05-11T10:00:00+01:00</p>
</li>
<li>
<p>env_broad_scale - The broad-scale environmental context of MIxS
    uses terminologies from Environment Ontology (EnvO). EnvO
    describes the broad-scale environmental context as environmental
    systems / biomes to which resident ecological communities have
    evolved adaptations. Biome possesses a degree of spatial and
    temporal stability that has allowed at least some of its
    constituent communities to adapt. In this field, report which
    major environmental system your sample or specimen came from. The
    systems identified should have a coarse spatial grain, to provide
    the general environmental context of where the sampling was</p>
<p>done (e.g. were you in the desert or a rainforest?).</p>
<p>Some of the broad-scale environmental context terms from EnvO that can
be used for soil biosamples are, terrestrial biome, anthropogenic
terrestrial biome, desert biome, cropland biome, forest biome, mixed
forest biome, grassland biome, tropical biome, tropical grassland
biome, tundra biome and urban biome.</p>
<p>We recommend using subclasses of ENVO’s biome class: Biome class
represents
<a href="http://purl.obolibrary.org/obo/ENVO_00000428"><strong><u>http://purl.obolibrary.org/obo/ENVO_00000428</u></strong></a>.</p>
<p>Section : environment<br />
Expected value : Add terms that identify the major environment type(s)
where your sample was collected. Recommend subclasses of biome
[ENVO:00000428]. Format for single term: termLabel [termID],Format
for multiple terms: termLabel [termID]|termLabel
[termID]|termLabel [termID].</p>
<p>Value syntax : {termLabel} {[termID]}<br />
Example:</p>
<p>annotating soil from permafrost: terrestrial biome [ENVO_00000446]
or</p>
<p>soil from meadow: grassland biome [ENVO_01000177]</p>
<p>terrestrial biome [ENVO_00000446]|urban biome[ENVO_01000249]</p>
</li>
<li>
<p>env_local_scale - The local environmental context of MIxS uses
    terminologies from Environment Ontology (EnvO). EnvO describes the
    local environmental context as environmental features that are in
    the vicinity of and have a strong causal influence on the entity;
    in this field, report the entity or entities which are in your
    sample or specimen’s local vicinity and which you believe have
    significant causal influences on your sample or specimen. Some of
    the MIxS local environmental context terms from EnvO that can be
    used describe soil feature are: agricultural field, desert, flood
    plain, garden, hill, paddy field and river bank etc. The MIxS
    local environmental context terms given in ENVO that are of
    smaller spatial grain than your entry for env_broad_scale.</p>
<p>If needed, request new terms on the ENVO tracker, identified here:
<a href="http://www.obofoundry.org/ontology/envo.html"><u>http://www.obofoundry.org/ontology/envo.html</u></a>.</p>
<p>Section : environment<br />
Expected value : Add terms that identify environmental entities having
causal influences upon the entity at time of sampling. Format for
single term: termLabel [termID]; Format for multiple terms:
termLabel [termID]|termLabel [termID]|termLabel [termID].</p>
<p>Value syntax : {termLabel} {[termID]}</p>
<p>Example:</p>
<p>annotating local environmental context of soil from permafrost active
layer: active permafrost layer [ENVO_04000009] or</p>
<p>soil from a biosphere reserve: biosphere reserve [ENVO_00000376]</p>
<p>agricultural field[ENVO_00000114]|banana
plantation[ENVO_00000161]</p>
</li>
<li>
<p>env_medium - The MIxS environmental medium context terms uses
    terminologies from Environment Ontology (EnvO). EnvO describes the
    environmental medium/material context terms as those terms that
    refers to masses, volumes, or other portions of some medium
    included in an environmental system; environmental material that
    is the substance surrounding or partially surrounding the entity.</p>
<p>Some of the MIxS env_medium terms from EnvO that can be used describe
soil biosamples are: agricultural soil, bulk soil, burned soil,
eucalyptus forest soil, forest soil, farm soil, fertilized soil,
forest soil, garden soil, grassland soil, greenhouse soil, heat
stressed soil, meadow soil, peat soil, soil, spruce forest soil,
surface soil etc.</p>
<p>In this field, report which environmental material or materials (pipe
separated) immediately surrounded your sample or specimen prior to
sampling, using one or more subclasses of ENVO’s environmental
material class:
<a href="http://purl.obolibrary.org/obo/ENVO_00010483"><u>http://purl.obolibrary.org/obo/ENVO_00010483</u></a>.</p>
<p>Section : environment<br />
Expected value : Add terms that identify the material displaced by the
entity at time of sampling. Recommend subclasses of environmental
material [ENVO:00010483]. Multiple terms can be separated by pipes
e.g.: estuarine water</p>
<p>Format (one term): termLabel [termID];</p>
<p>Format (multiple terms): termLabel [termID]|termLabel
[termID]|termLabel [termID].</p>
<p>Value syntax : {termLabel} {[termID]}<br />
Example:</p>
<p>Annotating env_medium (environmental medium context terms) of meadow
soil: meadow soil [ENVO_00005761].</p>
<p>When there are multiple terms, agricultural soil
[ENVO_00002259]|bulk soil [ENVO_00005802]|oil contaminated
soil [ENVO_00002875]</p>
</li>
<li>
<p>depth - Depth is defined as the vertical distance below local
    surface, e.g. For sediment or soil samples depth is measured from
    sediment or soil surface, respectively. Depth can be reported as
    an interval for subsurface samples.<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : meter<br />
    Value syntax : {float} {unit}<br />
    Example : 10 meter</p>
</li>
<li>
<p>elev - Elevation of the sampling site is its height above a fixed
    reference point, most commonly the mean sea level. Elevation is
    mainly used when referring to points on the earth's surface, while
    altitude is used for points above the surface, such as an aircraft
    in flight or a spacecraft in orbit.<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : meter<br />
    Value syntax : {float} {unit}<br />
    Example : 100 meter</p>
</li>
<li>
<p>submitted_to_insdc - Depending on the study (large-scale e.g. done
    with next generation sequencing technology, or small-scale)
    sequences have to be submitted to SRA (Sequence Read Archive), DRA
    (DDBJ Read Archive) or via the classical Webin/Sequin systems to
    Genbank, ENA and DDBJ. Although this field is mandatory, it is
    meant as a self-test field, therefore it is not necessary to
    include this field in contextual data submitted to databases.<br />
    Section : investigation<br />
    Expected value : boolean<br />
    Value syntax : {boolean}<br />
    Example : yes</p>
</li>
<li>
<p>seq_meth - Sequencing method used; e.g. Sanger, pyrosequencing,
    ABI-solid.<br />
    Section : sequencing<br />
    Expected value : enumeration<br />
    Value syntax : [MinION|GridION|PromethION|454 GS|454 GS
    20|454 GS FLX|454 GS FLX+|454 GS FLX Titanium|454 GS
    Junior|Illumina Genome Analyzer|Illumina Genome Analyzer
    II|Illumina Genome Analyzer IIx|Illumina HiSeq 4000|Illumina
    HiSeq 3000|Illumina HiSeq 2500|Illumina HiSeq 2000|Illumina
    HiSeq 1500|Illumina HiSeq 1000|Illumina HiScanSQ|Illumina
    MiSeq|Illumina HiSeq X Five|Illumina HiSeq X Ten|Illumina
    NextSeq 500|Illumina NextSeq 550|AB SOLiD System|AB SOLiD
    System 2.0|AB SOLiD System 3.0|AB SOLiD 3 Plus System|AB SOLiD
    4 System|AB SOLiD 4hq System|AB SOLiD PI System|AB 5500 Genetic
    Analyzer|AB 5500xl Genetic Analyzer|AB 5500xl-W Genetic Analysis
    System|Ion Torrent PGM|Ion Torrent Proton|Ion Torrent S5|Ion
    Torrent S5 XL|PacBio RS|PacBio RS II|Sequel|AB 3730xL Genetic
    Analyzer|AB 3730 Genetic Analyzer|AB 3500xL Genetic Analyzer|AB
    3500 Genetic Analyzer|AB 3130xL Genetic Analyzer|AB 3130 Genetic
    Analyzer|AB 310 Genetic Analyzer|BGISEQ-500]<br />
    Example : Illumina HiSeq 1500</p>
</li>
</ul>
<h2 id="unique-descriptors-46-in-mixs-soil-package"><strong>Unique descriptors (46) in MIxS Soil package</strong></h2>
<p>The MIxS Soil package has 46 unique descriptors when compared with other
MIxS packages. Name, definition, section of the MIxS package, expected
value, value syntax for all of these descriptors and preferred unit and
example value when available are listed below.</p>
<ul>
<li>
<p>agrochem_addition - Addition of fertilizers, pesticides, etc. -
    amount and time of applications.<br />
    Section : soil<br />
    Expected value : agrochemical name;agrochemical amount;timestamp<br />
    Preferred unit : gram, mole per liter, milligram per liter<br />
    Value syntax : {text};{float} {unit};{timestamp}<br />
    Example : roundup;5 milligram per liter;2018-06-21</p>
</li>
<li>
<p>al_sat - Aluminum saturation (esp. For tropical soils).<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : percentage<br />
    Value syntax : {float} {unit}</p>
</li>
<li>
<p>al_sat_meth - Reference or method used in determining Al
    saturation.<br />
    Section : soil<br />
    Expected value : PMID,DOI or URL<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>annual_precpt - The average of all annual precipitation values
    known, or an estimated equivalent value derived by such methods as
    regional indexes or Isohyetal maps. .<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : millimeter<br />
    Value syntax : {float} {unit}</p>
</li>
<li>
<p>annual_temp - Mean annual temperature.<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : degree Celsius<br />
    Value syntax : {float} {unit}<br />
    Example : 12.5 degree Celsius</p>
</li>
<li>
<p>crop_rotation - Whether or not crop is rotated, and if yes,
    rotation schedule.<br />
    Section : soil<br />
    Expected value : crop rotation status;schedule<br />
    Value syntax : {boolean};{Rn/start_time/end_time/duration}<br />
    Example : yes;R2/2017-01-01/2018-12-31/P6M</p>
</li>
<li>
<p>cur_land_use - Present state of sample site.<br />
    Section : soil<br />
    Expected value : enumeration<br />
    Value syntax : [cities|farmstead|industrial
    areas|roads/railroads|rock|sand|gravel|mudflats|salt
    flats|badlands|permanent snow or ice|saline
    seeps|mines/quarries|oil waste areas|small grains|row
    crops|vegetable crops|horticultural plants (e.g.
    tulips)|marshlands (grass,sedges,rushes)|tundra
    (mosses,lichens)|rangeland|pastureland (grasslands used for
    livestock grazing)|hayland|meadows
    (grasses,alfalfa,fescue,bromegrass,timothy)|shrub land (e.g.
    mesquite,sage-brush,creosote bush,shrub
    oak,eucalyptus)|successional shrub land (tree
    saplings,hazels,sumacs,chokecherry,shrub
    dogwoods,blackberries)|shrub crops (blueberries,nursery
    ornamentals,filberts)|vine crops (grapes)|conifers (e.g.
    pine,spruce,fir,cypress)|hardwoods (e.g.
    oak,hickory,elm,aspen)|intermixed hardwood and conifers|tropical
    (e.g. mangrove,palms)|rainforest (evergreen forest
    receiving &gt;406 cm annual rainfall)|swamp (permanent or
    semi-permanent water body dominated by woody plants)|crop trees
    (nuts,fruit,christmas trees,nursery trees)]<br />
    Example : conifers</p>
</li>
<li>
<p>cur_vegetation - Vegetation classification from one or more
    standard classification systems, or agricultural crop.<br />
    Section : soil<br />
    Expected value : current vegetation type<br />
    Value syntax : {text}</p>
</li>
<li>
<p>cur_vegetation_meth - Reference or method used in vegetation
    classification .<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>drainage_class - Drainage classification from a standard system
    such as the USDA system.<br />
    Section : soil<br />
    Expected value : enumeration<br />
    Value syntax : [very poorly|poorly|somewhat poorly|moderately
    well|well|excessively drained]<br />
    Example : well</p>
</li>
<li>
<p>extreme_event - Unusual physical events that may have affected
    microbial populations.<br />
    Section : soil<br />
    Expected value : date<br />
    Value syntax : {timestamp}</p>
</li>
<li>
<p>extreme_salinity - Measured salinity .<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : millisiemens per meter<br />
    Value syntax : {float} {unit}</p>
</li>
<li>
<p>fao_class - Soil classification from the FAO World Reference
    Database for Soil Resources. The list can be found at
    <a href="http://www.fao.org/nr/land/sols/soil/wrb-soil-maps/reference-groups"><u>http://www.fao.org/nr/land/sols/soil/wrb-soil-maps/reference-groups</u></a>.<br />
    Section : soil<br />
    Expected value : enumeration<br />
    Value syntax :
    [Acrisols|Andosols|Arenosols|Cambisols|Chernozems|Ferralsols|Fluvisols|Gleysols|Greyzems|Gypsisols|Histosols|Kastanozems|Lithosols|Luvisols|Nitosols|Phaeozems|Planosols|Podzols|Podzoluvisols|Rankers|Regosols|Rendzinas|Solonchaks|Solonetz|Vertisols|Yermosols]<br />
    Example : Luvisols</p>
</li>
<li>
<p>fire - Historical and/or physical evidence of fire.<br />
    Section : soil<br />
    Expected value : date<br />
    Value syntax : {timestamp}</p>
</li>
<li>
<p>flooding - Historical and/or physical evidence of flooding.<br />
    Section : soil<br />
    Expected value : date<br />
    Value syntax : {timestamp}</p>
</li>
<li>
<p>heavy_metals - Heavy metals present and concentrations any drug
    used by subject and the frequency of usage; can include multiple
    heavy metals and concentrations.<br />
    Section : soil<br />
    Expected value : heavy metal name;measurement value<br />
    Preferred unit : microgram per gram<br />
    Value syntax : {text};{float} {unit}</p>
</li>
<li>
<p>heavy_metals_meth - Reference or method used in determining heavy
    metals.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>horizon - Specific layer in the land area which measures parallel to
    the soil surface and possesses physical characteristics which
    differ from the layers above and beneath.<br />
    Section : soil<br />
    Expected value : enumeration<br />
    Value syntax : [O horizon|A horizon|E horizon|B horizon|C
    horizon|R layer|Permafrost]<br />
    Example : A horizon</p>
</li>
<li>
<p>horizon_meth - Reference or method used in determining the
    horizon.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>link_addit_analys - Link to additional analysis results performed
    on the sample.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>link_class_info - Link to digitized soil maps or other soil
    classification information.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>link_climate_info - Link to climate resource.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>local_class - Soil classification based on local soil
    classification system.<br />
    Section : soil<br />
    Expected value : local classification name<br />
    Value syntax : {text}</p>
</li>
<li>
<p>local_class_meth - Reference or method used in determining the
    local soil classification .<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>microbial_biomass - The part of the organic matter in the soil that
    constitutes living microorganisms smaller than 5-10 micrometer. If
    you keep this, you would need to have correction factors used for
    conversion to the final units.<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : ton, kilogram, gram per kilogram soil<br />
    Value syntax : {float} {unit}</p>
</li>
<li>
<p>microbial_biomass_meth - Reference or method used in determining
    microbial biomass.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>ph_meth - Reference or method used in determining ph.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>pool_dna_extracts - Indicate whether multiple DNA extractions were
    mixed. If the answer yes, the number of extracts that were pooled
    should be given.<br />
    Section : soil<br />
    Expected value : pooling status;number of pooled extracts<br />
    Value syntax : {boolean};{integer}<br />
    Example : yes;5</p>
</li>
<li>
<p>previous_land_use - Previous land use and dates.<br />
    Section : soil<br />
    Expected value : land use name;date<br />
    Value syntax : {text};{timestamp}</p>
</li>
<li>
<p>previous_land_use_meth - Reference or method used in determining
    previous land use and dates.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>profile_position - Cross-sectional position in the hillslope where
    sample was collected.sample area position in relation to
    surrounding areas.<br />
    Section : soil<br />
    Expected value : enumeration<br />
    Value syntax :
    [summit|shoulder|backslope|footslope|toeslope]<br />
    Example : summit</p>
</li>
<li>
<p>salinity_meth - Reference or method used in determining salinity.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>season_precpt - The average of all seasonal precipitation values
    known, or an estimated equivalent value derived by such methods as
    regional indexes or Isohyetal maps. .<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : millimeter<br />
    Value syntax : {float} {unit}</p>
</li>
<li>
<p>season_temp - Mean seasonal temperature.<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : degree Celsius<br />
    Value syntax : {float} {unit}<br />
    Example : 18 degree Celsius</p>
</li>
<li>
<p>sieving - Collection design of pooled samples and/or sieve size and
    amount of sample sieved.<br />
    Section : soil<br />
    Expected value : design name and/or size;amount<br />
    Value syntax : {{text}|{float} {unit}};{float} {unit}</p>
</li>
<li>
<p>slope_aspect - The direction a slope faces. While looking down a
    slope use a compass to record the direction you are facing
    (direction or degrees); e.g., nw or 315 degrees. This measure
    provides an indication of sun and wind exposure that will
    influence soil temperature and evapotranspiration.<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : degree<br />
    Value syntax : {float} {unit}</p>
</li>
<li>
<p>slope_gradient - Commonly called 'slope'. The angle between ground
    surface and a horizontal line (in percent). This is the direction
    that overland water would flow. This measure is usually taken with
    a hand level meter or clinometer.<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Preferred unit : percentage<br />
    Value syntax : {float} {unit}</p>
</li>
<li>
<p>soil_type - Soil series name or other lower-level classification.<br />
    Section : soil<br />
    Expected value : soil type name<br />
    Value syntax : {text}</p>
</li>
<li>
<p>soil_type_meth - Reference or method used in determining soil
    series name or other lower-level classification.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>store_cond - Explain how and for how long the soil sample was
    stored before DNA extraction.<br />
    Section : soil<br />
    Expected value : storage condition type;duration<br />
    Value syntax : {text};{duration}<br />
    Example : -20 degree Celsius freezer;P2Y10D</p>
</li>
<li>
<p>texture - The relative proportion of different grain sizes of
    mineral particles in a soil, as described using a standard system;
    express as % sand (50 um to 2 mm), silt (2 um to 50 um), and clay
    (&lt;2 um) with textural name (e.g., silty clay loam) optional..<br />
    Section : soil<br />
    Expected value : measurement value<br />
    Value syntax : {float} {unit}</p>
</li>
<li>
<p>texture_meth - Reference or method used in determining soil
    texture.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>tillage - Note method(s) used for tilling.<br />
    Section : soil<br />
    Expected value : enumeration<br />
    Value syntax : [drill|cutting disc|ridge till|strip
    tillage|zonal tillage|chisel|tined|mouldboard|disc plough]<br />
    Example : chisel</p>
</li>
<li>
<p>tot_nitro_content_meth - Reference or method used in determining
    the total nitrogen.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>tot_org_c_meth - Reference or method used in determining total
    organic carbon.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
<li>
<p>water_content_soil_meth - Reference or method used in determining
    the water content of soil.<br />
    Section : soil<br />
    Expected value : PMID,DOI or url<br />
    Value syntax : {PMID}|{DOI}|{URL}</p>
</li>
</ul>
<h2 id="other-descriptors-non-mandatory-and-non-unique-descriptors-from-mixs-soil-package"><strong>Other descriptors (non mandatory and non-unique descriptors) from MIxS Soil package</strong></h2>
<p>The MIxS Soil package has 89 descriptors that can also be found/used in
other MIxS environmental packages. Name, definition, section of the MIxS
package, expected value, value syntax for all of these descriptors and
preferred unit and example value when available are listed below.</p>
<ul>
<li>
<p>16s_recover - Can a 16S gene be recovered from the submitted SAG or
    MAG?.</p>
<p>Section : sequencing</p>
<p>Expected value : boolean</p>
<p>Value syntax : {boolean}</p>
<p>Example : yes</p>
</li>
<li>
<p>16s_recover_software - Tools used for 16S rRNA gene extraction.</p>
<p>Section : sequencing</p>
<p>Expected value : names and versions of software(s), parameters used</p>
<p>Value syntax : {software};{version};{parameters}</p>
<p>Example : rambl;v2;default parameters</p>
</li>
<li>
<p>adapters - Adapters provide priming sequences for both amplification
    and sequencing of the sample-library fragments. Both adapters
    should be reported; in uppercase letters.</p>
<p>Section : sequencing</p>
<p>Expected value : adapter A and B sequence</p>
<p>Value syntax : {dna};{dna}</p>
<p>Example : AATGATACGGCGACCACCGAGATCTACACGCT;CAAGCAGAAGACGGCATACGAGAT</p>
</li>
<li>
<p>annot - Tool used for annotation, or for cases where annotation was
    provided by a community jamboree or model organism database rather
    than by a specific submitter.</p>
<p>Section : sequencing</p>
<p>Expected value : name of tool or pipeline used, or annotation source
description</p>
<p>Value syntax : {text}</p>
<p>Example : prokka</p>
</li>
<li>
<p>assembly_name - Name/version of the assembly provided by the
    submitter that is used in the genome browsers and in the
    community.</p>
<p>Section : sequencing</p>
<p>Expected value : name and version of assembly</p>
<p>Value syntax : {text} {text}</p>
<p>Example : HuRef, JCVI_ISG_i3_1.0</p>
</li>
<li>
<p>assembly_qual - The assembly quality category is based on sets of
    criteria outlined for each assembly quality category. For
    MISAG/MIMAG; Finished: Single, validated, contiguous sequence per
    replicon without gaps or ambiguities with a consensus error rate
    equivalent to Q50 or better. High Quality Draft:Multiple fragments
    where gaps span repetitive regions. Presence of the 23S, 16S and
    5S rRNA genes and at least 18 tRNAs. Medium Quality Draft:Many
    fragments with little to no review of assembly other than
    reporting of standard assembly statistics. Low Quality Draft:Many
    fragments with little to no review of assembly other than
    reporting of standard assembly statistics. Assembly statistics
    include, but are not limited to total assembly size, number of
    contigs, contig N50/L50, and maximum contig length. For MIUVIG;
    Finished: Single, validated, contiguous sequence per replicon
    without gaps or ambiguities, with extensive manual review and
    editing to annotate putative gene functions and transcriptional
    units. High-quality draft genome: One or multiple fragments,
    totaling ≥ 90% of the expected genome or replicon sequence or
    predicted complete. Genome fragment(s): One or multiple fragments,
    totalling &lt; 90% of the expected genome or replicon sequence, or
    for which no genome size could be estimated.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [Finished genome|High-quality draft
genome|Medium-quality draft genome|Low-quality draft genome|Genome
fragment(s)]</p>
<p>Example : High-quality draft genome</p>
</li>
<li>
<p>assembly_software - Tool(s) used for assembly, including version
    number and parameters.</p>
<p>Section : sequencing</p>
<p>Expected value : name and version of software, parameters used</p>
<p>Value syntax : {software};{version};{parameters}</p>
<p>Example : metaSPAdes;3.11.0;kmer set 21,33,55,77,99,121, default
parameters otherwise</p>
</li>
<li>
<p>bin_param - The parameters that have been applied during the
    extraction of genomes from metagenomic datasets.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [homology search|kmer|coverage|codon
usage|combination]</p>
<p>Example : coverage and kmer</p>
</li>
<li>
<p>bin_software - Tool(s) used for the extraction of genomes from
    metagenomic datasets.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax :
[metabat|maxbin|concoct|groupm|esom|metawatt|combination|other]</p>
<p>Example : concoct and maxbin</p>
</li>
<li>
<p>biotic_relationship - Description of relationship(s) between the
    subject organism and other organism(s) it is associated with.
    E.g., parasite on species X; mutualist with species Y. The target
    organism is the subject of the relationship, and the other
    organism(s) is the object.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [free
living|parasitism|commensalism|symbiotic|mutualism]</p>
<p>Example : free living</p>
</li>
<li>
<p>chimera_check - A chimeric sequence, or chimera for short, is a
    sequence comprised of two or more phylogenetically distinct parent
    sequences. Chimeras are usually PCR artifacts thought to occur
    when a prematurely terminated amplicon reanneals to a foreign DNA
    strand and is copied to completion in the following PCR cycles.
    The point at which the chimeric sequence changes from one parent
    to the next is called the breakpoint or conversion point .</p>
<p>Section : sequencing</p>
<p>Expected value : name and version of software, parameters used</p>
<p>Value syntax : {software};{version};{parameters}</p>
<p>Example : uchime;v4.1;default parameters</p>
</li>
<li>
<p>compl_appr - The approach used to determine the completeness of a
    given SAG or MAG, which would typically make use of a set of
    conserved marker genes or a closely related reference genome. For
    UViG completeness, include reference genome or group used, and
    contig feature suggesting a complete genome.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [marker gene|reference based|other]</p>
<p>Example : other: UViG length compared to the average length of
reference genomes from the P22virus genus (NCBI RefSeq v83)</p>
</li>
<li>
<p>compl_score - Completeness score is typically based on either the
    fraction of markers found as compared to a database or the percent
    of a genome found as compared to a closely related reference
    genome. High Quality Draft: &gt;90%, Medium Quality
    Draft: &gt;50%, and Low Quality Draft: &lt; 50% should have the
    indicated completeness scores.</p>
<p>Section : sequencing</p>
<p>Expected value : quality;percent completeness</p>
<p>Value syntax : [high|med|low];{percentage}</p>
<p>Example : med;60%</p>
</li>
<li>
<p>compl_software - Tools used for completion estimate, i.e. checkm,
    anvi'o, busco.</p>
<p>Section : sequencing</p>
<p>Expected value : names and versions of software(s) used</p>
<p>Value syntax : {software};{version}</p>
<p>Example : checkm</p>
</li>
<li>
<p>contam_score - The contamination score is based on the fraction of
    single-copy genes that are observed more than once in a query
    genome. The following scores are acceptable for; High Quality
    Draft: &lt; 5%, Medium Quality Draft: &lt; 10%, Low Quality Draft:
    &lt; 10%. Contamination must be below 5% for a SAG or MAG to be
    deposited into any of the public databases.</p>
<p>Section : sequencing</p>
<p>Expected value : value</p>
<p>Value syntax : {float} percentage</p>
<p>Example : 0.01</p>
</li>
<li>
<p>contam_screen_input - The type of sequence data used as input.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [reads| contigs]</p>
<p>Example : contigs</p>
</li>
<li>
<p>contam_screen_param - Specific parameters used in the
    decontamination sofware, such as reference database, coverage, and
    kmers. Combinations of these parameters may also be used, i.e.
    kmer and coverage, or reference database and kmer.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration;value or name</p>
<p>Value syntax : [ref db|kmer|coverage|combination];{text|integer}</p>
<p>Example : kmer</p>
</li>
<li>
<p>decontam_software - Tool(s) used in contamination screening.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax :
[checkm/refinem|anvi'o|prodege|bbtools:decontaminate.sh|acdc|combination]</p>
<p>Example : anvi'o</p>
</li>
<li>
<p>detec_type - Type of UViG detection.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [independent sequence (UViG)|provirus (UpViG)]</p>
<p>Example : independent sequence (UViG)</p>
</li>
<li>
<p>encoded_traits - Should include key traits like antibiotic
    resistance or xenobiotic degradation phenotypes for plasmids,
    converting genes for phage.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : for plasmid: antibiotic resistance; for phage:
converting genes</p>
<p>Value syntax : {text}</p>
<p>Example : beta-lactamase class A</p>
</li>
<li>
<p>env_package - MIxS extension for reporting of measurements and
    observations obtained from one or more of the environments where
    the sample was obtained. All environmental packages listed here
    are further defined in separate subtables. By giving the name of
    the environmental package, a selection of fields can be made from
    the subtables and can be reported.</p>
<p>Section : mixs extension</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [air|built
environment|host-associated|human-associated|human-skin|human-oral|human-gut|human-vaginal|hydrocarbon
resources-cores|hydrocarbon resources-fluids/swabs|microbial
mat/biofilm|misc
environment|plant-associated|sediment|soil|wastewater/sludge|water]</p>
<p>Example : soil</p>
</li>
<li>
<p>estimated_size - The estimated size of the genome prior to
    sequencing. Of particular importance in the sequencing of
    (eukaryotic) genome which could remain in draft form for a long or
    unspecified period..</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : number of base pairs</p>
<p>Value syntax : {integer} bp</p>
<p>Example : 300000 bp</p>
</li>
<li>
<p>experimental_factor - Experimental factors are essentially the
    variable aspects of an experiment design which can be used to
    describe an experiment, or set of experiments, in an increasingly
    detailed manner. This field accepts ontology terms from
    Experimental Factor Ontology (EFO) and/or Ontology for Biomedical
    Investigations (OBI). For a browser of EFO (v 2.95) terms, please
    see
    <a href="http://purl.bioontology.org/ontology/EFO"><u>http://purl.bioontology.org/ontology/EFO</u></a>;
    for a browser of OBI (v 2018-02-12) terms please see
    <a href="http://purl.bioontology.org/ontology/OBI"><u>http://purl.bioontology.org/ontology/OBI</u></a>.</p>
<p>Section : investigation</p>
<p>Expected value : text or EFO and/or OBI</p>
<p>Value syntax : {termLabel} {[termID]}|{text}</p>
<p>Example : time series design [EFO:EFO_0001779]</p>
</li>
<li>
<p>extrachrom_elements - Do plasmids exist of significant phenotypic
    consequence (e.g. ones that determine virulence or antibiotic
    resistance). Megaplasmids? Other plasmids (borrelia has 15+
    plasmids).</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : number of extrachromosmal elements</p>
<p>Value syntax : {integer}</p>
<p>Example : 5</p>
</li>
<li>
<p>feat_pred - Method used to predict UViGs features such as ORFs,
    integration site, etc..</p>
<p>Section : sequencing</p>
<p>Expected value : names and versions of software(s), parameters used</p>
<p>Value syntax : {software};{version};{parameters}</p>
<p>Example : Prodigal;2.6.3;default parameters</p>
</li>
<li>
<p>health_disease_stat - Health or disease status of specific host at
    time of collection.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : enumeration</p>
<p>Value syntax :
[healthy|diseased|dead|disease-free|undetermined|recovering|resolving|pre-existing
condition|pathological|life threatening|congenital]</p>
<p>Example : dead</p>
</li>
<li>
<p>host_pred_appr - Tool or approach used for host prediction.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [provirus|host sequence similarity|CRISPR spacer
match|kmer similarity|co-occurrence|combination|other]</p>
<p>Example : CRISPR spacer match</p>
</li>
<li>
<p>host_pred_est_acc - For each tool or approach used for host
    prediction, estimated false discovery rates should be included,
    either computed de novo or from the literature.</p>
<p>Section : sequencing</p>
<p>Expected value : false discovery rate</p>
<p>Value syntax : {text}</p>
<p>Example : CRISPR spacer match: 0 or 1 mismatches, estimated 8% FDR at
the host genus rank (Edwards et al. 2016 doi:10.1093/femsre/fuv048)</p>
</li>
<li>
<p>host_spec_range - The NCBI taxonomy identifier of the specific
    host if it is known.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : NCBI taxid</p>
<p>Value syntax : {integer}</p>
<p>Example : 9606</p>
</li>
<li>
<p>isol_growth_condt - Publication reference in the form of pubmed ID
    (pmid), digital object identifier (doi) or url for isolation and
    growth condition specifications of the organism/material.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : PMID,DOI or URL</p>
<p>Value syntax : {PMID}|{DOI}|{URL}</p>
<p>Example : doi: 10.1016/j.syapm.2018.01.009</p>
</li>
<li>
<p>lib_layout - Specify whether to expect single, paired, or other
    configuration of reads.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [paired|single|vector|other]</p>
<p>Example : paired</p>
</li>
<li>
<p>lib_reads_seqd - Total number of clones sequenced from the
    library.</p>
<p>Section : sequencing</p>
<p>Expected value : number of reads sequenced</p>
<p>Value syntax : {integer}</p>
<p>Example : 20</p>
</li>
<li>
<p>lib_screen - Specific enrichment or screening methods applied
    before and/or after creating libraries.</p>
<p>Section : sequencing</p>
<p>Expected value : screening strategy name</p>
<p>Value syntax : {text}</p>
<p>Example : enriched, screened, normalized</p>
</li>
<li>
<p>lib_size - Total number of clones in the library prepared for the
    project.</p>
<p>Section : sequencing</p>
<p>Expected value : number of clones</p>
<p>Value syntax : {integer}</p>
<p>Example : 50</p>
</li>
<li>
<p>lib_vector - Cloning vector type(s) used in construction of
    libraries.</p>
<p>Section : sequencing</p>
<p>Expected value : vector</p>
<p>Value syntax : {text}</p>
<p>Example : Bacteriophage P1</p>
</li>
<li>
<p>mag_cov_software - Tool(s) used to determine the genome coverage
    if coverage is used as a binning parameter in the extraction of
    genomes from metagenomic datasets.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [bwa|bbmap|bowtie|other]</p>
<p>Example : bbmap</p>
</li>
<li>
<p>mid - Molecular barcodes, called Multiplex Identifiers (MIDs), that
    are used to specifically tag unique samples in a sequencing run.
    Sequence should be reported in uppercase letters.</p>
<p>Section : sequencing</p>
<p>Expected value : multiplex identifier sequence</p>
<p>Value syntax : {dna}</p>
<p>Example : GTGAATAT</p>
</li>
<li>
<p>misc_param - Any other measurement performed or parameter
    collected, that is not listed here.</p>
<p>Section : soil</p>
<p>Expected value : parameter name;measurement value</p>
<p>Value syntax : {text};{float} {unit}</p>
<p>Example : Bicarbonate ion concentration;2075 micromole per kilogram</p>
</li>
<li>
<p>nucl_acid_amp - A link to a literature reference, electronic
    resource or a standard operating procedure (SOP), that describes
    the enzymatic amplification (PCR, TMA, NASBA) of specific nucleic
    acids.</p>
<p>Section : sequencing</p>
<p>Expected value : PMID, DOI or URL</p>
<p>Value syntax : {PMID}|{DOI}|{URL}</p>
<p>Example :
<a href="https://phylogenomics.me/protocols/16s-pcr-protocol/"><u>https://phylogenomics.me/protocols/16s-pcr-protocol/</u></a></p>
</li>
<li>
<p>nucl_acid_ext - A link to a literature reference, electronic
    resource or a standard operating procedure (SOP), that describes
    the material separation to recover the nucleic acid fraction from
    a sample.</p>
<p>Section : sequencing</p>
<p>Expected value : PMID, DOI or URL</p>
<p>Value syntax : {PMID}|{DOI}|{URL}</p>
<p>Example :
<a href="https://mobio.com/media/wysiwyg/pdfs/protocols/12888.pdf"><u>https://mobio.com/media/wysiwyg/pdfs/protocols/12888.pdf</u></a></p>
</li>
<li>
<p>num_replicons - Reports the number of replicons in a nuclear genome
    of eukaryotes, in the genome of a bacterium or archaea or the
    number of segments in a segmented virus. Always applied to the
    haploid chromosome count of a eukaryote.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : for eukaryotes and bacteria: chromosomes (haploid
count); for viruses: segments</p>
<p>Value syntax : {integer}</p>
<p>Example : 2</p>
</li>
<li>
<p>number_contig - Total number of contigs in the cleaned/submitted
    assembly that makes up a given genome, SAG, MAG, or UViG.</p>
<p>Section : sequencing</p>
<p>Expected value : value</p>
<p>Value syntax : {integer}</p>
<p>Example : 40</p>
</li>
<li>
<p>pathogenicity - To what is the entity pathogenic.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : names of organisms that the entity is pathogenic to</p>
<p>Value syntax : {text}</p>
<p>Example : human, animal, plant, fungi, bacteria</p>
</li>
<li>
<p>pcr_cond - Description of reaction conditions and components of PCR
    in the form of 'initial denaturation:94degC_1.5min;
    annealing=...'.</p>
<p>Section : sequencing</p>
<p>Expected value : initial
denaturation:degrees_minutes;annealing:degrees_minutes;elongation:degrees_minutes;final
elongation:degrees_minutes;total cycles</p>
<p>Value syntax : initial
denaturation:degrees_minutes;annealing:degrees_minutes;elongation:degrees_minutes;final
elongation:degrees_minutes;total cycles</p>
<p>Example : initial
denaturation:94_3;annealing:50_1;elongation:72_1.5;final
elongation:72_10;35</p>
</li>
<li>
<p>pcr_primers - PCR primers that were used to amplify the sequence of
    the targeted gene, locus or subfragment. This field should contain
    all the primers used for a single PCR reaction if multiple forward
    or reverse primers are present in a single PCR reaction. The
    primer sequence should be reported in uppercase letters.</p>
<p>Section : sequencing</p>
<p>Expected value : FWD: forward primer sequence;REV:reverse primer
sequence</p>
<p>Value syntax : FWD:{dna};REV:{dna}</p>
<p>Example : FWD:GTGCCAGCMGCCGCGGTAA;REV:GGACTACHVGGGTWTCTAAT</p>
</li>
<li>
<p>ph - Ph measurement of the sample, or liquid portion of sample, or
    aqueous phase of the fluid.</p>
<p>Section : soil</p>
<p>Expected value : measurement value</p>
<p>Value syntax : {float}</p>
<p>Example : 7.2</p>
</li>
<li>
<p>ploidy - The ploidy level of the genome (e.g. allopolyploid,
    haploid, diploid, triploid, tetraploid). It has implications for
    the downstream study of duplicated gene and regions of the genomes
    (and perhaps for difficulties in assembly). For terms, please
    select terms listed under class ploidy (PATO:001374) of Phenotypic
    Quality Ontology (PATO), and for a browser of PATO (v 2018-03-27)
    please refer to
    <a href="http://purl.bioontology.org/ontology/PATO"><u>http://purl.bioontology.org/ontology/PATO</u></a>.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : PATO</p>
<p>Value syntax : {termLabel} {[termID]}</p>
<p>Example : allopolyploidy [PATO:0001379]</p>
</li>
<li>
<p>pred_genome_struc - Expected structure of the viral genome.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [segmented|non-segmented|undetermined]</p>
<p>Example : non-segmented</p>
</li>
<li>
<p>pred_genome_type - Type of genome predicted for the UViG.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [DNA|dsDNA|ssDNA|RNA|dsRNA|ssRNA|ssRNA
(+)|ssRNA (-)|mixed|uncharacterized]</p>
<p>Example : dsDNA</p>
</li>
<li>
<p>propagation - This field is specific to different taxa. For phages:
    lytic/lysogenic, for plasmids: incompatibility group, for
    eukaryotes: sexual/asexual (Note: there is the strong opinion to
    name phage propagation obligately lytic or temperate, therefore we
    also give this choice.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : for virus: lytic, lysogenic, temperate, obligately
lytic; for plasmid: incompatibility group; for eukaryote: asexual,
sexual</p>
<p>Value syntax : {text}</p>
<p>Example : lytic</p>
</li>
<li>
<p>reassembly_bin - Has an assembly been performed on a genome bin
    extracted from a metagenomic assembly?.</p>
<p>Section : sequencing</p>
<p>Expected value : boolean</p>
<p>Value syntax : {boolean}</p>
<p>Example : no</p>
</li>
<li>
<p>ref_biomaterial - Primary publication if isolated before genome
    publication; otherwise, primary genome report.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : PMID, DOI or URL</p>
<p>Value syntax : {PMID}|{DOI}|{URL}</p>
<p>Example : doi:10.1016/j.syapm.2018.01.009</p>
</li>
<li>
<p>ref_db - List of database(s) used for ORF annotation, along with
    version number and reference to website or publication.</p>
<p>Section : sequencing</p>
<p>Expected value : names, versions, and references of databases</p>
<p>Value syntax : {database};{version};{reference}</p>
<p>Example : pVOGs;5;
<a href="http://dmk-brain.ecn.uiowa.edu/pVOGs/"><u>http://dmk-brain.ecn.uiowa.edu/pVOGs/</u></a>
Grazziotin et al. 2017 doi:10.1093/nar/gkw975</p>
</li>
<li>
<p>rel_to_oxygen - Is this organism an aerobe, anaerobe? Please note
    that aerobic and anaerobic are valid descriptors for microbial
    environments.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : enumeration</p>
<p>Value syntax :
[aerobe|anaerobe|facultative|microaerophilic|microanaerobe|obligate
aerobe|obligate anaerobe]</p>
<p>Example : aerobe</p>
</li>
<li>
<p>samp_collect_device - The method or device employed for collecting
    the sample.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : type name</p>
<p>Value syntax : {text}</p>
<p>Example : biopsy, niskin bottle, push core</p>
</li>
<li>
<p>samp_mat_process - Any processing applied to the sample during or
    after retrieving the sample from environment. This field accepts
    OBI, for a browser of OBI (v 2018-02-12) terms please see
    <a href="http://purl.bioontology.org/ontology/OBI"><u>http://purl.bioontology.org/ontology/OBI</u></a>.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : text or OBI</p>
<p>Value syntax : {text}|{termLabel} {[termID]}</p>
<p>Example : filtering of seawater, storing samples in ethanol</p>
</li>
<li>
<p>samp_size - Amount or size of sample (volume, mass or area) that
    was collected.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : measurement value</p>
<p>Preferred unit : millliter, gram, milligram, liter</p>
<p>Value syntax : {float} {unit}</p>
<p>Example : 5 liter</p>
</li>
<li>
<p>samp_vol_we_dna_ext - Volume (ml), weight (g) of processed
    sample, or surface area swabbed from sample for DNA extraction.</p>
<p>Section : soil</p>
<p>Expected value : measurement value</p>
<p>Preferred unit : millliter, gram, milligram, square centimeter</p>
<p>Value syntax : {float} {unit}</p>
<p>Example : 1500 milliliter</p>
</li>
<li>
<p>seq_quality_check - Indicate if the sequence has been called by
    automatic systems (none) or undergone a manual editing procedure
    (e.g. by inspecting the raw data or chromatograms). Applied only
    for sequences that are not submitted to SRA,ENA or DRA.</p>
<p>Section : sequencing</p>
<p>Expected value : none or manually edited</p>
<p>Value syntax : [none|manually edited]</p>
<p>Example : none</p>
</li>
<li>
<p>sim_search_meth - Tool used to compare ORFs with database, along
    with version and cutoffs used.</p>
<p>Section : sequencing</p>
<p>Expected value : names and versions of software(s), parameters used</p>
<p>Value syntax : {software};{version};{parameters}</p>
<p>Example : HMMER3;3.1b2;hmmsearch, cutoff of 50 on score</p>
</li>
<li>
<p>single_cell_lysis_appr - Method used to free DNA from interior of
    the cell(s) or particle(s).</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [chemical|enzymatic|physical|combination]</p>
<p>Example : enzymatic</p>
</li>
<li>
<p>single_cell_lysis_prot - Name of the kit or standard protocol
    used for cell(s) or particle(s) lysis.</p>
<p>Section : sequencing</p>
<p>Expected value : kit, protocol name</p>
<p>Value syntax : {text}</p>
<p>Example : ambion single cell lysis kit</p>
</li>
<li>
<p>size_frac - Filtering pore size used in sample preparation.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : filter size value range</p>
<p>Value syntax : {float}-{float} {unit}</p>
<p>Example : 0-0.22 micrometer</p>
</li>
<li>
<p>sop - Standard operating procedures used in assembly and/or
    annotation of genomes, metagenomes or environmental sequences.</p>
<p>Section : sequencing</p>
<p>Expected value : reference to SOP</p>
<p>Value syntax : {PMID}|{DOI}|{URL}</p>
<p>Example :
<a href="http://press.igsb.anl.gov/earthmicrobiome/protocols-and-standards/its/"><u>http://press.igsb.anl.gov/earthmicrobiome/protocols-and-standards/its/</u></a></p>
</li>
<li>
<p>sort_tech - Method used to sort/isolate cells or particles of
    interest.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [flow cytometric cell
sorting|microfluidics|lazer-tweezing|optical
manipulation|micromanipulation|other]</p>
<p>Example : optical manipulation</p>
</li>
<li>
<p>source_mat_id - A unique identifier assigned to a material sample
    (as defined by
    <a href="http://rs.tdwg.org/dwc/terms/materialSampleID"><u>http://rs.tdwg.org/dwc/terms/materialSampleID</u></a>,
    and as opposed to a particular digital record of a material
    sample) used for extracting nucleic acids, and subsequent
    sequencing. The identifier can refer either to the original
    material collected or to any derived sub-samples. The INSDC
    qualifiers /specimen_voucher, /bio_material, or
    /culture_collection may or may not share the same value as the
    source_mat_id field. For instance, the /specimen_voucher
    qualifier and source_mat_id may both contain 'UAM:Herps:14' ,
    referring to both the specimen voucher and sampled tissue with the
    same identifier. However, the /culture_collection qualifier may
    refer to a value from an initial culture (e.g. ATCC:11775) while
    source_mat_id would refer to an identifier from some derived
    culture from which the nucleic acids were extracted (e.g. xatc123
    or ark:/2154/R2)..</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : for cultures of microorganisms: identifiers for two
culture collections; for other material a unique arbitrary identifer</p>
<p>Value syntax : {text}</p>
<p>Example : MPI012345</p>
</li>
<li>
<p>source_uvig - Type of dataset from which the UViG was obtained.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [metagenome (not viral targeted)|viral fraction
metagenome (virome)|sequence-targeted metagenome|metatranscriptome
(not viral targeted)|viral fraction RNA metagenome (RNA
virome)|sequence-targeted RNA metagenome|microbial single amplified
genome (SAG)|viral single amplified genome (vSAG)|isolate microbial
genome|other]</p>
<p>Example : viral fraction metagenome (virome)</p>
</li>
<li>
<p>specific_host - If there is a host involved, please provide its
    taxid (or environmental if not actually isolated from the dead or
    alive host - i.e. a pathogen could be isolated from a swipe of a
    bench etc) and report whether it is a laboratory or natural host).</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : host taxid, unknown, environmental</p>
<p>Value syntax : {NCBI taxid}|{text}</p>
<p>Example : 9606</p>
</li>
<li>
<p>subspecf_gen_lin - This should provide further information about
    the genetic distinctness of the sequenced organism by recording
    additional information e.g. serovar, serotype, biotype, ecotype,
    or any relevant genetic typing schemes like Group I plasmid. It
    can also contain alternative taxonomic information. It should
    contain both the lineage name, and the lineage rank, i.e.
    biovar:abc123.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : genetic lineage below lowest rank of NCBI taxonomy,
which is subspecies, e.g. serovar, biotype, ecotype</p>
<p>Value syntax : {rank name}:{text}</p>
<p>Example : serovar:Newport</p>
</li>
<li>
<p>target_gene - Targeted gene or locus name for marker gene studies.</p>
<p>Section : sequencing</p>
<p>Expected value : gene name</p>
<p>Value syntax : {text}</p>
<p>Example : 16S rRNA, 18S rRNA, nif, amoA, rpo</p>
</li>
<li>
<p>target_subfragment - Name of subfragment of a gene or locus.
    Important to e.g. identify special regions on marker genes like V6
    on 16S rRNA.</p>
<p>Section : sequencing</p>
<p>Expected value : gene fragment name</p>
<p>Value syntax : {text}</p>
<p>Example : V6, V9, ITS</p>
</li>
<li>
<p>tax_class - Method used for taxonomic classification, along with
    reference database used, classification rank, and thresholds used
    to classify new genomes.</p>
<p>Section : sequencing</p>
<p>Expected value : classification method, database name, and other
parameters</p>
<p>Value syntax : {text}</p>
<p>Example : vConTACT vContact2 (references from NCBI RefSeq v83, genus
rank classification, default parameters)</p>
</li>
<li>
<p>tax_ident - The phylogenetic marker(s) used to assign an organism
    name to the SAG or MAG.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [16S rRNA gene|multi-marker approach|other]</p>
<p>Example : other: rpoB gene</p>
</li>
<li>
<p>tot_nitro_content - Total nitrogen content of the sample.</p>
<p>Section : soil</p>
<p>Expected value : measurement value</p>
<p>Preferred unit : microgram per liter, micromole per liter, milligram
per liter</p>
<p>Value syntax : {float} {unit}</p>
</li>
<li>
<p>tot_org_carb - Definition for soil: total organic carbon content
    of the soil, definition otherwise: total organic carbon content.</p>
<p>Section : soil</p>
<p>Expected value : measurement value</p>
<p>Preferred unit : gram Carbon per kilogram sample material</p>
<p>Value syntax : {float} {unit}</p>
</li>
<li>
<p>trna_ext_software - Tools used for tRNA identification.</p>
<p>Section : sequencing</p>
<p>Expected value : names and versions of software(s), parameters used</p>
<p>Value syntax : {software};{version};{parameters}</p>
<p>Example : infernal;v2;default parameters</p>
</li>
<li>
<p>trnas - The total number of tRNAs identified from the SAG or MAG.</p>
<p>Section : sequencing</p>
<p>Expected value : value from 0-21</p>
<p>Value syntax : {integer}</p>
<p>Example : 18</p>
</li>
<li>
<p>trophic_level - Trophic levels are the feeding position in a food
    chain. Microbes can be a range of producers (e.g.
    chemolithotroph).</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : enumeration</p>
<p>Value syntax :
[autotroph|carboxydotroph|chemoautotroph|chemoheterotroph|chemolithoautotroph|chemolithotroph|chemoorganoheterotroph|chemoorganotroph|chemosynthetic|chemotroph|copiotroph|diazotroph|facultative|autotroph|heterotroph|lithoautotroph|lithoheterotroph|lithotroph|methanotroph|methylotroph|mixotroph|obligate|chemoautolithotroph|oligotroph|organoheterotroph|organotroph|photoautotroph|photoheterotroph|photolithoautotroph|photolithotroph|photosynthetic|phototroph]</p>
<p>Example : heterotroph</p>
</li>
<li>
<p>url.</p>
<p>Section : sequencing</p>
<p>Expected value : URL</p>
<p>Value syntax : {URL}</p>
<p>Example :
<a href="http://www.earthmicrobiome.org/"><u>http://www.earthmicrobiome.org/</u></a></p>
</li>
<li>
<p>vir_ident_software - Tool(s) used for the identification of UViG
    as a viral genome, software or protocol name including version
    number, parameters, and cutoffs used.</p>
<p>Section : sequencing</p>
<p>Expected value : software name, version and relevant parameters</p>
<p>Value syntax : {software};{version};{parameters}</p>
<p>Example : VirSorter; 1.0.4; Virome database, category 2</p>
</li>
<li>
<p>virus_enrich_appr - List of approaches used to enrich the sample
    for viruses, if any.</p>
<p>Section : nucleic acid sequence source</p>
<p>Expected value : enumeration</p>
<p>Value syntax :
[filtration|ultrafiltration|centrifugation|ultracentrifugation|PEG
Precipitation|FeCl Precipitation|CsCl density
gradient|DNAse|RNAse|targeted sequence capture|other|none]</p>
<p>Example : filtration + FeCl Precipitation + ultracentrifugation +
DNAse</p>
</li>
<li>
<p>votu_class_appr - Cutoffs and approach used when clustering new
    UViGs in “species-level” vOTUs. Note that results from standard
    95% ANI / 85% AF clustering should be provided alongside vOTUS
    defined from another set of thresholds, even if the latter are the
    ones primarily used during the analysis.</p>
<p>Section : sequencing</p>
<p>Expected value : cutoffs and method used</p>
<p>Value syntax : {ANI cutoff};{AF cutoff};{clustering method}</p>
<p>Example : 95% ANI;85% AF; greedy incremental clustering</p>
</li>
<li>
<p>votu_db - Reference database (i.e. sequences not generated as part
    of the current study) used to cluster new genomes in
    "species-level" vOTUs, if any.</p>
<p>Section : sequencing</p>
<p>Expected value : database and version</p>
<p>Value syntax : {database};{version}</p>
<p>Example : NCBI Viral RefSeq;83</p>
</li>
<li>
<p>votu_seq_comp_appr - Tool and thresholds used to compare
    sequences when computing "species-level" vOTUs.</p>
<p>Section : sequencing</p>
<p>Expected value : software name, version and relevant parameters</p>
<p>Value syntax : {software};{version};{parameters}</p>
<p>Example : blastn;2.6.0+;e-value cutoff: 0.001</p>
</li>
<li>
<p>water_content - Water content measurement.</p>
<p>Section : soil</p>
<p>Expected value : measurement value</p>
<p>Preferred unit : gram per gram or cubic centimeter per cubic
centimeter</p>
<p>Value syntax : {float}</p>
</li>
<li>
<p>wga_amp_appr - Method used to amplify genomic DNA in preparation
    for sequencing.</p>
<p>Section : sequencing</p>
<p>Expected value : enumeration</p>
<p>Value syntax : [pcr based|mda based]</p>
<p>Example : mda based</p>
</li>
<li>
<p>wga_amp_kit - Kit used to amplify genomic DNA in preparation for
    sequencing.</p>
<p>Section : sequencing</p>
<p>Expected value : kit name</p>
<p>Value syntax : {text}</p>
<p>Example : qiagen repli-g</p>
</li>
</ul>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../identifiers/" class="btn btn-neutral" title="Identifiers in NMDC"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../identifiers/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
