<h1 id="type-genomefeature">Type: GenomeFeature</h1>

<p>A feature localized to an interval along a genome</p>

<p>URI: <a href="https://microbiomedata/meta/GenomeFeature">nmdc:GenomeFeature</a></p>

<hr />

<p><img src="http://yuml.me/diagram/nofunky;dir:TB/class/[OntologyClass],[GeneProduct]%3Cencodes%200..1-%20[GenomeFeature%7Cseqid:string;start:integer;end:integer;strand:string%20%3F;phase:integer%20%3F],[OntologyClass]%3Ctype%200..1-%20[GenomeFeature],[Database]++-%20genome%20feature%20set%200..%2A%3E[GenomeFeature],[GeneProduct],[Database]" alt="img" /></p>

<hr />

<h2 id="referenced-by-class">Referenced by class</h2>

<ul>
  <li><strong><a href="Database.md">Database</a></strong> <em><a href="genome_feature_set.md">genome feature set</a></em>  <sub>0..*</sub>  <strong><a href="GenomeFeature.md">GenomeFeature</a></strong></li>
</ul>

<h2 id="attributes">Attributes</h2>

<h3 id="own">Own</h3>

<ul>
  <li><a href="genome_feature_encodes.md">genome feature➞encodes</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The gene product encoded by this feature. Typically this is used for a CDS feature or gene feature which will encode a protein. It can also be used by a nc transcript ot gene feature that encoded a ncRNA</li>
      <li>range: <a href="GeneProduct.md">GeneProduct</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_end.md">genome feature➞end</a>  <sub>REQ</sub>
    <ul>
      <li>Description: The end of the feature in positive 1-based integer coordinates</li>
      <li>range: <a href="types/Integer.md">Integer</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_phase.md">genome feature➞phase</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The phase for a coding sequence entity. For example, phase of a CDS as represented in a GFF3 with a value of 0, 1 or 2.</li>
      <li>range: <a href="types/Integer.md">Integer</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_seqid.md">genome feature➞seqid</a>  <sub>REQ</sub>
    <ul>
      <li>Description: The ID of the landmark used to establish the coordinate system for the current feature.</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_start.md">genome feature➞start</a>  <sub>REQ</sub>
    <ul>
      <li>Description: The start of the feature in positive 1-based integer coordinates</li>
      <li>range: <a href="types/Integer.md">Integer</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_strand.md">genome feature➞strand</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The strand on which a feature is located. Has a value of ‘+’ (sense strand or forward strand) or ‘-‘ (anti-sense strand or reverse strand).</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_type.md">genome feature➞type</a>  <sub>OPT</sub>
    <ul>
      <li>Description: A type from the sequence ontology</li>
      <li>range: <a href="OntologyClass.md">OntologyClass</a></li>
    </ul>
  </li>
</ul>

<h3 id="domain-for-slot">Domain for slot:</h3>

<ul>
  <li><a href="genome_feature_encodes.md">genome feature➞encodes</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The gene product encoded by this feature. Typically this is used for a CDS feature or gene feature which will encode a protein. It can also be used by a nc transcript ot gene feature that encoded a ncRNA</li>
      <li>range: <a href="GeneProduct.md">GeneProduct</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_end.md">genome feature➞end</a>  <sub>REQ</sub>
    <ul>
      <li>Description: The end of the feature in positive 1-based integer coordinates</li>
      <li>range: <a href="types/Integer.md">Integer</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_phase.md">genome feature➞phase</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The phase for a coding sequence entity. For example, phase of a CDS as represented in a GFF3 with a value of 0, 1 or 2.</li>
      <li>range: <a href="types/Integer.md">Integer</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_seqid.md">genome feature➞seqid</a>  <sub>REQ</sub>
    <ul>
      <li>Description: The ID of the landmark used to establish the coordinate system for the current feature.</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_start.md">genome feature➞start</a>  <sub>REQ</sub>
    <ul>
      <li>Description: The start of the feature in positive 1-based integer coordinates</li>
      <li>range: <a href="types/Integer.md">Integer</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_strand.md">genome feature➞strand</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The strand on which a feature is located. Has a value of ‘+’ (sense strand or forward strand) or ‘-‘ (anti-sense strand or reverse strand).</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="genome_feature_type.md">genome feature➞type</a>  <sub>OPT</sub>
    <ul>
      <li>Description: A type from the sequence ontology</li>
      <li>range: <a href="OntologyClass.md">OntologyClass</a></li>
    </ul>
  </li>
</ul>

<h2 id="other-properties">Other properties</h2>

<table>
  <thead>
    <tr>
      <th> </th>
      <th> </th>
      <th> </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>Comments:</strong></td>
      <td> </td>
      <td>corresponds to an entry in GFF3</td>
    </tr>
    <tr>
      <td><strong>See also:</strong></td>
      <td> </td>
      <td>https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md</td>
    </tr>
  </tbody>
</table>

