<p>Schema for National Microbiome Data Collaborative (NMDC). This schem is organized into 3 separate modules:</p>

<ul>
  <li>a set of core types for representing data values * the mixs schema (auto-translated from mixs excel) * the NMDC schema itself</li>
</ul>

<h3 id="classes">Classes</h3>

<ul>
  <li><a href="Activity.md">Activity</a> - a provence-generating activity
    <ul>
      <li><a href="WorkflowExecutionActivity.md">WorkflowExecutionActivity</a> - Represents an instance of an execution of a particular workflow
        <ul>
          <li><a href="MAGAnnotationActivity.md">MAGAnnotationActivity</a></li>
          <li><a href="MetabolomicsAnalysisActivity.md">MetabolomicsAnalysisActivity</a></li>
          <li><a href="MetagenomeAnnotationActivity.md">MetagenomeAnnotationActivity</a></li>
          <li><a href="MetagenomeAssembly.md">MetagenomeAssembly</a></li>
          <li><a href="MetaproteomicsAnalysisActivity.md">MetaproteomicsAnalysisActivity</a></li>
          <li><a href="ReadQCAnalysisActivity.md">ReadQCAnalysisActivity</a></li>
        </ul>
      </li>
    </ul>
  </li>
  <li><a href="Agent.md">Agent</a> - a provence-generating agent</li>
  <li><a href="AttributeValue.md">AttributeValue</a> - The value for any value of a attribute for a sample. This object can hold both the un-normalized atomic value and the structured value
    <ul>
      <li><a href="BooleanValue.md">BooleanValue</a> - A value that is a boolean</li>
      <li><a href="ControlledTermValue.md">ControlledTermValue</a> - A controlled term or class from an ontology</li>
      <li><a href="GeolocationValue.md">GeolocationValue</a> - A normalized value for a location on the earth’s surface</li>
      <li><a href="IntegerValue.md">IntegerValue</a> - A value that is an integer</li>
      <li><a href="PersonValue.md">PersonValue</a> - An attribute value representing a person</li>
      <li><a href="QuantityValue.md">QuantityValue</a> - A simple quantity, e.g. 2cm</li>
      <li><a href="TextValue.md">TextValue</a> - A basic string value</li>
      <li><a href="TimestampValue.md">TimestampValue</a> - A value that is a timestamp. The range should be ISO-8601</li>
      <li><a href="UrlValue.md">UrlValue</a> - A value that is a string that conforms to URL syntax</li>
    </ul>
  </li>
  <li><a href="Database.md">Database</a> - An abstract holder for any set of metadata and data. It does not need to correspond to an actual managed databse top level holder class. When translated to JSON-Schema this is the ‘root’ object. It should contain pointers to other objects of interest</li>
  <li><a href="FunctionalAnnotation.md">FunctionalAnnotation</a> - An assignment of a function term (e.g. reaction or pathway) that is executed by a gene product, or which the gene product plays an active role in. Functional annotations can be assigned manually by curators, or automatically in workflows. In the context of NMDC, all function annotation is performed automatically, typically using HMM or Blast type methods</li>
  <li><a href="GenomeFeature.md">GenomeFeature</a> - A feature localized to an interval along a genome</li>
  <li><a href="MetaboliteQuantification.md">MetaboliteQuantification</a> - This is used to link a metabolomics analysis workflow to a specific metabolite</li>
  <li><a href="NamedThing.md">NamedThing</a> - a databased entity or concept/class
    <ul>
      <li><a href="Biosample.md">Biosample</a> - A material sample. It may be environmental (encompassing many organisms) or isolate or tissue.   An environmental sample containing genetic material from multiple individuals is commonly referred to as a biosample.</li>
      <li><a href="BiosampleProcessing.md">BiosampleProcessing</a> - A process that takes one or more biosamples as inputs and generates one or as outputs. Examples of outputs include samples cultivated from another sample or data objects created by instruments runs.
        <ul>
          <li><a href="OmicsProcessing.md">OmicsProcessing</a> - The methods and processes used to generate omics data from a biosample or organism.</li>
        </ul>
      </li>
      <li><a href="DataObject.md">DataObject</a> - An object that primarily consists of symbols that represent information.   Files, records, and omics data are examples of data objects.</li>
      <li><a href="GeneProduct.md">GeneProduct</a> - A molecule encoded by a gene that has an evolved function</li>
      <li><a href="Instrument.md">Instrument</a> - A material entity that is designed to perform a function in a scientific investigation, but is not a reagent[OBI].</li>
      <li><a href="OntologyClass.md">OntologyClass</a>
        <ul>
          <li><a href="ChemicalEntity.md">ChemicalEntity</a> - An atom or molecule that can be represented with a chemical formula. Include lipids, glycans, natural products, drugs. There may be different terms for distinct acid-base forms, protonation states</li>
          <li><a href="EnvironmentalMaterialTerm.md">EnvironmentalMaterialTerm</a></li>
          <li><a href="FunctionalAnnotationTerm.md">FunctionalAnnotationTerm</a> - Abstract grouping class for any term/descriptor that can be applied to a functional unit of a genome (protein, ncRNA, complex).
            <ul>
              <li><a href="OrthologyGroup.md">OrthologyGroup</a> - A set of genes or gene products in which all members are orthologous</li>
              <li><a href="Pathway.md">Pathway</a> - A pathway is a sequence of steps/reactions carried out by an organism or community of organisms</li>
              <li><a href="Reaction.md">Reaction</a> - An individual biochemical transformation carried out by a functional unit of an organism, in which a collection of substrates are transformed into a collection of products. Can also represent transporters</li>
            </ul>
          </li>
        </ul>
      </li>
      <li><a href="Person.md">Person</a> - represents a person, such as a researcher</li>
      <li><a href="Study.md">Study</a> - A study summarizes the overall goal of a research initiative and outlines the key objective of its underlying projects.</li>
    </ul>
  </li>
  <li><a href="PeptideQuantification.md">PeptideQuantification</a> - This is used to link a metaproteomics analysis workflow to a specific peptide sequence and related information</li>
  <li><a href="ProteinQuantification.md">ProteinQuantification</a> - This is used to link a metaproteomics analysis workflow to a specific protein</li>
  <li><a href="ReactionParticipant.md">ReactionParticipant</a> - Instances of this link a reaction to a chemical entity participant</li>
</ul>
