<h1 id="type-metagenomeassembly">Type: MetagenomeAssembly</h1>

<p>URI: <a href="https://microbiomedata/meta/MetagenomeAssembly">nmdc:MetagenomeAssembly</a></p>

<hr />

<p><img src="http://yuml.me/diagram/nofunky;dir:TB/class/[WorkflowExecutionActivity],[WorkflowExecutionActivity]%5E-[MetagenomeAssembly%7Casm_score:float%20%3F;scaffolds:float%20%3F;scaf_logsum:float%20%3F;scaf_powsum:float%20%3F;scaf_max:float%20%3F;scaf_bp:float%20%3F;scaf_N50:float%20%3F;scaf_N90:float%20%3F;scaf_L50:float%20%3F;scaf_L90:float%20%3F;scaf_n_gt50K:float%20%3F;scaf_l_gt50K:float%20%3F;scaf_pct_gt50K:float%20%3F;contigs:float%20%3F;contig_bp:float%20%3F;ctg_N50:float%20%3F;ctg_L50:float%20%3F;ctg_N90:float%20%3F;ctg_L90:float%20%3F;ctg_logsum:float%20%3F;ctg_powsum:float%20%3F;ctg_max:float%20%3F;gap_pct:float%20%3F;gc_std:float%20%3F;gc_avg:float%20%3F;num_input_reads:float%20%3F;num_aligned_reads:float%20%3F;execution_resource(i):string%20%3F;git_url(i):string%20%3F;has_input(i):string%20%2A;has_output(i):string%20%2A;id(i):string;name(i):string%20%3F;type(i):string%20%3F;started_at_time(i):string%20%3F;ended_at_time(i):string%20%3F;used(i):string%20%3F],[Agent],[Activity]" alt="img" /></p>

<hr />

<h2 id="parents">Parents</h2>

<ul>
  <li>is_a: <a href="WorkflowExecutionActivity.md">WorkflowExecutionActivity</a> - Represents an instance of an execution of a particular workflow</li>
</ul>

<h2 id="attributes">Attributes</h2>

<h3 id="own">Own</h3>

<ul>
  <li><a href="asm_score.md">asm_score</a>  <sub>OPT</sub>
    <ul>
      <li>Description: A score for comparing metagenomic assembly quality from same sample.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="contig_bp.md">contig_bp</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Total size in bp of all contigs.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="contigs.md">contigs</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The sum of the (length*log(length)) of all contigs, times some constant.  Increase the contiguity, the score will increase</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="ctg_L50.md">ctg_L50</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Given a set of contigs, the N50 is defined as the sequence length of the shortest contig at 50% of the total genome length.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="ctg_L90.md">ctg_L90</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The L90 statistic is less than or equal to the L50 statistic; it is the length for which the collection of all contigs of that length or longer contains at least 90% of the sum of the lengths of all contigs.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="ctg_N50.md">ctg_N50</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Given a set of contigs, each with its own length, the L50 count is defined as the smallest number of contigs whose length sum makes up half of genome size.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="ctg_N90.md">ctg_N90</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Given a set of contigs, each with its own length, the L90 count is defined as the smallest number of contigs whose length sum makes up 90% of genome size.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="ctg_logsum.md">ctg_logsum</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Maximum contig length.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="ctg_max.md">ctg_max</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Maximum contig length.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="ctg_powsum.md">ctg_powsum</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Powersum of all contigs is the same as logsum except that it uses the sum of (length*(length^P)) for some power P (default P=0.25).</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="gap_pct.md">gap_pct</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The gap size percentage of all scaffolds.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="gc_avg.md">gc_avg</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Average of GC content of all contigs.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="gc_std.md">gc_std</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Standard deviation of GC content of all contigs.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="num_aligned_reads.md">num_aligned_reads</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The sequence count number of input reads aligned to assembled contigs.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="num_input_reads.md">num_input_reads</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The sequence count number of input reads for assembly.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_L50.md">scaf_L50</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Given a set of scaffolds, the N50 is defined as the sequence length of the shortest scaffold at 50% of the total genome length.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_L90.md">scaf_L90</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The L90 statistic is less than or equal to the L50 statistic; it is the length for which the collection of all scaffolds of that length or longer contains at least 90% of the sum of the lengths of all scaffolds.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_N50.md">scaf_N50</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Given a set of scaffolds, each with its own length, the L50 count is defined as the smallest number of scaffolds whose length sum makes up half of genome size.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_N90.md">scaf_N90</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Given a set of scaffolds, each with its own length, the L90 count is defined as the smallest number of scaffolds whose length sum makes up 90% of genome size.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_bp.md">scaf_bp</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Total size in bp of all scaffolds.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_l_gt50K.md">scaf_l_gt50K</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Total size in bp of all scaffolds greater than 50 KB.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_logsum.md">scaf_logsum</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The sum of the (length*log(length)) of all scaffolds, times some constant.  Increase the contiguity, the score will increase</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_max.md">scaf_max</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Maximum scaffold length.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_n_gt50K.md">scaf_n_gt50K</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Total sequence count of scaffolds greater than 50 KB.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_pct_gt50K.md">scaf_pct_gt50K</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Total sequence size percentage of scaffolds greater than 50 KB.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaf_powsum.md">scaf_powsum</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Powersum of all scaffolds is the same as logsum except that it uses the sum of (length*(length^P)) for some power P (default P=0.25).</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
  <li><a href="scaffolds.md">scaffolds</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Total sequence count of all scaffolds.</li>
      <li>range: <a href="types/Float.md">Float</a></li>
    </ul>
  </li>
</ul>

<h3 id="inherited-from-activity">Inherited from activity:</h3>

<ul>
  <li><a href="id.md">id</a>  <sub>REQ</sub>
    <ul>
      <li>Description: A unique identifier for a thing. Must be either a CURIE shorthand for a URI or a complete URI</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="name.md">name</a>  <sub>OPT</sub>
    <ul>
      <li>Description: A human readable label for an entity</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="type.md">type</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The type of entity that the json object represents.  This is used to allow for searches for different kinds of objects.</li>
      <li>range: <a href="types/String.md">String</a></li>
      <li>Example: nmdc:Biosample None</li>
      <li>Example: nmdc:Study None</li>
    </ul>
  </li>
  <li><a href="started_at_time.md">started at time</a>  <sub>OPT</sub>
    <ul>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="ended_at_time.md">ended at time</a>  <sub>OPT</sub>
    <ul>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="was_associated_with.md">was associated with</a>  <sub>OPT</sub>
    <ul>
      <li>range: <a href="Agent.md">Agent</a></li>
    </ul>
  </li>
  <li><a href="used.md">used</a>  <sub>OPT</sub>
    <ul>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
</ul>

<h3 id="inherited-from-agent">Inherited from agent:</h3>

<ul>
  <li><a href="acted_on_behalf_of.md">acted on behalf of</a>  <sub>OPT</sub>
    <ul>
      <li>range: <a href="Agent.md">Agent</a></li>
    </ul>
  </li>
  <li><a href="was_informed_by.md">was informed by</a>  <sub>OPT</sub>
    <ul>
      <li>range: <a href="Activity.md">Activity</a></li>
    </ul>
  </li>
</ul>

<h3 id="inherited-from-biosample-processing">Inherited from biosample processing:</h3>

<ul>
  <li><a href="biosample_processing_has_input.md">biosample processing➞has input</a>  <sub>0..*</sub>
    <ul>
      <li>range: <a href="Biosample.md">Biosample</a></li>
    </ul>
  </li>
</ul>

<h3 id="inherited-from-omics-processing">Inherited from omics processing:</h3>

<ul>
  <li><a href="omics_processing_part_of.md">omics processing➞part of</a>  <sub>0..*</sub>
    <ul>
      <li>range: <a href="Study.md">Study</a></li>
    </ul>
  </li>
  <li><a href="omics_processing_has_output.md">omics processing➞has output</a>  <sub>0..*</sub>
    <ul>
      <li>range: <a href="DataObject.md">DataObject</a></li>
    </ul>
  </li>
  <li><a href="omics_type.md">omics type</a>  <sub>OPT</sub>
    <ul>
      <li>Description: The type of omics data</li>
      <li>range: <a href="ControlledTermValue.md">ControlledTermValue</a></li>
      <li>Example: metatranscriptome None</li>
      <li>Example: metagenome None</li>
    </ul>
  </li>
  <li><a href="omics_processing_id.md">omics processing➞id</a>  <sub>REQ</sub>
    <ul>
      <li>Description: The primary identifier for the omics processing. E.g. GOLD:GpNNNN</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="omics_processing_name.md">omics processing➞name</a>  <sub>OPT</sub>
    <ul>
      <li>Description: A human readable name or description of the omics processing.</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="omics_processing_alternate_identifiers.md">omics processing➞alternate identifiers</a>  <sub>0..*</sub>
    <ul>
      <li>Description: The same omics processing may have distinct identifiers in different databases (e.g. GOLD and EMSL, as well as NCBI)</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
</ul>

<h3 id="inherited-from-workflow-execution-activity">Inherited from workflow execution activity:</h3>

<ul>
  <li><a href="execution_resource.md">execution resource</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Example: NERSC-Cori</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
  <li><a href="git_url.md">git url</a>  <sub>OPT</sub>
    <ul>
      <li>Description: Example: https://github.com/microbiomedata/mg_annotation/releases/tag/0.1</li>
      <li>range: <a href="types/String.md">String</a></li>
    </ul>
  </li>
</ul>

<h2 id="other-properties">Other properties</h2>

<table>
  <thead>
    <tr>
      <th> </th>
      <th> </th>
      <th> </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>In Subsets:</strong></td>
      <td> </td>
      <td>workflow subset</td>
    </tr>
  </tbody>
</table>

